---
title: "ATAC-seq"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(seriation)
source("../util/_plot_theme.R")
source("../util/_util.R")
source("../01_setup/assumptions.R")

# Keep track of thresholds for figure naming
thresh <- paste0("pval", pval_thresh)
```

```{r load, include=FALSE}
load("../01_setup/results/atacseq_data.RData", verbose = T)
```

```{r}
if(!file.exists("results/atacseq_firre_vs_control.RData")) {
  
  atac_dds <- DESeqDataSetFromMatrix(countData = atac_counts, 
                                     colData = atac_samples, 
                                     design = ~ condition)
  
  atac_dds <- DESeq(atac_dds)
  
  atac_lfc <- results(atac_dds) %>%
    as.data.frame() %>%
    rownames_to_column("interval_id") %>%
    merge(atac_consensus %>% 
            dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, 
                          Distance.to.TSS, Chr, Start, End)) %>% 
    unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
    unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)
  names(atac_lfc)[8] <- "gene_name"
  names(atac_lfc)[9] <- "gene_id"
  
  atac_shrnklfc <- lfcShrink(atac_dds, 
                             coef = "condition_firre_vs_control", 
                             type = "apeglm") %>%
    as.data.frame() %>%
    rownames_to_column("interval_id")
  
  atac_shrnklfc <- atac_shrnklfc %>%
    dplyr::select(interval_id, log2FoldChange) %>%
    dplyr::rename(l2fc_shrnk = log2FoldChange)
  
  atac_lfc <- atac_lfc %>% left_join(atac_shrnklfc)
  
  atac_rlog_counts <- rlog(atac_dds, blind = TRUE) %>%
    assay(rlog_counts) %>%
    as.data.frame() %>%
    rownames_to_column("interval_id") %>%
    pivot_longer(2:ncol(.), names_to = "sample_name", values_to = "count") %>%
    left_join(atac_samples)
  
  # fc_vs_zero
  fi_atac_samples <- atac_samples %>%
    filter(firre_induced == "firre_induced")
  fi_atac_counts <- atac_counts[,fi_atac_samples$sample_name]
  atac_vszero_dds <- DESeqDataSetFromMatrix(countData = fi_atac_counts, 
                                            colData = fi_atac_samples, 
                                            design = ~ timepoint_minutes)
  atac_vszero_dds <- DESeq(atac_vszero_dds)
  vs_zero_res <- resultsNames(atac_vszero_dds)
  vs_zero_res <- vs_zero_res[grepl("timepoint", vs_zero_res)]
  
  atac_vszero_lfc <- lapply(vs_zero_res, function(x) {
    results(atac_vszero_dds, name = x) %>%
      as.data.frame() %>%
      rownames_to_column("interval_id") %>% 
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_minutes_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  atac_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(atac_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("interval_id") %>% 
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_minutes_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  save(atac_lfc, atac_vszero_lfc, atac_vszero_shrnklfc, 
       atac_rlog_counts, file = "results/atacseq_firre_vs_control.RData")
}
load("results/atacseq_firre_vs_control.RData", verbose = T)
```

# Call sigificantly changing ATAC-seq peaks

```{r}
atac_dox_sig <- atac_lfc %>%
  filter(padj < pval_thresh)

atac_vszero_sig <- atac_vszero_lfc %>%
  filter(interval_id %in% atac_dox_sig$interval_id)

save(atac_vszero_sig, file = "results/atac_vszero_sig.RData")
```

# Look for overlaps of peaks and differentially expressed genes

We're going to take a peak centric view and look for promoters that overlap within a window of the peak center.

```{r}
load("../11_short_timecourse_combined/results/short_vszero_sig.RData")
sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% unique(atac_vszero_sig$interval_id)]

promoters_gr <- promoters(genes, upstream = 1, downstream = 1)
promoters_df <- promoters_gr %>% as.data.frame() %>%
  filter(gene_type %in% c("protein_coding", "processed_transcript", "bidirectional_promoter_lncRNA",
                          "antisense", "lincRNA", "sense_intronic"))

promoters_gr <- promoters_gr[promoters_gr$gene_id %in%promoters_df$gene_id]

peak_centers <- start(sig_atac_gr) + round(width(sig_atac_gr)/2)
sig_atac_center <- GRanges(seqnames = seqnames(sig_atac_gr),
                           ranges = IRanges(start = peak_centers,
                                            end = peak_centers),
                           interval_id = sig_atac_gr$interval_id)
offset <- 1e6
start(sig_atac_center) <- start(sig_atac_center) - offset
end(sig_atac_center) <- end(sig_atac_center) + offset

peak_tss_overlaps <- findOverlaps(promoters_gr, sig_atac_center)

atac_df <- sig_atac_center %>% as.data.frame()
promoters_df <- promoters_gr %>% as.data.frame()
overlapping_df <- atac_df[peak_tss_overlaps@to, ]
overlapping_df$gene_id <- promoters_df$gene_id[peak_tss_overlaps@from]
overlapping_df$gene_name <- promoters_df$gene_name[peak_tss_overlaps@from]
overlapping_df$gene_tss <- promoters_df$start[peak_tss_overlaps@from]
overlapping_df$gene_chr <- promoters_df$seqnames[peak_tss_overlaps@from]
overlapping_df$gene_strand <- promoters_df$strand[peak_tss_overlaps@from]


overlapping_df <- overlapping_df %>%
  mutate(dist_to_tss = gene_tss - (start+offset),
         dist_to_peak = (start+offset) - gene_tss)
# Flip the minus strand differences so that it's a TSS centric view
overlapping_df[overlapping_df$gene_strand == "-", "dist_to_peak"] <- -1* overlapping_df[overlapping_df$gene_strand == "-", "dist_to_peak"] 

overlapping_df$deg <- overlapping_df$gene_name %in%short_vszero_sig$gene_name
overlapping_df$y <- runif(nrow(overlapping_df))

# Gfap has a closer peak changes that explains its expression change.
overlapping_df$deg[overlapping_df$interval_id == "Interval_26047" &
                     overlapping_df$gene_name == "Gfap"] <- FALSE

ggplot(overlapping_df %>% filter(gene_name != "Firre", deg == TRUE), aes(x = dist_to_peak, fill = deg)) +
  geom_density(adjust = 1.2) +
  geom_vline(xintercept = 0, lty = 2) +
  geom_rug() +
  scale_fill_manual(values = c("#424242","#a8404c")) +
  scale_x_continuous(breaks = seq(from = -1e5, to = 1e5, by = 5e4),
                     labels = c("-100", "-50", "TSS", "50", "100"),
                     limits = c(-1e5,1e5))
ggsave("figures/atac_peak_tss_density.pdf", height = 2, width = 2.5)
# Intergenic cutoff
```


```{r}
atac_lfc_matrix <- atac_vszero_sig %>%
  dplyr::select(interval_id, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("interval_id") %>%
  as.matrix()
ordering <- atac_vszero_sig %>%
  filter(abs(log2FoldChange) > 0.2) %>%
  group_by(interval_id) %>%
  summarize(first_tp_de = min(timepoint),
            max_fc = max(log2FoldChange)) %>%
  arrange(first_tp_de,
          -max_fc)

atac_lfc_matrix <- cbind(matrix(0, nrow = nrow(atac_lfc_matrix), ncol = 1), atac_lfc_matrix)
colnames(atac_lfc_matrix)[[1]] <- "0"

atac_lfc_matrix <- atac_lfc_matrix[ordering$interval_id,]

row_ha = rowAnnotation(deg = as.numeric(rownames(atac_lfc_matrix) %in% overlapping_df$interval_id[overlapping_df$deg == TRUE]),
                       col = list(deg = c("1" = "black", "0" = "white")))

table(as.numeric(rownames(atac_lfc_matrix) %in% overlapping_df$interval_id[overlapping_df$deg == TRUE]))


pdf(paste0("figures/atac_peaks_heatmap_", thresh, ".pdf"), 
    width = 4, height = 3.5)
ht1 <- Heatmap(atac_lfc_matrix, 
               name = "l2fc",
               cluster_columns = FALSE, show_row_names = FALSE, 
               cluster_rows = TRUE,
               left_annotation = row_ha,
               col = colorRamp2(seq(-2,2,length.out = 100), col_pal10))
draw(ht1)
dev.off()
draw(ht1)
```

```{r}
# Let's check out how many ATAC peaks are near genes
# Gosh, how to get the right TSS? Oh man. 
# gencode <- rtracklayer::import(file.path("../../../../genomes/Mus_musculus/",
#                                            "Gencode/M25/",
#                                            "gencode.vM25.annotation.gtf"))

sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% unique(atac_vszero_sig$interval_id)]
length(sig_atac_gr)
promoters_gr <- promoters(genes, upstream = 1, downstream = 1)
promoters_df <- promoters_gr %>% as.data.frame() %>%
  filter(gene_type %in% c("protein_coding", "processed_transcript", "bidirectional_promoter_lncRNA",
                          "antisense", "lincRNA", "sense_intronic"))

promoters_gr <- promoters_gr[promoters_gr$gene_id %in%promoters_df$gene_id]

table(promoters_df$gene_type)
sig_atac_center <- GRanges(seqnames = seqnames(sig_atac_gr),
                           ranges = IRanges(start = start(sig_atac_gr) + ((end(sig_atac_gr) - start(sig_atac_gr))/2),
                                            end = start(sig_atac_gr) + ((end(sig_atac_gr) - start(sig_atac_gr))/2) + 1),
                           interval_id = sig_atac_gr$interval_id)
offset <- 1e6
start(sig_atac_center) <- start(sig_atac_center) - offset
end(sig_atac_center) <- end(sig_atac_center) + offset

peak_tss_overlaps <- findOverlaps(promoters_gr, sig_atac_center)

atac_df <- sig_atac_center %>% as.data.frame()
promoters_df <- promoters_gr %>% as.data.frame() %>% dplyr::select(gene_name, gene_id, seqnames, start, end)
overlapping_df <- atac_df[peak_tss_overlaps@to, ]
overlapping_df$gene_id <- promoters_df$gene_id[peak_tss_overlaps@from]
overlapping_df$gene_name <- promoters_df$gene_name[peak_tss_overlaps@from]
overlapping_df$gene_tss <- promoters_df$start[peak_tss_overlaps@from]
overlapping_df$gene_chr <- promoters_df$seqnames[peak_tss_overlaps@from]

overlapping_df <- overlapping_df %>%
  mutate(dist_to_tss = gene_tss - (start+offset))
length(unique(overlapping_df$interval_id))

overlapping_df$deg <- overlapping_df$gene_name %in%short_vszero_sig$gene_name
overlapping_df$y <- runif(nrow(overlapping_df))

# Gfap has a closer peak changes that explains its expression change.
overlapping_df$deg[overlapping_df$interval_id == "Interval_26047" &
                     overlapping_df$gene_name == "Gfap"] <- FALSE

ggplot(overlapping_df %>% filter(gene_name != "Firre"), aes(x = dist_to_tss, y = y, color = deg)) +
  geom_point() +
  facet_grid(deg~.)

# Intergenic cutoff
```


```{r}
# INtergenic cutoff
cutoffs <- seq(1e6,0, by = -10000)
x <- cutoffs[[100]]
pl <- sapply(cutoffs, function(x) {
  cut_df <- overlapping_df[abs(overlapping_df$dist_to_tss) <= x,]
  length(unique(cut_df$interval_id))
})

gl <- sapply(cutoffs, function(x) {
  cut_df <- overlapping_df[abs(overlapping_df$dist_to_tss) <= x,]
  cut_df <- cut_df[cut_df$deg == TRUE,]
  length(unique(cut_df$gene_name))
})


ig_cutoffs <- tibble(cutoff = cutoffs,
                     num_peaks = pl,
                     num_deg = gl) %>%
  mutate(ratio = num_deg / num_peaks)

```

```{r}
# Let's take a look at which genes started "off"
load("../01_setup/results/rnaseq_data.RData", verbose = T)

sig_gene_tpm <- tpm %>%
  filter(gene_id %in% short_vszero_sig$gene_id) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC",firre_induced == "firre_induced", as.numeric(as.character(timepoint)) == 0,
         timecourse_length == "short") %>%
  group_by(gene_id, gene_name) %>%
  summarize(tpm = mean(tpm))


sig_gene_tpm$has_atac_peak <- sig_gene_tpm$gene_name %in% overlapping_df$gene_name[overlapping_df$deg == TRUE]
sig_gene_tpm$has_atac_peak[sig_gene_tpm$gene_name == "Firre"] <- FALSE
ggplot(sig_gene_tpm %>% filter(gene_name != "Firre"), aes(x = tpm, fill = has_atac_peak)) +
  geom_density(alpha = 0.2)
```



```{r}

# Find out which DE genes have an ATAC peak changing nearby
load("../11_short_timecourse_combined/results/short_vszero_sig.RData")
sig_genes_gr <- genes[genes$gene_name %in% short_vszero_sig$gene_name]

sig_genes_df <- sig_genes_gr %>% as.data.frame()
length(sig_genes_gr)
# Extend 10kb
start(sig_genes_gr) <- start(sig_genes_gr) - 10000
end(sig_genes_gr) <- end(sig_genes_gr) + 10000

sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% unique(atac_vszero_sig$interval_id)]
length(sig_atac_gr)
peak_gene_overlaps <- findOverlaps(sig_genes_gr, sig_atac_gr)
atac_df <- sig_atac_gr %>% as.data.frame()
sig_genes_df <- sig_genes_gr %>% as.data.frame() %>% dplyr::select(gene_name, gene_id, seqnames, start, end)
overlapping_df <- atac_df[peak_gene_overlaps@to, ]
overlapping_df$gene_id <- sig_genes_df$gene_id[peak_gene_overlaps@from]
overlapping_df$gene_name <- sig_genes_df$gene_name[peak_gene_overlaps@from]
overlapping_df$gene_tss <- sig_genes_df$start[peak_gene_overlaps@from] + 10000
overlapping_df$gene_chr <- sig_genes_df$seqnames[peak_gene_overlaps@from]
length(unique(overlapping_df$gene_name))


overlapping_df <- overlapping_df %>%
  mutate(peak_center = start + (end - start),
         dist_to_tss = peak_center - gene_tss)

overlapping_df <- overlapping_df %>%
  mutate(sig_peak = interval_id %in% sig_atac_peaks$interval_id)
```



```{r}
sig_peak_counts <- atac_rlog_counts %>%
  filter(interval_id %in% sig_atac_peaks$interval_id)

sig_peak_count_summary <- sig_peak_counts %>%
  filter(firre_induced == "firre_induced") %>%
  group_by(interval_id, timepoint) %>%
  summarize(count = mean(count))

zero_tp <- sig_peak_count_summary %>%
  filter(timepoint == 0) %>%
  dplyr::select(-timepoint) %>%
  dplyr::rename(zero_val = count)

sig_peak_count_summary <- sig_peak_count_summary %>%
  left_join(zero_tp) %>%
  mutate(fc = count / zero_val,
         l2fc = log2(fc))

sp_matrix <- sig_peak_count_summary %>%
  dplyr::select(interval_id, timepoint, l2fc) %>%
  pivot_wider(names_from = timepoint, names_sort = T, values_from = l2fc) %>%
  column_to_rownames("interval_id") %>%
  as.matrix()

sp_scaled <- t(scale(t(sig_peak_count_summary)))
pheatmap::pheatmap(sp_matrix, cluster_cols = F, show_rownames = F, breaks = seq(-0.25, 0.25, length.out = length(col_pal10)),
                   color = col_pal10)


short_lfc <- short_vszero_sig %>%
  dplyr::select(gene_name, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()
```



```{r}
ggplot(rlog_df %>% filter(interval_id == "Interval_76872", firre_induced == "firre_induced"), aes(x = timepoint, y = count, color = firre_ko)) +
  geom_point()


summary(mod1)
sig_atac_peaks <- atac_lfc %>%
  filter(padj <= 0.05)

# abs(log2FoldChange) > log2(1.5)
```

```{r}

load("../09_short_timecourse_overexpression/results/wt_overexp_short_vszero_sig.RData", verbose = T)
load("../06_short_timecourse_rescue/results/ko_rescue_short_vszero_sig.RData")

ko_rescue_1 <- unique(ko_rescue_short_vszero_sig$gene_name)
wt_overexp_2 <- unique(wt_overexp_short_vszero_sig$gene_name)
short_3 <- unique(short_vszero_sig$gene_name)

individual_overlaps_12 <- intersect(ko_rescue_1, wt_overexp_2)
all_overlaps_123 <- intersect(individual_overlaps_12, short_3)
all_sig_genes <- unique(c(ko_rescue_1, wt_overexp_2, short_3))
gencode_genes <- rtracklayer::import("../util/gencode.vM25.annotation.genes.gtf")

# sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% sig_atac_peaks$interval_id]







# What is the difference between the genes changing without an ATAC-seq peak change
# and those that do? 
# Are we over controlling for dox? Are there ATAC-seq windows in these regions that have 
# linearly increasing trends, but are not significant when compared to the dox control? 




# Let's retrieve the peak counts for the ATAC-seq peaks in the region
# of these genes.
peaks_of_interest <- unique(overlapping_df$interval_id)

library(broom)

peak_id <- peaks_of_interest[[1]]
# Run linear model
peak_lm <- function(peak_id) {
  peak1 <- rlog_df %>% filter(interval_id == peak_id, 
                              firre_induced == "firre_induced")
  mod1 <- lm(peak1$count~peak1$timepoint)
  res1 <- tidy(mod1) 
  res1$intercept <- res1$estimate[[1]]
  res1 <- res1[2,2:ncol(res1)]
  res1 <- res1 %>%
    mutate(interval_id = peak_id) %>%
    dplyr::rename(slope = estimate)
  #WT
  peak1 <- rlog_df %>% filter(interval_id == peak_id, 
                              firre_induced == "firre_induced",
                              firre_ko == "WT")
  mod2 <- lm(peak1$count~peak1$timepoint)
  res2 <- tidy(mod2) 
  res2$intercept <- res2$estimate[[1]]
  res2 <- res2[2,2:ncol(res2)]
  res2 <- res2 %>%
    dplyr::select(estimate, p.value, intercept) %>%
    dplyr::rename(WT_slope = estimate,
                  WT_pval = p.value,
                  WT_intercept = intercept)
  #KO
  peak1 <- rlog_df %>% filter(interval_id == peak_id, 
                              firre_induced == "firre_induced",
                              firre_ko == "KO")
  mod3 <- lm(peak1$count~peak1$timepoint)
  res3 <- tidy(mod3) 
  res3$intercept <- res3$estimate[[1]]
  res3 <- res3[2,2:ncol(res3)]
  res3 <- res3 %>%
    dplyr::select(estimate, p.value, intercept) %>%
    dplyr::rename(KO_slope = estimate,
                  KO_pval = p.value,
                  KO_intercept = intercept)
  res <- cbind(res1, res2)
  res <- cbind(res, res3)
  return(res)
  
  
}

peak_lm_res <- lapply(peaks_of_interest, peak_lm) %>%
  bind_rows()
save(peak_lm_res, file = "results/peak_lm_res.RData")


hmm <- overlapping_df %>%
  left_join(peak_lm_res)
max(peak_lm_res$p.value)
max(hmm$p.value)
hmm$padj <- p.adjust(hmm$p.value)

huh <- hmm %>% filter(p.value < 0.05)
```

# Gene expresion changes vs atac-seq

```{r}
# Let's do rapgef4
huh1 <- hmm %>% filter(gene_name %in% short_3)
goi_atac_peaks <-huh1 %>% filter(gene_name == "Duox1") %>%
  mutate(sig = p.value < 0.05)
unique(goi_atac_peaks$interval_id)
# goi_atac_slopes <- goi_atac_peaks %>%
#   dplyr::select(interval_id, dist_to_tss, slope, p.value, )
huh1_batch <- short_3[1:10]
ggplot(huh1, aes(x = dist_to_tss, y = KO_slope, color = sig_peak)) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point() +
  facet_wrap(~gene_name, scales = "free_x")
ggsave("figures/atacpeak_locations_ko.pdf", height = 7, width = 7)
ggplot(goi_atac_slopes, aes(x = slope, y = -log10(p.value))) +
  geom_point()

poi <- goi_atac_peaks %>% arrange(p.value) %>%
  dplyr::slice(1) %>%
  pull(interval_id)

poi_vals <- rlog_df %>% filter(interval_id == poi, 
                               firre_induced == "firre_induced")

ggplot(poi_vals, aes(x = timepoint, y = count)) +
  geom_point()

poi_mean <- poi_vals %>%
  group_by(timepoint) %>%
  summarize(count = mean(count))
ggplot(poi_mean, aes(x = timepoint, y = count)) +
  geom_point()
ggplot(short_vszero_sig %>% filter(gene_name == "Shf"), 
       aes(x = timepoint, y = log2FoldChange)) +
  geom_point()
range01 <- function(x){(x-min(x))/(max(x)-min(x))}
goi_vals <- short_vszero_sig %>% filter(gene_name == "Shf")
goi_vals$relexp <- range01(goi_vals$log2FoldChange)

ggplot(goi_vals, 
       aes(x = timepoint, y = relexp)) +
  geom_point()
poi_mean$relexp <- range01(poi_mean$count)
ggplot(poi_mean, 
       aes(x = timepoint, y = relexp)) +
  geom_point()


# Combine
combined_vals <- bind_rows(poi_mean %>% dplyr::select(timepoint, relexp) %>%
                             mutate(type = "atac_signal"),
                           goi_vals %>% dplyr::select(timepoint, relexp) %>%
                             mutate(type = "gene_expression"))

ggplot(combined_vals, aes(x = timepoint, y = relexp, color = type, group = type)) +
  geom_point() +
  geom_smooth(se = FALSE, span = 1.1)

ggplot(combined_vals, aes(x = timepoint, y = type, fill = relexp)) +
  geom_tile() +
  scale_fill_gradientn(colors = col_pal10[50:100])

```



```{r}
table(huh$gene_name)
length(unique(huh$gene_name))

table(short_3 %in% unique(huh$gene_name))


huh2 <- hmm %>% filter(WT_pval < 0.05)
table(wt_overexp_2 %in% huh2$gene_name)
table(all_overlaps_123 %in% huh2$gene_name)

huh3 <- hmm %>% filter(KO_pval < 0.05)
table(ko_rescue_1 %in% huh3$gene_name)
table(all_overlaps_123 %in% huh3$gene_name)


table(all_overlaps_123 %in% unique(huh$gene_name))
sig_ov_peaks <- hmm %>% filter(padj < 0.05)

short_vszero_sig <- short_vszero_sig %>%
  mutate(has_atac_peak = gene_id %in% overlapping_df$gene_id)

ggplot(short_vszero_sig, aes(x = timepoint, y= log2FoldChange, color = has_atac_peak)) +
  geom_point() +
  facet_wrap(~has_atac_peak)
ggplot(short_vszero_sig, aes(x = log2FoldChange, y= -log10(padj), color = has_atac_peak)) +
  geom_point() +
  facet_wrap(~has_atac_peak)
load("../01_setup/results/rnaseq_data.RData")
```

```{r}
load("../06_short_timecourse_rescue/results/ko_rescue_short_vszero_sig.RData")
ko_rescue_short_sig <- unique(ko_rescue_short_vszero_sig$gene_id)
load("../09_short_timecourse_overexpression/results/wt_overexp_short_vszero_sig.RData")
wt_overexp_short_sig <- unique(wt_overexp_short_vszero_sig$gene_id)

sig_genes <- intersect(wt_overexp_short_sig, ko_rescue_short_sig)

sig_genes_gr <- gencode_genes[gencode_genes$gene_id %in% sig_genes]
# Extend 10kb
start(sig_genes_gr) <- start(sig_genes_gr) - 50000
end(sig_genes_gr) <- end(sig_genes_gr) + 50000

peak_gene_overlaps <- findOverlaps(sig_genes_gr, sig_atac_gr)
sig_atac_df <- sig_atac_gr %>% as.data.frame()

sig_genes_df <- sig_genes_gr %>% as.data.frame() %>% dplyr::select(gene_name, gene_id, seqnames, start, end)

overlapping_df <- sig_atac_df[peak_gene_overlaps@to, ]
overlapping_df$gene_id <- sig_genes_df$gene_id[peak_gene_overlaps@from]
overlapping_df$gene_name <- sig_genes_df$gene_name[peak_gene_overlaps@from]
overlapping_df$gene_tss <- sig_genes_df$start[peak_gene_overlaps@from] + 50000
overlapping_df$gene_chr <- sig_genes_df$seqnames[peak_gene_overlaps@from]
```






