---
title: "ATAC-seq"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

```{r}
atac_consensus <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.annotatePeaks.txt", 
                             sep = "\t", header = T)
colnames(atac_consensus)[[1]] <- "interval_id"

atac_c_fc <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.featureCounts.txt", 
                        skip = 1, sep = "\t", header = T)
colnames(atac_c_fc) <- gsub(".mLb.clN.bam","",colnames(atac_c_fc))
rownames(atac_c_fc) <- atac_c_fc$Geneid
interval_table <- atac_c_fc[, 1:6]

counts <- atac_c_fc[, 7:ncol(atac_c_fc), drop = FALSE] %>% 
  as.matrix()
```

```{r}
samples <- data.frame("sample_name" = colnames(counts)) %>%
  mutate(sample_name = gsub("firre_induced", "firreinduced", sample_name)) %>%
  separate(sample_name, 
           into = c("cell_type", "firre_ko", "firre_induced", 
                    "timepoint_minutes", "replicate"), remove = FALSE) %>%
  mutate(sample_name = gsub("firreinduced", "firre_induced", sample_name),
         firre_induced = gsub("firreinduced", "firre_induced", firre_induced),
         firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         firre_induced = factor(firre_induced, levels = c("control", "firre_induced")),
         timepoint_minutes = factor(timepoint_minutes, levels = seq(0, 150, 30))) %>%
  dplyr::select(-replicate)
rownames(samples) <- samples$sample_name
stopifnot(all(colnames(counts) == rownames(samples)))
samples$condition <- "control"
samples$timepoint <- as.numeric(as.character(samples$timepoint_minutes))
samples$condition[which(samples$timepoint > 0 & 
                          samples$firre_induced == "firre_induced")] <- "firre"
samples$condition <- factor(samples$condition, levels = c("control", "firre"))

counts <- counts[,rownames(samples)]
stopifnot(all(colnames(counts) == rownames(samples)))
mode(counts) <- "integer"
```

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = samples, 
                              design = ~ condition)

dds <- DESeq(dds)
res_df <- results(dds) %>%
  as.data.frame() %>%
  rownames_to_column("interval_id") %>%
  merge(atac_consensus %>% 
          dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, 
                        Distance.to.TSS, Chr, Start, End)) %>% 
  unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)
names(res_df)[8] <- "gene_name"
names(res_df)[9] <- "gene_id"

sig_atac_peaks <- res_df %>%
  filter(padj <= 0.05)

 # abs(log2FoldChange) > log2(1.5)
```

```{r}
load("../11_short_timecourse_combined/results/short_vszero_sig.RData")
# short_vszero_sig <- short_vszero_sig %>% filter(max_fc > 1)
sig_genes <- unique(short_vszero_sig$gene_id)
gencode_genes <- rtracklayer::import("../util/gencode.vM25.annotation.genes.gtf")
sig_genes_gr <- gencode_genes[gencode_genes$gene_id %in% sig_genes]
# Extend 10kb
start(sig_genes_gr) <- start(sig_genes_gr) - 50000
end(sig_genes_gr) <- end(sig_genes_gr) + 50000

atac_consensus_gr <- GRanges(seqnames = atac_consensus$Chr,
                             IRanges(start = atac_consensus$Start,
                                     end = atac_consensus$End),
                             interval_id = atac_consensus$interval_id)

sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% sig_atac_peaks$interval_id]

peak_gene_overlaps <- findOverlaps(sig_genes_gr, sig_atac_gr)


sig_atac_df <- sig_atac_gr %>% as.data.frame()

sig_genes_df <- sig_genes_gr %>% as.data.frame() %>% dplyr::select(gene_name, gene_id, seqnames, start, end)
# What is the difference between the genes changing without an ATAC-seq peak change
# and those that do? 
# Are we over controlling for dox? Are there ATAC-seq windows in these regions that have 
# linearly increasing trends, but are not significant when compared to the dox control? 
overlapping_df <- sig_atac_df[peak_gene_overlaps@to, ]
overlapping_df$gene_id <- sig_genes_df$gene_id[peak_gene_overlaps@from]
overlapping_df$gene_name <- sig_genes_df$gene_name[peak_gene_overlaps@from]
overlapping_df$gene_tss <- sig_genes_df$start[peak_gene_overlaps@from] + 50000
overlapping_df$gene_chr <- sig_genes_df$seqnames[peak_gene_overlaps@from]

overlapping_df <- overlapping_df %>%
  mutate(peak_center = start + (end - start),
         dist_to_tss = peak_center - gene_tss)
short_vszero_sig <- short_vszero_sig %>%
  mutate(has_atac_peak = gene_id %in% overlapping_df$gene_id)

ggplot(short_vszero_sig, aes(x = timepoint, y= log2FoldChange, color = has_atac_peak)) +
  geom_point() +
  facet_wrap(~has_atac_peak)
ggplot(short_vszero_sig, aes(x = log2FoldChange, y= -log10(padj), color = has_atac_peak)) +
  geom_point() +
  facet_wrap(~has_atac_peak)
load("../01_setup/results/rnaseq_data.RData")
```

```{r}
load("../06_short_timecourse_rescue/results/ko_rescue_short_vszero_sig.RData")
ko_rescue_short_sig <- unique(ko_rescue_short_vszero_sig$gene_id)
load("../09_short_timecourse_overexpression/results/wt_overexp_short_vszero_sig.RData")
wt_overexp_short_sig <- unique(wt_overexp_short_vszero_sig$gene_id)

sig_genes <- intersect(wt_overexp_short_sig, ko_rescue_short_sig)

sig_genes_gr <- gencode_genes[gencode_genes$gene_id %in% sig_genes]
# Extend 10kb
start(sig_genes_gr) <- start(sig_genes_gr) - 50000
end(sig_genes_gr) <- end(sig_genes_gr) + 50000

peak_gene_overlaps <- findOverlaps(sig_genes_gr, sig_atac_gr)
sig_atac_df <- sig_atac_gr %>% as.data.frame()

sig_genes_df <- sig_genes_gr %>% as.data.frame() %>% dplyr::select(gene_name, gene_id, seqnames, start, end)

overlapping_df <- sig_atac_df[peak_gene_overlaps@to, ]
overlapping_df$gene_id <- sig_genes_df$gene_id[peak_gene_overlaps@from]
overlapping_df$gene_name <- sig_genes_df$gene_name[peak_gene_overlaps@from]
overlapping_df$gene_tss <- sig_genes_df$start[peak_gene_overlaps@from] + 50000
overlapping_df$gene_chr <- sig_genes_df$seqnames[peak_gene_overlaps@from]
```






