---
title: "Overview plots"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(umap)
library(gganimate)
library(scales)
source("../util/_plot_theme.R")
```

# Let's make a umap for the Firre response.

```{r}

genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id

g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

res_shrunkendf <- read_csv(file.path("../01_firre_induction_vs_control",
                                     "results/deseq_res_shrunken.csv"))
dynamic_res <- res_shrunkendf %>%
  filter(grepl(".firre_inducedfirre_induced", result_name))
unique(dynamic_res$result_name)
dynamic_res$timepoint <- gsub("timepoint_minutes", "", dynamic_res$result_name)
dynamic_res$timepoint <- gsub(".firre_inducedfirre_induced", "", dynamic_res$timepoint)
dynamic_res$timepoint <- as.numeric(dynamic_res$timepoint)

dynamic_res <- dynamic_res %>%
  unite(experiment, cell_type, firre_ko, remove = F)
experiments <- unique(dynamic_res$experiment)

exp1 <- dynamic_res %>% filter(experiment == experiments[1]) %>%
   filter(!is.na(padj))

# Now let's add in the Dox response for the same experiment.
dox_res <- read_csv(file.path("../02_doxycycline_responders/results",
                              "deseq_dox_control_res_shrunken.csv")) %>%
  unite(experiment, cell_type, firre_ko, remove = F) %>%
  arrange(experiment)
dox_res$timepoint <- gsub("timepoint_minutes_", "", dox_res$result_name)
dox_res$timepoint <- gsub("_vs_0", "", dox_res$timepoint)
dox_res$timepoint <- as.numeric(dox_res$timepoint)
unique(dox_res$timepoint)

dox_exp1 <- dox_res %>% filter(experiment == experiments[1]) %>%
  filter(!is.na(padj)) %>%
  dplyr::select(gene_id, log2FoldChange, timepoint) %>%
  pivot_wider(names_from = timepoint, values_from = log2FoldChange)
names(dox_exp1)[2:12] <- paste0("dox_", names(dox_exp1)[2:12])

fr_exp1 <- exp1 %>% 
  filter(!is.na(padj)) %>%
  dplyr::select(gene_id, log2FoldChange, timepoint) %>%
  pivot_wider(names_from = timepoint, values_from = log2FoldChange) 
names(fr_exp1)[2:12] <- paste0("fr_", names(fr_exp1)[2:12])

dox_sig <- dox_res %>% filter(experiment == experiments[1]) %>% 
  filter(padj < 0.1)
fr_sig <- dynamic_res %>% filter(experiment == experiments[1]) %>%
  filter(padj < 0.1)

combined_sig <- unique(c(dox_sig$gene_id, fr_sig$gene_id))

fr_exp1 <- fr_exp1[fr_exp1$gene_id %in% combined_sig,]
dox_exp1 <- dox_exp1[dox_exp1$gene_id %in% combined_sig,]
exp1 <- merge(fr_exp1, dox_exp1) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()

pca <- prcomp(exp1)


umap_config <- umap.defaults
umap_config$n_neighbors <- 100
umap_config$min_dist <- 0.1
umap_config$metric <- "pearson"
exp1_umap <- umap(exp1, config = umap_config)
exp1_umap_layout <- exp1_umap$layout %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s)
names(exp1_umap_layout)[2:3] <- c("umap_x", "umap_y")

exp1_umap_layout$pc1 <- pca$x[,1]
exp1_umap_layout$pc2 <- pca$x[,2]
g <- ggplot(exp1_umap_layout, aes(x = umap_x, y = umap_y, label = gene_name))
g + geom_point()

g <- ggplot(exp1_umap_layout, aes(x = pc1, y = pc2, label = gene_name))
g + geom_point()

unique(exp1df$timepoint)
exp1df <- merge(exp1, exp1_umap_layout)
exp1df <- exp1df %>% arrange(abs(log2FoldChange))
p <- ggplot(exp1df, 
            aes(x = umap_x, y = umap_y, color = log2FoldChange, group = gene_id)) + 
  geom_point() + 
  scale_color_gradient2(mid = "gray80", limits = c(-0.5, 0.5), oob = squish) +
  labs(title = 'ESC KO: {closest_state} min') +
  transition_states(timepoint, transition_length = 10, state_length = 30) +
  ease_aes('cubic-in-out')
animate(p, renderer = gifski_renderer(),
        nframes = 100)
anim_save("figures/esc_ko_umap_by_lfc.gif")
```

by tpm 
```{r}
tpm <- read.csv("../../rnaseq/results/salmon/salmon_merged_gene_tpm.csv") %>%
  merge(g2s) %>%
  dplyr::select(gene_id, gene_name, everything()) 
samples <- read.csv("../../rnaseq/samplesheet.csv") %>%
  filter(cell_type != "pMEF",
         date_sequenced != "18-Sep",
         timepoint_minutes < 400,
         timecourse_length == "short") %>%
  mutate(firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         timepoint_minutes = factor(timepoint_minutes, 
                                    levels = c(seq(0,360,30))),
         firre_induced = factor(firre_induced, 
                                levels = c("control", "firre_induced")))
samples <- samples %>%
  unite(experiment, cell_type, firre_ko, remove = F)
exp1_samples <- samples %>% filter(experiment == experiments[1]) %>%
  arrange(firre_induced, timepoint_minutes)
exp1_tpm_matrix <- tpm %>% dplyr::select(-gene_name) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()
exp1_tpm_matrix <- exp1_tpm_matrix[,as.character(exp1_samples$sample_id)]

exp1_tpm_subset <- exp1_tpm_matrix[sample(5000),]

# exp1_subset <- exp1_tpm_scaled[!is.na(rowSums(exp1_tpm_scaled)),]
# exp1_subset <- exp1_tpm_scaled[sample(5000),]
# summary(rowSums(exp1_tpm_scaled, na.rm = T))
# summary(rowSums(exp1_tpm_matrix))
umap_config <- umap.defaults
umap_config$n_neighbors <- 500
umap_config$min_dist <- 0.7
exp1_tpm_umap <- umap(exp1_tpm_subset, scale = "Z",
                      config = umap_config)

# saveRDS(exp1_tpm_umap, "results/exp_tpm_umap.rds")

exp1_tpm_umap_layout <- exp1_tpm_umap$layout %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s)
names(exp1_tpm_umap_layout)[2:3] <- c("umap_x", "umap_y")

g <- ggplot(exp1_tpm_umap_layout, aes(x = umap_x, y = umap_y))
g + geom_point()


exp1_subset <- exp1 %>% filter(gene_id %in% rownames(exp1_tpm_subset))
exp1df <- merge(exp1_subset, exp1_tpm_umap_layout)

exp1df <- exp1df %>% arrange(abs(log2FoldChange))
p <- ggplot(exp1df %>% filter(timepoint == 30), 
            aes(x = umap_x, y = umap_y, color = log2FoldChange, group = gene_id)) + 
  geom_point() + 
  scale_color_gradient2(mid = "gray80", limits = c(-0.5, 0.5), oob = squish) 

p <- ggplot(exp1df, 
            aes(x = umap_x, y = umap_y, color = log2FoldChange, group = gene_id)) + 
  geom_point() + 
  scale_color_gradient2(mid = "gray80", limits = c(-0.5, 0.5), oob = squish) +
  labs(title = 'ESC KO: {closest_state} min') +
  transition_states(timepoint, transition_length = 10, state_length = 30) +
  ease_aes('cubic-in-out')
animate(p, renderer = gifski_renderer(),
        nframes = 100)
anim_save("figures/esc_ko_umap_by_tpm.gif")
# pheatmap(exp1_tpm_matrix)
```


