---
title: "Setup"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
```

```{r}
# Check raw data file locations
chipseq_samples <- read_csv("../../chipseq/chipseq_samplesheet.csv")
rnaseq_samples <- read_csv("../../rnaseq/rnaseq_samplesheet.csv")
atacseq_samples <- read_csv("../../atacseq/atacseq_samplesheet.csv")
proseq_samples <- read_csv("../../proseq/proseq_samplesheet.csv")

table(file.exists(chipseq_samples$raw_data_read1, chipseq_samples$raw_data_read2))
table(file.exists(rnaseq_samples$raw_data_read1, rnaseq_samples$raw_data_read2))
table(file.exists(atacseq_samples$raw_data_read1, atacseq_samples$raw_data_read2))
table(file.exists(proseq_samples$raw_data_read1, proseq_samples$raw_data_read2))
```


# RNA-seq count data setup

```{r}
### ANNOTATIONS
if(!file.exists("../util/gencode.vM25.annotation.genes.gtf")) {
  gencode <- rtracklayer::import(file.path("../../../../genomes/Mus_musculus/",
                                           "Gencode/M25/",
                                           "gencode.vM25.annotation.gtf"))
  genes <- gencode[gencode$type == "gene"]
  rtracklayer::export(genes, 
                      "../util/gencode.vM25.annotation.genes.gtf")
}
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id
# Remove pseudogenes
genes <- genes[!grepl("pseudogene", genes$gene_type), ]

hmm <- genes %>% as.data.frame() %>% dplyr::select(gene_id, gene_type, gene_name)


g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

tx2gene <- read.csv("../../rnaseq/results/salmon/tx2gene.csv")


### COUNTS
tpm <- read.csv("../../rnaseq/results/salmon/salmon_merged_gene_tpm.csv") %>%
  merge(g2s) %>%
  dplyr::select(gene_id, gene_name, everything()) %>%
  filter(gene_id %in% genes$gene_id)

# There are some genes in this list which were not quantified by salmon. 
genes <- genes[genes$gene_id %in% tpm$gene_id,]

salmon_gene_counts <- read.csv(file.path("../../rnaseq/results/salmon/",
                                         "salmon_merged_gene_counts.csv")) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()
mode(salmon_gene_counts) <- "integer"

# Filter to relevant genes
salmon_gene_counts <- salmon_gene_counts[rownames(salmon_gene_counts) %in% genes$gene_id,]
# Put in order
genes <- genes[rownames(salmon_gene_counts)]

### SAMPLE INFO
samples <- read.csv("../../rnaseq/rnaseq_samplesheet.csv") %>%
  filter(cell_type != "pMEF",
         date_sequenced != "2018-09") %>%
  mutate(firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         timepoint = factor(timepoint, 
                            levels = c(seq(0,360,30), 720, 1440, 2880, 5760)),
         firre_induced = factor(firre_induced, 
                                levels = c("control", "firre_induced")),
         timecourse_length = factor(timecourse_length, levels = c("long", "short")),
         cell_type = factor(cell_type, levels = c("ESC", "NPC")),
         experiment_replicate = factor(experiment_replicate, levels = c(1,2))) %>%
  unite(experiment, cell_type, firre_ko, remove = FALSE)
rownames(samples) <- samples$sample_id

counts <- salmon_gene_counts[,as.character(samples$sample_id)]
stopifnot(all(rownames(samples) == colnames(counts)))
stopifnot(all(rownames(salmon_gene_counts) == genes$gene_id))

nonzero_genes <- counts[rowSums(counts) > 0,]
num_measured_genes <- nrow(nonzero_genes)

# Save as a bundle of objects.
save(genes, g2s, tx2gene, tpm, salmon_gene_counts, samples, counts, num_measured_genes,
     file = "results/rnaseq_data.RData")
```

## ATAC-seq data

```{r}
base_path <- "../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/"
atac_consensus <- read.table(paste0(base_path,
                                    "consensus_peaks.mLb.clN.annotatePeaks.txt"),
                             sep = "\t", header = T)
colnames(atac_consensus)[[1]] <- "interval_id"

atac_consensus_gr <- GRanges(seqnames = atac_consensus$Chr,
                             IRanges(start = atac_consensus$Start,
                                     end = atac_consensus$End),
                             interval_id = atac_consensus$interval_id)

atac_c_fc <- read.table(paste0(base_path,
                               "consensus_peaks.mLb.clN.featureCounts.txt"), 
                        skip = 1, sep = "\t", header = T)
colnames(atac_c_fc) <- gsub(".mLb.clN.bam","",colnames(atac_c_fc))
rownames(atac_c_fc) <- atac_c_fc$Geneid
interval_table <- atac_c_fc[, 1:6]

atac_counts <- atac_c_fc[, 7:ncol(atac_c_fc), drop = FALSE] %>% 
  as.matrix()

atac_samples <- data.frame("sample_name" = colnames(atac_counts)) %>%
  mutate(sample_name = gsub("firre_induced", "firreinduced", sample_name)) %>%
  separate(sample_name, 
           into = c("cell_type", "firre_ko", "firre_induced", 
                    "timepoint_minutes", "replicate"), remove = FALSE) %>%
  mutate(sample_name = gsub("firreinduced", "firre_induced", sample_name),
         firre_induced = gsub("firreinduced", "firre_induced", firre_induced),
         firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         firre_induced = factor(firre_induced, levels = c("control", "firre_induced")),
         timepoint_minutes = factor(timepoint_minutes, levels = seq(0, 150, 30))) %>%
  dplyr::select(-replicate)
rownames(atac_samples) <- atac_samples$sample_name
stopifnot(all(colnames(atac_counts) == rownames(atac_samples)))
atac_samples$condition <- "control"
atac_samples$timepoint <- as.numeric(as.character(atac_samples$timepoint_minutes))
atac_samples$condition[which(atac_samples$timepoint > 0 & 
                               atac_samples$firre_induced == "firre_induced")] <- "firre"
atac_samples$condition <- factor(atac_samples$condition, levels = c("control", "firre"))

atac_counts <- atac_counts[,rownames(atac_samples)]
stopifnot(all(colnames(atac_counts) == rownames(atac_samples)))
mode(atac_counts) <- "integer"

save(genes, g2s, tpm, atac_samples, atac_counts, atac_consensus, atac_consensus_gr,
     file = "results/atacseq_data.RData")
```

# GEO Submission


```{r}
# save sample info for GEO submission
geo_rnaseq_samples <- samples %>%
  filter(cell_type == "ESC") %>%
  rowwise() %>%
  mutate(replicate_numeric = replicate, 
         replicate = paste0("rep", replicate),
         timepoint_numeric = as.numeric(as.character(timepoint)),
         timepoint = paste0(timepoint, "min"),
         treatment = "dox") %>%
  unite(geo_sample_name, experiment, firre_induced, timepoint, treatment, 
        timecourse_length, replicate, remove = FALSE) %>%
  unite(og_geo_sample_name, experiment, firre_induced, timepoint, treatment, replicate, remove = FALSE) %>%
  mutate(geo_sample_name = gsub("ESC_KO", "mESC_Firre_Rescue", geo_sample_name),
         geo_sample_name = gsub("ESC_WT", "mESC_Firre_OE", geo_sample_name)) %>%
  arrange(firre_ko, timecourse_length, firre_induced, timepoint_numeric, replicate_numeric) %>%
  mutate(fastq1 = paste0(sample_id, "_read1.fastq.gz"),
         fastq2 = paste0(sample_id, "_read2.fastq.gz"),
         geo_fastq1 = paste0(geo_sample_name, "_read1.fastq.gz"),
         geo_fastq2 = paste0(geo_sample_name, "_read2.fastq.gz")) %>%
  mutate(type = "RNAseq",
         raw_read1_exists = file.exists(raw_data_read1),
         raw_read2_exists = file.exists(raw_data_read2))

# Gather and format the ATAC-seq samples
geo_atacseq_samples <- read_csv("/scratch/Shares/rinn/Michael/firre_timecourse/atacseq/atacseq_samplesheet.csv") %>%
  mutate(timepoint_min = paste0(timepoint_minutes, "min"),
         type = "ATACseq") %>%
  unite(geo_sample_name, cell_type, firre_ko, firre_induced, timepoint_min, type, remove = FALSE) %>%
  mutate(geo_sample_name = gsub("ESC_KO", "mESC_Firre_Rescue", geo_sample_name),
         geo_sample_name = gsub("ESC_WT", "mESC_Firre_OE", geo_sample_name)) %>%
  arrange(firre_ko, firre_induced, timepoint_minutes) %>%
  rowwise() %>%
  mutate(fastq1 = paste0(sample_id, "_read1.fastq.gz"),
         fastq2 = paste0(sample_id, "_read2.fastq.gz"),
         geo_fastq1 = paste0(geo_sample_name, "_read1.fastq.gz"),
         geo_fastq2 = paste0(geo_sample_name, "_read2.fastq.gz")) %>%
  mutate(raw_read1_exists = file.exists(raw_data_read1),
         raw_read2_exists = file.exists(raw_data_read2))



write_csv(geo_rnaseq_samples, "geo_rnaseq_samples.csv")
write_csv(geo_atacseq_samples, "geo_atacseq_samples.csv")


# Make md5 check sums for each file (using slurm to parallelize)
md5_file_list <- list.files("../../geo_md5sums", full.names = TRUE, pattern = "*.md5.txt")
md5_list <- sapply(md5_file_list, readLines)
md5_df <- data.frame(md5sum = md5_list) %>%
  separate(md5sum, into = c("md5sum", "filename"), sep = "  ../geo_upload/rnaseq/")
# Let's merge this into the geo_sample data
md5_r1 <- md5_df %>%
  filter(grepl("read1", filename)) %>%
  dplyr::rename(geo_fastq1 = filename,
                md5_geo_read1 = md5sum)
md5_r2 <- md5_df %>%
  filter(grepl("read2", filename)) %>%
  dplyr::rename(geo_fastq2 = filename,
                 md5_geo_read2 = md5sum)
geo_rnaseq_samples <- geo_rnaseq_samples %>%
  left_join(md5_r1) %>%
  left_join(md5_r2)
geo_rnaseq_samples <- geo_rnaseq_samples %>%
  mutate(md5_r1_equal = md5_read1 == md5_geo_read1,
         md5_r2_equal = md5_read2 == md5_geo_read2)

table(geo_rnaseq_samples$md5_r1_equal)
table(geo_rnaseq_samples$md5_r2_equal)

write_csv(md5_df, "geo_md5sum.csv")

# Let's also check the md5sums seperately for the ATAC-seq
md5_file_list <- list.files("../../geo_atacseq_md5sums", full.names = TRUE, pattern = "*.md5.txt")
md5_list <- sapply(md5_file_list, readLines)
md5_df <- data.frame(md5sum = md5_list) %>%
  separate(md5sum, into = c("md5sum", "filename"), sep = "  ../geo_upload/geo_submission_atacseq_may4/")
# Let's merge this into the geo_sample data
md5_r1 <- md5_df %>%
  filter(grepl("read1", filename)) %>%
  dplyr::rename(geo_fastq1 = filename,
                md5_geo_read1 = md5sum)
md5_r2 <- md5_df %>%
  filter(grepl("read2", filename)) %>%
  dplyr::rename(geo_fastq2 = filename,
                 md5_geo_read2 = md5sum)
geo_atacseq_samples <- geo_atacseq_samples %>%
  left_join(md5_r1) %>%
  left_join(md5_r2) %>%
  mutate(md5_r1_equal = md5_read1 == md5_geo_read1,
         md5_r2_equal = md5_read2 == md5_geo_read2)

table(geo_atacseq_samples$md5_r1_equal)
table(geo_atacseq_samples$md5_r2_equal)

write_csv(md5_df, "geo_atacseq_md5sum.csv")


# Let's get some processed data files ready for the GEO submission. 
geo_rnaseq_counts <- read.csv("../../rnaseq/results/salmon/salmon_merged_gene_counts.csv") %>%
  merge(g2s) %>%
  dplyr::select(gene_id, gene_name, everything()) %>%
  filter(gene_id %in% genes$gene_id) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "count") %>%
  left_join(geo_rnaseq_samples %>% dplyr::select(sample_id, geo_sample_name)) %>%
  filter(!is.na(geo_sample_name)) %>%
  dplyr::rename(sample_name = geo_sample_name) %>%
  dplyr::select(gene_id, gene_name, sample_name, count)
write_csv(geo_rnaseq_counts, "mESC_Firre_induction_timecourse_merged_gene_counts.csv")

geo_rnaseq_tpm <- read.csv("../../rnaseq/results/salmon/salmon_merged_gene_tpm.csv") %>%
  merge(g2s) %>%
  dplyr::select(gene_id, gene_name, everything()) %>%
  filter(gene_id %in% genes$gene_id) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "TPM") %>%
  left_join(geo_rnaseq_samples %>% dplyr::select(sample_id, geo_sample_name)) %>%
  filter(!is.na(geo_sample_name)) %>%
  dplyr::rename(sample_name = geo_sample_name) %>%
  dplyr::select(gene_id, gene_name, sample_name, TPM)

write_csv(geo_rnaseq_tpm, "mESC_Firre_induction_timecourse_merged_gene_tpm.csv")

# Now let's get the processed data files in order for the ATAC-seq
base_path <- "../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/"
atac_consensus <- read.table(paste0(base_path,
                                    "consensus_peaks.mLb.clN.annotatePeaks.txt"),
                             sep = "\t", header = T)
colnames(atac_consensus)[[1]] <- "interval_id"

atac_consensus <- atac_consensus %>%
  separate(interval_id, into = c("interval", "interval_number"), sep = "_", remove = FALSE) %>%
  mutate(interval_number = as.numeric(interval_number)) %>%
  arrange(interval_number) %>%
  dplyr::select(-interval, -interval_number, -Peak.Score, -Strand)



write_csv(atac_consensus, "atacseq_consensus_peaks.csv")

atac_c_fc <- read.table(paste0(base_path,
                               "consensus_peaks.mLb.clN.featureCounts.txt"), 
                        skip = 1, sep = "\t", header = T)
colnames(atac_c_fc) <- gsub(".mLb.clN.bam","",colnames(atac_c_fc))
rownames(atac_c_fc) <- atac_c_fc$Geneid
interval_table <- atac_c_fc[, 1:6]



geo_atac_counts <- atac_c_fc[, 7:ncol(atac_c_fc), drop = FALSE] %>%
  rownames_to_column("interval_id") %>%
  pivot_longer(2:ncol(.), names_to = "orig_sample_name", values_to = "count")

geo_atacseq_samples <- geo_atacseq_samples %>%
  mutate(append = "R1") %>%
  unite(orig_sample_name, cell_type, firre_ko, firre_induced, timepoint_minutes, append, sep = "_", remove = FALSE)

geo_atac_counts <- geo_atac_counts %>%
  left_join(geo_atacseq_samples %>% dplyr::select(orig_sample_name, geo_sample_name)) %>%
  dplyr::select(interval_id, geo_sample_name, count) %>%
  dplyr::rename(sample_name = geo_sample_name)

write_csv(geo_atac_counts, "atacseq_consensus_peak_counts.csv")

```



