---
title: "Vs zero"
author: "Michael Smallegan"
date: "10/1/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(gplots)
library(ggrepel)
library(ggpubr)
source("../util/_plot_theme.R")
source("../util/_util.R")
library(BiocParallel)
register(MulticoreParam(12))
```

```{r import}
### ANNOTATIONS
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id

g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

tx2gene <- read.csv("../../rnaseq/results/salmon/tx2gene.csv")

### COUNTS
tpm <- read.csv("../../rnaseq/results/salmon/salmon_merged_gene_tpm.csv") %>%
  merge(g2s) %>%
  dplyr::select(gene_id, gene_name, everything())

salmon_gene_counts <- read.csv(file.path("../../rnaseq/results/salmon/",
                                         "salmon_merged_gene_counts.csv")) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()
mode(salmon_gene_counts) <- "integer"

# Put in order
genes <- genes[rownames(salmon_gene_counts)]

### SAMPLE INFO
# Somehow the long timepoints didn't make it into this run.
# TODO: re-run with the long timepoints. Gaaah.
samples <- read.csv("../../rnaseq/rnaseq_samplesheet.csv") %>%
  filter(cell_type != "pMEF",
         date_sequenced != "2018-09",
         timecourse_length == "short") %>%
  mutate(firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         timepoint_minutes = factor(timepoint_minutes, 
                                    levels = c(seq(0,360,30), 720, 1440, 2880, 5760)),
         firre_induced = factor(firre_induced, 
                                levels = c("control", "firre_induced")))
rownames(samples) <- samples$sample_id

tpml <- tpm %>% pivot_longer(cols = 3:ncol(.), names_to = "sample_id",
                             values_to = "tpm") 
```

```{r}
timepoints <- seq(30,330,30)
vs_zero_timepoints <- rep(list(c("timepoint_minutes")),length(timepoints))
for(i in 1:length(vs_zero_timepoints)) {
  vs_zero_timepoints[[i]] <- c(vs_zero_timepoints[[i]],timepoints[[i]], "0")
}
timepoints <- seq(30,330,30)
vs_prev_timepoints <- rep(list(c("timepoint_minutes")), length(timepoints)-1)

for(i in 1:length(vs_prev_timepoints)) {
  vs_prev_timepoints[[i]] <- c(vs_prev_timepoints[[i]],timepoints[[i+1]])
  vs_prev_timepoints[[i]] <- c(vs_prev_timepoints[[i]],timepoints[[i]])
}
contrasts_list <- c(vs_zero_timepoints, vs_prev_timepoints)

names(contrasts_list) <- sapply(contrasts_list, function(x) {
  paste(x[[1]], x[[2]], "vs", x[[3]], sep = "_")
})


```


```{r}



run_vs_zero_deseq <- function(experiment,
                              counts, samples, genes,
                              contrasts_list,
                              ncores = 12,
                              save_dds = FALSE,
                              save_res = FALSE) {
  
  ct <- unlist(strsplit(experiment, "_"))[[1]]
  fko <- unlist(strsplit(experiment, "_"))[[2]]
  fi <- unlist(strsplit(experiment, "_"))[[3]]
  if(fi == "fi") {
    fi <- "firre_induced"
  } else {
    fi <- "control"
  }
  
  samples <- samples %>% filter(cell_type == ct,
                                firre_ko == fko,
                                firre_induced == fi)
  
  counts <- counts[,samples$sample_id]
  mode(counts) <- "integer"
  
  # Reorder gencode genes
  genes <- genes[rownames(counts)]
  
  # Check ordering
  stopifnot(all(rownames(samples) == colnames(counts)))
  stopifnot(all(names(genes) == rownames(counts)))
  
  dds <- DESeqDataSetFromMatrix(countData = counts, 
                                colData = samples, 
                                design = ~ timepoint_minutes,
                                rowData = genes)
  
  # Filter low counts
  dds <- dds[rowSums(counts(dds)) >= 10,]
  
  
  dds <- DESeq(dds,
               parallel=TRUE,
               BPPARAM=MulticoreParam(ncores))
  
  if(save_dds == TRUE) {
    saveRDS(dds, paste0("results/",
                        ct, "_",
                        fko, "_dds.rds"))
  }
  
  
  res_list_unshrunken <- mapply(
    FUN = results,
    contrast = contrasts_list,
    MoreArgs = list(
      object = dds),
    SIMPLIFY = FALSE,
    USE.NAMES = FALSE)
  
  
  
  res_list_shrunken <- mapply(
    FUN = lfcShrink,
    res = res_list_unshrunken,
    contrast = contrasts_list,
    MoreArgs = list(
      dds = dds,
      type = "ashr",
      parallel = TRUE),
    SIMPLIFY = FALSE,
    USE.NAMES = TRUE)
  
  # Convert to data.frame
  for(i in 1:length(res_list_unshrunken)) {
    res_list_unshrunken[[i]] <- res_list_unshrunken[[i]] %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>%
      merge(g2s) %>%
      mutate(result_name = names(contrasts_list)[[i]])
  }
  res_unshrunken <- bind_rows(res_list_unshrunken)
  
  for(i in 1:length(res_list_shrunken)) {
    res_list_shrunken[[i]] <- res_list_shrunken[[i]] %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>%
      merge(g2s) %>%
      mutate(result_name = names(contrasts_list)[[i]])
  }
  res_shrunken <- bind_rows(res_list_shrunken)
  
  # Label results
  res_shrunken$cell_type <- ct
  res_shrunken$firre_ko <- fko
  res_shrunken$firre_induced <- fi
  res_unshrunken$cell_type <- ct
  res_unshrunken$firre_ko <- fko
  res_unshrunken$firre_induced <- fi
  
  combined_results <- list("res" = res_unshrunken, "res_shrunken" = res_shrunken)
  
  return(combined_results)
}


```


```{r}
if(!file.exists("results/vs_zero_deseq_res.csv")) {
  
  ## EXPERIMENTS
  # We're going to make 4 comparisons
  experiments <- c("ESC_KO_fi","ESC_WT_fi",
                   "ESC_KO_dox","ESC_WT_dox")
  
  ## RUN DESEQ2
  deseq_res <- lapply(experiments, run_vs_zero_deseq,
                      counts = salmon_gene_counts, 
                      samples = samples,
                      genes = genes,
                      contrasts_list = contrasts_list)
  
  res_list <- lapply(deseq_res, "[[","res")
  res_shrunken_list <- lapply(deseq_res, "[[","res_shrunken")
  
  resdf <- bind_rows(res_list)
  res_shrunkendf <- bind_rows(res_shrunken_list)
  
  # Annotate the comparison
  resdf$comparison <- sapply(resdf$result_name, function(x) {
    if(grepl("_vs_0", x)) {
      "vs_zero"
    } else { "vs_prev" }
  })
  
  res_shrunkendf$comparison <- sapply(res_shrunkendf$result_name, function(x) {
    if(grepl("_vs_0", x)) {
      "vs_zero"
    } else { "vs_prev" }
  })
  
  
  # Split into separate comparisons
  
  vs_zero_res <- resdf %>%
    filter(comparison == "vs_zero")
  vs_zero_res$timepoint_vs_zero <- sapply(vs_zero_res$result_name, function(x) {
    unlist(strsplit(x, "_"))[[3]] %>% as.numeric()
  })
  
  vs_zero_res_shrunken <- res_shrunkendf %>%
    filter(comparison == "vs_zero")
  vs_zero_res_shrunken$timepoint_vs_zero <- sapply(vs_zero_res_shrunken$result_name, function(x) {
    unlist(strsplit(x, "_"))[[3]] %>% as.numeric()
  })
  
  vs_prev_res <- resdf %>%
    filter(comparison == "vs_prev")
  
  vs_prev_res_shrunken <- res_shrunkendf %>%
    filter(comparison == "vs_prev")
  
  write_csv(vs_zero_res, "results/vs_zero_deseq_res.csv")
  write_csv(vs_prev_res, "results/vs_prev_deseq_res.csv")
  
  write_csv(vs_zero_res_shrunken, "results/vs_zero_deseq_res_shrunken.csv")
  write_csv(vs_prev_res_shrunken, "results/vs_prev_deseq_res_shrunken.csv")
}
```

```{r}
vs_zero_res <- read_csv("results/vs_zero_deseq_res.csv")
vs_prev_res <- read_csv("results/vs_prev_deseq_res.csv")

vs_zero_res_shrunken <- read_csv("results/vs_zero_deseq_res_shrunken.csv")
vs_prev_res_shrunken <- read_csv("results/vs_prev_deseq_res_shrunken.csv")


gene_res <- vs_zero_res_shrunken %>% 
  group_by(gene_id, gene_name, cell_type, firre_ko, firre_induced) %>%
  mutate(n_sig = length(which(padj < 0.05)),
         sig_tps = paste(timepoint_vs_zero[which(padj < 0.05)], collapse = ";")) %>%
  ungroup() %>%
  mutate(abs_l2fc = abs(log2FoldChange)) %>%
  arrange(-abs_l2fc) %>%
  mutate(timepoint_vs_zero = factor(timepoint_vs_zero, 
                          levels = c(seq(30,360,30), 720, 1440, 2880, 5760)),
         sig = padj < 0.05)

#TODO: Whoops, it looks like I put the comparisons in the wrong order. 
# Need to flip the sign here.
gene_res$log2FoldChange <- gene_res$log2FoldChange * -1
gene_res$firre_ko <- factor(gene_res$firre_ko, levels = c("WT", "KO"))
gene_res$sig[is.na(gene_res$sig)] <- FALSE

hmm <- gene_res[997,]
sixty_sig_tp <- gene_res %>% filter(n_sig == 4) %>%
  arrange(sig_tps) %>%
  mutate(sig = padj < 0.05)


one_tp_only <- gene_res %>% filter(n_sig == 1)
hi_base_mean <- one_tp_only %>% filter(baseMean > 100)


ex_gene <- "ENSMUSG00000062960.10"
g <- ggplot(gene_res %>% filter(gene_id == ex_gene), 
            aes(x = timepoint_vs_zero, y = log2FoldChange, color = sig))
g + geom_point() + facet_grid(firre_induced ~ firre_ko)



tpm_goi <- tpml %>% filter(gene_id == ex_gene) %>%
  merge(samples) %>%
  filter(cell_type == "ESC",
         as.numeric(as.character(timepoint_minutes)) < 700)
g <- ggplot(tpm_goi, aes(x = timepoint_minutes, y = tpm, color = firre_induced))
g + geom_point() + 
  facet_grid(firre_induced ~ firre_ko)
```



