---
title: "Firre responder functional analysis"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        readr.show_progress = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
# library(DESeq2)
# library(gplots)
# library(ggrepel)
# library(ggpubr)
source("../util/_plot_theme.R")
source("../util/_util.R")
```

## Shared TFs between Shf and Adgrg1

```{r}
# http://dbtoolkit.cistrome.org/?specie=mm10&distance=10k&factor=factor&keyword=chr2%3A122348891%3A122369167%3AXM_006499877%3AShf
shf_tfs <- read.csv("data/chr2_122348891_122369167_XM_006499877_Shfmm1010k_factor_result.csv") %>%
  group_by(Factor) %>%
  summarize(shf_mean_rp_score = mean(RP_score),
            shf_biosources = paste(Biosource, collapse = ";"))
adgrg1_tfs <- read.csv("data/chr8_94974755_95014207_XM_006530695_Adgrg1mm1010k_factor_result.csv") %>%
    group_by(Factor) %>%
  summarize(adgrg1_mean_rp_score = mean(RP_score),
            adgrg1_biosources = paste(Biosource, collapse = ";"))

combined_tfs <- merge(shf_tfs, adgrg1_tfs)

incprint <- read.table("../archive/02_differential_exp/data/incprint_firre_interactors.txt")
combined_tfs$firre_binder <- FALSE
combined_tfs[which(tolower(combined_tfs$Factor) %in% tolower(incprint$V1)), "firre_binder"] <- TRUE

g <- ggplot(combined_tfs, aes(x = shf_mean_rp_score, y = adgrg1_mean_rp_score,
                              color = firre_binder))
g + geom_point() + 
  theme_paperwhite() 


combined_tfs$es_expressed <- FALSE
combined_tfs[which(tolower(combined_tfs$Factor) %in% tolower(es_expressed_factors$gene_name)), "es_expressed"] <- TRUE

g <- ggplot(combined_tfs %>% filter(es_expressed == TRUE), 
            aes(x = shf_mean_rp_score, y = adgrg1_mean_rp_score,
                              color = firre_binder))
g + geom_point() + 
  theme_paperwhite()

combined_tf_matrix <- combined_tfs %>%
  filter(es_expressed == TRUE) %>%
  dplyr::select(Factor, shf_mean_rp_score, adgrg1_mean_rp_score) %>%
  column_to_rownames("Factor") %>%
  as.matrix()

library(pheatmap)
pheatmap(combined_tf_matrix)  

sig_tfs <- res_sig  %>% filter(tolower(gene_name) %in% tolower(combined_tfs$Factor))
```

### Cistrome GO

```{r}
# CISTROME Giggle results for peaks overlapping 
# ESC firre responders responding in both WT and KO
fr_tfs <- read.csv("data/esc_firre_responding_promoters.bed_result.csv") %>%
  group_by(Factor) %>%
  summarize(mean_giggle_score = mean(GIGGLE_score))

# fr_tfs_shf <- fr_tfs %>% filter(tolower(Factor) %in% tolower(combined_tfs$Factor))
fr_tfs$shf_adgrg_regulator <- FALSE
fr_tfs[tolower(fr_tfs$Factor) %in% tolower(combined_tfs$Factor), "shf_adgrg_regulator"] <- TRUE

fr_tfs$es_expressed <- FALSE
fr_tfs[tolower(fr_tfs$Factor) %in% tolower(es_expressed_factors$gene_name),"es_expressed"] <- TRUE
table(fr_tfs$shf_adgrg_regulator)

fr_tfs$firre_binder <- FALSE
fr_tfs[tolower(fr_tfs$Factor) %in% tolower(incprint$V1), "firre_binder"] <- TRUE
table(fr_tfs$firre_binder)
fr_tfs <- fr_tfs %>% arrange(mean_giggle_score)
fr_tfs$Factor <- factor(fr_tfs$Factor, levels = fr_tfs$Factor)


sig_tfs <- res_sig  %>% filter(tolower(gene_name) %in% tolower(fr_tfs$Factor))
g <- ggplot(fr_tfs %>% filter(es_expressed == TRUE), aes(x = Factor, y = mean_giggle_score, color = shf_adgrg_regulator))
g + geom_point() + 
  theme_paperwhite() + 
  coord_flip() + 
  geom_segment(aes(xend = Factor, yend = 40))

fr_tfs_expressed <- fr_tfs %>% filter(es_expressed == TRUE)
cat(paste(fr_tfs_expressed$Factor, collapse = "\n"))
```


```{r}
# Expressed transcription factors.
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id

g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

samples <- read.csv("../../rnaseq/samplesheet.csv") %>%
  filter(cell_type != "pMEF",
         date_sequenced != "18-Sep",
         timepoint_minutes < 400,
         timecourse_length == "short") %>%
  mutate(firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         timepoint_minutes = factor(timepoint_minutes, 
                                    levels = c(seq(0,360,30))),
         firre_induced = factor(firre_induced, 
                                levels = c("control", "firre_induced")))

tpm <- read.csv("../../rnaseq/results/salmon/salmon_merged_gene_tpm.csv") %>%
  merge(g2s) %>%
  dplyr::select(gene_id, gene_name, everything()) %>%
  pivot_longer(cols = 3:ncol(.), names_to = "sample_id", values_to = "tpm") 

tpm <- merge(tpm, samples)


tpm_by_condition <- tpm %>% 
  group_by(gene_id, gene_name, cell_type, firre_ko, firre_induced) %>%
  summarize(mean_tpm = mean(tpm, na.rm = T))
tpm_by_conditionw <- tpm_by_condition %>%
  unite("condition", cell_type, firre_ko, firre_induced) %>%
  pivot_wider(id_cols = c(gene_id, gene_name), names_from = condition, 
              names_prefix = "tpm", values_from = mean_tpm)
write_csv(tpm_by_conditionw, "results/tpm_by_condition.csv")

cistrome_factors <- read.table("data/cistrome_mouse_factors.txt")

factors_tpm <- tpm_by_conditionw[tolower(tpm_by_conditionw$gene_name) %in% tolower(cistrome_factors$V1),]

es_expressed_factors <- factors_tpm %>% 
  filter(tpmESC_KO_firre_induced > 1,
         tpmESC_WT_firre_induced > 1)
write_csv(es_expressed_factors, "results/esc_expressed_factors.csv")

# How many Firre interactors are also TFs?
incprint <- read.table("../archive/02_differential_exp/data/incprint_firre_interactors.txt")

es_fi_factors <- es_expressed_factors %>% filter(tolower(gene_name) %in% tolower(incprint$V1))
cat(paste0(es_fi_factors$gene_name, collapse = "\n"))



```



```{r}
# Firre responder results
res_shrunkendf <- read_csv(file.path("../01_firre_induction_vs_control",
                                     "results/deseq_res_shrunken.csv"))
res_shrunkendf[grep("timepoint_minutes_",res_shrunkendf$result_name),"comparison"] <- "vs_zero"
res_shrunkendf[res_shrunkendf$result_name == "firre_induced_firre_induced_vs_control",
    "comparison"] <- "static_firre_induction_vs_control"
res_shrunkendf[res_shrunkendf$result_name == "firre_ko_KO_vs_WT",
    "comparison"] <- "KO_vs_WT"
res_shrunkendf[grep(".firre_induced",res_shrunkendf$result_name,
         fixed = T),"comparison"] <- "dynamic_firre_induction_vs_control"

# We'd like to organize this in a rank-ordered fashion per timepoint
# in fact we'd like to get it into whatever format webgestalt would like
# for GSEA.
#
# So we need a two column table with the gene_name and the rank one table 
# per timepoint.

res_dynamic <- res_shrunkendf %>% filter(comparison == "dynamic_firre_induction_vs_control")

res_dynamic$timepoint <- sapply(res_dynamic$result_name,
                                function(x) {
                                  x <- gsub("timepoint_minutes", "", x)
                                  gsub(".firre_inducedfirre_induced", "", x,
                                       fixed = T)
                                }) %>%
  as.numeric()


timepoints <- unique(res_dynamic$timepoint) 
experiments <- c("ESC_KO","ESC_WT", "NPC_KO", "NPC_WT")
dir.create("results/shrunken_lfc_rankings")


for(j in 1:length(experiments)) {
  dir.create(paste0("results/shrunken_lfc_rankings/", experiments[j]))
  ct <- unlist(strsplit(experiments[j], "_"))[[1]]
  fko <- unlist(strsplit(experiments[j], "_"))[[2]]
  
  res_exp <- res_dynamic %>% filter(cell_type == ct,
                                    firre_ko == fko)
  timepoints <- unique(res_exp$timepoint)
  
  for(i in 1:length(timepoints)) {
    
    tp <- res_exp %>% filter(timepoint == timepoints[i]) %>%
      filter(!is.na(padj)) %>%
      arrange(-log2FoldChange) %>%
      dplyr::select(gene_id, log2FoldChange)
    tp$gene_id <- sapply(tp$gene_id, function(x) {
      unlist(strsplit(x, ".", fixed = T))[[1]]
    })
    write.table(tp, 
                paste0("results/shrunken_lfc_rankings/", experiments[j],
                       "/", experiments[j], "_slfc_", timepoints[i],".rnk"),
                row.names = FALSE, col.names = FALSE, sep = "\t", quote = FALSE)
  }
}
```

```{r}
library(WebGestaltR)
dir.create("results/gsea_enrichment")
gene_sets_all <- listGeneSet()
gene_sets <- gene_sets_all$name[c(1,3,5,7:12,53,54,58:61)]
gene_sets[1]
gsea <- WebGestaltR(
  enrichMethod = "GSEA",
  organism = "mmusculus",
  enrichDatabase = gene_sets[1],
  interestGeneFile = "results/shrunken_lfc_rankings/ESC_KO/ESC_KO_slfc_30.rnk",
  interestGeneType = "ensembl_gene_id",
  sigMethod = "top",
  isParallel = T,
  nThreads = 8,
  outputDirectory = "results/gsea_enrichment"
)

```

```{r}
# What are the ESC firre responder regions of interest to look for TF enrichement?
genes <- rtracklayer::import("../util/gencode.vM25.annotation.genes.gtf")
names(genes) <- genes$gene_id
res_sig <- read.csv("../01_firre_induction_vs_control/results/firre_induced_genes.csv")



overlapping <- res_sig %>% filter(ESC_KO_padj0.05 == 1,
                                  ESC_WT_padj0.05 == 1)

cat(paste(overlapping$gene_name, collapse = "\n"))


# Firre responder gene regions
fr_genes <- genes[overlapping$gene_id] %>%
  as.data.frame() %>%
  dplyr::select(seqnames, start, end)
write.table(fr_genes, 
            "results/esc_firre_responding_genes.bed",
            sep = "\t", quote = FALSE, col.names = F, row.names = F)  

fr_promoters <- genes[overlapping$gene_id] %>%
  promoters(upstream = 5000, downstream = 5000) %>%
  as.data.frame() %>%
  dplyr::select(seqnames, start, end)
write.table(fr_promoters, 
            "results/esc_firre_responding_promoters5kb.bed",
            sep = "\t", quote = FALSE, col.names = F, row.names = F)  

```



```{r}
library(GenomicRanges)
genes <- rtracklayer::import("../util/gencode.vM25.annotation.genes.gtf")
genes$score <- 0
rtracklayer::export(genes, "results/universe_all_genes.bed")
gene_proms <- promoters(genes)
gene_proms$score <- 0
rtracklayer::export(gene_proms, "results/universe_all_promoters.bed")


resdf <- read.csv("../01_firre_induction_vs_control/results/deseq_res.csv")
expressed_genes <- resdf %>% filter(!is.na(padj),
                                    baseMean > 100)

exp_genes <- genes[unique(expressed_genes$gene_id)]
exp_genes$score <- 0
rtracklayer::export(exp_genes, "results/universe_expressed_genes.bed")


res_sig <- read.csv("../01_firre_induction_vs_control/results/firre_induced_genes.csv")
overlapping <- res_sig %>% filter(ESC_KO_padj0.05 == 1,
                                  ESC_WT_padj0.05 == 1)

names(gene_proms) <- gene_proms$gene_id
esc_firre_responding_promoters <- gene_proms[overlapping$gene_id] %>%
  as.data.frame() %>%
  dplyr::select(seqnames, start, end)
write.table(esc_firre_responding_promoters, 
            "results/esc_firre_responding_promoters.bed",
            sep = "\t", quote = FALSE, col.names = F, row.names = F)

all_npc <- res_sig %>% filter(NPC_KO_padj0.05 == 1 | NPC_WT_padj0.05 == 1)

res_res <- resdf %>% filter(gene_id %in% overlapping$gene_id) %>%
  arrange(baseMean)
min(res_res$baseMean)
names(genes) <- genes$gene_id
all_deg <- genes[res_sig$gene_id]
all_deg$score <- 0
rtracklayer::export(all_deg,"results/all_deg_genes.bed")

all_deg_promoters <- promoters(all_deg)
rtracklayer::export(all_deg_promoters, "results/all_deg_promoters.bed")

overlapping_deg <- all_deg[overlapping$gene_id]
rtracklayer::export(overlapping_deg, "results/esc_overlapping_deg_genes.bed")

npc_deg <- all_deg[all_npc$gene_id]
rtracklayer::export(npc_deg, "results/npc_deg_genes.bed")
```

```{r}
library("LOLA")
regionDB = loadRegionDB("data/LOLACore/mm10")

regionSetA = readBed("results/all_deg_genes.bed")
regionSetB <- readBed("results/esc_overlapping_deg_genes.bed")
regionSetC <- readBed("results/all_deg_promoters.bed")
userSets = GRangesList(regionSetA, regionSetB, regionSetC)
genes_all <- readBed("results/universe_all_genes.bed")
locResults = runLOLA(userSets, genes_all, regionDB, cores=6)

promoters_all <- readBed("results/universe_all_promoters.bed")
locResults = runLOLA(regionSetC, promoters_all, regionDB, cores=6)



exp_genes <- readBed("results/universe_expressed_genes.bed")
locResults = runLOLA(userSets, exp_genes, regionDB, cores=6)
locResults$padj <- p.adjust(10^-locResults$pValueLog)
incprint <- read.table("../archive/02_differential_exp/data/incprint_firre_interactors.txt")
hmm <- locResults[which(tolower(locResults$antibody) %in% tolower(incprint$V1)),] %>%
  filter(padj < 0.05)

npc_deg_bed <- readBed("results/npc_deg_genes.bed")
npc_lola_res <- runLOLA(npc_deg_bed, exp_genes, regionDB, cores=6)
oneResult = locResults[2,]
extractEnrichmentOverlaps(oneResult, regionSetA, regionDB)
```



