---
title: "Impulse model"
author: "Michael Smallegan"
date: "10/14/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(DESeq2)
library(broom)
source("../util/_plot_theme.R")
```

Okay, for now, since I haven't actually seen any profiles that look quite like 
the impulse model, I'll just fit the sigmoid model. 

It makes sense to use the rlog counts sense they are normalized and the
dispersion for low counts is taken care of. 

We'll start with generating those counts.

```{r}
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id

g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

salmon_gene_counts <- read.csv(file.path("../../rnaseq/results/salmon/",
                                         "salmon_merged_gene_counts.csv")) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()
mode(salmon_gene_counts) <- "integer"


samples <- read.csv("../../rnaseq/rnaseq_samplesheet.csv") %>%
  filter(cell_type != "pMEF",
         date_sequenced != "2018-09",
         timepoint < 400,
         timecourse_length == "short") %>%
  mutate(firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         timepoint = factor(timepoint, 
                                    levels = c(seq(0,360,30))),
         firre_induced = factor(firre_induced, 
                                levels = c("control", "firre_induced")))
rownames(samples) <- samples$sample_id

# Probably makes sense to do this separately for ESC and NPC
esc_samples <- samples %>% 
  filter(cell_type == "ESC")
# Filter to just the data we're using
esc_counts <- salmon_gene_counts[,esc_samples$sample_id]

genes <- genes[rownames(esc_counts)]
```

```{r}
stopifnot(all(rownames(esc_samples) == colnames(esc_counts)))
stopifnot(all(names(genes) == rownames(esc_counts)))
  
dds <- DESeqDataSetFromMatrix(countData = esc_counts, 
                              colData = esc_samples, 
                              design = ~ firre_induced + firre_ko,
                              rowData = genes)

# Filter low counts
dds <- dds[rowSums(counts(dds)) >= 10,]

esc_rlog_counts <- rlog(dds, blind = FALSE)

esc_rlog_countsl <- assay(esc_rlog_counts) %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(2:ncol(.), names_to = "sample_id", values_to = "rlog_count") 
esc_rlog_countsl <- esc_rlog_countsl %>%
  merge(esc_samples %>% dplyr::select(sample_id, cell_type, firre_ko, 
                                      firre_induced, timepoint)) %>%
  merge(g2s)
esc_rlog_countsl$timepoint <- as.numeric(as.character(esc_rlog_countsl$timepoint))
esc_rlog_countsl <- esc_rlog_countsl %>%
  arrange(gene_id, timepoint)

write_rds(esc_rlog_countsl, "results/esc_rlog_counts.RDS")
esc_rlog_countsl <- read_rds("results/esc_rlog_counts.RDS")
esc_rlog_countsl$timepoint <- as.numeric(as.character(esc_rlog_countsl$timepoint))
```



```{r}
# By subtracting the zero timepoint rlog counts from all other timepoints,
# it's effectively like the log2 fold change from zero, since rlog counts
# are roughly the log2 counts
# From the DEseq2 vignette: "[rlog] computes a variance stabilizing 
# transformation which is roughly similar to putting the data on the 
# log2 scale, while also dealing with the sampling variability of low counts.

# For now, we'll also remove this sample (JR2432)
# which has a lot of outlier genes
# TODO: Figure out what's going on with this sample.
esc_rlog_countsl <- esc_rlog_countsl %>% 
  filter(sample_id != "JR2432")

esc_rlog_counts_zero_centered <- esc_rlog_countsl %>%
  group_by(gene_id, gene_name, firre_ko, firre_induced) %>%
  mutate(rlog_count = rlog_count - mean(rlog_count[timepoint == 0])) %>%
  ungroup()

# Triage genes that are more or less not changing, so that we
# don't have to compute the curve fitting for so many genes. 

dynamic <- esc_rlog_counts_zero_centered %>%
  group_by(gene_id, gene_name, firre_ko, firre_induced, timepoint) %>%
  # Take the mean to deal with outliers
  summarize(lfc = mean(rlog_count)) %>%
  group_by(gene_id, gene_name) %>%
  summarize(max_lfc = max(abs(lfc), na.rm = T)) %>%
  # 0.08 is ~50th percentile and is a very small change
  filter(max_lfc > stats::quantile(.$max_lfc, 0.5))


rlog_counts <- esc_rlog_counts_zero_centered %>%
  filter(gene_id %in% dynamic$gene_id)


# If each gene takes 5 seconds to curve fit, let's consider the best way to 
# divide. I estimate a total of about 22 hours, so if we divide into 
# about twenty batches, each should take an hour. That's about
# 800 genes per batch
n_batches <- 20
batches <- rlog_counts %>% 
  dplyr::select(gene_id) %>%
  distinct() %>%
  mutate(batch = sample(1:n_batches, nrow(.), replace=T))

rlog_counts <- rlog_counts %>%
  merge(batches)

rlog_counts %>% 
  nest_by(batch) %>%
  pmap(~write_rds(x = .y, 
                  path = paste0("results/rlog_counts/batch_", .x, ".rds"))) %>%
  invisible()

```

```{r}
batches <- 1:n_batches

create_sbatch_script <- function(batch) {

line <- paste0("#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=fit_timecourse_", batch, "
#SBATCH --mail-type=NONE
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=30gb
#SBATCH --time=4:00:00
#SBATCH --output=fit_tc_", batch, ".out
#SBATCH --error=fit_tc_", batch, ".err
date
Rscript fit_timecourse.R ", batch, "
date
")
writeLines(line, 
           con = paste0("bin/fit_tc_", batch, ".sbatch"))
}

lapply(batches, create_sbatch_script) %>% 
  invisible()
# Create submit script
cat(paste0("#!/bin/bash \n",
              paste(paste0("sbatch ", paste0("fit_tc_", batches, ".sbatch")),
              collapse = "\n")),
    file = "bin/submit.sh")
# TODO: this is a really cheap and messy way to do caching
if(!file.exists("results/curve_fits/batch_1.rds")) {
  system("source /etc/profile; cd bin; chmod u+x submit.sh; ./submit.sh")
}
```

```{r}
# Import sig
fl <- list.files("../01_threshold_differential_genes/results/", 
                 pattern = "_sig.csv", full.names = T)
experiment <- sapply(fl, function(x) {
  gsub("_sig.csv", "", unlist(strsplit(x,"results//", fixed = T))[2])
})
deg <- lapply(fl, read_csv)
for(i in 1:length(fl)) {
  deg[[i]]$experiment <- experiment[[i]]
}
deg <- deg %>% bind_rows()

deg <- deg %>% dplyr::select(-result_name, -sig) %>%
  distinct()
esc_short_deg <- deg %>% filter(grepl("ESC", experiment),
                                grepl("short", experiment))

esc_short_deg <- esc_short_deg %>%
  mutate(experiment = gsub("firre_induced", "fi", experiment)) %>%
  separate(experiment, into = c("cell_type", "firre_ko",
                                "firre_induced", "timecourse_length", "experiment_replicate"),
           remove = F) %>%
  mutate(firre_induced = gsub("fi", "firre_induced", firre_induced)) %>%
  unite(condition, firre_ko, firre_induced, remove = F)
  

esc_short_deg <- esc_short_deg %>%
  unite(gene_condition, gene_id, condition, remove = F)
fl <- list.files("results/curve_fits/", full.names = T)
curve_fits <- read_rds("results/curve_fits/batch_1.rds") 
curve_fits <-curve_fits %>% 
  unite(gene_condition, gene_id, condition, remove = F)
curve_fits <-curve_fits %>%
  filter(gene_condition %in% esc_short_deg$gene_condition)

for(i in 2:length(fl)) {
  print(i)
  batch <- read_rds(fl[i]) %>%
  unite(gene_condition, gene_id, condition, remove = F) %>%
  filter(gene_condition %in% esc_short_deg$gene_condition)
  curve_fits <- bind_rows(curve_fits, batch)
}

table(curve_fits$condition)
# Import results.
# curve_fits <- tibble(batch = 1:n_batches) %>%
#   mutate(curve_fits = map(batch, ~ read_rds(paste0("results/curve_fits/batch_", .x, ".rds")))) %>%
#   unnest(curve_fits)



format(object.size(curve_fits), units = "GB")
# write_rds(curve_fits, "results/curve_fits.rds")

curve_fits <- read_rds("results/curve_fits.rds")

curve_fitsl <- curve_fits %>%
  mutate(lm_tidy = map(lm_fit, tidy),
         sigmoid_tidy = map(sigmoid_fit, tidy),
         impulse_tidy = map(impulse_fit, tidy),
         lm_bic = map(lm_fit, BIC),
         sigmoid_bic = map(sigmoid_fit, BIC),
         impulse_bic = map(impulse_fit, BIC))

choose_best_model <- function(bic_list) {
  unlist(strsplit(names(bic_list)[which.min(bic_list)], "_"))[[1]]
}
curve_fitsl$best_model <- apply(curve_fitsl[,c("lm_bic","sigmoid_bic", "impulse_bic")],
                            1, choose_best_model)
colnames(curve_fits)
sig_curve_fits <- curve_fits %>% filter(gene_id %in% unique(hmm$gene_id))
write_rds(sig_curve_fits, "results/sig_curve_fits.rds")
curve_fitsl_sigmoid <- curve_fitsl %>%
  unnest(sigmoid_tidy)
curve_fitsl <- curve_fitsl %>%
  unnest(impulse_tidy)
plot_timecourse_fits("Klf4", curve_fitsl)
plot_timecourse_fits("Kdr", curve_fitsl)
plot_timecourse_fits("Arih1", curve_fitsl)
plot_timecourse_fits("Firre", curve_fitsl)
plot_timecourse_fits("Nceh1", curve_fitsl)
plot_timecourse_fits("Nceh1", curve_fitsl)
plot_timecourse_fits("Shf", curve_fitsl)
plot_timecourse_fits("Dpf2", curve_fitsl)
plot_timecourse_fits("Umps", curve_fitsl)	
plot_timecourse_fits("Fst", curve_fitsl)
plot_timecourse_fits("Clock", curve_fitsl)
plot_timecourse_fits("Adgrg1", curve_fitsl)
plot_timecourse_fits("Apod", curve_fitsl)
plot_timecourse_fits("Kdr", curve_fitsl)
plot_timecourse_fits("Duox1", curve_fitsl)
plot_timecourse_fits("Pde6d", curve_fitsl)
plot_timecourse_fits("Ppp1r2", curve_fitsl)
plot_timecourse_fits("Ddhd1", curve_fitsl)
plot_timecourse_fits("Hacd3", curve_fitsl)


Kdr

	Clock

	

Dpf2
Nceh1
Arih1
# plot_timecourse_fits("Slf29a3", curve_fitsl)
unique(curve_fitsl_sigmoid$term)
good_ones <- curve_fitsl_sigmoid %>%
  filter(term == "rate") %>%
  filter(p.value < 0.05)
# Cool gene list
# Slc29a3
# Ahdc1
# Fv1

```

```{r}
# Determine sig genes
fl <- list.files("../01_threshold_differential_genes/results/", 
                 # pattern = "_sig.csv", 
                 full.names = T)
fl <- fl[!grepl("sig",fl)]
fl <- fl[!grepl("rds",fl)]
fl <- fl[grep("ESC", fl)]
fl <- fl[grep("short", fl)]
fl <- fl[!grepl("ESC-WT-firre_induced-short-1.csv", fl)]
experiment <- sapply(fl, function(x) {
  gsub(".csv", "", unlist(strsplit(x,"results//", fixed = T))[2])
})
res <- lapply(fl, read_csv)
for(i in 1:length(fl)) {
  res[[i]]$experiment <- experiment[[i]]
}
res <- res %>% bind_rows()

res <- res %>%
  mutate(experiment = gsub("firre_induced", "fi", experiment)) %>%
  separate(experiment, into = c("cell_type", "firre_ko",
                                "firre_induced", "timecourse_length", "experiment_replicate"),
           remove = F) %>%
  mutate(firre_induced = gsub("fi", "firre_induced", firre_induced)) %>%
  unite(condition, firre_ko, firre_induced, remove = F)

fi_vs_control <- read_csv("../01_firre_induction_vs_control/results/firre_induced_genes.csv") %>%
  dplyr::select(gene_id, gene_name, ESC_KO_padj0.05,ESC_WT_padj0.05) %>%
  pivot_longer(cols = c(ESC_KO_padj0.05,ESC_WT_padj0.05), names_to = "condition", values_to = "sig") %>%
  filter(sig == 1)
fi_vs_control$condition <- gsub("ESC_KO_padj0.05", "KO_firre_induced", fi_vs_control$condition)
fi_vs_control$condition <- gsub("ESC_WT_padj0.05", "WT_firre_induced", fi_vs_control$condition)
table(fi_vs_control$condition)
res_summary <- res %>% filter(grepl("_vs_0",result_name)) %>%
  group_by(gene_id, experiment, condition, firre_ko, firre_induced) %>%
  summarize(mean_fc = mean(l2fc))

res_pval <- res %>%
  group_by(gene_id, experiment, condition, firre_ko, firre_induced) %>%
  summarize(min_pval = min(padj))
res_pval <- res_pval %>%
  pivot_wider(id_cols = c(gene_id, firre_ko), names_from = firre_induced, values_from = min_pval)
res_summary <- res_summary %>%
  pivot_wider(id_cols = c(gene_id, firre_ko), names_from = firre_induced, values_from = mean_fc)


res_pval <- res_pval %>% as.data.frame()
res_summary <- res_summary %>% as.data.frame()
names(res_pval)[3:4] <- c("ctrl_pval", "fi_pval")
res_res <- merge(res_summary, res_pval)

res_res$ctrl_sig <- as.numeric(res_res$ctrl_pval < 0.05)
res_res$fi_sig <- as.numeric(res_res$fi_pval < 0.05)
res_res[is.na(res_res$ctrl_sig),"ctrl_sig"] <- 0L
res_res[is.na(res_res$fi_sig),"fi_sig"] <- 0L

res_res$fi_sign <- sign(res_res$firre_induced)
res_res$ctrl_sign <- sign(res_res$control)
res_res$diff <- res_res$fi_sign - res_res$ctrl_sign
table(res_res$diff)
res_res_fi <- res_res %>%
  filter(fi_sig == 1)

res_res_fi$fi_deg <- 0L
res_res_fi[which(res_res_fi$fi_sig == 1 & res_res_fi$ctrl_sig == 0),"fi_deg"] <- 1L
res_res_fi[which(res_res_fi$fi_sig == 1 & res_res_fi$ctrl_sig == 1 & res_res_fi$diff != 0),"fi_deg"] <- 1L

write_csv(res_res_fi, "results/firre_deg_genes.csv")
res_res_fi <- read_csv("results/firre_deg_genes.csv")
res_res_fi <- res_res_fi %>%
  filter(fi_deg == 1)
res_res_fi$condition <- paste0(res_res_fi$firre_ko, "_firre_induced")
res_res_fi <- res_res_fi %>%
  unite(gene_condition, gene_id, condition, remove = F)

fi_vs_control <- fi_vs_control %>%
  unite(gene_condition, gene_id, condition, remove = F)
sig_gene_conditions <- c(fi_vs_control$gene_condition, res_res_fi$gene_condition)
```

```{r}
# Curve fits
table(curve_fits$condition)
curve_fits_fi <- curve_fits %>% filter(gene_condition %in% sig_gene_conditions)
curve_fits_fi <- curve_fits_fi %>%
  mutate(lm_tidy = map(lm_fit, tidy),
         sigmoid_tidy = map(sigmoid_fit, tidy),
         impulse_tidy = map(impulse_fit, tidy),
         lm_bic = map(lm_fit, BIC),
         sigmoid_bic = map(sigmoid_fit, BIC),
         impulse_bic = map(impulse_fit, BIC))

choose_best_model <- function(bic_list) {
  unlist(strsplit(names(bic_list)[which.min(bic_list)], "_"))[[1]]
}
curve_fits_fi$best_model <- apply(curve_fits_fi[,c("lm_bic","sigmoid_bic", "impulse_bic")],
                            1, choose_best_model)
table(curve_fits_fi$best_model)

curve_fits_fi_sigmoid <-  curve_fits_fi %>%
  unnest(sigmoid_tidy)
cf_fi_sig <- curve_fits_fi_sigmoid %>%
  filter(term == "t_rise")  %>%
  filter(p.value < 0.05)
table(cf_fi_sig$best_model)
# curve_fits_fi_sigmoid <- curve_fits_fi_sigmoid %>%
#   filter(best_model == "sigmoid")



cf_fi_imp <- curve_fits_fi %>%
  unnest(impulse_tidy) %>%
  # filter(best_model == "impulse")
 filter(term == "t_rise")  %>%
  filter(p.value < 0.05)

table(cf_fi_imp$best_model)

covered_genes <- unique(c(cf_fi_imp$gene_condition, cf_fi_sig$gene_condition))
length(which(curve_fits_fi$gene_condition %in% covered_genes))
length(unique(curve_fits_fi$gene_condition))
table(curve_fits_fi$best_model)

cf_fi_lm <- curve_fits_fi %>%
  unnest(lm_tidy)

lm_good <- cf_fi_lm %>%
  filter(term == "timepoint_minutes",
         best_model == "lm") %>%
  arrange(p.value)
imp_good <- cf_fi_imp %>%
 
  arrange(p.value)

for(i in 1:20) {
  show(plot_timecourse_fits(lm_good$gene_name[[i]], curve_fitsl))
  Sys.sleep(2)
}
?Sys.sleep
show(plot_timecourse_fits("Adgrg1", curve_fitsl))



cf_fi_sig <- curve_fits_fi %>%
  unnest(sigmoid_tidy) %>%
  filter(term == "t_rise",
         p.value < 0.01) 
hist(cf_fi_sig$p.value)
Psmd11
```

```{r}
# let's merge the estimates of t_rise
cf_sig <- cf_fi_sig %>% dplyr::select(gene_condition, gene_name, estimate, p.value)
cf_imp <- cf_fi_imp %>% dplyr::select(gene_condition, gene_name, estimate, p.value)

names(cf_sig)[3:4] <- c("sig_trise", "sig_p")
names(cf_imp)[3:4] <- c("imp_trise", "imp_p")
cfm <- merge(cf_sig, cf_imp, all = T)
library(ggpubr)
g <- ggplot(cfm %>% filter(sig_p < 0.1, imp_p < 0.1), aes(x = sig_trise, y = imp_trise))
g + geom_point() + stat_cor() +
  geom_abline(slope = 1) + 
  ggtitle("t_rise model correlation") + 
  xlab("Sigmoid") +
  ylab("Impulse")

cfm$t_rise <- cfm$sig_trise
cfm[which(!is.na(cfm$imp_trise) | cfm$imp_p < cfm$sig_p), "t_rise"] <- cfm[which(!is.na(cfm$imp_trise) | cfm$imp_p < cfm$sig_p), "imp_trise"]

ress <- res %>% filter(grepl("_vs_0",result_name)) %>%
  group_by(gene_id, experiment, condition, firre_ko, firre_induced) %>%
  summarize(max_fc = l2fc[which.max(abs(l2fc))])
names(cfm)
cfm$gene_condition
cfm_tr <- cfm  %>%
  mutate(gene_condition = gsub("firre_induced", "fi", gene_condition)) %>%
  separate(gene_condition, into = c("gene_id", "firre_ko", "firre_induced"), sep = "_") %>%
  mutate(firre_induced = gsub("fi", "firre_induced", firre_induced)) %>%
  dplyr::select(gene_id, firre_ko, firre_induced, gene_name, t_rise)
hmm <- merge(cfm_tr, ress, all.x = T)
library(ggrepel)
library(scico)
g <- ggplot(hmm %>% filter(firre_ko == "KO"), aes(x = t_rise, y = max_fc, color = t_rise))
g + geom_hline(yintercept = 0) +geom_point() +
  # geom_text_repel(aes(label = gene_name), data = hmm %>% filter(firre_ko == "KO", max_fc > 2)) +
  geom_text_repel(aes(label = gene_name), data = hmm %>% filter(firre_ko == "KO", t_rise < 60)) +
    xlim(20,100) +
  ylim(-1,1) +
  ggtitle("ESC WT Firre timecouse") + 
  scale_color_scico(direction = -1, begin = 0.1)

g <- ggplot(hmm %>% filter(firre_ko == "WT"), aes(x = t_rise, y = max_fc, color = t_rise))
g + geom_hline(yintercept = 0) +geom_point() +
  # geom_text_repel(aes(label = gene_name), data = hmm %>% filter(firre_ko == "WT", max_fc > 1 | max_fc < -0.9)) + 
  geom_text_repel(aes(label = gene_name), data = hmm %>% filter(firre_ko == "WT", t_rise < 50)) +
  xlim(20,70) +
  ylim(-1,1) +
   ggtitle("ESC KO Firre timecouse") +
  scale_color_scico(direction = -1, begin = 0.1)
```





```{r}


goi <- c("Firre", "Sox2", "Gfap", "Adgrg1", "Duox1")
goi <- "Firre"

curve_fits <- rlog_counts  %>%
  filter(gene_name %in% goi) %>%
  dplyr::select(gene_id, gene_name, timepoint, 
                rlog_count, firre_ko, firre_induced) %>%
  unite(condition, firre_ko, firre_induced) %>%
  # Center zero timepoint at zero
  group_by(gene_id, gene_name, condition) %>%
  ungroup() %>%
  nest(timecourse = c(timepoint, rlog_count)) %>%
  mutate(lm_fit = map(timecourse, ~ lm(rlog_count ~ 0 + timepoint, data = .)),
         sigmoid_fit = map(timecourse, ~ cmp_fit_sigmoid_wrapper(.$rlog_count, .$timepoint)),
         impulse_fit = map(timecourse, ~ cmp_fit_impulse_wrapper(.$rlog_count, .$timepoint))) 

curve_fits <- curve_fits %>%
  mutate(lm_tidy = map(lm_fit, tidy),
         sigmoid_tidy = map(sigmoid_fit, tidy),
         impulse_tidy = map(impulse_fit, tidy),
         lm_bic = map(lm_fit, BIC),
         sigmoid_bic = map(sigmoid_fit, BIC),
         impulse_bic = map(impulse_fit, BIC))

choose_best_model <- function(bic_list) {
  unlist(strsplit(names(bic_list)[which.min(bic_list)], "_"))[[1]]
}
curve_fits$best_model <- apply(curve_fits[,c("lm_bic","sigmoid_bic", "impulse_bic")],
                            1, choose_best_model)


plot_timecourse_fits <- function(gene, fit_df) {
  
og <- fit_df %>%
  filter(gene_name == gene)

og_augmented <- og %>%
  mutate(lm_augmented = map(lm_fit, augment),
         sigmoid_augmented = map(sigmoid_fit, augment),
         impulse_augmented = map(impulse_fit, augment))%>%
  unnest(c(lm_augmented, sigmoid_augmented, impulse_augmented), names_sep = "_") %>%
  dplyr::select(gene_id, gene_name, condition, lm_augmented_timepoint_minutes,
                lm_augmented_rlog_count, lm_augmented_.fitted, impulse_augmented_.fitted,
                sigmoid_augmented_.fitted, best_model) %>%
  pivot_longer(cols = c(lm_augmented_.fitted, 
                        sigmoid_augmented_.fitted,
                        impulse_augmented_.fitted),
               names_to = "model", values_to = "fitted_counts") %>%
  mutate(model = sub("\\_.*", "", model),
         is_best = ifelse(model == best_model, "solid", "dashed")) 

g <- ggplot(og_augmented, aes(x = lm_augmented_timepoint_minutes, 
                          y = lm_augmented_rlog_count, 
                          group = model, color = model, linetype = is_best)) + 
  geom_hline(yintercept = 0, color = "#E0E2E1") +
  geom_vline(xintercept = 0, color = "#E0E2E1") +
  geom_line(aes(y = fitted_counts), size = 1.1, alpha = 0.9) + 
  geom_point(color = "#424242") +
  facet_wrap(~condition) +
  scale_linetype_identity() + 
  scale_color_manual(values = c("#5BBA62", "#a8404c", "#024059")) + 
  ggtitle(paste(unique(og_augmented$gene_name), " | ", 
                           unique(og_augmented$gene_id))) + 
  xlab("t") +
  ylab("Abundance")

return(g)
}

hmm <- curve_fitsl %>% filter(gene_name == "Firre")
plot_timecourse_fits("Nanog", curve_fitsl)
# Lars2 outlier in sample JR2432
# Also RN7sk in sample JR2432
```

```{r}
one_tc <- esc_rlog_countsl %>%
  filter(firre_induced == "firre_induced", 
         firre_ko == "WT",
         gene_name == "Firre") %>%
  arrange(timepoint)
zero_tp_val <- mean(one_tc[one_tc$timepoint == 0,"rlog_count"])
x <- as.numeric(as.character(one_tc$timepoint))
y <- one_tc$rlog_count
y <- y - zero_tp_val
plot(y~x)

sigmoid_fit <- fit_sigmoid_wrapper(y, x)
impulse_fit <- fit_impulse_wrapper(y, x)
lm_fit <- lm(y~0+x)
plot(y~x)
# lines(x, sigmoid_fun(x,v_inter = -0.1, t_rise =180, rate = 0.9))
lines(unique(x), 
      predict(lm_fit, newdata = data.frame(x = unique(x))), col = "red")
lines(unique(x), 
      predict(impulse_fit, newdata = data.frame(x = unique(x))), col = "green")
lines(unique(x), 
      predict(sigmoid_fit, newdata = data.frame(x = unique(x))), col = "blue")

```

