---
title: "fun with faces"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(plotly)
library(GenomeInfoDb)
library(tidyverse)
```

## What is the face of mouse genome?

Using the Gencode annotation, what does the gene distribution of the mouse genome look like? I would of course actually like to do this for the Firre timecourse data, but I'm just trying out some ideas here.

If you lay the chromosomes end-to-end (as is often done with genomic tiling methods), the genome is a 1-dimensional vector. If you consider each chromosome as a vector, then you have a list of vectors of different lengths corresponding to each chromosome. You could line these up starting at the first index of each chromosome, but there's no a priori reason why the N-th letter of chromosome 2 should be lined up with the same letter on chromosome 14. To lay things out with some biological basis, we could use Hi-C models, but that is dynamic across cellular states. You could also use gene co-expression networks to layout the genome. This is intriguing, but comlicated in practice. For these reasons, it is both justifiable and so simple to layout the genome in an arbirtrary (but ordered) fashion. 

That's what I'll do here. First creating a 1D vector, then binning it, then threading the bins through 2 -- no 3! -- dimensions. For fun and profit!

```{r import}
dat <- read.csv("../08_de_long/results/shrunkenlfc_ftc_diffinwtko.csv") %>% select(-X)
gff <- rtracklayer::import("../../../genomes/mus_musculus/gencode/m23/gencode.vM23.annotation.gff3")
genes <- gff[gff$type == "gene"]

#genes in data
dat <- dat %>% filter(!is.na(log2FoldChange))
genes_exp <- genes[genes$gene_id %in% unique(dat$gene_id)]




chr_lengths <- c("chr1" = 195471971,
"chr2" = 182113224,
"chr3" = 160039680,
"chr4" = 156508116,
"chr5" = 151834684,
"chr6" = 149736546,
"chr7" = 145441459,
"chr8" = 129401213,
"chr9" = 124595110,
"chr10" = 130694993,
"chr11" = 122082543,
"chr12" = 120129022,
"chr13" = 120421639,
"chr14" = 124902244,
"chr15" = 104043685,
"chr16" = 98207768,
"chr17" = 94987271,
"chr18" = 90702639,
"chr19" = 61431566,
"chrX" = 171031299,
"chrY" = 91744698,
"chrM" = 16299)
seqlengths(genes) <- chr_lengths
```

### Genomic grid

```{r}
# the resolution of the face as it were.
grid_size <- 50
nbins <- grid_size * grid_size * grid_size
nbins
```



```{r}
bins <- GenomicRanges::tileGenome(chr_lengths, n = nbins)
# For some reason the lengths are wrong in the gff.
# The lengths above were determined with:
# `samtools faidx GRCm38.p6.genome.fa`
# `cut -f1-2 GRCm38.p6.genome.fa.fai`
# so the lengths are from gencode m23.


geneov <- findOverlaps(genes,bins , type ="any", select = "first", ignore.strand = TRUE)

genes$bin <- geneov

# merge into dat
genesdf <- genes %>% as.data.frame() %>% dplyr::select(gene_id, bin)

# dat <- dat %>% merge(genesdf)
```

For some fun, let's just plot the genomic density (annotation density; irrespective of expression in this experiment).

```{r, fig.width=12, fig.height=12, warning=FALSE}
# genesdf$genomic_bin <- genesdf$genomic_bin - 1
# genesdf$x <- floor(genesdf$genomic_bin / grid_size) + 1
# genesdf$y <- genesdf$genomic_bin %% grid_size + 1
# genomic_density <- genesdf %>% group_by(x,y) %>%
#   summarize(ngenes = n())
# gd_dense <- expand.grid(x =  1:grid_size, y = 1:grid_size)
# hmm <- merge(gd_dense, genomic_density, all.x = TRUE)
# hmm[is.na(hmm$ngenes),"ngenes"] <- 0


td <- expand.grid(x = 1:grid_size, y = 1:grid_size, z = 1:grid_size)
td$bin <- 1:nrow(td)

gd <- genesdf %>% group_by(bin) %>%
  summarize(ngenes = n())


gd <- merge(gd, td)

# Create lists for axis properties


axis <- list(
  title = "",
  showticklabels=FALSE,
  showgrid = F
)

scene = list(
  xaxis = axis,
  yaxis = axis,
  zaxis = axis,
  camera = list(eye = list(x = -1.25, y = 1.25, z = 1.25)))

plot_ly(gd %>% filter(ngenes > 1), type = 'scatter3d', mode = 'markers', x = ~x, y = ~y, z = ~z, color = ~ngenes,
        text = ~bin,
        hoverinfo = 'text',
        size = ~ngenes, sizes = c(1, 3000),
        colors = rev(pals::magma(100)), alpha = 0.9) %>% layout(title = "mm10 gene density", scene = scene)

```

Man what is happening with 48400 through 48388?

```{r}
bins[48400:48388] %>% as.data.frame()
```

Oh wow, it's still the SNHG14 region, just spread across many bins now.

What about

```{r}
bins[9447:9448] %>% as.data.frame()
bins[125000] %>% as.data.frame()
```

So cool, it's a rodent specific miRNA cluster.
https://www.sciencedirect.com/science/article/pii/S221112471730493X


```{r}
# g <- ggplot(hmm, aes(x = x, y = y, fill = ngenes, color = ngenes))
# g + geom_raster() + scale_fill_viridis_c(option = "C") + 
#   theme_void() + 
#   ggtitle("mm10 Gene Density")
# 
# g <- ggplot(hmm %>% filter(ngenes != 0), aes(x = x, y = y, color = ngenes, size = ngenes))
# g + geom_point() + 
#   theme_void() + 
#   scale_size(range = c(1,3)) + 
#   scale_color_viridis_c(option = "A") + 
#   ggtitle("mm10 Gene Density")


```


```{r}
# Let's assign a column and row
# dat$x <- floor(dat$genomic_bin / grid_size)
# dat$y <- dat$genomic_bin %% grid_size
# 
# dat$timepoint <- sapply(dat$contrast, function(x) {
#   unlist(strsplit(unlist(strsplit(x, "oint"))[2], ".firr", fixed = T))[1]
# })
```



