---
title: "Firre knockout vs wildtype"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(seriation)
source("../util/_plot_theme.R")
source("../util/_util.R")
```

```{r load, include=FALSE}
load("../00_setup/results/rnaseq_data.RData")
```

```{r thresholds}
pval_thresh <- 0.05
l2fc_thresh <- log2(1.5)
```

# Static Firre knockout vs wildtype

The cell lines derived from the mouse have had a lot of time to equilibrate to the effect of the loss of the Firre locus. This is a true static snapshot of the state of the cell without Firre.

```{r wtko}
# NOTE: The short timecourse had WT and KO cell lines
# sequenced at different times. Therefore, it is 
# Better to use the results of the long timecourse
# to compare KO vs WT.
wtko_samples <- samples %>% 
  filter(cell_type == "ESC", 
         timepoint == 0, 
         timecourse_length == "long")

wtko_counts <- salmon_gene_counts[,wtko_samples$sample_id]

# Check the ordering
stopifnot(all(rownames(wtko_samples) == colnames(wtko_counts)))
stopifnot(all(rownames(wtko_counts) == genes$gene_id))

# DESeq2
wtko_dds <- DESeqDataSetFromMatrix(countData = wtko_counts, 
                                   colData = wtko_samples, 
                                   design = ~ firre_ko)
wtko_dds <- DESeq(wtko_dds)

# Results
wtko_res_lfc <- results(wtko_dds, name = "firre_ko_KO_vs_WT") %>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
  merge(g2s) 
  

wtko_res_shrnklfc <- lfcShrink(wtko_dds, coef = "firre_ko_KO_vs_WT", type = "apeglm") %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s)

wtko_res_shrnklfc <- wtko_res_shrnklfc %>%
  mutate(sig = padj <= pval_thresh & abs(log2FoldChange) > l2fc_thresh)

wtko_sig <- wtko_res_shrnklfc %>% 
  filter(sig == TRUE)
```

```{r wtko_plots}
ggplot(wtko_res_shrnklfc %>% filter(!is.na(padj)), 
       aes(x = padj)) +
  geom_histogram(bins = 30)

# Number of DEG
wtko_num_deg <- wtko_res_shrnklfc %>%
  filter(sig == TRUE) %>%
  mutate(direction = ifelse(log2FoldChange > 0, "up", "down")) %>%
  group_by(direction) %>%
  summarize(num_deg = n())

ggplot(wtko_res_shrnklfc %>% filter(!is.na(padj)), 
       aes(x = log2FoldChange, y = -log10(padj), color = sig)) +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  geom_hline(yintercept = -log10(pval_thresh), lty = 2) +
  geom_point(data = wtko_res_shrnklfc %>% filter(sig == TRUE), size = 1)  +
  geom_point(data = wtko_res_shrnklfc %>% filter(sig == FALSE, !is.na(padj)), size = 1, alpha = 0.7) +
  annotate("text", x = 2, y = 100, label = paste0("n = ", wtko_num_deg %>% filter(direction == "up") %>%
             pull(num_deg))) +
   annotate("text", x = -2, y = 100, label = paste0("n = ", wtko_num_deg %>% filter(direction == "down") %>%
             pull(num_deg))) +
  geom_text_repel(data = wtko_res_shrnklfc %>% filter(gene_name == "Firre"), aes(label = gene_name)) +
  theme(legend.position = "none") +
  xlim(-5,5) +
  ylim(0,200)
ggsave("figures/firre_ko_wt_volcano.pdf", height = 3, width = 3, useDingbats = FALSE)

ggplot(wtko_res_shrnklfc %>% filter(!is.na(padj)), 
       aes(x = log10(baseMean), y = log2FoldChange, color = sig)) +
  geom_hline(yintercept = 0) +
  geom_point(data = wtko_res_shrnklfc %>% filter(sig == TRUE), size = 1)  +
  geom_point(data = wtko_res_shrnklfc %>% filter(sig == FALSE, !is.na(padj)), size = 1, alpha = 0.7) +
  geom_text_repel(data = wtko_res_shrnklfc %>% filter(gene_name == "Firre"), aes(label = gene_name)) +
  theme(legend.position = "none")
ggsave("figures/firre_ko_wt_ma.pdf", height = 3, width = 3, useDingbats = FALSE)

```

# Firre induction rescue experiment

For this experiment, we induced the expression of Firre with an rTTA element by adding doxycycline to the cells. We see that Firre is indeed expressed in the KO background after the addition of doxycycline. The drug does instigate some gene expression changes on its own, so we will control for the effects by using a linear model which accounts for the effect of dox.

```{r long_timecourse}
# Filter to ESC KO long timecourse
ko_rescue_samples <- samples %>%
  filter(cell_type == "ESC",
         timecourse_length == "long",
         firre_ko == "KO")
ko_rescue_counts <- salmon_gene_counts[,ko_rescue_samples$sample_id]

# Check ordering
stopifnot(all(rownames(ko_rescue_samples) == colnames(ko_rescue_counts)))
stopifnot(all(rownames(ko_rescue_counts) == genes$gene_id))

# DESeq2 -- controlling for doxycycline; likelihood ratio test
ko_rescue_dds <- DESeqDataSetFromMatrix(countData = ko_rescue_counts, 
                                        colData = ko_rescue_samples, 
                                        design = ~ firre_induced + timepoint + timepoint*firre_induced)
ko_rescue_dds <- DESeq(ko_rescue_dds, test = "LRT", reduced = ~ firre_induced + timepoint)

# Compile results
res_names <- resultsNames(ko_rescue_dds)
dynamic_res <- res_names[grepl("firre_inducedfirre_induced.timepoint", res_names)]

ko_rescue_res <- lapply(dynamic_res, function(x) {
  results(ko_rescue_dds, 
          name = x) %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>% 
    merge(g2s) %>%
    mutate(result_name = x,
           timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
}) %>% bind_rows()

# Shrunken LFC results
ko_rescue_shrnklfc <- lapply(dynamic_res, function(x) {
  lfcShrink(ko_rescue_dds, 
          coef = x,
          type = "apeglm") %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>% 
    merge(g2s) %>%
    mutate(result_name = x,
           timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
}) %>% bind_rows()

# Calculate the maximum fold-change in any one timepoint
ko_rescue_maxfc <- ko_rescue_shrnklfc %>%
  group_by(gene_id) %>%
  summarize(max_fc = max(abs(log2FoldChange))) 
ko_rescue_shrnklfc <- ko_rescue_shrnklfc %>%
  left_join(ko_rescue_maxfc) %>%
  mutate(sig = padj <= pval_thresh & max_fc > l2fc_thresh)

ko_rescue_sig <- ko_rescue_shrnklfc %>% 
  filter(sig == TRUE)
```


```{r}
ko_rescue_only_samples <- samples %>%
  filter(cell_type == "ESC",
         timecourse_length == "long",
         firre_ko == "KO",
         firre_induced == "firre_induced")
ko_rescue_only_counts <- salmon_gene_counts[,ko_rescue_only_samples$sample_id]

# Check ordering
stopifnot(all(rownames(ko_rescue_only_samples) == colnames(ko_rescue_only_counts)))
stopifnot(all(rownames(ko_rescue_only_counts) == genes$gene_id))

# DESeq2 -- controlling for doxycycline; likelihood ratio test
ko_rescue_only_dds <- DESeqDataSetFromMatrix(countData = ko_rescue_only_counts,
                                        colData = ko_rescue_only_samples,
                                        design = ~ timepoint)
ko_rescue_only_dds <- DESeq(ko_rescue_only_dds)
res_names <- resultsNames(ko_rescue_only_dds)

vs_zero_res <- res_names[grepl("_vs_0", res_names)]
ko_rescue_only_vs_zero <- lapply(vs_zero_res, function(x) {
  lfcShrink(ko_rescue_only_dds, 
          coef = x,
          type = "apeglm") %>%
    as.data.frame() %>%
    rownames_to_column("gene_id") %>% 
    merge(g2s) %>%
    mutate(result_name = x,
           timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
}) %>% bind_rows()



# Let's figure out which genes are rescued
rescued_genes <- wtko_sig$gene_id[wtko_sig$gene_id %in% (ko_rescue_shrnklfc %>%
                                                           filter(padj <= pval_thresh) %>% pull(gene_id))]
# rescued_genes <- wtko_sig$gene_id[wtko_sig$gene_id %in% (ko_rescue_sig %>% pull(gene_id))]

rescue_wtko <- wtko_sig %>%
  filter(gene_id %in% rescued_genes) %>%
  dplyr::select(gene_id, gene_name, log2FoldChange) %>%
  dplyr::rename(l2fc_wtko = log2FoldChange)
rescue_ltc <- ko_rescue_only_vs_zero %>%
  filter(gene_id %in% rescued_genes) %>%
  dplyr::select(gene_id, gene_name, log2FoldChange, timepoint) %>%
  dplyr::rename(l2fc_ltc = log2FoldChange)

rescued_df <- rescue_wtko %>%
  left_join(rescue_ltc)

rescued_df <- rescued_df %>%
  mutate(dist = abs((l2fc_wtko * -1) - l2fc_ltc))

rescued_df_closest <- rescued_df %>%
  group_by(gene_id, gene_name) %>%
  mutate(closest = min(dist) == dist,
         max_fc = max(abs(l2fc_ltc)) == abs(l2fc_ltc)) %>%
  filter(closest == TRUE)

mod <- lm(l2fc_ltc ~ 1 + offset(-1*l2fc_wtko), data=rescued_df_closest)
summary(mod)
cooksd <- cooks.distance(mod)
plot(cooksd, pch="*", cex=2, main="Influential Obs by Cooks distance")
abline(h = 4*mean(cooksd, na.rm=T), col="red")  # add cutoff line
text(x=1:length(cooksd)+1, y=cooksd, labels=ifelse(cooksd>4*mean(cooksd, na.rm=T),names(cooksd),""), col="red")  # add labels

cooksd_cutoff <-0.5*mean(cooksd, na.rm=T)

rescued_df_closest$cooks_dist <- cooksd
rescued_df_closest$residual <- residuals(mod)
rescued_df_closest <- rescued_df_closest %>%
  mutate(cooks_cut = cooks_dist < cooksd_cutoff,
         resid_cut = abs(residual) < 0.5)

rescued_df_closest <-  rescued_df_closest %>%
  mutate(twenty_percent = l2fc_wtko * 0.2,
         rescued = (-1*l2fc_wtko - twenty_percent)>= l2fc_ltc & l2fc_ltc>= (-1*l2fc_wtko + twenty_percent) |
           (-1*l2fc_wtko + twenty_percent)>= l2fc_ltc & l2fc_ltc>= (-1*l2fc_wtko - twenty_percent) )

ggplot(rescued_df_closest, aes(x = l2fc_wtko, y = l2fc_ltc, color = rescued)) +
  geom_abline(slope = -1) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point(alpha = 0.7) +
  geom_abline(slope = -0.8, lty = 2) +
  geom_abline(slope = -1.2, lty = 2) +
  xlim(-2,2) +
  ylim(-2,2)
ggsave("figures/rescue_lfc_scatter.pdf", height = 3, width = 3, useDingbats = FALSE)

table(rescued_df_closest$rescued)
```



```{r}
# Heatmap of fold-changes for DEGs in the rescue
# Check that there are no duplicate row names.
stopifnot(all(length(unique(ko_rescue_sig$gene_id)) == length(unique(ko_rescue_sig$gene_name))))



ko_rescue_lfc <- ko_rescue_sig %>% 
  dplyr::select(gene_name, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()
# Add a zero column.
ko_rescue_lfc <- cbind(matrix(0, nrow = nrow(ko_rescue_lfc), ncol = 1), ko_rescue_lfc)
colnames(ko_rescue_lfc)[[1]] <- "0"
table(rescued_df_closest$rescued)
library(ComplexHeatmap)
library(circlize)
rescued_genes_20 <- rescued_df_closest %>% filter(rescued == TRUE) %>% pull(gene_name)
row_ha = rowAnnotation(sig_in_ko = as.numeric(rownames(ko_rescue_lfc) %in% wtko_sig$gene_name),
                       rescued = as.numeric(rownames(ko_rescue_lfc) %in% rescued_genes_20),
                       col = list(sig_in_ko = c("1" = "black", "0" = "white"),
                                  rescued = c("1" = "black", "0" = "white")))

table(as.numeric(rownames(ko_rescue_lfc) %in% wtko_sig$gene_name))
table(as.numeric(rownames(ko_rescue_lfc) %in% rescued_genes_20))
pdf("figures/ltc_rescue_heatmap_with_anno_allsig_nolfc.pdf", width = 3, height = 7)
Heatmap(ko_rescue_lfc, 
        name = "l2fc",
        cluster_columns = FALSE, show_row_names = FALSE, col = colorRamp2(seq(-1.5,1.5,length.out = 100), col_pal10),
        left_annotation = row_ha)
dev.off()
?Heatmap
pdf("figures/ltc_rescue_heatmap.pdf", width = 3, height = 7)
pheatmap::pheatmap(ko_rescue_lfc, cluster_cols = FALSE, 
                   color = col_pal10, border_color = NA,
                    breaks = seq(-3, 3, length.out = length(col_pal10)),
                   fontsize_row = 8)
dev.off()
```


# Firre rescued genes

We can visualize the expression levels of the genes that were significantly differentially expressed in the Firre knockout and dynamically changing in response to Firre's induction in a knockout background.

```{r rescue_lfc_plots}

```


```{r rescue_heatmap}
# Retrieve the set of genes that is DE in the KO and is significantly changing
# in the KO rescue experiment. 
# rescued_genes <- wtko_sig$gene_id[wtko_sig$gene_id %in% (ko_rescue_res %>% 
#                                                            filter(padj <= pval_thresh) %>% pull(gene_id))]

rescued_genes_20 <- rescued_df_closest %>% filter(rescued == TRUE) %>% pull(gene_id)
# Retrieve the TPM -- take the mean TPM in each condition.
wtko_matrix <- tpm %>% 
  filter(gene_id %in% rescued_genes_20) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(wtko_samples) %>%
  drop_na() %>%
  group_by(gene_name, firre_ko) %>%
  summarize(tpm = mean(tpm)) %>%
  pivot_wider(names_from = "firre_ko", 
              values_from = "tpm")

ko_rescue_matrix <- tpm %>% 
  filter(gene_id %in% rescued_genes_20) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(ko_rescue_samples) %>%
  drop_na() %>%
  filter(timepoint != 0) %>%
  group_by(gene_name, timepoint) %>%
  summarize(tpm = mean(tpm)) %>%
  pivot_wider(names_from = "timepoint", 
              values_from = "tpm")

comb_matrix <- wtko_matrix %>% 
  left_join(ko_rescue_matrix) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()

comb_matrix <- comb_matrix[,c("KO", "720", "1440", "2880", "5760", "WT")]

comb_matrix_scaled <- t(scale(t(comb_matrix)))
```

```{r rescue_heatmap}
pdf("figures/longtimecourse_rescue.pdf", width = 3.5, height = 5)
pheatmap::pheatmap(comb_matrix_scaled, cluster_cols = FALSE, cluster_rows = TRUE, show_rownames = TRUE,
                   color = col_pal10,
                   clustering_callback = olo_seriate,
                   clustering_distance_rows = "euclidean",
                   border_color = NA,
                   fontsize_row = 8,
                   breaks = seq(-1.5, 1.5, length.out = length(col_pal10)),
                   treeheight_row = 25)
dev.off()
```

```{r rescue_pca}
pca_dat <- prcomp(t(comb_matrix))
proportion_of_variance <- summary(pca_dat)$importance[2,1:2]

pca_df <- pca_dat$x[,1:2] %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  mutate(PC1 = PC1 * -1)

ggplot(pca_df, aes(x = PC1, y = PC2, label = sample)) +
  geom_point() +
  geom_text_repel() +
  xlab(paste0("PC1: ", round(proportion_of_variance["PC1"]*100), "%")) +
  ylab(paste0("PC2: ", round(proportion_of_variance["PC2"]*100), "%"))
ggsave("figures/longtimecourse_pca.pdf", 
       width = 3.5, height = 0.9, 
       useDingbats = FALSE)
```

# Firre's induction

We would like to check that Firre's expression is actually induced by the addition of
doxycycline. 

```{r firre_profile}
firre_ko_rescue <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", firre_ko == "KO", timecourse_length == "long")

firre_means <- firre_ko_rescue %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_ko_rescue, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/firre_esc_ko_profile.pdf", height = 3, width = 3, useDingbats = FALSE)

```
