---
title: "tf_dox_cascade"
author: "Michael Smallegan"
date: "9/24/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Purpose

The goal here is to understand the cascade of transcription factor activation that follows from the addition of doxycycline. There could be direct results of dox hitting TFs, there could be signalling changes that result in changes to the transcriptome, and there could be changes that are a response of having a lot more RNA turned on with a quickly transcribing promoter at the site of the transgene. 

## Tasks
- get the dox timecourse shrunken lfcs. 
- take just the significant gene lists.
- run rcis target 
- connect the results for each timepoint to the targets in the next timepoint.
- What percentage of changes in the next timepoint can be potentially explained by the change in the level of a TF in the previous timepoint?



```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
library(RcisTarget)
```

```{r}
# read in the dox timepoint data.
tc <- read.csv("../timecourse_shrunken_lfc_results.csv") %>%
  filter(grepl("dox", contrast))
tcsig <- tc %>% filter(padj < 0.1 & abs(log2FoldChange) > 0.1)

gl <- list()

contrasts <- unique(tc$contrast)
for(i in 1:length(contrasts)) {
  tdf <- tcsig[which(tcsig$contrast == contrasts[[i]]), c("gene_name", "gene_id")]
  tdf <- tdf %>% as_tibble()
  gl <- c(gl, tdf[which(tdf$gene_name != ""),"gene_name"])
  names(gl)[length(gl)] <- contrasts[[i]]
}

```


```{r}
# now let's run the motif enrichment
motifRankings <- importRankings("../mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.feather")

# Motif enrichment analysis:
motif_enrichment <- cisTarget(gl, motifRankings, motifAnnot = motifAnnotations_mgi)
```

```{r}
# let's see if any of the high conf TFs are even in our data
tfs <- motif_enrichment$TF_highConf
tfs <- tfs[which(tfs != "")]
tfs <- sapply(tfs, function(x) {
  unlist(strsplit(x, " \\("))[[1]]
})

tfs <- sapply(tfs, function(x) {
  strsplit(x, "; ")
})

tfs <- unlist(tfs)
tfs <- unique(tfs)

difex <- unique(unlist(gl))

length(which(tfs %in% difex))
# okay great, so this will be worthwile. 
# next wee need a list of all the TFs and which associated timepoint they're enriched in.
# We'll start with just the high confidence TFs.
```

```{r}
tfsig <- motif_enrichment %>% select(geneSet, TF_highConf) %>%
  filter(TF_highConf != "")
tfsig$TF_highConf <- sapply(tfsig$TF_highConf, function(x) {
  unlist(strsplit(x, " \\("))[[1]]
})

tfsig <- tfsig %>% mutate(TF_highConf = strsplit(TF_highConf, "; ")) %>%
  unnest(TF_highConf)
tfsig <- tfsig %>% distinct()


# let's now add in the lower confidence
tflc <- motif_enrichment %>% select(geneSet, TF_lowConf) %>%
  filter(TF_lowConf  != "")
tflc <- tflc %>% mutate(TF_lowConf = strsplit(TF_lowConf, "; ")) %>%
  unnest(TF_lowConf)
tflc <- tflc %>% mutate(TF_lowConf = strsplit(TF_lowConf, ".", fixed = T)) %>%
  unnest(TF_lowConf)
tflc$TF_lowConf <- sapply(tflc$TF_lowConf, function(x) {
  unlist(strsplit(x, " \\("))[[1]]
})

tflc <- tflc  %>% distinct()

names(tfsig)[2] <- "tf"
names(tflc)[2] <- "tf"

tf_enriched <- bind_rows(tfsig, tflc) %>% distinct()

tps <- seq(from = 30, to = 330, by =30)
offsets <- c(0,30,60)
comparison_list <- list()
for(i in 1:length(tps)) {
  for(j in 1:length(offsets)) {
    comparison_list <- c(comparison_list, list(c(tps[[i]], tps[[i]]+offsets[[j]])))
  }
}
# let's remove those timepoints that don't exist

valid_timepoints <- sapply(comparison_list, function(x) {
  all(unlist(x) <= 330)
})
comparison_list <- comparison_list[valid_timepoints]
 


# okay, this is going to be offset by one timepopint

tfgs_list <- list()
for(i in 1:length(comparison_list)) {
  # let's do ko_dox first
  # we want the motifs enriched at t+30
  tf <- tf_enriched %>% filter(geneSet == paste0("ko_dox_shrunken/timepoint_", comparison_list[[i]][[2]],"_vs_0"))
  gs <- tcsig %>% filter(contrast == paste0("ko_dox_shrunken/timepoint_", comparison_list[[i]][[1]],"_vs_0"))
  # 
  tfgs_list <- c(tfgs_list, list(tf$tf[which(tf$tf %in% gs$gene_name)]))
  names(tfgs_list)[length(tfgs_list)] <- paste0("ko_dox|tf", comparison_list[[i]][[2]],"|deg", comparison_list[[i]][[1]])
  
    tf <- tf_enriched %>% filter(geneSet == paste0("wt_dox_shrunken/timepoint_", comparison_list[[i]][[2]],"_vs_0"))
  gs <- tcsig %>% filter(contrast == paste0("wt_dox_shrunken/timepoint_", comparison_list[[i]][[1]],"_vs_0"))
  # 
  tfgs_list <- c(tfgs_list, list(gs$gene_name[which(gs$gene_name %in% tf$tf)]))
  names(tfgs_list)[length(tfgs_list)] <- paste0("wt_dox|tf", comparison_list[[i]][[2]],"|deg", comparison_list[[i]][[1]])
}


# let's turn this into a data frame
tfdf <- tibble("contrast" = names(tfgs_list), "tf" = tfgs_list) %>%
  unnest(tf) %>% 
  separate(contrast, into = c("condition", "tf_tp", "deg_tp"), sep = "\\|")

tfdf$tf_tp <- as.numeric(gsub("tf", "", tfdf$tf_tp))
tfdf$deg_tp <- as.numeric(gsub("deg", "", tfdf$deg_tp))

tfdf$offset <- tfdf$tf_tp - tfdf$deg_tp


nrow(tfdf)
nrow(tfdf %>% distinct())
# okay, let's group by offset and then see what the overlap is for each.
offset_summary <- tfdf %>% group_by(condition, offset) %>%
  summarize(count = n())

# zero offset
wttfs <- tfdf[which(tfdf$condition == "wt_dox" & tfdf$offset == 0), "tf"]
wttfs <- unique(wttfs$tf)
kotfs <- tfdf[which(tfdf$condition == "ko_dox" & tfdf$offset == 0), "tf"]
kotfs <- unique(kotfs$tf)
length(which(wttfs %in% kotfs))
zero_off <- wttfs[which(wttfs %in% kotfs)]

# 9 tfs overlap

wttfs <- tfdf[which(tfdf$condition == "wt_dox" & tfdf$offset == 30), "tf"]
wttfs <- unique(wttfs$tf)
kotfs <- tfdf[which(tfdf$condition == "ko_dox" & tfdf$offset == 30), "tf"]
kotfs <- unique(kotfs$tf)
length(which(wttfs %in% kotfs))
thirty_off <- wttfs[which(wttfs %in% kotfs)]

# 7 tfs overlap

wttfs <- tfdf[which(tfdf$condition == "wt_dox" & tfdf$offset == 60), "tf"]
wttfs <- unique(wttfs$tf)
kotfs <- tfdf[which(tfdf$condition == "ko_dox" & tfdf$offset == 60), "tf"]
kotfs <- unique(kotfs$tf)
length(which(wttfs %in% kotfs))
sixty_off <- wttfs[which(wttfs %in% kotfs)]

# 8 tfs overlap. 




```

Okay, well since it's just (0 offset = 9 overlaps, 30 min offset = 7 overlaps, and 60 min offset is 8 overlaps), for now I'm going to move forward with the 30 minutes for simplicity.
But a good way to address this, might be which offsets are most explanatory of the changes in the data?

I would also like to check whether these correspond to different motifs or the same.


```{r}
# let's make a map of motifs to tfs
# mt2tf <- 
hc_mt2tf <- motif_enrichment %>% select(motif, TF_highConf) %>%
  filter(TF_highConf != "")
hc_mt2tf$TF_highConf <- sapply(hc_mt2tf$TF_highConf, function(x) {
  unlist(strsplit(x, " \\("))[[1]]
})
hc_mt2tf <- hc_mt2tf %>% mutate(TF_highConf = strsplit(TF_highConf, "; ")) %>%
  unnest(TF_highConf)
hc_mt2tf <- hc_mt2tf %>% distinct()
names(hc_mt2tf)[2] <- "tf"


lc_mt2tf <- motif_enrichment %>% select(motif, TF_lowConf) %>%
  filter(TF_lowConf  != "")
lc_mt2tf <- lc_mt2tf %>% mutate(TF_lowConf = strsplit(TF_lowConf, "; ")) %>%
  unnest(TF_lowConf)
lc_mt2tf <- lc_mt2tf %>% mutate(TF_lowConf = strsplit(TF_lowConf, ".", fixed = T)) %>%
  unnest(TF_lowConf)
lc_mt2tf$TF_lowConf <- sapply(lc_mt2tf$TF_lowConf, function(x) {
  unlist(strsplit(x, " \\("))[[1]]
})

lc_mt2tf <- lc_mt2tf  %>% distinct()
names(lc_mt2tf)[2] <- "tf"

mt2tf <- bind_rows(hc_mt2tf, lc_mt2tf)
mt2tf <- mt2tf %>% distinct()

tfs_of_interest <- unique(c(zero_off, thirty_off, sixty_off))
motifs_of_i <- mt2tf[which(mt2tf$tf %in% tfs_of_interest),]
sum_motifs <- motifs_of_i %>% group_by(motif) %>% summarize(count = length(unique(tf)))

table(sum_motifs$count)
```

Hmm, there are a few motifs that have six of those TFs listed. I wonder if it's that motif that is predominantly driving these results. 

```{r}
# let's now look at the overlap of target genes. 
hc_tf2tg <- motif_enrichment %>% select(TF_highConf, enrichedGenes) %>%
  filter(TF_highConf != "")
hc_tf2tg$TF_highConf <- sapply(hc_tf2tg$TF_highConf, function(x) {
  unlist(strsplit(x, " \\("))[[1]]
})
hc_tf2tg <- hc_tf2tg %>% mutate(TF_highConf = strsplit(TF_highConf, "; ")) %>%
  unnest(TF_highConf)
hc_tf2tg <- hc_tf2tg %>% distinct()
names(hc_tf2tg)[2] <- "tf"


lc_tf2tg <- motif_enrichment %>% select(TF_lowConf, enrichedGenes) %>%
  filter(TF_lowConf  != "")
lc_tf2tg <- lc_tf2tg %>% mutate(TF_lowConf = strsplit(TF_lowConf, "; ")) %>%
  unnest(TF_lowConf)
lc_tf2tg <- lc_tf2tg %>% mutate(TF_lowConf = strsplit(TF_lowConf, ".", fixed = T)) %>%
  unnest(TF_lowConf)
lc_tf2tg$TF_lowConf <- sapply(lc_tf2tg$TF_lowConf, function(x) {
  unlist(strsplit(x, " \\("))[[1]]
})

lc_tf2tg <- lc_tf2tg  %>% distinct()
names(lc_tf2tg)[2] <- "tf"


tf2tg <- bind_rows(hc_tf2tg, lc_tf2tg)

tf2tg <- tf2tg %>% mutate(enrichedGenes = strsplit(enrichedGenes, ";")) %>%
  unnest(enrichedGenes)


tf2tg <- tf2tg %>% distinct()
# save(tf2tg, file = "tf2tg.Rds")
```

```{r}
# now let's go entry by entry and see which genes are changing as a result of the tf changing.
tfdf[1,]
for(i in 1:nrow(tfdf)) {
  degs <- tcsig %>% filter(grepl(paste0(tfdf[i,"condition"], "_shrunken/timepoint_", tfdf[i,"deg_tp"], "_vs_0"), contrast))
  tgs <- tf2tg[which(tf2tg$tf == tfdf[[i,"tf"]]),"enrichedGenes"]
  tfdf[i,"target_genes_diffex"] <- list(list(tgs[which(tgs %in% degs$gene_name)]))
}

tfdf$target_count <- sapply(tfdf$target_genes_diffex, length)


tfdfl <- tfdf %>% unnest(target_genes_diffex)

toi <- tfdfl %>% filter(tf %in% tfs_of_interest)
toi <- toi %>% filter(target_genes_diffex %in% tfs_of_interest)


#### Hmm, I think now I just need to plot these genes.
```


