---
title: "PRO-seq"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(Rsubread)
library(DESeq2)
library(ggpubr)
source("../util/_util.R")
source("../util/_plot_theme.R")
```

## Nascent transcription upon Firre induction

PRO-seq libraries were sequenced on a HiSeq 4000 using paired-end 150bp reads to an average read depth of 120M reads. Mapping and QC were handled by the NascentFlow Nextflow pipeline (DOI 10.17605/OSF.IO/NDHJ2). Reads were mapped to mm10 using HISAT2

```{r}
# Create SAF file for featurecounts

# 
# chr_sizes <- read_tsv("/scratch/Shares/rinn/genomes/Mus_musculus/hisat2/mm10_chrom_sizes.txt",
#                       col_names = c("Chr", "chr_length"))
mumerge_bed <- read_tsv("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/mumerge/split_bidir_predictions_mumerge_MUMERGE.bed",
                        col_names = c("Chr", "Start", "End"),
                        col_types = list(Chr = col_character(), Start = col_integer(), End = col_integer())) %>%
  mutate(GeneID = paste0("bidir_", 1:nrow(.)), .before = 1)%>%
  mutate(Strand = ".")

write.table(mumerge_bed, "results/mumerge_bidirs.saf", 
            quote = FALSE, sep = "\t",
            col.names = TRUE, row.names = FALSE)

mumerge_bed <- mumerge_bed %>%
  mutate(width = End - Start)

ggdensity(mumerge_bed, "width", add = "mean") +
  stat_central_tendency(type = "median")

median(mumerge_bed$width)
mean(mumerge_bed$width)  
min(mumerge_bed$width, na.rm = T)
# Find the center and make +/- 300bp up and down
mumerge_uniform_saf <- mumerge_bed %>%
  mutate(center = Start + round(width/2),
         strt = center - 300,
         end = center + 300,
         new_width = end - strt) %>%
  dplyr::select(GeneID, Chr, strt, end, Strand) %>%
  dplyr::rename(Start = strt,
                End = end)

write_tsv(mumerge_uniform_saf, "results/mumerge_bidirs_600bp.saf")


mumerge_bed_positive_strand <- mumerge_bed %>%
  mutate(Strand = "+")
write.table(mumerge_bed_positive_strand, "results/mumerge_bidirs_positive_strand.saf", 
            quote = FALSE, sep = "\t",
            col.names = TRUE, row.names = FALSE)

mumerge_bed_negative_strand <- mumerge_bed %>%
  mutate(Strand = "-")
write.table(mumerge_bed_negative_strand, "results/mumerge_bidirs_negative_strand.saf", 
            quote = FALSE, sep = "\t",
            col.names = TRUE, row.names = FALSE)

bidir_coords <- mumerge_bed %>%
  dplyr::rename(region_id = GeneID,
                chrom = Chr,
                start = Start,
                end = End) %>%
  dplyr::select(-Strand) %>%
  dplyr::select(chrom, start, end, region_id)
write_tsv(bidir_coords, "results/bidir_coords.bed", col_names = F)


library(Rsubread)
library(tidyverse)
fl <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
                 pattern = "*.bam$",
                 full.names = TRUE)
bidir_counts <- featureCounts(fl, 
                              annot.ext = "/scratch/Shares/rinn/Michael/firre_timecourse/analysis/14_proseq/results/mumerge_bidirs.saf",
                              isPairedEnd = TRUE,
                              nthreads = 22)
write_rds(bidir_counts, "../results/bidir_counts_pe.rds")

bidir_counts <- read_rds("results/bidir_counts.rds")


# Positive strand read counts
library(Rsubread)
library(tidyverse)
fl <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
                 pattern = "*.bam$",
                 full.names = TRUE)
bidir_counts <- featureCounts(fl, 
                              annot.ext = "/scratch/Shares/rinn/Michael/firre_timecourse/analysis/14_proseq/results/mumerge_bidirs_positive_strand.saf",
                              isPairedEnd = TRUE,
                              strandSpecific = 1,
                              nthreads = 22)
write_rds(bidir_counts, "../results/bidir_counts_positive_strand.rds")

# Negative strand read counts
library(Rsubread)
library(tidyverse)
fl <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
                 pattern = "*.bam$",
                 full.names = TRUE)
bidir_counts <- featureCounts(fl, 
                              annot.ext = "/scratch/Shares/rinn/Michael/firre_timecourse/analysis/14_proseq/results/mumerge_bidirs_negative_strand.saf",
                              isPairedEnd = TRUE,
                              strandSpecific = 1,
                              nthreads = 18)
write_rds(bidir_counts, "../results/bidir_counts_negative_strand.rds")


bidir_counts <- read_rds("results/bidir_counts.rds")
# srun --pty -p short --ntasks 22 --time=12:00:00 --mem 16G /bin/bash 
library(Rsubread)
library(tidyverse)
fl <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
                 pattern = "*.bam$",
                 full.names = TRUE)
bidir_counts <- featureCounts(fl, 
                              annot.ext = "/scratch/Shares/rinn/Michael/firre_timecourse/analysis/14_proseq/results/mumerge_bidirs_600bp.saf",
                              isPairedEnd = TRUE,
                              strandSpecific = 0,
                              nthreads = 21,
                              tmpDir = "/tmp")
write_rds(bidir_counts, "/scratch/Shares/rinn/Michael/firre_timecourse/analysis/14_proseq/results/bidir_counts_600bp_unstranded.rds")
```

## ATAC-seq peak quantification

1. What's happening in the 55 differential ATAC peaks?

```{r}


# Read in the differential ATAC-seq peaks 
atac_up <- read_tsv("../12_atacseq/results/atac_up.bed", col_names = c("Chr", "Start", "End")) %>%
  mutate(GeneID = paste0("up_", 1:n()), .before = 1)
atac_down <- read_tsv("../12_atacseq/results/atac_down.bed", col_names = c("Chr", "Start", "End")) %>%
  mutate(GeneID = paste0("down_", 1:n()), .before = 1)

de_atac <- bind_rows(atac_up,
                     atac_down) %>%
  mutate(Strand = ".")

write_tsv(de_atac, "results/de_atac.saf")
?featureCounts
library(Rsubread)
library(tidyverse)
fl <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
                 pattern = "*.bam$",
                 full.names = TRUE)
bidir_counts <- featureCounts(fl, 
                              annot.ext = "results/de_atac.saf",
                              isPairedEnd = TRUE,
                              strandSpecific = 0,
                              nthreads = 9,
                              tmpDir = "/tmp")
write_rds(bidir_counts, "results/de_atac_proseq_counts.rds")

```

2. Let's count nascent reads over all ATAC-seq peaks

```{r}
load("../01_setup/results/atacseq_data.RData", verbose = T)
atac_saf <- atac_consensus %>%
  dplyr::select(1:5) %>%
  dplyr::rename(GeneID = interval_id)
write_tsv(atac_saf, "results/atac.saf")

library(Rsubread)
library(tidyverse)
fl <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
                 pattern = "*.bam$",
                 full.names = TRUE)
bidir_counts <- featureCounts(fl, 
                              annot.ext = "results/atac.saf",
                              isPairedEnd = TRUE,
                              strandSpecific = 0,
                              nthreads = 9,
                              tmpDir = "/tmp")
write_rds(bidir_counts, "results/atac_proseq_counts.rds")
```

```{r}
hmm <- read_rds("results/atac_proseq_counts.rds")
colnames(hmm$counts)
```


### Differential expression of nascent transcripts -- unstranded

```{r}
bidir_counts <- read_rds("results/bidir_counts_600bp_unstranded.rds")

# Let's normalize based on the number of mapped reads in each sample.
# We don't need to normalize based on length since each region was 600bp wide.
mapped_list <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/qc/mapstats",
                          full.names = T,
                          pattern = "*.millionsmapped") %>%
  sapply(read_lines) %>%
  as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  mutate(sample_id = gsub("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/qc/mapstats/|.millionsmapped", "", sample_id)) %>%
  dplyr::rename(reads_mapped = ".") %>%
  mutate(reads_mapped = as.numeric(reads_mapped))
  

bidir_counts_l <- bidir_counts$counts %>%
    as.data.frame() %>%
  rownames_to_column("bidir_id") %>%
  pivot_longer(2:ncol(.), names_to = "sample_id", values_to = "count") %>%
  mutate(sample_id = gsub(".sorted.bam", "", sample_id))

bidir_counts_l <- bidir_counts_l %>%
  left_join(mapped_list)


bidir_counts_l <- bidir_counts_l %>%
  mutate(CPM = (count / reads_mapped) *1e6)

sinfo <- read_tsv("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/mumerge/sample_info.txt") %>%
  dplyr::rename(sample_id = sampid) %>%
  dplyr::select(sample_id, group) %>%
    mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                             TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))
bidir_counts_l <- bidir_counts_l %>%
  left_join(sinfo)


bidir_30_vs_0 <- bidir_counts_l %>%
  dplyr::filter(timepoint %in% c("0","30"))

library(rstatix)
library(tidyverse)
bidir_30_vs_0_ttest <- bidir_30_vs_0 %>%
  group_by(bidir_id) %>%
  t_test(data = ., CPM ~ timepoint) %>%
    adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj")



bidir_15_vs_0 <- bidir_counts_l %>%
  dplyr::filter(timepoint %in% c("0","15"))
library(rstatix)
library(tidyverse)
bidir_15_vs_0_ttest <- bidir_15_vs_0 %>%
  group_by(bidir_id) %>%
  t_test(data = ., CPM ~ timepoint) %>%
    adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj")

bidir_62 <-  bidir_30_vs_0 %>%
  filter(bidir_id == "bidir_62")

bidir_15_vs_0 <- bidir_counts_l %>%
  dplyr::filter(timepoint %in% c("0","15"))

```


### Strand-specific differential expression of nascent transcripts

It would be interesting to see the best changing bidirectional nascent transcripts. Therefore, what we'll do here is run DEseq on the +/- strand counts separately. Then look at the correlation and choose those that are the most correlated with the highest fold-change.

```{r}
sinfo <- read_tsv("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/mumerge/sample_info.txt")
counts_matrix <- bidir_counts$counts
samples <- tibble(file = colnames(counts_matrix)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                             TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))
```

```{r}
# Positive strand counts 
pcounts <- read_rds("results/bidir_counts_positive_strand.rds")
pcounts_matrix <- pcounts$counts

rownames(pcounts_matrix)
samples <- tibble(file = colnames(pcounts_matrix)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                             TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))

pdds <- DESeqDataSetFromMatrix(countData = pcounts_matrix,
                              colData = samples,
                              design = ~ timepoint)
# Each timepoint vs zero
pdds <- DESeq(pdds)
# write_rds(dds, "results/deseq_dds_vs_zero.rds")
pres_vs_zero <- bind_rows(results(pdds, name = "timepoint_30_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_30_vs_0"),
  results(pdds, name = "timepoint_15_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_15_vs_0")) %>%
  dplyr::rename(region_id = row) %>%
  left_join(bidir_coords) %>%
  mutate(strand = "+")
write_csv(pres_vs_zero, "results/res_vs_zero_positive_strand.csv")

# Negative strand counts
ncounts <- read_rds("results/bidir_counts_negative_strand.rds")
ncounts_matrix <- ncounts$counts
samples <- tibble(file = colnames(ncounts_matrix)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                             TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))

ndds <- DESeqDataSetFromMatrix(countData = ncounts_matrix,
                              colData = samples,
                              design = ~ timepoint)
# Each timepoint vs zero
ndds <- DESeq(ndds)
# write_rds(dds, "results/deseq_dds_vs_zero.rds")
nres_vs_zero <- bind_rows(results(ndds, name = "timepoint_30_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_30_vs_0"),
  results(ndds, name = "timepoint_15_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_15_vs_0")) %>%
  dplyr::rename(region_id = row) %>%
  left_join(bidir_coords) %>%
  mutate(strand = "-")
write_csv(nres_vs_zero, "results/res_vs_zero_negative_strand.csv")

pres_vs_zero <- read_csv("results/res_vs_zero_positive_strand.csv")
nres_vs_zero <- read_csv("results/res_vs_zero_negative_strand.csv")
# Combine into one data.frame
pres_vs_zero <- pres_vs_zero %>%
  dplyr::rename(l2fc_pos = log2FoldChange,
                padj_pos = padj) %>%
  dplyr::select(region_id, comparison, chrom, start, end, l2fc_pos, padj_pos)

nres_vs_zero <- nres_vs_zero %>%
  dplyr::rename(l2fc_neg = log2FoldChange,
                padj_neg = padj) %>%
  dplyr::select(region_id, comparison, l2fc_neg, padj_neg)

combined_strands_vs_zero <- pres_vs_zero %>%
  left_join(nres_vs_zero)

combined_strands_vs_zero <- combined_strands_vs_zero %>%
  mutate(sig_in_both = (padj_pos < 0.05 & padj_neg < 0.05))
table(combined_strands_vs_zero$sig_in_both)
ggplot(combined_strands_vs_zero, aes(x = l2fc_pos, y = l2fc_neg, color = sig_in_both)) +
  facet_wrap(~comparison) +
  geom_point()


ggplot(combined_strands_vs_zero %>%
         filter(sig_in_both == TRUE), aes(x = l2fc_pos, y = l2fc_neg, color = sig_in_both)) +
  facet_wrap(~comparison) +
  geom_point()


hmm <- combined_strands_vs_zero %>%
  filter(sig_in_both == TRUE,
         l2fc_pos > 1, l2fc_neg > 1)
```

\

```{r}



dds <- DESeqDataSetFromMatrix(countData = counts_matrix,
                              colData = samples,
                              design = ~ timepoint)
# Each timepoint vs zero
dds <- DESeq(dds)
# write_rds(dds, "results/deseq_dds_vs_zero.rds")
res_vs_zero <- bind_rows(results(dds, name = "timepoint_30_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_30_vs_0"),
  results(dds, name = "timepoint_15_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_15_vs_0")) %>%
  dplyr::rename(region_id = row) %>%
  left_join(bidir_coords)
write_csv(res_vs_zero, "results/res_vs_zero.csv")

ggplot(res_vs_zero, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point() +
  facet_wrap(~comparison)

# LRT 
res_lrt <- DESeq(dds, test = "LRT", reduced = ~1)
write_rds(res_lrt, "results/deseq_dds_vs_zero_lrt.rds")

res_lrt_res <- bind_rows(results(res_lrt, name = "timepoint_30_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_30_vs_0_lrt")) %>%
  dplyr::rename(region_id = row) %>%
  left_join(bidir_coords)

sig_lrt_res <- res_lrt_res %>%
  filter(padj < 0.05) %>%
  arrange(padj)
```

## Gene level differential expression

```{r}


fl_unstranded_gene_counts <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/bidirflow/featurecounts_genes", full.names = T, pattern = "*unstranded*")
# Let's just get those with the 5' part of the gene removed
fl_unstranded_gene_counts <- fl_unstranded_gene_counts[grepl("5ptrunc_gene_counts.txt", fl_unstranded_gene_counts)]

ugc <- lapply(fl_unstranded_gene_counts,function(x)
  read_tsv(x)[,4])
ugc <- do.call(cbind, ugc)

# The gene ID order is the same in all the count files. 
# However, there are duplicated gene names and transcript ids.
# ugc_gene_ids <- lapply(fl_unstranded_gene_counts, function(x)
#   read_tsv(x)[,2])
# ugc_gene_ids <- do.call(cbind, ugc_gene_ids)
ugc_gid <- read_tsv(fl_unstranded_gene_counts[[1]]) %>%
  mutate(index = 1:nrow(.)) %>%
  unite(col = "id", GeneID, TranscriptID, index)
# ugc_gid <- ugc_gid$id

rownames(ugc) <- ugc_gid$id


# Calc TPM
ugc_tpm <- counts_to_tpm(ugc, ugc_gid$Length, rep(250,9))

tpm_5ptrunc <- ugc_tpm %>%
  as.data.frame() %>%
  rownames_to_column("id") %>%
  pivot_longer(2:ncol(.), names_to="sampid", values_to = "tpm") %>%
  mutate(sampid = gsub(".sorted.bam", "", sampid)) %>%
  left_join(samples) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
write_csv(tpm_5ptrunc, "results/tpm_5ptrunc.csv")
tpm_5ptrunc <- read_csv("results/tpm_5ptrunc.csv")

ggplot(tpm_5ptrunc %>% filter(grepl("Duox1", id)),
       aes(x = timepoint, y = tpm)) +
  geom_point() +
  geom_smooth(method = "lm")

library(ggpubr)
ggplot(tpm_5ptrunc %>% filter(grepl("Firre", id)),
       aes(x = timepoint, y = tpm)) +
 geom_hline(yintercept = 0) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "black") +
  ggtitle("Firre expression") +
  ylab("TPM") +
  xlab("Time (Min.)") +
  stat_regline_equation(size = 8)
 
sinfo <- read_tsv("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/mumerge/sample_info.txt")
samples <- tibble(file = colnames(ugc)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                             TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))

dds <- DESeqDataSetFromMatrix(countData = ugc,
                              colData = samples,
                              design = ~ timepoint)

dds <- DESeq(dds)

res_vs_zero <- bind_rows(results(dds, name = "timepoint_30_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_30_vs_0"),
  results(dds, name = "timepoint_15_vs_0", tidy = T) %>%
  mutate(comparison = "timepoint_15_vs_0")) %>%
  dplyr::rename(gene_id = row)
write_csv(res_vs_zero, "results/res_vs_zero_gene5ptrunc.csv")


res_vs_zero <- res_vs_zero %>%
  rowwise() %>%
  mutate(gene_name = unlist(strsplit(gene_id, "_"))[[1]])

res_vs_zero <- res_vs_zero %>%
  mutate(gene_name = gsub("Gm35612", "CrossFirre", gene_name))
library(ggrepel)
ggplot(res_vs_zero, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point() + 
  facet_wrap(~ comparison) +
  geom_text_repel(data = res_vs_zero %>% filter(log2FoldChange > 4 | -log10(padj) > 75),
                  aes(label = gene_name))

gene_sig <- res_vs_zero %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1)

gs_summary <- gene_sig %>%
  mutate(deg = ifelse(log2FoldChange > 0, 1, -1)) %>%
  group_by(comparison, deg) %>%
  summarize(ndeg = n())

table(gene_sig$comparison)


# Overlaps with previous Firre results
load("../11_short_timecourse_combined/results/short_vszero_sig.RData", verbose = T)
svz_sig_genes <- unique(short_vszero_sig$gene_name)
length(svz_sig_genes)
length(which(svz_sig_genes %in% gene_sig$gene_name))
cat(unique(gene_sig$gene_name[which(gene_sig$gene_name %in% svz_sig_genes)]))


load("../03_doxycycline_effect/results/esc_ko_control_short.RData", verbose = T)
dox_sig <- esc_ko_control_short %>%
  filter(padj < 0.05) %>%
  group_by(gene_id, gene_name) %>%
  summarize(max_lfc = max(log2FoldChange)) %>%
  filter(abs(max_lfc) > log2(1.5))


dox_sig_genes <- dox_sig$gene_name
length(which(dox_sig_genes %in% gene_sig$gene_name))
```
