---
title: "Nascent transcription upon *Firre* induction"
author: "Michael Smallegan"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_notebook: 
    css: /scratch/Shares/rinn/Michael/firre_timecourse/analysis/util/style.css
    theme: simplex
    toc: yes
    fig_caption: yes
    number_sections: yes
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(Rsubread)
library(DESeq2)
library(ggpubr)
library(broom)
source("../util/_util.R")
source("../util/_plot_theme.R")

proseq_dir <- "/scratch/Shares/rinn/Michael/firre_timecourse/proseq"
genome_dir <- "/scratch/Shares/rinn/genomes/Mus_musculus/hisat2"
```

# Pipeline

PRO-seq libraries were sequenced on a HiSeq 4000 using paired-end 150bp reads to an average read depth of 120M reads. Mapping and QC were handled by the [NascentFlow Nextflow pipeline](https://github.com/Dowell-Lab/Nascent-Flow) (DOI 10.17605/OSF.IO/NDHJ2). Reads were mapped to mm10 using HISAT2.

Bidirectional transcripts were called with Tfit and Dreg using the [Bidirectional-Flow Nextflow pipeline](https://github.com/Dowell-Lab/Bidirectional-Flow). Annotations of bidirectional transcripts were merged across all nine samples using [muMerge](https://github.com/Dowell-Lab/mumerge) (PMID: 34079046). Read counts across consensus bidirectional transcripts are calculated here using featureCounts.

# Read counts over bidirectional transcripts

There are two methods for selecting consensus regions to quantify over: 1.) Take the muMerge regions directly. 2.) Take regions that are a uniform size +/- 300bp from the muMerge center.

The reason for the two approaches is that muMerge regions represent the confidence interval in which you're likely to have the center of the bidirectional transcript. Therefore, this region shrinks when you add more samples & have higher confidence.

## Method 1: Quantify directly over muMerge outputted regions

```{r create-mumerge-saf, warning=FALSE, message=FALSE, include=FALSE}
# Create SAF file for featurecounts
if(!(file.exists("results/mumerge_bidirs.saf") &
     file.exists("results/bidir_coords.bed"))) {
  
  chr_sizes <- read_tsv(file.path(genome_dir, "mm10_chrom_sizes.txt"),
                        col_names = c("Chr", "chr_length"))
  
  mumerge_bed <- read_tsv(file.path(proseq_dir, 
                                    "mumerge/split_bidir_predictions_mumerge_MUMERGE.bed"),
                          col_names = c("Chr", "Start", "End"),
                          col_types = list(Chr = col_character(), 
                                           Start = col_integer(), 
                                           End = col_integer())) %>%
    mutate(GeneID = paste0("bidir_", 1:nrow(.)), .before = 1) %>%
    mutate(Strand = ".")
  
  write.table(mumerge_bed, "results/mumerge_bidirs.saf", 
              quote = FALSE, sep = "\t",
              col.names = TRUE, row.names = FALSE)
  
  bidir_coords <- mumerge_bed %>%
    dplyr::rename(region_id = GeneID,
                  chrom = Chr,
                  start = Start,
                  end = End) %>%
    dplyr::select(-Strand) %>%
    dplyr::select(chrom, start, end, region_id)
  
  write_tsv(bidir_coords, "results/bidir_coords.bed", col_names = F)
}

bidir_coords <- read_tsv("results/bidir_coords.bed",
                         col_names = c("chrom", "start", "end", "bidir_id"),
                         col_types = list("chrom" = col_character(),
                                          "start" = col_integer(),
                                          "end" = col_integer(),
                                          "bidir_id" = col_character()))
```

```{r import-mumerge-saf}
mumerge_bed <- read_tsv("results/mumerge_bidirs.saf",
                        col_types = list(
                          GeneID = col_character(),
                          Chr = col_character(),
                          Start = col_double(),
                          End = col_double(),
                          Strand = col_character()
                        )) %>%
  mutate(width = End - Start)

tidy(summary(mumerge_bed$width))
```

```{r mumerge-region-width, warning=FALSE, message=FALSE}
ggdensity(mumerge_bed, "width", add = "mean") +
  stat_central_tendency(type = "median", color = "red") +
  ggtitle("muMerged bidirectional widths")
```

The mean region width corresponds well to what the Dowell lab has seen across other PRO-seq datasets -- a bidirectional transcript lenth of \~600bp. This justifies a window size of +/-300bp. And as we can see, there are some potentially high confidence bidirectional calls with a muMerge region size of 4bp.

We'll ultimately use the 600bp fixed window size since it has good theoretical justification, but we'll compare that with the unmodified muMerge regions.

```{r counts-mumerge-windows}
# Generate counts over muMerge windows
proseq_bams <- list.files(file.path(proseq_dir,
                                    "results/mapped/bams"),
                          pattern = "*.bam$",
                          full.names = TRUE)

if(!file.exists("results/bidir_counts_pe.rds")) {
  bidir_counts <- featureCounts(proseq_bams, 
                                annot.ext = "results/mumerge_bidirs.saf",
                                isPairedEnd = TRUE,
                                nthreads = 22)
  write_rds(bidir_counts, "results/bidir_counts_pe.rds")
}
```

## Method 2: Fixed width of 600bp centered on muMerge region centers

```{r}
# Find the center and make +/- 300bp up and down
mumerge_uniform_saf <- mumerge_bed %>%
  mutate(center = Start + round(width/2),
         strt = center - 300,
         end = center + 300,
         new_width = end - strt) %>%
  dplyr::select(GeneID, Chr, strt, end, Strand) %>%
  dplyr::rename(Start = strt,
                End = end)

write_tsv(mumerge_uniform_saf, "results/mumerge_bidirs_600bp.saf")

if(!file.exists("results/bidir_counts_600bp_unstranded.rds")) {
  bidir_counts <- featureCounts(proseq_bams, 
                                annot.ext = "results/mumerge_bidirs_600bp.saf",
                                isPairedEnd = TRUE,
                                nthreads = 22)
  write_rds(bidir_counts, "results/bidir_counts_600bp_unstranded.rds")
}

proseq_samples <- read_csv("../../proseq/proseq_samplesheet.csv")
bidir_counts <- read_rds("results/bidir_counts.rds")
bidir_count_table <- bidir_counts$counts %>%
  as.data.frame() %>%
  rownames_to_column("bidir_id") %>%
  pivot_longer(2:ncol(.), names_to = "sample_id", values_to = "count") %>%
  mutate(sample_id = gsub(".sorted.bam", "", sample_id)) %>%
  left_join(proseq_samples %>% dplyr::select(sample_id, geo_sample_name)) %>%
  dplyr::select(bidir_id, geo_sample_name, count) %>%
  dplyr::rename(sample_name = geo_sample_name)
write_csv(bidir_count_table, "../../geo_upload/geo_submission_proseq_oct18/bidirectional_600bp_counts.csv")

tfit_mumerge_bidirectionals_600bp <- read_tsv("results/mumerge_bidirs_600bp.saf", col_types = cols(
  "GeneID" = col_character(),
  "Chr" = col_character(),
  "Start" = col_double(),
  "End" = col_double(),
  "Strand" = col_character()
)) %>%
  rowwise() %>%
  mutate(chr_prefix = "chr",
         chromosome = paste0(chr_prefix, Chr),
         score = ".") %>%
  dplyr::select(chromosome, Start, End, GeneID, score, Strand)
write_tsv(tfit_mumerge_bidirectionals_600bp, 
          "../../geo_upload/geo_submission_proseq_oct18/tfit_mumerge_bidirectionals_600bp.bed", col_names = FALSE)

all_bidir_coords <- mumerge_bed <- read_tsv("results/mumerge_bidirs.saf",
                        col_types = list(
                          GeneID = col_character(),
                          Chr = col_character(),
                          Start = col_double(),
                          End = col_double(),
                          Strand = col_character()
                        )) %>%
  rowwise() %>%
  mutate(chr_prefix = "chr",
         chromosome = paste0(chr_prefix, Chr),
         score = ".") %>%
  dplyr::select(chromosome, Start, End, GeneID, score, Strand)
write_tsv(all_bidir_coords, 
          "../../geo_upload/geo_submission_proseq_oct18/tfit_mumerge_bidirectionals.bed", col_names = FALSE)
```

### Stranded bidirectional quantification

Additionally, we'll quantify the reads on each strand in these fixed 600bp bidirectional regions.

```{r stranded-counts}
# Positive strand SAF
if(!file.exists("results/bidir_counts_positive_strand_600bp.rds")) {
  
  mumerge_bed_positive_strand <- mumerge_uniform_saf %>%
    mutate(Strand = "+")
  
  write.table(mumerge_bed_positive_strand, "results/mumerge_bidirs_positive_strand_600bp.saf", 
              quote = FALSE, sep = "\t",
              col.names = TRUE, row.names = FALSE)
  
  bidir_counts <- featureCounts(proseq_bams, 
                                annot.ext = "results/mumerge_bidirs_positive_strand_600bp.saf",
                                isPairedEnd = TRUE,
                                nthreads = 22)
  write_rds(bidir_counts, "results/bidir_counts_positive_strand_600bp.rds")
}

# Negative strand SAF
if(!file.exists("results/bidir_counts_negative_strand_600bp.rds")) {
  
  mumerge_bed_negative_strand <- mumerge_uniform_saf %>%
    mutate(Strand = "-")
  
  write.table(mumerge_bed_negative_strand, "results/mumerge_bidirs_negative_strand_600bp.saf", 
              quote = FALSE, sep = "\t",
              col.names = TRUE, row.names = FALSE)
  
  bidir_counts <- featureCounts(proseq_bams, 
                                annot.ext = "mumerge_bidirs_negative_strand_600bp.saf",
                                isPairedEnd = TRUE,
                                nthreads = 22)
  write_rds(bidir_counts, "results/bidir_counts_negative_strand_600bp.rds")
}
```

```{r run-counts-slurm, include=FALSE, eval=FALSE}
# # Run jobs in R session on fiji compute node using slurm interactive job -- since they 
# # take a while, but not long enough to justifiy a script? Questionable, but I'm feeling lazy.
#
# # srun --pty -p short --ntasks 22 --time=12:00:00 --mem 16G /bin/bash 
# # Positive strand read counts
# library(Rsubread)
# library(tidyverse)
# proseq_bams <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
#                           pattern = "*.bam$",
#                           full.names = TRUE)
# bidir_counts <- featureCounts(proseq_bams, 
#                               annot.ext = "/scratch/Shares/rinn/Michael/firre_timecourse/analysis/14_proseq/results/mumerge_bidirs_positive_strand_600bp.saf",
#                               isPairedEnd = TRUE,
#                               strandSpecific = 1,
#                               nthreads = 22,
#                               tmpDir = "/tmp")
# write_rds(bidir_counts, "results/bidir_counts_positive_strand_600bp.rds")
# 
# # Negative strand read counts
# library(Rsubread)
# library(tidyverse)
# proseq_bams <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
#                           pattern = "*.bam$",
#                           full.names = TRUE)
# bidir_counts <- featureCounts(proseq_bams, 
#                               annot.ext = "/scratch/Shares/rinn/Michael/firre_timecourse/analysis/14_proseq/results/mumerge_bidirs_negative_strand_600bp.saf",
#                               isPairedEnd = TRUE,
#                               strandSpecific = 1,
#                               nthreads = 18,
#                               tmpDir = "/tmp")
# write_rds(bidir_counts, "../results/bidir_counts_negative_strand_600bp.rds")
```

Having done so, we can calculate the strand bias on bidirectional regions. According to the Dowell lab, their expectation based on previous PRO-seq data is to have a neutral strand bias, which is what we see here in this data. However, we are seeing the Firre transgene exons being called as bidirectionals since although it is a nacscent transcript, it lacks introns, so when it's mapped to the reference genome, the reads map only to exons and look more like a bidirectional nascent transcript than a nascent gene body.

```{r}
# Let's calculate strand bias. 
ncounts <- read_rds("results/bidir_counts_negative_strand_600bp.rds")
pcounts <- read_rds("results/bidir_counts_positive_strand_600bp.rds")

ncounts_m <- ncounts$counts %>%
  as.data.frame() %>%
  rownames_to_column("bidir_id") %>%
  pivot_longer(cols = 2:ncol(.), 
               names_to = "sample_id",
               values_to = "neg_count") %>%
  mutate(sample_id = gsub(".sorted.bam", "", sample_id))

pcounts_m <- pcounts$counts %>%
  as.data.frame() %>%
  rownames_to_column("bidir_id") %>%
  pivot_longer(cols = 2:ncol(.), 
               names_to = "sample_id",
               values_to = "pos_count") %>%
  mutate(sample_id = gsub(".sorted.bam", "", sample_id))

counts_stranded <- pcounts_m %>%
  left_join(ncounts_m)
```

```{r, class.source = 'fold-show'}
counts_stranded <- counts_stranded %>%
  mutate(strand_bias = log2((pos_count+0.1) / (neg_count+0.1)),
         abs_strand_bias = abs(strand_bias)) %>%
  left_join(bidir_coords) %>%
  mutate(width = end - start)

quantile(counts_stranded$abs_strand_bias)
```

We'll impose a read depth cutoff of at least 5 reads on the positive strand and 5 reads on the negative strand.

```{r, message=FALSE, warning=FALSE}
bidir_counts_filtered <- counts_stranded %>% 
  dplyr::filter(pos_count >= 5, neg_count >= 5)
```

This leaves us with `r length(unique(bidir_counts_filtered$bidir_id))` bidirectionals detected. Let's plot the strand bias.

```{r, fig.cap="Bidirectional transcript strand bias"}
ggdensity(bidir_counts_filtered, x = "strand_bias")
```

```{r}
bidir_counts_sbfiltered <- counts_stranded %>% 
  dplyr::filter(pos_count >= 5, neg_count >= 5, abs_strand_bias <= 1)
```

If we additionally filter on a low strand bias (true "bi"directionals), then we have `r length(unique(bidir_counts_sbfiltered$bidir_id))` bidirectionals.

```{r}
# Bidir regions to include
bidir_filtered <- unique(bidir_counts_sbfiltered$bidir_id)
length(bidir_filtered)
```

# Differential expression of bidirectional transcripts

```{r}
bidir_counts <- read_rds("results/bidir_counts_pe.rds")
sinfo <- read_tsv("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/mumerge/sample_info.txt")

# bidir_counts <- read_rds("results/bidir_counts_600bp_unstranded.rds")
counts_matrix <- bidir_counts$counts
counts_matrix <- counts_matrix[bidir_filtered,]
samples <- tibble(file = colnames(counts_matrix)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                               TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))

dds <- DESeqDataSetFromMatrix(countData = counts_matrix,
                              colData = samples,
                              design = ~ timepoint)
# LRT 
dds <- DESeq(dds, test = "LRT", reduced = ~1)
# write_rds(res_lrt, "results/deseq_dds_vs_zero_lrt.rds")

res_lrt_res <- bind_rows(results(dds, name = "timepoint_30_vs_0", tidy = T) %>%
                           mutate(comparison = "timepoint_30_vs_0_lrt")) %>%
  dplyr::rename(bidir_id = row) %>%
  left_join(bidir_coords) %>%
  mutate(width = end - start)

sig_lrt_res <- res_lrt_res %>%
  dplyr::filter(padj < 0.05) %>%
  arrange(padj)

length(unique(sig_lrt_res$bidir_id))
```



```{r}
dds <- DESeqDataSetFromMatrix(countData = counts_matrix,
                              colData = samples,
                              design = ~ timepoint)

dds <- DESeq(dds)
# write_rds(res_lrt, "results/deseq_dds_vs_zero_lrt.rds")
# TODO: apeglm is not working
# library(apeglm)
res_vs_zero <- bind_rows(lfcShrink(dds, coef = "timepoint_30_vs_0", type = "ashr") %>%
                           as.data.frame() %>%
                           rownames_to_column("bidir_id") %>%
                           mutate(comparison = "timepoint_30_vs_0_shrnk"),
                         lfcShrink(dds, coef = "timepoint_15_vs_0", type = "ashr") %>%
                           as.data.frame() %>%
                           rownames_to_column("bidir_id") %>%
                           mutate(comparison = "timepoint_15_vs_0_shrnk")) %>%
  left_join(mumerge_uniform_saf %>% dplyr::rename(bidir_id = GeneID))

length(unique(res_vs_zero$bidir_id))
# Let's filter to just those that were sig in LRT
sig_res_vs_zero <- res_vs_zero %>%
  dplyr::filter(bidir_id %in% sig_lrt_res$bidir_id)

bidir_maxfc <- sig_res_vs_zero %>%
  group_by(bidir_id) %>%
  summarize(max_fc = max(abs(log2FoldChange)))

sig_res_vs_zero <- sig_res_vs_zero %>%
  left_join(bidir_maxfc) %>%
  dplyr::filter(max_fc >= log2(1.5))
length(unique(sig_res_vs_zero$bidir_id))
write_csv(sig_res_vs_zero, "results/proseq_bidir_de_sig.csv")


sig_summary <- sig_res_vs_zero %>%
  filter(comparison == "timepoint_15_vs_0_shrnk") %>%
  mutate(deg = ifelse(log2FoldChange > 0, "up", "down")) %>%
  group_by(deg) %>%
  summarize(count = n())
```

```{r}
length(unique(sig_res_vs_zero$bidir_id))
sig_res_vs_zero <- sig_res_vs_zero %>%
  mutate(timepoint = gsub("timepoint_|_vs_0_shrnk", "", comparison),
         timepoint = as.numeric(timepoint))

zero_tp <- sig_res_vs_zero %>%
  dplyr::select(bidir_id) %>%
  distinct() %>%
  mutate(timepoint = 0,
         log2FoldChange = 0)

sig_res_vs_zero <- sig_res_vs_zero %>%
  bind_rows(sig_res_vs_zero, zero_tp)

ggplot(sig_res_vs_zero, aes(x = timepoint, y = log2FoldChange, group = bidir_id)) +
  geom_hline(yintercept = 0) +
  geom_point(alpha = 0.3) + 
  geom_line(alpha = 0.1)
ggsave("figures/proseq_sig_vs_zero_scatter.pdf")

# Let's look at those that continue their trajectory
hmm <- sig_res_vs_zero %>%
  dplyr::filter(timepoint %in% c(15,30)) %>%
  dplyr::select(-comparison) %>%
  distinct() %>%
  pivot_wider(id_cols = c("bidir_id", "baseMean", "Chr", "Start", "End", "Strand", "max_fc"), names_from = "timepoint", 
              values_from = c("log2FoldChange","lfcSE", "pvalue", "padj")) %>%
  mutate(monotonic = abs(log2FoldChange_30) > abs(log2FoldChange_15)) %>%
  dplyr::select(bidir_id, monotonic) %>%
  distinct()

sig_res_vs_zero <- sig_res_vs_zero %>%
  left_join(hmm)

ggplot(sig_res_vs_zero %>% filter(monotonic == TRUE), aes(x = timepoint, y = log2FoldChange, group = bidir_id)) +
  geom_hline(yintercept = 0) +
  geom_point(alpha = 0.3) + 
  geom_line(alpha = 0.1)

monoton <- sig_res_vs_zero %>%
  filter(monotonic == TRUE)

length(unique(monoton$bidir_id))

```

Are the bidirs that continue their trajectory more likely to be those associated with ATAC-seq peaks? Is there a correlation with read counts? In other words, may this be an artifact of the DESeq2 dispersion normalization?

```{r}
ggplot(res_vs_zero %>% dplyr::filter(comparison == "timepoint_15_vs_0_shrnk"), aes(x = log10(baseMean), y = log2FoldChange)) +
  geom_point()

ggplot(res_vs_zero %>% dplyr::filter(comparison == "timepoint_30_vs_0_shrnk"), aes(x = log10(baseMean), y = log2FoldChange)) +
  geom_point()
```

```{r}
ggplot(res_vs_zero %>% filter(bidir_id %in% sig_lrt_res$bidir_id), 
       aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point() +
  facet_wrap(~comparison)

# Each timepoint vs zero
dds <- DESeq(dds)

lfcShrink()
# write_rds(dds, "results/deseq_dds_vs_zero.rds")
res_vs_zero <- bind_rows(results(dds, name = "timepoint_30_vs_0", tidy = T) %>%
                           mutate(comparison = "timepoint_30_vs_0"),
                         results(dds, name = "timepoint_15_vs_0", tidy = T) %>%
                           mutate(comparison = "timepoint_15_vs_0")) %>%
  dplyr::rename(bidir_id = row) %>%
  left_join(mumerge_uniform_saf %>% dplyr::rename(bidir_id = GeneID))
write_csv(res_vs_zero, "results/res_vs_zero.csv")

res_vs_zero_sig <- res_vs_zero %>%
  filter(padj < 0.05) %>%
  mutate(width = End - Start)
ggplot(res_vs_zero, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point() +
  facet_wrap(~comparison)


# Maybe next do a ranked list with the mean FC and submit to TFEA.
# Or separately for 15 min and 30 min. 
```

## ATAC-seq peak quantification

1.  What's happening in the 55 differential ATAC peaks?

```{r}


# Read in the differential ATAC-seq peaks 
atac_up <- read_tsv("../12_atacseq/results/atac_up.bed", col_names = c("Chr", "Start", "End")) %>%
  mutate(GeneID = paste0("up_", 1:n()), .before = 1)
atac_down <- read_tsv("../12_atacseq/results/atac_down.bed", col_names = c("Chr", "Start", "End")) %>%
  mutate(GeneID = paste0("down_", 1:n()), .before = 1)

de_atac <- bind_rows(atac_up,
                     atac_down) %>%
  mutate(Strand = ".")

write_tsv(de_atac, "results/de_atac.saf")
?featureCounts
library(Rsubread)
library(tidyverse)
proseq_bams <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
                          pattern = "*.bam$",
                          full.names = TRUE)
bidir_counts <- featureCounts(proseq_bams, 
                              annot.ext = "results/de_atac.saf",
                              isPairedEnd = TRUE,
                              strandSpecific = 0,
                              nthreads = 9,
                              tmpDir = "/tmp")
write_rds(bidir_counts, "results/de_atac_proseq_counts.rds")

de_atac_fc <- read_rds("results/de_atac_proseq_counts.rds")



mapped_list <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/qc/mapstats",
                          full.names = T,
                          pattern = "*.millionsmapped") %>%
  sapply(read_lines) %>%
  as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  mutate(sample_id = gsub("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/qc/mapstats/|.millionsmapped", "", sample_id)) %>%
  dplyr::rename(reads_mapped = ".") %>%
  mutate(reads_mapped = as.numeric(reads_mapped))


de_atac_counts <- de_atac_fc$counts  %>%
  as.data.frame() %>%
  rownames_to_column("atac_id") %>%
  pivot_longer(2:ncol(.), names_to = "sample_id", values_to = "count") %>%
  mutate(sample_id = gsub(".sorted.bam", "", sample_id)) %>%
  left_join(mapped_list) %>%
  left_join(sinfo %>% dplyr::rename(sample_id = sampid)) %>%
  unite(col = "sample_name", sample_id, group) %>%
  mutate(CPM = (count / reads_mapped) *1e6)

de_atac_cpm <- de_atac_counts %>%
  dplyr::select(atac_id, sample_name, CPM) %>%
  pivot_wider(names_from = "sample_name", values_from = "CPM") %>%
  column_to_rownames("atac_id") %>%
  as.matrix()
cat(colnames(de_atac_cpm_scaled))
de_atac_cpm_scaled <- t(scale(t(de_atac_cpm)))
de_atac_cpm_scaled <- de_atac_cpm_scaled[,c("JR3113_Control",
                                            "JR3116_Control",
                                            "JR3119_Control",
                                            "JR3114_15minDOX",
                                            "JR3120_15minDOX",
                                            "JR3117_15minDOX",
                                            "JR3118_30minDOX",
                                            "JR3121_30minDOX",
                                            "JR3115_30minDOX")]
pheatmap::pheatmap(de_atac_cpm , cluster_cols = F)

```

2.  Let's count nascent reads over all ATAC-seq peaks

```{r}
load("../01_setup/results/atacseq_data.RData", verbose = T)
atac_saf <- atac_consensus %>%
  dplyr::select(1:5) %>%
  dplyr::rename(GeneID = interval_id)
write_tsv(atac_saf, "results/atac.saf")

library(Rsubread)
library(tidyverse)
proseq_bams <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/mapped/bams", 
                          pattern = "*.bam$",
                          full.names = TRUE)
bidir_counts <- featureCounts(fl, 
                              annot.ext = "results/atac.saf",
                              isPairedEnd = TRUE,
                              strandSpecific = 0,
                              nthreads = 9,
                              tmpDir = "/tmp")
write_rds(bidir_counts, "results/atac_proseq_counts.rds")
```

```{r}
hmm <- read_rds("results/atac_proseq_counts.rds")
colnames(hmm$counts)
```

### Differential expression of nascent transcripts -- unstranded

```{r}
bidir_counts <- read_rds("results/bidir_counts_600bp_unstranded.rds")

# Let's normalize based on the number of mapped reads in each sample.
# We don't need to normalize based on length since each region was 600bp wide.
mapped_list <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/qc/mapstats",
                          full.names = T,
                          pattern = "*.millionsmapped") %>%
  sapply(read_lines) %>%
  as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  mutate(sample_id = gsub("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/results/qc/mapstats/|.millionsmapped", "", sample_id)) %>%
  dplyr::rename(reads_mapped = ".") %>%
  mutate(reads_mapped = as.numeric(reads_mapped))


bidir_counts_l <- bidir_counts$counts %>%
  as.data.frame() %>%
  rownames_to_column("bidir_id") %>%
  pivot_longer(2:ncol(.), names_to = "sample_id", values_to = "count") %>%
  mutate(sample_id = gsub(".sorted.bam", "", sample_id))

bidir_counts_l <- bidir_counts_l %>%
  left_join(mapped_list)


bidir_counts_l <- bidir_counts_l %>%
  mutate(CPM = (count / reads_mapped) *1e6)

sinfo <- read_tsv("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/mumerge/sample_info.txt") %>%
  dplyr::rename(sample_id = sampid) %>%
  dplyr::select(sample_id, group) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                               TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))
bidir_counts_l <- bidir_counts_l %>%
  left_join(sinfo)


bidir_30_vs_0 <- bidir_counts_l %>%
  dplyr::filter(timepoint %in% c("0","30"))

library(rstatix)
library(tidyverse)
bidir_30_vs_0_ttest <- bidir_30_vs_0 %>%
  group_by(bidir_id) %>%
  t_test(data = ., CPM ~ timepoint) %>%
  adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj")
write_rds(bidir_30_vs_0_ttest, "results/bidir_30_vs_0_ttest.rds")


bidir_15_vs_0 <- bidir_counts_l %>%
  dplyr::filter(timepoint %in% c("0","15"))
library(rstatix)
library(tidyverse)
bidir_15_vs_0_ttest <- bidir_15_vs_0 %>%
  group_by(bidir_id) %>%
  t_test(data = ., CPM ~ timepoint) %>%
  adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj")
write_rds(bidir_15_vs_0_ttest, "results/bidir_15_vs_0_ttest.rds")
bidir_62 <-  bidir_30_vs_0 %>%
  filter(bidir_id == "bidir_62")

bidir_15_vs_0 <- bidir_counts_l %>%
  dplyr::filter(timepoint %in% c("0","15"))


bidir_15_vs_0_ttest$padj <- p.adjust(bidir_15_vs_0_ttest$p, method = "BH")
bidir_30_vs_0_ttest$padj <- p.adjust(bidir_30_vs_0_ttest$p, method = "BH")
```

So this result is that there are no significantly changing bidirectionals... Before we go back to trying DESeq2 or something, let's just calculate the lm of the CPM per bidir. Another thing to consider is starting from the coverage filtered bidirectionals so that the number of statistical tests performed is smaller.

```{r}
bidir_counts_l_test <- bidir_counts_l[1:27,]
library(multidplyr)

cl <- new_cluster(16)

bclm1 <- bidir_counts_l %>%
  group_by(bidir_id) %>%
  partition(cluster = cl) %>%
  do(mod = broom::tidy(lm(CPM ~ timepoint, data = .))) %>%
  collect() %>%
  unnest(mod)

bclm1_tp <- bclm1 %>%
  filter(term == "timepoint")
bclm1_tp$padj <- p.adjust(bclm1_tp$p.value, method = "BH")

bcpm_sig <- bclm1_tp %>%
  filter(padj < 0.1) %>%
  left_join(mumerge_uniform_saf %>% dplyr::rename(bidir_id = GeneID))

bclm <- bidir_counts_l_test %>%
  mutate(timepoint = as.numeric(timepoint)) %>%
  group_by(bidir_id) %>%
  nest() %>%
  mutate(test = map(data, ~ lm(.x$CPM ~ .x$timepoint)),
         tidied = map(test, tidy)) %>%
  unnest(tidied)
bidir_counts_l <- bidir_counts_l %>%
  mutate(timepoint = as.numeric(timepoint))
ggplot(bidir_counts_l %>% filter(bidir_id == "bidir_106735"), aes(x = timepoint, y = CPM)) +
  geom_point() +
  geom_smooth(method = "lm") +
  stat_regline_equation()
```

### Strand-specific differential expression of nascent transcripts

It would be interesting to see the best changing bidirectional nascent transcripts. Therefore, what we'll do here is run DEseq on the +/- strand counts separately. Then look at the correlation and choose those that are the most correlated with the highest fold-change.

```{r}
sinfo <- read_tsv("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/mumerge/sample_info.txt")
counts_matrix <- bidir_counts$counts
samples <- tibble(file = colnames(counts_matrix)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                               TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))
```

```{r}
# Positive strand counts 
pcounts <- read_rds("results/bidir_counts_positive_strand.rds")
pcounts_matrix <- pcounts$counts
# Filter to just those with counts on positive and negative strand

rownames(pcounts_matrix)
pcounts_matrix <- pcounts_matrix[bidir_filtered,]

samples <- tibble(file = colnames(pcounts_matrix)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                               TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))

pdds <- DESeqDataSetFromMatrix(countData = pcounts_matrix,
                               colData = samples,
                               design = ~ timepoint)
# Each timepoint vs zero
# pdds <- DESeq(pdds)
pdds <- DESeq(pdds, test = "LRT", reduced = ~1)
# write_rds(dds, "results/deseq_dds_vs_zero.rds")
pres_vs_zero <- bind_rows(results(pdds, name = "timepoint_30_vs_0", tidy = T) %>%
                            mutate(comparison = "timepoint_30_vs_0"),
                          results(pdds, name = "timepoint_15_vs_0", tidy = T) %>%
                            mutate(comparison = "timepoint_15_vs_0")) %>%
  dplyr::rename(bidir_id = row) %>%
  left_join(bidir_coords) %>%
  mutate(strand = "+")
write_csv(pres_vs_zero, "results/res_vs_zero_positive_strand.csv")

# Negative strand counts
ncounts <- read_rds("results/bidir_counts_negative_strand.rds")
ncounts_matrix <- ncounts$counts
ncounts_matrix <- ncounts_matrix[bidir_filtered,]

samples <- tibble(file = colnames(ncounts_matrix)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                               TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))

ndds <- DESeqDataSetFromMatrix(countData = ncounts_matrix,
                               colData = samples,
                               design = ~ timepoint)
# Each timepoint vs zero
# ndds <- DESeq(ndds)
ndds <- DESeq(ndds, test = "LRT", reduced = ~1)
# write_rds(dds, "results/deseq_dds_vs_zero.rds")
nres_vs_zero <- bind_rows(results(ndds, name = "timepoint_30_vs_0", tidy = T) %>%
                            mutate(comparison = "timepoint_30_vs_0"),
                          results(ndds, name = "timepoint_15_vs_0", tidy = T) %>%
                            mutate(comparison = "timepoint_15_vs_0")) %>%
  dplyr::rename(bidir_id = row) %>%
  left_join(bidir_coords) %>%
  mutate(strand = "-")
write_csv(nres_vs_zero, "results/res_vs_zero_negative_strand.csv")

pres_vs_zero <- read_csv("results/res_vs_zero_positive_strand.csv")
nres_vs_zero <- read_csv("results/res_vs_zero_negative_strand.csv")
# Combine into one data.frame
pres_vs_zero <- pres_vs_zero %>%
  dplyr::rename(l2fc_pos = log2FoldChange,
                padj_pos = padj) %>%
  dplyr::select(bidir_id, comparison, chrom, start, end, l2fc_pos, padj_pos)

nres_vs_zero <- nres_vs_zero %>%
  dplyr::rename(l2fc_neg = log2FoldChange,
                padj_neg = padj) %>%
  dplyr::select(bidir_id, comparison, l2fc_neg, padj_neg)

combined_strands_vs_zero <- pres_vs_zero %>%
  left_join(nres_vs_zero)

combined_strands_vs_zero <- combined_strands_vs_zero %>%
  mutate(sig_in_both = (padj_pos < 0.05 & padj_neg < 0.05))
table(combined_strands_vs_zero$sig_in_both)


ggplot(combined_strands_vs_zero, aes(x = l2fc_pos, y = l2fc_neg, color = sig_in_both)) +
  facet_wrap(~comparison) +
  geom_point()


ggplot(combined_strands_vs_zero %>%
         filter(sig_in_both == TRUE), aes(x = l2fc_pos, y = l2fc_neg, color = sig_in_both)) +
  facet_wrap(~comparison) +
  geom_point()


hmm <- combined_strands_vs_zero %>%
  filter(sig_in_both == TRUE,
         abs(l2fc_pos) > log2(1.5), abs(l2fc_neg) > log2(1.5))

ggplot(hmm, aes(x = l2fc_pos, y = l2fc_neg, color = sig_in_both)) +
  facet_wrap(~comparison) +
  geom_point()

sig_bidirs <- unique(hmm$bidir_id)


combined_strands_vs_zero_sig_w <- combined_strands_vs_zero %>%
  filter(bidir_id %in% sig_bidirs) %>%
  mutate(timepoint = gsub("timepoint_|_vs_0", "", comparison)) %>%
  dplyr::select(-comparison, -sig_in_both) %>%
  pivot_wider(id_cols = c(bidir_id, chrom, start, end),
              names_from = timepoint,
              values_from = l2fc_pos:padj_neg)

combined_strands_vs_zero_sig_w <- combined_strands_vs_zero_sig_w %>%
  mutate(continued_pos = ifelse(l2fc_pos_15 > 0, l2fc_pos_30 > l2fc_pos_15,
                                l2fc_pos_30 < l2fc_pos_15),
         continued_neg = ifelse(l2fc_neg_15 > 0, l2fc_neg_30 > l2fc_neg_15,
                                l2fc_neg_30 < l2fc_neg_15))


bidir_cont <- combined_strands_vs_zero_sig_w %>%
  filter(continued_pos == TRUE,
         continued_neg == TRUE)

bidir_contl <- bidir_cont %>%
  dplyr::select(bidir_id, chrom, start, end, contains("l2fc")) %>%
  pivot_longer(cols = l2fc_pos_30:l2fc_neg_15, names_to = "l2fc_condition", values_to = "l2fc") %>%
  separate(col = l2fc_condition, into = c("none", "strand", "timepoint")) %>%
  mutate(timepoint = as.numeric(timepoint))

ggplot(bidir_contl, aes(x = timepoint, y = l2fc, color = strand, group = bidir_id)) +
  geom_point()+geom_line() +
  facet_wrap(~strand)
ggplot(bidir_cont, aes(x = l2fc_pos, y = l2fc_neg, color = sig_in_both)) +
  facet_wrap(~comparison) +
  geom_point()

# bidir_68169
# Does this region come up when we take an unstranded approach?
```

## Gene level differential expression

```{r}


fl_unstranded_gene_counts <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/bidirflow/featurecounts_genes", full.names = T, pattern = "*unstranded*")
# Let's just get those with the 5' part of the gene removed
fl_unstranded_gene_counts <- fl_unstranded_gene_counts[grepl("5ptrunc_gene_counts.txt", fl_unstranded_gene_counts)]

ugc <- lapply(fl_unstranded_gene_counts,function(x)
  read_tsv(x)[,4])
ugc <- do.call(cbind, ugc)

# The gene ID order is the same in all the count files. 
# However, there are duplicated gene names and transcript ids.
# ugc_gene_ids <- lapply(fl_unstranded_gene_counts, function(x)
#   read_tsv(x)[,2])
# ugc_gene_ids <- do.call(cbind, ugc_gene_ids)
ugc_gid <- read_tsv(fl_unstranded_gene_counts[[1]]) %>%
  mutate(index = 1:nrow(.)) %>%
  unite(col = "id", GeneID, TranscriptID, index)
# ugc_gid <- ugc_gid$id

rownames(ugc) <- ugc_gid$id


# Calc TPM
ugc_tpm <- counts_to_tpm(ugc, ugc_gid$Length, rep(250,9))

tpm_5ptrunc <- ugc_tpm %>%
  as.data.frame() %>%
  rownames_to_column("id") %>%
  pivot_longer(2:ncol(.), names_to="sampid", values_to = "tpm") %>%
  mutate(sampid = gsub(".sorted.bam", "", sampid)) %>%
  left_join(samples) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
write_csv(tpm_5ptrunc, "results/tpm_5ptrunc.csv")
tpm_5ptrunc <- read_csv("results/tpm_5ptrunc.csv")

expressed <- tpm_5ptrunc %>%
  filter(tpm > 1)

length(unique(expressed$id))
length(unique(expressed$id))/length(unique(tpm_5ptrunc$id))
ggplot(tpm_5ptrunc %>% filter(grepl("Duox1", id)),
       aes(x = timepoint, y = tpm)) +
  geom_point() +
  geom_smooth(method = "lm")

library(ggpubr)
ggplot(tpm_5ptrunc %>% filter(grepl("Firre", id)),
       aes(x = timepoint, y = tpm)) +
  geom_hline(yintercept = 0) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "black") +
  ggtitle("Firre expression") +
  ylab("TPM") +
  xlab("Time (Min.)") +
  stat_regline_equation(size = 8)

sinfo <- read_tsv("/scratch/Shares/rinn/Michael/firre_timecourse/proseq/mumerge/sample_info.txt")
samples <- tibble(file = colnames(ugc)) %>%
  mutate(sampid = gsub(".sorted.bam", "", file)) %>%
  left_join(sinfo) %>%
  mutate(timepoint = case_when(group == "Control" ~ "0",
                               group == "15minDOX" ~ "15",
                               TRUE ~ "30"),
         timepoint = factor(timepoint, levels = c("0", "15", "30")))

dds <- DESeqDataSetFromMatrix(countData = ugc,
                              colData = samples,
                              design = ~ timepoint)

dds <- DESeq(dds)

res_vs_zero <- bind_rows(results(dds, name = "timepoint_30_vs_0", tidy = T) %>%
                           mutate(comparison = "timepoint_30_vs_0"),
                         results(dds, name = "timepoint_15_vs_0", tidy = T) %>%
                           mutate(comparison = "timepoint_15_vs_0")) %>%
  dplyr::rename(gene_id = row)
write_csv(res_vs_zero, "results/res_vs_zero_gene5ptrunc.csv")


res_vs_zero <- res_vs_zero %>%
  rowwise() %>%
  mutate(gene_name = unlist(strsplit(gene_id, "_"))[[1]])

res_vs_zero <- res_vs_zero %>%
  mutate(gene_name = gsub("Gm35612", "CrossFirre", gene_name))
library(ggrepel)
ggplot(res_vs_zero, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point() + 
  facet_wrap(~ comparison) +
  geom_text_repel(data = res_vs_zero %>% filter(log2FoldChange > 4 | -log10(padj) > 75),
                  aes(label = gene_name))

gene_sig <- res_vs_zero %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1)

gs_summary <- gene_sig %>%
  mutate(deg = ifelse(log2FoldChange > 0, 1, -1)) %>%
  group_by(comparison, deg) %>%
  summarize(ndeg = n())

table(gene_sig$comparison)


# Overlaps with previous Firre results
load("../11_short_timecourse_combined/results/short_vszero_sig.RData", verbose = T)
svz_sig_genes <- unique(short_vszero_sig$gene_name)
length(svz_sig_genes)
length(which(svz_sig_genes %in% gene_sig$gene_name))
cat(unique(gene_sig$gene_name[which(gene_sig$gene_name %in% svz_sig_genes)]))


load("../03_doxycycline_effect/results/esc_ko_control_short.RData", verbose = T)
dox_sig <- esc_ko_control_short %>%
  filter(padj < 0.05) %>%
  group_by(gene_id, gene_name) %>%
  summarize(max_lfc = max(log2FoldChange)) %>%
  filter(abs(max_lfc) > log2(1.5))


dox_sig_genes <- dox_sig$gene_name
length(which(dox_sig_genes %in% gene_sig$gene_name))
```




## Heatmap of differential bidirectionals

```{r}
# We jut need to plot the 0 and 15 minute heatmaps. 
# In order to do that we need to:
# 1. Get a list of the differential pro-seq regions. 
# 2. Retrieve the read count vectors from the positive and negative strands.
# 3. Look at how the heatmap is created in the seurat code.
# 4. Explore how to plot it in ggplot so that it looks like. https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.researchgate.net%2Ffigure%2FGDNF-activates-transcriptional-regulatory-elements-A-Heatmap-depicting-PRO-seq_fig1_324211287&psig=AOvVaw02DJqeN5uiXTmmJrFC_2o2&ust=1664153481979000&source=images&cd=vfe&ved=0CAwQjRxqFwoTCNiukf_crvoCFQAAAAAdAAAAABAD

de_bidirs <- read_csv("results/proseq_bidir_de_sig.csv")

de_15_min <- de_bidirs %>%
  filter(comparison == "timepoint_15_vs_0_shrnk")

de_15_up <- de_15_min %>%
  filter(log2FoldChange > log2(1.5))

de_15_down <- de_15_min %>%
  filter(log2FoldChange < -log2(1.5))




de_bidir_coords_up <- de_15_up %>%
  dplyr::select(bidir_id, Chr, Start, End) %>%
  distinct() %>%
  mutate(bidir_number = 1:nrow(.))
write_csv(de_bidir_coords_up, "results/de_bidir_coords_up.csv")


de_bidir_coords_down <- de_15_down %>%
  dplyr::select(bidir_id, Chr, Start, End) %>%
  distinct() %>%
  mutate(bidir_number = 1:nrow(.))
write_csv(de_bidir_coords_down, "results/de_bidir_coords_down.csv")

de_bidir_gr_up <- GRanges(seqnames = de_bidir_coords_up$Chr,
                       ranges = IRanges(start = de_bidir_coords_up$Start,
                                        end = de_bidir_coords_up$End))

de_bidir_gr_down <- GRanges(seqnames = de_bidir_coords_down$Chr,
                       ranges = IRanges(start = de_bidir_coords_down$Start,
                                        end = de_bidir_coords_down$End))

# Let's think about taking the mean signal (or maybe the max across the replicates)?

source("../15_coverage_tracks/bin/track_location_lists.R")

# What are the steps here?
# Import the pos and neg strand and make into a df
# Number the rows so that pos strand is 1, neg is 2
# pos 3, neg 2
de_proseq_signal_up <- lapply(1:length(proseq_bw_all), function(i) {
  
  pos_strand_coverage <- rtracklayer::import(
    con = proseq_bw_all[[i]][["pos"]],
    which = de_bidir_gr_up,
    as = "NumericList"
  ) 
  
  pos_strand_df <- do.call(rbind, pos_strand_coverage@listData) %>%
    as.data.frame() %>%
    rownames_to_column("bidir_number") %>%
    pivot_longer(2:ncol(.),
                 names_to = "bp",
                 values_to = "coverage") %>%
    mutate(bp = as.numeric(gsub("V", "", bp)),
           bidir_number = as.numeric(bidir_number),
           strand = "+",
           y_index = (2 * bidir_number)-1,
           sample_name = names(proseq_bw_all)[[i]]) 
  
  neg_strand_coverage <- rtracklayer::import(
    con = proseq_bw_all[[i]][["neg"]],
    which = de_bidir_gr_up,
    as = "NumericList"
  ) 
  
  neg_strand_df <- do.call(rbind, neg_strand_coverage@listData) %>%
    as.data.frame() %>%
    rownames_to_column("bidir_number") %>%
    pivot_longer(2:ncol(.),
                 names_to = "bp",
                 values_to = "coverage") %>%
    mutate(bp = as.numeric(gsub("V", "", bp)),
           bidir_number = as.numeric(bidir_number),
           strand = "-",
           y_index = 2 * bidir_number,
           sample_name = names(proseq_bw_all)[[i]]) 
  
  cov_df <- bind_rows(pos_strand_df,
                      neg_strand_df) %>%
    mutate(deg = "up")
  return(cov_df)
}) %>%
  bind_rows()

write_rds(de_proseq_signal_up, "results/de_proseq_signal_600bp_up.rds")

```

```{r}
de_proseq_signal_down <- lapply(1:length(proseq_bw_all), function(i) {
  
  pos_strand_coverage <- rtracklayer::import(
    con = proseq_bw_all[[i]][["pos"]],
    which = de_bidir_gr_down,
    as = "NumericList"
  ) 
  
  pos_strand_df <- do.call(rbind, pos_strand_coverage@listData) %>%
    as.data.frame() %>%
    rownames_to_column("bidir_number") %>%
    pivot_longer(2:ncol(.),
                 names_to = "bp",
                 values_to = "coverage") %>%
    mutate(bp = as.numeric(gsub("V", "", bp)),
           bidir_number = as.numeric(bidir_number),
           strand = "+",
           y_index = (2 * bidir_number)-1,
           sample_name = names(proseq_bw_all)[[i]]) 
  
  neg_strand_coverage <- rtracklayer::import(
    con = proseq_bw_all[[i]][["neg"]],
    which = de_bidir_gr_down,
    as = "NumericList"
  ) 
  
  neg_strand_df <- do.call(rbind, neg_strand_coverage@listData) %>%
    as.data.frame() %>%
    rownames_to_column("bidir_number") %>%
    pivot_longer(2:ncol(.),
                 names_to = "bp",
                 values_to = "coverage") %>%
    mutate(bp = as.numeric(gsub("V", "", bp)),
           bidir_number = as.numeric(bidir_number),
           strand = "-",
           y_index = 2 * bidir_number,
           sample_name = names(proseq_bw_all)[[i]]) 
  
  cov_df <- bind_rows(pos_strand_df,
                      neg_strand_df) %>%
    mutate(deg = "down")
  return(cov_df)
}) %>%
  bind_rows()

write_rds(de_proseq_signal_down, "results/de_proseq_signal_600bp_down.rds")
```


```{r}

de_proseq_mean_up <- de_proseq_signal_up %>%
  separate(col = sample_name, into = c("timepoint", "replicate")) %>%
  group_by(bidir_number, bp, strand, timepoint) %>%
  summarize(mean_coverage = mean(coverage)) %>%
  group_by(bidir_number, bp, timepoint) %>%
  summarize(cov = mean_coverage[which.max(abs(mean_coverage))])
write_rds(de_proseq_mean_up, "results/de_proseq_mean_600bp_up.rds")

de_proseq_mean_down <- de_proseq_signal_down %>%
  separate(col = sample_name, into = c("timepoint", "replicate")) %>%
  group_by(bidir_number, bp, strand, timepoint) %>%
  summarize(mean_coverage = mean(coverage)) %>%
  group_by(bidir_number, bp, timepoint) %>%
  summarize(cov = mean_coverage[which.max(abs(mean_coverage))])
write_rds(de_proseq_mean_down, "results/de_proseq_mean_600bp_up.rds")



de_proseq_0_up <- de_proseq_mean_up %>%
  filter(timepoint == 0) %>%
  pivot_wider(names_from = bp, values_from = cov) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint) %>%
  as.matrix()

de_proseq_15_up <- de_proseq_mean_up %>%
  filter(timepoint == 15) %>%
  pivot_wider(names_from = bp, values_from = cov) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint) %>%
  as.matrix()

de_proseq_0_down <- de_proseq_mean_down %>%
  filter(timepoint == 0) %>%
  pivot_wider(names_from = bp, values_from = cov) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint) %>%
  as.matrix()

de_proseq_15_down <- de_proseq_mean_down %>%
  filter(timepoint == 15) %>%
  pivot_wider(names_from = bp, values_from = cov) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint) %>%
  as.matrix()

# ord_decreasing_15 <- rev(order(rowSums(de_proseq_15[,1:50])))


# library(seriation)
# o1 = seriate(dist(de_proseq_0), method = "HC_average")
col_fun = colorRamp2(c(-2, 0, 2), c("#7F8999", "white", "#AC404D"))
ht0_up <- Heatmap(de_proseq_0_up, 
               # row_order = get_order(o1),
               show_row_names = FALSE, 
               show_column_names = FALSE,
               cluster_columns = FALSE,
               name = "0'", 
               column_title = "0'",
               row_title = "Up",
               col = col_fun)


ht15_up <- Heatmap(de_proseq_15_up, 
                # row_order = get_order(o1),
                # cluster_rows = TRUE,
                show_row_names = FALSE, 
                show_column_names = FALSE,
                cluster_columns = FALSE,
                name = "15'", 
                column_title = "15'",
                row_title = "Up",
                col = col_fun)

ht0_down <- Heatmap(de_proseq_0_down, 
               # row_order = get_order(o1),
               show_row_names = FALSE, 
               show_column_names = FALSE,
               cluster_columns = FALSE,
               name = "0'", 
               column_title = "0'",
               row_title = "Down",
               col = col_fun)


ht15_down <- Heatmap(de_proseq_15_down, 
                # row_order = get_order(o1),
                # cluster_rows = TRUE,
                show_row_names = FALSE, 
                show_column_names = FALSE,
                cluster_columns = FALSE,
                name = "15'", 
                column_title = "15'",
                row_title = "Down",
                col = col_fun)

pdf("figures/proseq_15_up.pdf", height = 2, width = 4)
ht0_up + ht15_up
dev.off()

pdf("figures/proseq_15_down.pdf", height = 2, width = 4)
ht0_down + ht15_down
dev.off()




ht0_up %v% ht0_down  
ht15_up %v% ht15_down  
htu %v% htd

ht_list1 <- ht1 %v% ht2

col_fun_pos = colorRamp2(c(-1.5, 0, 1.5), c("white", "#AC404D"))
ht1_pos <- Heatmap(de_proseq_0_pos, show_row_names = FALSE, 
                   show_column_names = FALSE,
                   cluster_columns = FALSE,
                   name = "0' Pos", col = col_fun_pos)

de_proseq_0_pos <- de_proseq_mean %>%
  filter(timepoint == 0, strand == "+") %>%
  pivot_wider(names_from = bp, values_from = mean_coverage) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint, -strand) %>%
  as.matrix()

col_fun_pos = colorRamp2(c(0, 1.5), c("white", "#AC404D"))
ht1_pos <- Heatmap(de_proseq_0_pos, show_row_names = FALSE, 
                   show_column_names = FALSE,
                   cluster_columns = FALSE,
                   name = "0' Pos", col = col_fun_pos)

de_proseq_15_pos <- de_proseq_mean %>%
  filter(timepoint == 15, strand == "+") %>%
  pivot_wider(names_from = bp, values_from = mean_coverage) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint, -strand) %>%
  as.matrix()


ht2_pos <- Heatmap(de_proseq_15_pos, show_row_names = FALSE, 
                   show_column_names = FALSE,
                   cluster_columns = FALSE,
                   name = "15' Pos", col = col_fun_pos)

ht1_pos + ht2_pos




de_proseq_0_neg <- de_proseq_mean %>%
  filter(timepoint == 0, strand == "-") %>%
  pivot_wider(names_from = bp, values_from = mean_coverage) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint, -strand) %>%
  as.matrix()

col_fun_neg = colorRamp2(c(0, -1.5), c("white", "#7F8999"))
ht1_neg <- Heatmap(de_proseq_0_neg, show_row_names = FALSE, 
                   show_column_names = FALSE,
                   cluster_columns = FALSE,
                   name = "0' neg", col = col_fun_neg)

de_proseq_15_neg <- de_proseq_mean %>%
  filter(timepoint == 15, strand == "-") %>%
  pivot_wider(names_from = bp, values_from = mean_coverage) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint, -strand) %>%
  as.matrix()


ht2_neg <- Heatmap(de_proseq_15_neg, show_row_names = FALSE, 
                   show_column_names = FALSE,
                   cluster_columns = FALSE,
                   name = "15' neg", col = col_fun_neg)

ht1_pos + ht1_neg + ht2_pos + ht2_neg
```

```{r}
pheatmap_2layer <- function( hmat.plus, hmat.minus, breaks.plus=NULL, breaks.minus=NULL, cluster_rows = FALSE, cluster_cols = FALSE, legend=FALSE, show_rownames=FALSE, show_colnames=FALSE, silent=T )
{
  get_color_map <- function(pMat, mMat)
  {
    cMap <- matrix("", nrow=NROW(pMat), ncol=NCOL(pMat));
    
    if(is.null(breaks.minus)) breaks.minus <- hmat.minus;
    mMat <-  ( mMat -  min(breaks.minus)) / (max(breaks.minus) -  min(breaks.minus));
    if ( sum(mMat<0)>0 ) mMat[mMat<0] <- 0;
    if ( sum(mMat>1)>0 ) mMat[mMat>1] <- 1;
    
    if(is.null(breaks.plus)) breaks.plus <- hmat.plus;
    pMat <-  ( pMat -  min(breaks.plus)) / (max(breaks.plus) -  min(breaks.plus));
    if ( sum(pMat<0)>0 ) pMat[pMat<0] <- 0;
    if ( sum(pMat>1)>0 ) pMat[pMat>1] <- 1;
    
    for(i in 1:NROW(pMat))
      for(j in 1:NCOL(pMat))
      {
        B <- max( mMat[i,j], 1-pMat[i,j] );
        G <- 1 - max( mMat[i,j], pMat[i,j] );
        R <- max( 1- mMat[i,j], pMat[i,j] );
        
        cMap[i,j] <- rgb(R, G,B);
      }
    return(cMap);
  }
  
  gt <- pheatmap( hmat.plus, cluster_rows = cluster_rows, cluster_cols = cluster_cols, col= rgb( 0, 0, 1 - breaks.plus/max(breaks.plus)), breaks = breaks.plus, legend=legend, show_rownames=show_rownames, show_colnames=show_colnames, silent=silent );
  gt$gtable$grobs[[1]]$children[[1]]$gp$fill <- get_color_map( hmat.plus, hmat.minus);
  
  return(gt);
}


pheatmap_2layer(de_proseq_0_pos, de_proseq_0_neg, breaks.plus = c(0,1.5),
                breaks.minus = c(-1.5,0))
```



```{r}


de_proseq_15 <- de_proseq_mean %>%
  filter(timepoint == 15) %>%
  pivot_wider(names_from = bp, values_from = cov) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint) %>%
  as.matrix()
table(de_proseq_mean$timepoint)
de_proseq_30 <- de_proseq_mean %>%
  filter(timepoint == 30) %>%
  pivot_wider(names_from = bp, values_from = cov) %>%
  column_to_rownames("bidir_number") %>%
  dplyr::select(-timepoint) %>%
  as.matrix()

# de_0_clust <- hclust(dist(de_proseq_0))
# de_proseq_0 <- de_proseq_0[de_0_clust$order,]
# de_proseq_15 <- de_proseq_15[de_0_clust$order,]
# de_proseq_30 <- de_proseq_30[de_0_clust$order,]
rownames(de_proseq_0) == rownames(de_proseq_30)
library(ComplexHeatmap)
dim(de_proseq_0)
library(circlize)
col_fun = colorRamp2(c(-1.5, 0, 1.5), c("#7F8999", "white", "#AC404D"))
ht1 <- Heatmap(de_proseq_0, show_row_names = FALSE, 
               show_column_names = FALSE,
               cluster_columns = FALSE,
               name = "0'", col = col_fun)

ht2 <- Heatmap(de_proseq_15, show_row_names = FALSE, 
               show_column_names = FALSE,
               cluster_columns = FALSE,
               name = "15'", col = col_fun)

ht3 <- Heatmap(de_proseq_30, show_row_names = FALSE, 
               show_column_names = FALSE,
               cluster_columns = FALSE,
               name = "30'", col = col_fun)

ht1+ht2+ht3


library(RColorBrewer)
breaksList = seq(-0.5, 0.5, by = 0.1)
pheatmap(de_proseq_0, cluster_cols = FALSE, cluster_rows = FALSE,
         show_rownames = FALSE, 
         show_colnames = FALSE,  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList)), # Defines the vector of colors for the legend (it has to be of the same lenght of breaksList)
         breaks = breaksList)




library(RColorBrewer)
breaksList = seq(-0.5, 0.5, by = 0.1)
pheatmap(de_proseq_15, cluster_cols = FALSE, cluster_rows = FALSE,
         show_rownames = FALSE, 
         show_colnames = FALSE,  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList)), # Defines the vector of colors for the legend (it has to be of the same lenght of breaksList)
         breaks = breaksList)

library(RcppRoll)
max(hmm$bp)
4000

hmm <- de_proseq_mean %>%
  filter(bidir_number == 256)
region_size <- 300
ggplot(hmm, aes(x = bp, y = mean_coverage, color = timepoint)) +
  geom_point() +
  scale_color_discrete() +
  xlim(4300-region_size,4300+region_size)

ggplot(cov_df %>% filter(bidir_number %in% 1:10), aes(x = bp, y = y_index, fill = coverage)) +
  geom_tile() +
  scale_fill_gradient2(limits = c(-1.5,1.5))


huh_mat <- do.call(rbind,hmm@listData)

huh_df <- huh_mat %>%
  as.data.frame() %>%
  rownames_to_column("bidir_number") %>%
  pivot_longer(2:ncol(.),
               names_to = "bp",
               values_to = "coverage") %>%
  mutate(bp = as.numeric(gsub("V", "", bp)),
         bidir_number = as.numeric(bidir_number)) 

ggplot(huh_df, aes(x = bp, y = bidir_number, fill = coverage)) +
  geom_tile()

library(pheatmap)
dim(hmm_mat)
pheatmap(hmm_mat, cluster_cols = FALSE, show_colnames = FALSE, show_rownames = FALSE)

pheatmap(huh_mat, cluster_cols = FALSE, show_colnames = FALSE, show_rownames = FALSE)
plot(region_data)
```




