---
title: "Identifying doxycycline responders"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(gplots)
library(ggrepel)
library(ggpubr)
source("../util/_plot_theme.R")
source("../util/_util.R")
```

```{r}
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id

g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

tx2gene <- read.csv("../../rnaseq/results/salmon/tx2gene.csv")

### COUNTS
tpm <- read.csv("../../rnaseq/results/salmon/salmon_merged_gene_tpm.csv") %>%
  merge(g2s) %>%
  dplyr::select(gene_id, gene_name, everything())

salmon_gene_counts <- read.csv(file.path("../../rnaseq/results/salmon/",
                                         "salmon_merged_gene_counts.csv")) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()
mode(salmon_gene_counts) <- "integer"

# Put in order
genes <- genes[rownames(salmon_gene_counts)]

### SAMPLE INFO
# Somehow the long timepoints didn't make it into this run.
# TODO: re-run with the long timepoints. Gaaah.
samples <- read.csv("../../rnaseq/samplesheet.csv") %>%
  filter(cell_type != "pMEF",
         firre_induced == "control",
         date_sequenced != "18-Sep",
         timepoint_minutes < 400,
         timecourse_length == "short") %>%
  mutate(firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         timepoint_minutes = factor(timepoint_minutes, 
                                    levels = c(seq(0,360,30))),
         firre_induced = factor(firre_induced, 
                                levels = c("control", "firre_induced")))
rownames(samples) <- samples$sample_id
```

```{r}


if(!file.exists("results/deseq_dox_control_res.csv")) {
  
  ## FULL MODEL
  design_formula <- formula("~ timepoint_minutes")
  ## REDUCED MODEL
  reduced_formula <- formula("~ 1")
  
  ## EXPERIMENTS
  # We're going to make 4 comparisons
  experiments <- c("ESC_KO","ESC_WT", "NPC_KO", "NPC_WT")
  
  ## RUN DESEQ2
  deseq_res <- lapply(experiments, run_control_deseq,
                      counts = salmon_gene_counts, 
                      samples = samples,
                      genes = genes,
                      design_formula = design_formula,
                      reduced_formula = reduced_formula,
                      independent_filtering = TRUE)
  
  # Extract sublists -- shrunken lfc and unshrunken
  res_list <- lapply(deseq_res, "[[","res")
  res_shrunken_list <- lapply(deseq_res, "[[","res_shrunken")
  resdf <- bind_rows(res_list)
  write_csv(resdf, "results/deseq_dox_control_res.csv")
    res_shrunkendf <- bind_rows(res_shrunken_list)
  write_csv(res_shrunkendf, "results/deseq_dox_control_res_shrunken.csv")
}
resdf <- read.csv("results/deseq_dox_control_res.csv")
res_shrunkendf <- read.csv("results/deseq_dox_control_res_shrunken.csv")
```

```{r}
g <- ggplot(resdf, aes(x = pvalue)) 
g + geom_histogram() + facet_grid(cell_type~firre_ko, scales = "free_y") +
  theme_paperwhite() + 
  ggtitle("p-value histogram")
```

```{r}
sig_summary <- resdf %>% 
  mutate(pthresh_01 = padj < 0.01,
         pthresh_05 = padj < 0.05,
         pthresh_10 = padj < 0.10) %>%
  dplyr::select(gene_id, gene_name, cell_type, firre_ko,
                pthresh_01, pthresh_05, pthresh_10) %>%
  unite("experiment", cell_type, firre_ko) %>%
  group_by(experiment) %>%
  distinct() %>%
  summarize(pthresh_01 = length(which(pthresh_01 == TRUE)),
            pthresh_05 = length(which(pthresh_05 == TRUE)),
            pthresh_10 = length(which(pthresh_10 == TRUE))) %>%
  pivot_longer(cols = 2:4, names_to = "pthreshold",
               values_to = "num_DEG") %>%
  mutate(pthreshold = gsub("pthresh_", "", pthreshold),
         pthreshold = as.numeric(pthreshold)/100)
knitr::kable(sig_summary)
ggtexttable(sig_summary)

g <- ggplot(sig_summary, aes(x = experiment, y = num_DEG, fill = factor(pthreshold)))
g + geom_bar(stat = "identity", position = "dodge") + 
  scale_fill_manual(values = c("#424242","#a8404c","#71969F")) +
  ggtitle("DEG count: Dox only")
```

```{r}
pthresh <- 0.05

sig_matrix <- resdf %>% 
  filter(padj < pthresh) %>%
  dplyr::select(gene_id, gene_name, cell_type, firre_ko) %>%
  unite("experiment", cell_type, firre_ko) %>%
  mutate(sig = 1) %>%
  distinct() %>%
  pivot_wider(names_from = experiment, values_from = sig, values_fill = 0) %>%
  dplyr::select(-gene_name) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()

sigdf <- sig_matrix %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>%
  merge(g2s)
write_csv(sigdf, "results/deg_control.csv")

gplots::venn(sig_matrix, simplify = TRUE)
```

```{r}




sig_dyn <- res_shrunkendf %>%
  filter(padj < 0.05) %>%
  mutate(timepoint = gsub("timepoint_minutes_", "", result_name),
         timepoint = gsub("_vs_0", "", timepoint),
         timepoint = as.numeric(timepoint)) %>%
  group_by(gene_id, gene_name, cell_type, firre_ko) %>%
  mutate(max_lfc = log2FoldChange[which.max(abs(log2FoldChange))],
         max_abs_lfc = max(abs(log2FoldChange)))

## We're having a lot of really small fold-changes here.
# Let's calculate the max lfc and filter on that
sig_max_lfc <- sig_dyn %>%
  dplyr::select(gene_id, gene_name, cell_type, 
                firre_ko, max_lfc, max_abs_lfc) %>%
  distinct()

g <- ggplot(sig_max_lfc, aes(x = max_abs_lfc))
g + geom_histogram(bins = 40) + 
  theme_paperwhite() + 
  ggtitle("Max magnitude fold-change")


# sig_dyn <- sig_dyn %>% filter(gene_id %in% genes_overlapping$gene_id)
# sig_dyn <- sig_dyn %>% filter(max_abs_lfc > 0.5)
# Let's label the two genes with the highest fold-changes in the
# last time point.
ltp <- sig_dyn %>% 
  # filter(timepoint == 330 | timepoint == 360) %>%
  filter(timepoint == 330 | timepoint == 360) %>%
  arrange(-abs(log2FoldChange)) %>%
  group_by(cell_type, firre_ko) %>%
  dplyr::slice(1:3)


res_shrunkendf <- res_shrunkendf %>%
  mutate(timepoint = gsub("timepoint_minutes_", "", result_name),
         timepoint = gsub("_vs_0", "", timepoint),
         timepoint = as.numeric(timepoint))

# early_timepoint_genes <- res_shrunkendf %>% filter(gene_name %in% genes_overlapping$gene_name,
#                                                    padj < 0.01,
#                                                    timepoint < 180)
g <- ggplot(res_shrunkendf %>% filter(gene_name %in% genes_overlapping$gene_name),
            aes(x = timepoint, y = log2FoldChange, group = gene_id))
g + geom_hline(yintercept = 0) +
  geom_point(alpha = 0.7) + 
  geom_line(color = "#424242", alpha = 0.2) + 
  facet_wrap(cell_type~firre_ko, scales = "free_y") +
  theme_paperwhite() +
  geom_text_repel(data = ltp,
                  aes(label = gene_name))

g <- ggplot(sig_dyn %>% filter(cell_type == "ESC", gene_name != "Rn7s1"),
            aes(x = timepoint, y = log2FoldChange, group = gene_name))
g + geom_hline(yintercept = 0) +
  geom_point(alpha = 0.7) + 
  geom_line(color = "#424242", alpha = 0.2) + 
  facet_wrap(~firre_ko, scales = "free_y") +
  theme_paperwhite() +
  geom_text_repel(data = ltp,
                  aes(label = gene_name))


esc_wt_pre <- sig_dyn %>%
  filter(cell_type == "ESC", firre_ko == "KO", timepoint < 180,
         abs(log2FoldChange) > 0.3)
paste(esc_wt_pre$gene_name[grep("mt-",esc_wt_pre$gene_name)], collapse = " ")


g <- ggplot(sig_dyn %>% filter(cell_type == "ESC", 
                               firre_ko == "KO",
                               gene_name != "Rn7s1",
                               gene_name %in% esc_wt_pre$gene_name),
            aes(x = timepoint, y = log2FoldChange, group = gene_name))
g + geom_hline(yintercept = 0) +
  geom_point(alpha = 0.7) + 
  geom_line(color = "#424242", alpha = 0.2) + 
  # facet_wrap(~firre_ko, scales = "free_y") +
  theme_paperwhite() +
  geom_text_repel(data = sig_dyn %>% filter(cell_type == "ESC",
                                            firre_ko == "KO",
                                            gene_name != "Rn7s1",
                               gene_name %in% esc_wt_pre$gene_name) %>%
                    filter(timepoint == "90",
                           abs(log2FoldChange) > 0.3),
                  aes(label = gene_name))

g <- ggplot(sig_dyn %>% filter(cell_type == "ESC", gene_name != "Rn7s1",
                               grepl("mt-", gene_name)),
            aes(x = timepoint, y = log2FoldChange, group = gene_name))
g + geom_hline(yintercept = 0) +
  geom_point(alpha = 0.7) + 
  geom_line(color = "#424242", alpha = 0.2) + 
  facet_wrap(~firre_ko, scales = "free_y") +
  theme_paperwhite() +
  geom_text_repel(data = ltp,
                  aes(label = gene_name))

```


```{r}
genes_overlapping <- res_shrunkendf %>% 
  mutate(timepoint = gsub("timepoint_minutes_", "", result_name),
         timepoint = gsub("_vs_0", "", timepoint),
         timepoint = as.numeric(timepoint)) %>%
  filter(padj < 0.05) %>%
  dplyr::select(gene_id, gene_name, cell_type, firre_ko) %>%
  unite("experiment", cell_type, firre_ko) %>%
  mutate(sig = 1) %>%
  distinct() %>%
  pivot_wider(names_from = experiment, values_from = sig, values_fill = 0) %>%
  filter(ESC_KO == 1, ESC_WT == 1, NPC_KO == 1, NPC_WT == 1)

paste(genes_overlapping$gene_name, collapse = " ")
goi_gene_stats <- gene_stats %>% filter(gene_id %in% genes_overlapping$gene_name)

if(plot_gene_profiles == TRUE) {
  
  dir.create("figures/gene_profile_plots", showWarnings = FALSE)
  
  genes_to_plot <- unique(genes_overlapping$gene_id)
genes_to_plot <- c("ENSMUSG00000099021.1")
  
  for(i in 1:length(genes_to_plot)) {
    
    goi <- genes_to_plot[i]
    gene_name <- g2s$gene_name[g2s$gene_id == goi]
    
    gene_stats <- resdf %>% 
      filter(gene_id %in% genes_to_plot,
             result_name != "Intercept") %>%
      dplyr::select(gene_id, gene_name, cell_type, 
                    firre_ko, padj, log2FoldChange) %>%
      group_by(gene_id, gene_name, cell_type, firre_ko, padj) %>%
      summarize(max_abs_l2fc = log2FoldChange[which.max(abs(log2FoldChange))]) %>%
      mutate(padj = paste0("padj = ", signif(padj, digits = 3)))
    
    goi_stats <- gene_stats %>% filter(gene_id == goi)
    
    tpm_goi <- tpm %>% filter(gene_id == goi) %>%
      pivot_longer(3:ncol(.), names_to = "sample_id",
                   values_to = "tpm") %>%
      merge(samples) %>%
      unite("condition", cell_type, firre_ko, firre_induced, sep = ";",
            remove = FALSE)
    
    res_goi <- resdf %>% filter(gene_id == goi)
    
    g_tpm <- ggplot(tpm_goi, aes(x = timepoint_minutes, y = tpm)) + 
      geom_point(aes(color = firre_induced)) + 
      geom_smooth(aes(group = condition, color = firre_induced), method = "loess", 
                  formula = "y ~ x", se = FALSE) +
      facet_grid(cell_type~firre_ko, scales = "free_y") +
      ggtitle(gene_name) +
      theme_paperwhite() + 
      geom_text(data = goi_stats,  aes(x=Inf, y = Inf, 
                                       label = padj), 
                vjust=1.5, hjust=1.1) + 
      theme(legend.title = element_blank(),
            legend.position="bottom")
    show(g_tpm)
    ggsave(paste0("figures/gene_profile_plots/", gene_name, ".png"),
           width = 12, height = 10)
  }
}
```

```{r}
# Looking at the first timepoint
samples_tp <- samples %>%
  filter(cell_type == "ESC",
         firre_ko == "WT")



  counts <- salmon_gene_counts[,samples_tp$sample_id]
  mode(counts) <- "integer"
  
  
  
  # Reorder gencode genes
  genes <- genes[rownames(counts)]
  
  # Check ordering
  stopifnot(all(rownames(samples_tp) == colnames(counts)))
  stopifnot(all(names(genes) == rownames(counts)))
  
  dds <- DESeqDataSetFromMatrix(countData = counts, 
                                colData = samples_tp, 
                                design = formula("~ timepoint_minutes"),
                                rowData = genes)
  
  # Filter low counts
  dds <- dds[rowSums(counts(dds)) >= 10,]
  
  
  dds <- DESeq(dds)
  
  # dds <- DESeq(dds,
  #              test="LRT",
  #              reduced = formula("~ 1"),
  #              parallel=TRUE,
  #              BPPARAM=MulticoreParam(ncores))
  
  
  # Compile results
  res_names <- resultsNames(dds)
  
  
  res_vs_zero <- results(dds, name = res_names[1]) %>%
        as.data.frame() %>%
    rownames_to_column(var = "gene_id") %>%
    merge(g2s) %>%
    filter(padj < 0.05) %>%
    mutate(result_name = res_names[1])
  
  for(i in 2:length(res_names)) {
    print(i)
  }
  
  res_esc_ko <- results(dds, name = "timepoint_minutes_30_vs_0") %>%
        as.data.frame() %>%
    rownames_to_column(var = "gene_id") %>%
    merge(g2s) %>%
    filter(padj < 0.05)
  
    res_npc_ko <- results(dds, name = "timepoint_minutes_30_vs_0") %>%
        as.data.frame() %>%
    rownames_to_column(var = "gene_id") %>%
    merge(g2s) %>%
    filter(padj < 0.05)
    
    res_npc_wt <- results(dds, name = "timepoint_minutes_30_vs_0") %>%
        as.data.frame() %>%
    rownames_to_column(var = "gene_id") %>%
    merge(g2s) %>%
    filter(padj < 0.05)

    
```






