---
title: "ATAC-seq quant: Part 1 through 150 min"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(Rsubread)
source("../util/_setup.R")
```

```{r}
# # let's first grab the consensus peaks to quantify over
# consensus_peaks <- rtracklayer::import("~/ftc_atac/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.saf")


# let's grab all the bam files. 
# fl <- list.files("~/ftc_atac/results/bwa/mergedLibrary/", full.names = TRUE)
# fl <- fl[grep(".bam$", fl)]
# # Let's quantify one file and see how long it takes...
# Sys.time()
# fc <- featureCounts(fl[1:2],
#                     annot.ext = "~/ftc_atac/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.saf",
#                     fracOverlap = 0.2,
#                     useMetaFeatures=TRUE,
#                     isPairedEnd = TRUE,
#                     autosort = FALSE,
#                     # might want to increase this as well. 
#                     nthreads = 6)
# Sys.time()

# Did this step in a separate script since it took 30 minutes.
```

```{r}
load("fc.RData")

consensus_peaks <- rtracklayer::import("~/ftc_atac/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.bed")
genes <- rtracklayer::import("/scratch/Shares/rinn/Michael/genomes/references/Mus_musculus/Gencode/M23/annotation/gencode.vM23.annotation.genes.gtf")

chr2_peaks <- consensus_peaks[which(consensus_peaks@seqnames == "chr2")]

q=GRanges(seqnames="chr2",
          ranges=IRanges(start = 122189685, end = 122506587))
roi_peaks <- subsetByOverlaps(chr2_peaks,q)
122299685
122250000
122396587-122299685
122396587
122430000

1122430000-122250000
roi_peaksg <- chr2_peaks[which(chr2_peaks$name %in% roi_peaks$name)]

unique(consensus_peaks@seqnames)
shf <- genes[which(genes$gene_name == "Shf")]
shf

duox1 <- genes[which(genes$gene_name == "Duox1")]
duox1

roi <- rtracklayer::import("roi.bed")



totals <- rowSums(counts)
hist(log10(totals))

# Let's filter out the zeros
length(which(totals < 20))


```

```{r}
nbinom

require(foreign)
require(ggplot2)
require(MASS)

# Let's get the data in proper order here.

library(DESeq2)
```

```{r}
# Okay, let's just run deseq2 with treating the different genetic backgrounds as replicates.. which they kind of are.
counts <- fc$counts

# Filter to just chr2 peaks
counts <- counts[which(rownames(counts) %in% roi_peaksg$name),]

keep <- rowSums(counts) >= 100
counts <- counts[keep, ]

colnames(counts) <- sapply(colnames(counts), function(x) {
  unlist(strsplit(unlist(strsplit(x, "rary.", fixed = TRUE))[[2]], ".mLb.clN"))[[1]]
})
samples <- data.frame("sample_name" = colnames(counts))
samples <- samples %>% separate(sample_name, into = c("genotype", "timepoint", "replicate"), remove = FALSE)
samples$condition <- "control"
samples[grep("iTg", samples$genotype), "condition"] <- "firre_induced"

samples$timepoint <- factor(samples$timepoint, levels = seq(0,150,30))
samples$condition <- factor(samples$condition, levels = c("control", "firre_induced"))

rownames(samples) <- samples$sample_name

counts <- counts[,samples$sample_name]
stopifnot(all(rownames(samples) == colnames(counts)))


dds <- DESeqDataSetFromMatrix(counts,
                              colData = samples,
                              design = formula("~ condition + timepoint + condition*timepoint"))
dds <- DESeq(dds, 
             test="LRT", 
             reduced = ~ condition + timepoint,
             parallel=TRUE,
             BPPARAM=MulticoreParam(12))

resultsNames(dds)
res <- results(dds, name = "conditionfirre_induced.timepoint120") %>%
  as.data.frame() %>%
  rownames_to_column(var = "peak_id")




peak_bed <- consensus_peaks[which(consensus_peaks$name == "Interval_141244")]
rtracklayer::export(peak_bed, "duox_shf_peak.bed")
```


```{r}
# Okay, so we achieve no significant peaks for this test. This could be for a number of reasons.
# Let's just plot one that has a high fold change.

	
# Interval_125350


dds_counts <- counts(dds, normalized = T)
peak_id <- "Interval_141244"
poi_counts <- dds_counts[which(row.names(dds_counts) == peak_id),] %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample_name")
names(poi_counts)[2] <- "normalized_count"
poi_counts <- poi_counts %>% merge(samples)
g <- ggplot(poi_counts, aes(x = timepoint, y = normalized_count, color = condition, group = genotype))
g + geom_point() + geom_line() + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(peak_id)


peak_id <- "Interval_141234"
poi_counts <- dds_counts[which(row.names(dds_counts) == peak_id),] %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample_name")
names(poi_counts)[2] <- "normalized_count"
poi_counts <- poi_counts %>% merge(samples)
g <- ggplot(poi_counts, aes(x = timepoint, y = normalized_count, color = condition, group = genotype))
g + geom_point() + geom_line() + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(peak_id)


peak_id <- "Interval_206953"
poi_counts <- dds_counts[which(row.names(dds_counts) == peak_id),] %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample_name")
names(poi_counts)[2] <- "normalized_count"
poi_counts <- poi_counts %>% merge(samples)
g <- ggplot(poi_counts, aes(x = timepoint, y = normalized_count, color = condition, group = genotype))
g + geom_point() + geom_line() + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(peak_id)

peak_id <- "Interval_242421"
poi_counts <- dds_counts[which(row.names(dds_counts) == peak_id),] %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample_name")
names(poi_counts)[2] <- "normalized_count"
poi_counts <- poi_counts %>% merge(samples)
g <- ggplot(poi_counts, aes(x = timepoint, y = normalized_count, color = condition, group = genotype))
g + geom_point() + geom_line() + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(peak_id)


peak_id <- "Interval_72902"
poi_counts <- dds_counts[which(row.names(dds_counts) == peak_id),] %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample_name")
names(poi_counts)[2] <- "normalized_count"
poi_counts <- poi_counts %>% merge(samples)
g <- ggplot(poi_counts, aes(x = timepoint, y = normalized_count, color = condition, group = genotype))
g + geom_point() + geom_line() + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(peak_id)


peak_id <- "Interval_146810"
poi_counts <- dds_counts[which(row.names(dds_counts) == peak_id),] %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample_name")
names(poi_counts)[2] <- "normalized_count"
poi_counts <- poi_counts %>% merge(samples)
g <- ggplot(poi_counts, aes(x = timepoint, y = normalized_count, color = condition, group = genotype))
g + geom_point() + geom_line() + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(peak_id)


peak_id <- "Interval_141248"

	
peak_id <- "Interval_146810"
poi_counts <- dds_counts[which(row.names(dds_counts) == peak_id),] %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample_name")
names(poi_counts)[2] <- "normalized_count"
poi_counts <- poi_counts %>% merge(samples)
g <- ggplot(poi_counts, aes(x = timepoint, y = normalized_count, color = condition, group = genotype))
g + geom_point() + geom_line() + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(peak_id)

peak_id<- "Interval_144154"
poi_counts <- dds_counts[which(row.names(dds_counts) == peak_id),] %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample_name")
names(poi_counts)[2] <- "normalized_count"
poi_counts <- poi_counts %>% merge(samples)
g <- ggplot(poi_counts, aes(x = timepoint, y = normalized_count, color = condition, group = genotype))
g + geom_point() + geom_line() + 
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(peak_id)

res_sig <- res %>% filter(pvalue < 0.1)


# Let's see if there's a difference in baseMean for the peaks that are not sig
bm <- res %>% filter(pvalue > 0.9) %>% dplyr::select(baseMean)
bm$condition <- "highp"
bml <- res %>% filter(pvalue < 0.01) %>% dplyr::select(baseMean)
bml$condition <- "lowp"
bmt <- bind_rows(bm, bml)
g <- ggplot(bmt, aes(x = log10(baseMean), fill = condition))
g + geom_density(alpha = 0.1)
```

