---
title: "Delta Differential Expression: Firre WT mESC dox"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
params:
  design: !r formula("~ timepoint")
  contrasts: !r list(
       c(factor = "timepoint",
        numerator = "30",
        denominator = "0"),
        c(factor = "timepoint",
        numerator = "60",
        denominator = "30"),
        c(factor = "timepoint",
        numerator = "90",
        denominator = "60"),
        c(factor = "timepoint",
        numerator = "120",
        denominator = "90"),
        c(factor = "timepoint",
        numerator = "150",
        denominator = "120"),
        c(factor = "timepoint",
        numerator = "180",
        denominator = "150"),
        c(factor = "timepoint",
        numerator = "210",
        denominator = "180"),
        c(factor = "timepoint",
        numerator = "240",
        denominator = "210"),
        c(factor = "timepoint",
        numerator = "270",
        denominator = "240"),
        c(factor = "timepoint",
        numerator = "300",
        denominator = "270"),
        c(factor = "timepoint",
        numerator = "330",
        denominator = "300"))
  alpha: 0.1
  lfc_threshold: 0
  data_dir: !r file.path("data")
  results_dir: !r file.path("results")
editor_options: 
  chunk_output_type: console
always_allow_html: yes
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, message=FALSE, warning=FALSE}
source("../_setup.R")
source("../helper_functions.R")
library(tidyverse); library(tximport); library(bcbioRNASeq); library(pals)
library(lemon); library(scales); library(ggrepel); library(colorspace)
```


```{r load_data, message=FALSE}
# # Gene annotation data from script 00.
# load(paste0(params$data_dir,"/gene_annotations.RData"))
counts_combined <- read.table("../../results/featureCounts/merged_gene_counts.txt",
                              header = T)
g2s <- counts_combined[,c(1,2)]
names(g2s) <- c("gene_id", "gene_name")

genes <- rtracklayer::import("../Mus_musculus.GRCm38.81_genes.gtf")

# Read in sample annotations
samples <- read.csv("../../samplesheet.csv") %>%
# Filter to just the non-inducible mESC knockout samples.
 filter(cell_type == "mESCs", 
        firre == "WT",
        contains_itg_dna == FALSE) %>%
   mutate(timepoint = factor(timepoint, 
         levels = as.character(seq(from = 0, to = 330, by = 30))))
rownames(samples) <- samples$id

# Read in the featureCounts.
counts <- fCountReader("../../results/featureCounts/gene_counts/",
                       samples$id, 
                       "_read1Aligned.sortedByCoord.out_gene.featureCounts.txt")


# Ensure that the ordering of the columns in the counts matrix 
# is the same as in the sample sheet.
samples <- samples[colnames(counts$count),]
stopifnot(all(rownames(samples) == colnames(counts$count)))


# Ensure that the ordering of the rows (gene info) 
# is the same between the counts and the gencode
names(genes) <- genes$gene_id
genes <- genes[rownames(counts$count)]
stopifnot(all(names(genes) == rownames(counts$count)))
```

We don't have the option to test the interaction term here 
(Firre:Transgene) since we don't have every condition. In particular,
we don't have the uninducible transgene in the Firre WT background.

```{r create_DESeqDataSet, message=FALSE}
# Create DESeq2 experiment objects.
dds <- DESeqDataSetFromMatrix(countData = counts$count, 
                              colData = samples, 
                              design = params$design,
                              rowData = genes)
```

#### DESeq dataset

```{r prefilter}
# Note that this criteria can be made more stringent
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep, ]
print(dds)
```


```{r deseq, results='hide', message=FALSE}
# dds <- DESeq(dds, test="LRT", reduced = ~1)
dds <- DESeq(dds)
# suppressMessages(saveData(dds, dir = params$data_dir))
```

```{r}
stopifnot(is.list(params$contrasts))
contrasts <- params$contrasts
names <- vapply(
    X = contrasts,
    FUN = function(x) {
        paste(x[[1]], x[[2]], "vs", x[[3]], sep = "_")
        
    },
    FUN.VALUE = "character"
)
names(contrasts) <- names
```

```{r res_unshrunken}
res_list_unshrunken <- mapply(
    FUN = results,
    contrast = contrasts,
    MoreArgs = list(
        object = dds,
        alpha = params$alpha
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = FALSE
)
names(res_list_unshrunken) <- names
# saveData(res_list_unshrunken, dir = params$data_dir)
```

```{r res_shrunken}
library(BiocParallel)
register(MulticoreParam(16))
res_list_shrunken <- mapply(
    FUN = lfcShrink,
    res = res_list_unshrunken,
    contrast = contrasts,
    MoreArgs = list(
        dds = dds,
        type = "normal",
        parallel = TRUE
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = TRUE
)
```

```{r results_tables, results="asis"}
# Here we're creating subset tables of the DEGs, and adding the normalized
# counts used by DESeq2 for the differential expression analysis.
res_tbl_list <- mapply(
    FUN = resultsTables,
    results = res_list_shrunken,
    MoreArgs = list(
        counts = dds,
        alpha = params$alpha,
        lfcThreshold = params$lfc,
        summary = TRUE,
        headerLevel = 2,
        write = TRUE,
        dir = paste0(params$results_dir,"/wt_dox_shrunken_deltas", collapse = "")
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = TRUE
)
```

```{r}
fl <- list.files("results/wt_dox_shrunken_deltas/", full.names = T)
fl <- fl[grep("all", fl)]
dat <- read.csv(gzfile(fl[[1]]))
dat$contrast <- fl[[1]]
for(i in 2:length(fl)) {
  tdat <- read.csv(gzfile(fl[[i]]))
  tdat$contrast <- fl[[i]]
  dat <- bind_rows(dat, tdat)
}

write.csv(dat, "wt_dox_shrunken_deltas.csv", row.names = FALSE)
```

```{r}
# Let's put each of these significant genes in a table with which timepoint
# they were significant in.
deg <- data.frame("gene_id" = character(),
                  "gene_name" = character(),
                  "baseMean" = numeric(),
                  "log2FoldChange" = numeric(),
                  "lfcSE" = numeric(),
                  "stat" = numeric(),
                  "pvalue" = numeric(),
                  "padj" = numeric(),
                  "contrast" = character())

for(i in 1:length(res_list_unshrunken)) {
  temp_deg <- res_list_unshrunken[[i]][which(res_list_unshrunken[[i]]$padj < params$alpha & 
                   abs(res_list_unshrunken[[i]]$log2FoldChange) > params$lfc_threshold),] %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene_id") %>%
  merge(., g2s) %>%
  select("gene_id", "gene_name", everything())
  if(nrow(temp_deg) > 0) {
  temp_deg$contrast <- names(res_list_unshrunken)[[i]]
  deg <- bind_rows(deg, temp_deg)
  }
}


deg_matrix <- deg %>% select(gene_name, log2FoldChange, contrast) %>%
  spread(key = contrast, value = log2FoldChange)




deg$significant <- "sig"

deg_profiles <- data.frame("gene_id" = character(),
                  "gene_name" = character(),
                  "baseMean" = numeric(),
                  "log2FoldChange" = numeric(),
                  "lfcSE" = numeric(),
                  "stat" = numeric(),
                  "pvalue" = numeric(),
                  "padj" = numeric(),
                  "contrast" = character())


sig_genes <- unique(deg$gene_name)
to_fetch <- unique(deg[which(deg$gene_name %in% sig_genes),"gene_id"])

for(i in 1:length(res_list_unshrunken)) {
  temp_deg <- res_list_unshrunken[[i]][rownames(res_list_unshrunken[[i]]) %in% to_fetch,] %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene_id") %>%
  merge(., g2s) %>%
  select("gene_id", "gene_name", everything())
  if(nrow(temp_deg) > 0) {
  temp_deg$contrast <- names(res_list_unshrunken)[[i]]
  deg_profiles <- bind_rows(deg_profiles, temp_deg)
  }
}

deg_profiles <- merge(deg_profiles, deg %>% select(gene_id, contrast, significant),
                      all.x = TRUE)
deg_profiles[which(is.na(deg_profiles$significant)), "significant"] <- "ns"
```

```{r}
# Okay, let's make a plot of how many genes change in each timepoint up or down.

deg$direction <- "up"
deg[which(deg$log2FoldChange < 0), "direction"] <- "down"

deg_summary <- deg %>%
  group_by(contrast, direction) %>%
  summarize(deg_count = n())
# let's add those that are not included.
cts <- unique(deg_profiles$contrast)
missing_contrasts_up <- data.frame("contrast" = cts[which(!(cts %in% deg_summary$contrast))],
                                direction = "up",
                                deg_count = 0)
missing_contrasts_down <- data.frame("contrast" = cts[which(!(cts %in% deg_summary$contrast))],
                                direction = "down",
                                deg_count = 0)
deg_summary <- bind_rows(deg_summary, missing_contrasts_up)
deg_summary <- bind_rows(deg_summary, missing_contrasts_down)



deg_summary$contrast <- gsub("timepoint_", "", deg_summary$contrast)
print(deg_summary$contrast)
deg_summary$contrast <- factor(deg_summary$contrast, levels = c("30_vs_0",
                                                                "60_vs_30",
                                                                "90_vs_60",
                                                                "120_vs_90",
                                                                "150_vs_120",
                                                                "180_vs_150",
                                                                "210_vs_180",
                                                                "240_vs_210",
                                                                "270_vs_240",
                                                                "300_vs_270",
                                                                "330_vs_300"))

g <- ggplot(deg_summary, aes(x = contrast, y = deg_count, fill = direction))
g + geom_bar(stat = "identity", position = "dodge") + 
  ggtitle("Dox mESC WT | delta DEG count")
# ggsave("dox_mesc_wt_delta_deg_count.pdf")
```
