---
title: "ATAC-seq"
author: "Michael Smallegan"
date: "9/29/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(DESeq2)
source("../util/_plot_theme.R")
```

```{r}
# Import data
atac_consensus <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.annotatePeaks.txt", sep = "\t", header = T)
atac_consensus_boolean <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.boolean.annotatePeaks.txt", sep = "\t", header = T)
colnames(atac_consensus_boolean)
atac_consensus_boolean$ESC_KO_control_120_R1.mLb.clN.fc

atac_c_fc <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.featureCounts.txt", skip = 1, sep = "\t", header = T)

```

```{r}
# Let's run a linear model and determine which peaks are different about
# Apparently this consensus peak approach is prone to false positives
# https://support.bioconductor.org/p/108458/#108469
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4066778/
# https://support.bioconductor.org/p/109154/
colnames(atac_c_fc) <- gsub(".mLb.clN.bam","",colnames(atac_c_fc))
rownames(atac_c_fc) <- atac_c_fc$Geneid
interval_table <- atac_c_fc[,1:6]
counts <- atac_c_fc[,7:ncol(atac_c_fc),drop=FALSE] %>% as.matrix()

samples <- data.frame("sample_name" = colnames(counts)) %>%
  mutate(sample_name = gsub("firre_induced", "firreinduced", sample_name)) %>%
  separate(sample_name, into = c("cell_type", "firre_ko", "firre_induced", "timepoint_minutes", "replicate"), remove = FALSE) %>%
  mutate(sample_name = gsub("firreinduced", "firre_induced", sample_name),
         firre_induced = gsub("firreinduced", "firre_induced", firre_induced)) %>%
  dplyr::select(-replicate)

samples$firre_ko <- factor(samples$firre_ko, levels = c("WT", "KO"))
samples$firre_induced <- factor(samples$firre_induced, levels = c("control", "firre_induced"))
samples$timepoint_minutes <- factor(samples$timepoint_minutes, 
                                    levels = seq(0,150,30))
rownames(samples) <- samples$sample_name

stopifnot(all(colnames(counts) == rownames(samples)))


```

```{r}
## FULL MODEL
design_formula <- formula(paste0("~ timepoint_minutes + firre_induced + firre_ko +",
                                   "firre_induced*timepoint_minutes"))
## REDUCED MODEL
reduced_formula <- formula("~ timepoint_minutes + firre_induced + firre_ko")

dds <- DESeqDataSetFromMatrix(countData = round(counts), 
                               colData = samples, 
                               design = design_formula)

dds <- DESeq(dds, test="LRT", 
             reduced = reduced_formula)

(res_names <- resultsNames(dds))

res <- results(dds, name = res_names[2]) %>% 
    as.data.frame() %>%
    rownames_to_column(var = "interval_id") %>%
    mutate(result_name = res_names[2]) 

 for(i in 3:length(res_names)) {
    tmp_res <- results(dds, 
                       name = res_names[i]) %>%
      as.data.frame() %>%
      rownames_to_column(var = "interval_id") %>%
      mutate(result_name = res_names[i]) 
    res <- bind_rows(res, tmp_res)
 }

write_csv(res, "results/atac_timecourse_de.csv")

sig_res <- res %>% filter(padj < 0.1)
```


```{r}
# Let's check out what's happening near genes that are changing in response to 
# Firre. 
fr <- read_csv("../01_firre_induction_vs_control/results/deseq_res.csv")
unique(fr$comparison)
fr_sig <- fr %>% filter(padj < 0.05, cell_type == "ESC") %>%
  group_by(gene_id, gene_name, firre_ko) %>%
  summarize(max_fc = log2FoldChange[which.max(abs(log2FoldChange))]) %>%
  pivot_wider(names_from = firre_ko, values_from = max_fc)

 

atac_fr_sig <- atac_consensus[atac_consensus$Nearest.PromoterID %in% fr_sig$gene_id,]
names(atac_fr_sig)[1] <- "interval_id"

fr_res <- res %>% filter(res$interval_id %in% atac_fr_sig$interval_id)

fr_res <- merge(fr_res, atac_fr_sig %>% 
                  dplyr::select(interval_id, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End))
names(fr_res)[9] <- "gene_id"

fr_res <- merge(fr_res, fr_sig)
fr_res$abs_dist_tss <- abs(fr_res$Distance.to.TSS)
fr_res <- fr_res %>% unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)
```

```{r}
# Prepare for UCSC
consensus_bed <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.bed")
consensus_bed <- consensus_bed %>% filter(V1 %in% c(paste0("chr",1:19), "chrM", "chrX", "chrY"))
write.table(consensus_bed, "results/consensus_peaks.mLb.clN.bed", 
            quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "\t")
unique(consensus_bed$V1)

# Prepare trackhub trackDB
atac_bigwigs <- list.files("../../atacseq/results/bwa/mergedLibrary/bigwig/",
                           pattern = "*.bigWig")
atac_bigwig_name <- gsub("_R1.mLb.clN.bigWig", "", atac_bigwigs)
atac_bigwig_name <- gsub("ESC_", "", atac_bigwig_name)

lines <- list()
for(i in 1:length(atac_bigwigs)) {
  write(paste0("track  ", atac_bigwig_name[[i]], "
bigDataUrl https://biof-trackhub.colorado.edu/rinnlab/micahel/firre_timecourse/firre_atacseq/", atac_bigwigs[[i]], "
shortLabel ",atac_bigwig_name[[i]] , "
longLabel ", atac_bigwig_name[[i]], "
type bigWig
visibility full\n\n"), "results/atacseq_trackDb.txt", append = T)
}

```



