---
title: "ATAC-seq"
author: "Michael Smallegan"
date: "9/29/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(DESeq2)
source("../util/_plot_theme.R")
```

```{r}
# Import data
atac_consensus <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.annotatePeaks.txt", sep = "\t", header = T)
atac_consensus_boolean <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.boolean.annotatePeaks.txt", sep = "\t", header = T)
colnames(atac_consensus_boolean)
atac_consensus_boolean$ESC_KO_control_120_R1.mLb.clN.fc

atac_c_fc <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.featureCounts.txt", skip = 1, sep = "\t", header = T)

```

```{r}
library(bamsignals)

library(Rsamtools)

atac_consensus_gr <- GRanges(seqnames = atac_consensus$Chr,
                             IRanges(start = atac_consensus$Start,
                                     end = atac_consensus$End))

fl <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary/", pattern = ".bam$", full.names = T)
peak_coverage <- sapply(fl, function(x) bamCount(x, atac_consensus_gr[92261], paired.end = "midpoint"))
peak_coverage
peak_coverage_df <- data.frame(counts = peak_coverage) %>%
  rownames_to_column("filepath") %>%
  mutate(sample_name = gsub("/scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//",
                            "", filepath),
         sample_name = gsub(".mLb.clN.sorted.bam", "", sample_name)) %>%
  merge(samples)

hmm <- merge(peak_coverage_df, test_counts)
plot(hmm$feature_counts, hmm$counts)
# plot(peak_coverage[[1]]@signals[[1]])
# plot(peak_coverage[[4]]@signals[[1]])

?bamProfile
# awk '{print NF}' firre_atac_treat_pileup_sorted.bedGraph > ncols.txt
hmm <- read.table("/scratch/Shares/rinn/Michael/firre_timecourse/analysis/07_atacseq/bin/ncols.txt")
```


```{r}
# Let's run a linear model and determine which peaks are different about
# Apparently this consensus peak approach is prone to false positives
# https://support.bioconductor.org/p/108458/#108469
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4066778/
# https://support.bioconductor.org/p/109154/
colnames(atac_c_fc) <- gsub(".mLb.clN.bam","",colnames(atac_c_fc))
rownames(atac_c_fc) <- atac_c_fc$Geneid
interval_table <- atac_c_fc[,1:6]
counts <- atac_c_fc[,7:ncol(atac_c_fc),drop=FALSE] %>% as.matrix()

samples <- data.frame("sample_name" = colnames(counts)) %>%
  mutate(sample_name = gsub("firre_induced", "firreinduced", sample_name)) %>%
  separate(sample_name, into = c("cell_type", "firre_ko", "firre_induced", "timepoint_minutes", "replicate"), remove = FALSE) %>%
  mutate(sample_name = gsub("firreinduced", "firre_induced", sample_name),
         firre_induced = gsub("firreinduced", "firre_induced", firre_induced)) %>%
  dplyr::select(-replicate)

samples$firre_ko <- factor(samples$firre_ko, levels = c("WT", "KO"))
samples$firre_induced <- factor(samples$firre_induced, levels = c("control", "firre_induced"))
samples$timepoint_minutes <- factor(samples$timepoint_minutes, 
                                    levels = seq(0,150,30))
rownames(samples) <- samples$sample_name

stopifnot(all(colnames(counts) == rownames(samples)))
wt_samples <- samples %>% filter(firre_ko == "WT")
ko_samples <- samples %>% filter(firre_ko == "KO")


wt_counts <- counts[,rownames(wt_samples)]
ko_counts <- counts[,rownames(ko_samples)]
```

```{r}
# Let's plot some things
# Normalize to total mapped reads
mapped_reads <- read_tsv("data/samtools_alignments.tsv")
names(mapped_reads)[1] <- "sample_name"
mapped_reads <- mapped_reads %>% 
  dplyr::select(-Unmapped) %>%
  mutate(mapped_M = Mapped / 10e6) %>%
  dplyr::select(-Mapped)

mapped_reads$sample_name <- gsub("_T1", "", mapped_reads$sample_name)
mapped_reads <- mapped_reads %>%
  filter(!grepl("_T2", sample_name))

counts_df <- counts %>% 
  as.data.frame() %>%
  rownames_to_column("interval_id") %>%
  pivot_longer(2:ncol(.), names_to = "sample_name", values_to = "counts") %>%
  merge(samples)

mapped_reads$sample_name %in% unique(counts_df$sample_name)
length(unique(mapped_reads$sample_name))
nrow(mapped_reads)
counts_df <- counts_df %>% merge(mapped_reads)
counts_df <- counts_df %>% 
  mutate(cpm = counts / mapped_M)

test_counts <- counts_df %>% filter(interval_id == "Interval_76872") %>%
  dplyr::select(sample_name, counts) %>%
  dplyr::rename(feature_counts = counts)

# 132119
# 76872
ggplot(roi, aes(x = counts, y = cpm)) + geom_point()

roi <- counts_df %>% filter(grepl("76872", interval_id))
ggplot(roi, aes(x = timepoint_minutes, y = cpm, group = interval_id)) + 
  geom_point() + 
  facet_grid(firre_ko~firre_induced) + 
  geom_line()


```

ATAC-seq normalization
https://epigeneticsandchromatin.biomedcentral.com/articles/10.1186/s13072-020-00342-y


```{r}
# ## FULL MODEL
# design_formula <- formula(paste0("~ timepoint_minutes + firre_induced + firre_ko +",
#                                    "firre_induced*timepoint_minutes"))
# ## REDUCED MODEL
# reduced_formula <- formula("~ timepoint_minutes + firre_induced + firre_ko")

samples$condition <- "control"
samples$timepoint <- as.numeric(as.character(samples$timepoint_minutes))
samples$condition[which(samples$timepoint> 0 & samples$firre_induced == "firre_induced")] <- "firre"
colnames(counts)
counts <- counts[,rownames(samples)]
stopifnot(all(colnames(counts) == rownames(samples)))
mode(counts) <- "integer"
samples$condition <- factor(samples$condition, levels = c("control", "firre"))
dds <- DESeqDataSetFromMatrix(countData = counts, 
                               colData = samples, 
                               design = ~ condition)

dds <- DESeq(dds)
res_df <- results(dds) %>%
  as.data.frame() %>%
  rownames_to_column("interval_id") %>%
  merge(atac_consensus %>% 
                  dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End)) %>% 
  unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)
names(res_df)[8] <- "gene_name"
names(res_df)[9] <- "gene_id"



# From manual annotation in the browser
res_df$gene_name[which(res_df$interval_id == "Interval_4124")] <- "Pde6d"
res_df$gene_name[which(res_df$interval_id == "Interval_138095")] <- "Arih1"
	
res_df$gene_name[which(res_df$interval_id == "Interval_76872")] <- "Rapgef4"
res_df$gene_name[which(res_df$interval_id == "Interval_43032")] <- "Ddhd1"
res_df$firre_responder <- FALSE
res_df$firre_responder[which(res_df$gene_name %in% unique(sig_fr$gene_name))] <- TRUE

ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), color = firre_responder)) + geom_point() + 
  geom_text_repel(data = res_df %>% filter(firre_responder == TRUE,
                                           padj < 0.05), aes(label = gene_name))

wt_dds <- DESeqDataSetFromMatrix(countData = round(wt_counts), 
                               colData = wt_samples, 
                               design = ~ firre_induced)

wt_dds <- DESeq(wt_dds)
(res_names <- resultsNames(wt_dds))

wt_res <- results(wt_dds, name = res_names[2]) %>% 
    as.data.frame() %>%
    rownames_to_column(var = "interval_id") %>%
    mutate(result_name = res_names[2]) 
names(atac_consensus)[1] <- "interval_id"
wt_res <- merge(wt_res, atac_consensus %>% 
                  dplyr::select(interval_id, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End)) %>% 
  unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)

wt_res_shrunken <- lfcShrink(dds = wt_dds,
                            coef = res_names[2],
                            type = "apeglm") %>% 
    as.data.frame() %>%
    rownames_to_column(var = "interval_id") %>%
    mutate(result_name = res_names[2]) 

wt_res_shrunken <- wt_res_shrunken %>%
  merge(atac_consensus %>% 
                  dplyr::select(interval_id, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End)) %>% 
  unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)

wt_res_sig <- wt_res %>% filter(padj < 0.05)
names(wt_res_sig)[9] <- "gene_id"

write_csv(wt_res_shrunken, "results/wt_atac_de_intervals_shrunken.csv")
write_csv(wt_res, "results/wt_atac_de_intervals.csv")


ko_dds <- DESeqDataSetFromMatrix(countData = round(ko_counts), 
                               colData = ko_samples, 
                               design = ~ firre_induced)

ko_dds <- DESeq(ko_dds)
(res_names <- resultsNames(ko_dds))

ko_res <- results(ko_dds, name = res_names[2]) %>% 
    as.data.frame() %>%
    rownames_to_column(var = "interval_id") %>%
    mutate(result_name = res_names[2]) 
names(atac_consensus)[1] <- "interval_id"
ko_res <- merge(ko_res, atac_consensus %>% 
                  dplyr::select(interval_id, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End)) %>% 
  unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)

ko_res_sig <- ko_res %>% filter(padj < 0.05)
names(ko_res_sig)[9] <- "gene_id"

ko_res_shrunken <- lfcShrink(dds = ko_dds,
                            coef = res_names[2],
                            type = "apeglm") %>% 
    as.data.frame() %>%
    rownames_to_column(var = "interval_id") %>%
    mutate(result_name = res_names[2]) 

ko_res_shrunken <- ko_res_shrunken %>%
  merge(atac_consensus %>% 
                  dplyr::select(interval_id, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End)) %>% 
  unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)


write_csv(ko_res_shrunken, "results/ko_atac_de_intervals_shrunken.csv")
write_csv(ko_res, "results/ko_atac_de_intervals.csv")



# write_csv(wt_res, "results/wt_atac_timecourse_de.csv")
# 
# sig_res <- res %>% filter(padj < 0.1)

ko_wt_overlapping <- ko_res_sig[which(ko_res_sig$interval_id %in% wt_res_sig$interval_id),]


```


```{r}
ko_res_shrunken <- read_csv("results/ko_atac_de_intervals_shrunken.csv") %>%
  filter(padj < 0.1)
names(ko_res_shrunken)[8] <- "gene_id"
wt_res_shrunken <- read_csv("results/wt_atac_de_intervals_shrunken.csv") %>%
  filter(padj < 0.1)
names(wt_res_shrunken)[8] <- "gene_id"

# Let's check out what's happening near genes that are changing in response to 
# Firre. 
fr <- read_csv("../01_firre_induction_vs_control/results/deseq_res.csv")

sig_fr <- fr %>% filter(comparison == "dynamic_firre_induction_vs_control", cell_type == "ESC") %>%
  filter(padj < 0.05, log2FoldChange > 0.2)


ggplot(sig_fr, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()

unique(fr$comparison)
fr <- fr %>% filter(cell_type == "ESC") %>%
  group_by(gene_id, gene_name, firre_ko) %>%
  summarize(max_fc = log2FoldChange[which.max(abs(log2FoldChange))]) %>%
  pivot_wider(names_from = firre_ko, values_from = max_fc)

wt_res_sig_fr <- merge(wt_res_shrunken, fr, all.x = T)
ko_res_sig_fr <- merge(ko_res_shrunken, fr, all.x = T)

# atac_fr_sig <- atac_consensus[atac_consensus$Nearest.PromoterID %in% fr_sig$gene_id,]
# names(atac_fr_sig)[1] <- "interval_id"

fr_res <- wt_res %>% filter(wt_res$interval_id %in% atac_fr_sig$interval_id)

fr_res <- merge(fr_res, atac_fr_sig %>% 
                  dplyr::select(interval_id, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End))
names(fr_res)[9] <- "gene_id"

fr_res <- merge(fr_res, fr_sig)
fr_res$abs_dist_tss <- abs(fr_res$Distance.to.TSS)
fr_res <- fr_res %>% unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE) %>%
  filter(padj < 0.05)
```

```{r}
# Prepare for UCSC
consensus_bed <- read.table("../../atacseq/results/bwa/mergedLibrary/macs/broadPeak/consensus/consensus_peaks.mLb.clN.bed")
consensus_bed <- consensus_bed %>% filter(V1 %in% c(paste0("chr",1:19), "chrM", "chrX", "chrY"))
write.table(consensus_bed, "results/consensus_peaks.mLb.clN.bed", 
            quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "\t")
unique(consensus_bed$V1)

# Prepare trackhub trackDB
atac_bigwigs <- list.files("../../atacseq/results/bwa/mergedLibrary/bigwig/",
                           pattern = "*.bigWig")
atac_bigwig_name <- gsub("_R1.mLb.clN.bigWig", "", atac_bigwigs)
atac_bigwig_name <- gsub("ESC_", "", atac_bigwig_name)

lines <- list()
for(i in 1:length(atac_bigwigs)) {
  write(paste0("track  ", atac_bigwig_name[[i]], "
bigDataUrl https://biof-trackhub.colorado.edu/rinnlab/micahel/firre_timecourse/firre_atacseq/", atac_bigwigs[[i]], "
shortLabel ",atac_bigwig_name[[i]] , "
longLabel ", atac_bigwig_name[[i]], "
type bigWig
visibility full\n\n"), "results/atacseq_trackDb.txt", append = T)
}

```

```{r}
# lm 
ko_fi_samples <- samples %>% filter(firre_ko == "KO", firre_induced == "firre_induced")
ko_fi_samples$timepoint_minutes <- as.numeric(ko_fi_samples$timepoint_minutes)
ko_fi_samples <- ko_fi_samples %>% arrange(timepoint_minutes)
ko_fi_counts <- counts[,ko_fi_samples$sample_name]
library(broom)

ko_fi_lm_df <- tidy(lm(seq(0,150,30)~ko_fi_counts[1,]))
ko_fi_lm_df$interval_id <- rownames(ko_fi_counts)[1]
for(i in 51425:nrow(ko_fi_counts)) {
  if((i %% 10000) == 0) print(i)
  tmp_df <- tidy(lm(seq(0,150,30)~ko_fi_counts[i,]))
  tmp_df$interval_id <- rownames(ko_fi_counts)[i]
  ko_fi_lm_df <- bind_rows(ko_fi_lm_df, tmp_df)
}




```

```{r}
fl <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary/", pattern = ".bam$", full.names = T)
cat(fl)
```

```{bash}
# Combine all data and call peaks on everything combined.
#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=atac_macs
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=24:00:00
#SBATCH --output=atac_macs.out
#SBATCH --error=atac_macs.err
date; hostname

source ~/anaconda3/bin/activate macs3

macs3 callpeak -f BAMPE -t /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_control_0_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_control_120_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_control_150_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_control_30_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_control_60_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_control_90_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_firre_induced_0_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_firre_induced_120_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_firre_induced_150_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_firre_induced_30_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_firre_induced_60_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_KO_firre_induced_90_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_control_0_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_control_120_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_control_150_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_control_30_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_control_60_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_control_90_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_firre_induced_0_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_firre_induced_120_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_firre_induced_150_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_firre_induced_30_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//ESC_WT_firre_induced_90_R1.mLb.clN.sorted.bam -g mm -n firre_atac -B -q 0.1
date
```

```{bash}
# Convert bedGraph to bigwig
sort -k1,1 -k2,2n firre_atac_treat_pileup.bdg > firre_atac_treat_pileup_sorted.bedGraph
  sed '78360883d' firre_atac_treat_pileup_sorted.bedGraph > firre_atac_treat_pileup_sorted_cleaned.bedGraph
  sed -n '113645386,113645388p' firre_atac_treat_pileup_sorted_cleaned_sorted_again.bedGraph
  bedSort firre_atac_treat_pileup_sorted_cleaned_sorted_again.bedGraph firre_atac_treat_pileup_sorted_cleaned_sorted_again.bedGraph
bedGraphToBigWig firre_atac_treat_pileup_sorted_cleaned.bedGraph /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/genome/GRCm38.p6.genome.fa.sizes firre_atac_treat_pileup.bigWig

bedSort firre_atac_control_lambda.bdg firre_atac_control_lambda_sorted.bedGraph



bedGraphToBigWig firre_atac_control_lambda_sorted.bedGraph /scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/genome/GRCm38.p6.genome.fa.sizes firre_atac_control_lambda_sorted.bigWig
 

```

```{r}
# Make file ready for UCSC
atac_peaks <- rtracklayer::import("bin/firre_atac_peaks.narrowPeak") %>%
  as.data.frame() %>%
  filter(seqnames %in% c(paste0("chr", 1:19), "chrX", "chrY", "chrM")) %>%
  dplyr::select(seqnames, start, end, name, score, strand, signalValue, pValue, qValue, peak) %>%
  mutate(strand = ".")
# atac_peaks <- atac_peaks[which(seqnames(atac_peaks) %in% c(paste0("chr", 1:19), "chrX", "chrY", "chrM"))]  
# rtracklayer::export(atac_peaks, "results/atac_peaks_merge.narrowPeak", format = "narrowPeak")
?export


write_tsv(atac_peaks, "results/atac_peaks_merge.narrowPeak")
```

```{r}
# Let's get the counts for these peak regions
atac_peaks_gr <- GRanges(seqnames = atac_peaks$seqnames,
                         IRanges(start = atac_peaks$start,
                                 end = atac_peaks$end))


saf <- atac_peaks %>% select(name, seqnames, start, end, strand)
names(saf) <- c("GeneID", "Chr", "Start", "End", "Strand")
write.table(saf, "results/atac_peaks.saf", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")


peak_counts <- sapply(bams, function(x) bamCount(x, atac_peaks_gr, paired.end = "midpoint"))

```


```{r}
bams <- list.files("/scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary/", pattern = ".bam$", full.names = T)
names(bams) <- gsub("/scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//|.mLb.clN.sorted.bam", "", bams)
for(i in 1:length(bams)) {
script <- paste0("#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=atac_fc_",names(bams)[i],"
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --mem=20gb
#SBATCH --time=3:00:00
#SBATCH --output=atac_fc_",names(bams)[i],".out
#SBATCH --error=atac_fc_",names(bams)[i],".err

date; hostname
module load subread/1.6.2
featureCounts \\
-F SAF \\
-O \\
--fracOverlap 0.2 \\
-T 4 \\
-p --donotsort \\
-a /scratch/Shares/rinn/Michael/firre_timecourse/analysis/07_atacseq/results/atac_peaks.saf \\
-o /scratch/Shares/rinn/Michael/firre_timecourse/analysis/07_atacseq/results/featureCounts/",names(bams)[i],".featureCounts.txt \\
",bams[i],"
date
")
write_lines(script, paste0("bin/featureCounts/fc_", names(bams)[i], ".sh"))
}

scripts <- c("#!/bin/bash", paste0("sbatch ", list.files("bin/featureCounts/")))
write_lines(scripts, "bin/featureCounts/submit.sh")
system("source /etc/profile; cd bin/featureCounts; chmod u+x submit.sh; ./submit.sh")
```

```{r}


fc <- lapply(paste0("results/featureCounts/", names(bams), ".featureCounts.txt"),
             read_tsv, skip = 1)
counts <- sapply(fc, function(x) x[[7]])
colnames(counts) <- gsub("scratch/Shares/rinn/Michael/firre_timecourse/atacseq/results/bwa/mergedLibrary//|.mLb.clN.sorted.bam", "", sapply(fc, function(x) colnames(x)[[7]]))
colnames(counts) <- gsub("/","",colnames(counts))
rownames(counts) <- fc[[1]]$Geneid
mode(counts) <- "integer"

samples <- data.frame("sample_name" = colnames(counts)) %>%
  mutate(sample_name = gsub("firre_induced", "firreinduced", sample_name)) %>%
  separate(sample_name, into = c("cell_type", "firre_ko", "firre_induced", "timepoint", "r"), remove = FALSE) %>%
  mutate(sample_name = gsub("firreinduced", "firre_induced", sample_name),
         firre_induced = gsub("firreinduced", "firre_induced", firre_induced),
         timepoint = as.numeric(timepoint),
         condition = ifelse(timepoint > 0 & firre_induced == "firre_induced", "firre", "control"),
         rownames = sample_name,
         condition = factor(condition, levels = c("control", "firre"))) %>%
  column_to_rownames("rownames") %>%
  dplyr::select(-r)
stopifnot(all(colnames(counts) == rownames(samples)))

dds <- DESeqDataSetFromMatrix(countData = counts, 
                               colData = samples, 
                               design = ~ condition)

dds <- DESeq(dds)

atac_annotated <- read.table("bin/firre_atac_peaks.annotatePeaks.txt", sep = "\t", header = T)
names(atac_annotated)[1] <- "name"

res_df <- results(dds) %>%
  as.data.frame() %>%
  rownames_to_column("name") %>%
  merge(atac_annotated %>% 
          dplyr::select(name, Gene.Name, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End))

write_csv(res_df, "results/atac_merged_deseq.csv")
  # merge(atac_consensus %>% 
  #                 dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, Distance.to.TSS, Chr, Start, End)) %>% 
  # unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
  # unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)
# firre_atac_peak_83137 -- RAPGEF4
# firre_atac_peak_4531 -- PDE6D
# firre_atac_peak_149640 -- ARIH1
# firre_atac_peak_46566 -- Ddhd1
# firre_atac_peak_147750 -- Pknox2?
man_annotation <- read_csv("data/firre_peaks_manually_annotated.csv", col_names = c("name", "Gene.Name"))
# Update res_df
for(i in 1:nrow(man_annotation)) {
res_df$Gene.Name[which(res_df$name == man_annotation$name[i])] <- man_annotation$Gene.Name[i]
}
fr <- read_csv("../01_firre_induction_vs_control/results/deseq_res.csv")
sig_fr <- fr %>% filter(comparison == "dynamic_firre_induction_vs_control", cell_type == "ESC") %>%
  filter(padj < 0.05, log2FoldChange > 0.2)

res_df$firre_responder <- FALSE
res_df$firre_responder[which(res_df$Gene.Name %in% unique(sig_fr$gene_name))] <- TRUE
library(ggrepel)
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), color = firre_responder)) +  
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), lty = 2) +
  geom_text_repel(data = res_df %>% filter(firre_responder == TRUE,
                                           padj < 0.05), aes(label = Gene.Name)) + 
  ggtitle("Firre ATAC-seq differential peaks") 
ggsave("figures/firre_atacseq_differential_peaks.pdf", useDingbats = FALSE)

ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()

ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point() + 
  geom_smooth()
```

```{bash}
annotatePeaks.pl \
/scratch/Shares/rinn/Michael/firre_timecourse/analysis/07_atacseq/bin/firre_atac_peaks.narrowPeak \
/scratch/Shares/rinn/genomes/Mus_musculus/Gencode/M25/GRCm38.p6.genome.fa \
-gid \
-gtf /scratch/Shares/rinn/genomes/Mus_musculus/Gencode/M25/gencode.vM25.annotation.gtf \
-cpu 6 \
> firre_atac_peaks.annotatePeaks.txt
```


```{r}
# Motifs in top peaks
top_peaks <- res_df %>% filter(padj < 0.05)
top_peaks_gr <- GRanges(seqnames = top_peaks$Chr,
                        IRanges(start = top_peaks$Start,
                                end = top_peaks$End))
top_peaks_gr$name <- top_peaks$name
rtracklayer::export(top_peaks_gr, "results/top_peaks.bed")


```

```{bash}
module load bedtools
bedtools getfasta -fi /scratch/Shares/rinn/genomes/Mus_musculus/Gencode/M25/GRCm38.p6.genome.fa -fo top_peaks.fa -bed top_peaks.bed -name+ 


fimo --verbosity 1 --text --thresh 0.05 \
--bfile mm10_background_frequencies.txt /scratch/Shares/rinn/Michael/firre_timecourse/analysis/07_atacseq/data/motif_databases/MOUSE/HOCOMOCOv11_core_MOUSE_mono_meme_format.meme \
top_peaks.fa > top_peaks_fimo.txt
```


```{r}
fimo_res <- read_tsv("results/top_peaks_fimo.txt")
fimo_sig <- fimo_res %>% filter(`p-value` < 1e-6)
fimo_sig$name <- sapply(fimo_sig$sequence_name, function(x) {
  unlist(strsplit(x, "::"))[[1]]
})
fimo_sig$dbp <- sapply(fimo_sig$motif_id, function(x) {
  unlist(strsplit(x, "_"))[[1]]
})

fr_peaks <- res_df %>% filter(firre_responder == TRUE)
fr_peaks <- unique(fr_peaks$name)
fr_peaks <- fr_peaks[fr_peaks %in% fimo_sig$name]

motif_peaks <- fimo_sig %>%
  group_by(dbp, name) %>%
  summarize(count = n(),
            occurrence = as.numeric(count > 0)) %>%
  dplyr::select(-count) %>%
  filter(name %in% fr_peaks) %>%
  pivot_wider(names_from = name, values_from = occurrence, values_fill = 0) %>%
  column_to_rownames("dbp") %>%
  as.matrix()

tfs <- data.frame(count = rowSums(motif_peaks)) %>%
  rownames_to_column("dbp")


motif_merge <- fimo_sig %>%
  filter(name %in% fr_peaks) %>%
  group_by(dbp, name) %>%
  summarize(count = n()) %>%
  pivot_wider(names_from = dbp, values_from = count, values_fill = 0)

hmm <- merge(res_df, motif_merge)

fr_motif_matrix <- hmm %>%
  dplyr::select(Gene.Name, ASCL1:ZN335) %>%
  column_to_rownames("Gene.Name") %>%
  as.matrix()

pheatmap::pheatmap(fr_motif_matrix)
names(hmm)
up_peaks <- res_df %>% filter(log2FoldChange > 0)
up_peaks <- up_peaks$name
up_peaks <- up_peaks[up_peaks %in% colnames(motif_peaks)]

up_motif_peaks <- motif_peaks[,up_peaks]

up_tfs <- data.frame(count = rowSums(up_motif_peaks)) %>%
  rownames_to_column("dbp") %>%
  mutate(percent = count / ncol(up_motif_peaks) *100) %>%
  filter(count > 4)
ncol(up_motif_peaks)
hist(up_tfs$count)

motif_merge <- fimo_sig %>%
  filter(dbp %in% up_tfs$dbp) %>%
  group_by(dbp, name) %>%
  summarize(count = n()) %>%
  pivot_wider(names_from = dbp, values_from = count, values_fill = 0)

hmm <- merge(res_df, motif_merge)

ggplot(hmm, aes(x = log2FoldChange, y = MAZ)) + 
  geom_point()

ggplot(hmm, aes(x = log2FoldChange, y = ZBT17)) + 
  geom_point()

motif_merge <- fimo_sig %>%
  filter(dbp %in% up_tfs$dbp) %>%
  group_by(dbp, name) %>%
  summarize(count = n(),
            occurrence = as.numeric(count > 0)) %>%
  dplyr::select(-count) %>%
  pivot_wider(names_from = dbp, values_from = occurrence, values_fill = 0) %>%
  column_to_rownames("name") %>%
  as.matrix()

pheatmap::pheatmap(motif_merge)
fr_tfs <- data.frame(count = rowSums(fr_motif_peaks)) %>%
  rownames_to_column("dbp") %>%
  filter(count > 0)



```








