---
title: "Firre binders"
author: "Michael Smallegan"
date: "9/25/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(readxl)
library(DESeq2)
library(pheatmap)
library(ggrepel)
source("../util/_plot_theme.R")
```

## Which Firre binders are expressed in these cells?

```{r}
system("cd data; wget https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-019-13235-w/MediaObjects/41467_2019_13235_MOESM4_ESM.xlsx")
incprint <- read_excel("data/41467_2019_13235_MOESM4_ESM.xlsx") %>%
  filter(`FIRRE interactor` == "Firre") %>%
  dplyr::select(`Protein name`, `Ensembl Gene ID`) %>%
  rename("Protein name" = "gene_name",
         "Ensembl Gene ID"  = "gene_id")
write_csv(incprint, "results/firre_binders.csv")

cat(paste(incprint$gene_name, collapse = "\n"))
samples <- read.csv("../../rnaseq/rnaseq_samplesheet.csv")
tpm <- read_csv("../../rnaseq/results/salmon/salmon_merged_gene_tpm.csv")
genes <- rtracklayer::import("../util/gencode.vM25.annotation.genes.gtf")
g2s <- genes[genes$type == "gene"] %>% as.data.frame() %>% dplyr::select(gene_id, gene_name)
tpm <- merge(tpm, g2s)


tpm_gene_id <- sapply(tpm$gene_id, function(x) {
  unlist(strsplit(x, ".", fixed = T))[[1]]
})
fb_tpm <- tpm[which(tolower(tpm$gene_name) %in% tolower(incprint$gene_name)),] %>%
  dplyr::select(gene_id, gene_name, everything()) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id",
               values_to = "tpm") %>%
  merge(samples) %>%
  group_by(gene_id, gene_name, cell_type, firre_ko, firre_induced) %>%
  summarize(mean_log10_tpm = log10(mean(tpm, na.rm = T)+0.1)) %>%
  unite(condition, cell_type, firre_ko, firre_induced) %>%
  pivot_wider(names_from = condition, values_from = mean_log10_tpm)
write_csv(fb_tpm, "results/firre_binders_tpm.csv")

```

```{r}

# Deal with duplicated gene name
fb_tpm$gene_name[duplicated(fb_tpm$gene_name)] <- paste0(fb_tpm$gene_name[duplicated(fb_tpm$gene_name)],"_2")
fb_tpm_matrix <- fb_tpm %>% 
  ungroup() %>%
  dplyr::select(-gene_id) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()
pheatmap(fb_tpm_matrix)

```

```{r}
# Decide on an expression cutoff
hist(fb_tpm_matrix, breaks = seq(0,max(fb_tpm_matrix)+10,1), xlim = c(0,100))

```

```{r}
# Which genes are different between cell types?
salmon_gene_counts <- read.csv(file.path("../../rnaseq/results/salmon/",
                                         "salmon_merged_gene_counts.csv")) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()
mode(salmon_gene_counts) <- "integer"
names(genes) <- genes$gene_id
genes <- genes[rownames(salmon_gene_counts)]

ko_samples <- samples %>% filter(cell_type %in% c("ESC", "NPC"),
                                 firre_ko == "KO")
rownames(ko_samples) <- ko_samples$sample_id

ko_counts <- salmon_gene_counts[,as.character(ko_samples$sample_id)]
stopifnot(all(rownames(ko_samples) == colnames(ko_counts)))


ko_dds <- DESeqDataSetFromMatrix(countData = ko_counts, 
                                colData = ko_samples, 
                                design = ~ cell_type + timecourse_length + firre_induced,
                                rowData = genes)
ko_dds <- ko_dds[rowSums(counts(ko_dds)) >= 20,]
ko_dds <- DESeq(ko_dds)
res_names <- resultsNames(ko_dds)
ko_res <- results(ko_dds, 
               name = res_names[2])
ko_resdf <- ko_res %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
  merge(g2s) %>%
  arrange(padj)



wt_samples <- samples %>% filter(cell_type %in% c("ESC", "NPC"),
                                 firre_ko == "WT")
rownames(wt_samples) <- wt_samples$sample_id

wt_counts <- salmon_gene_counts[,as.character(wt_samples$sample_id)]
stopifnot(all(rownames(wt_samples) == colnames(wt_counts)))


wt_dds <- DESeqDataSetFromMatrix(countData = wt_counts, 
                                colData = wt_samples, 
                                design = ~ cell_type + timecourse_length + firre_induced,
                                rowData = genes)
wt_dds <- wt_dds[rowSums(counts(wt_dds)) >= 20,]
wt_dds <- DESeq(wt_dds)
res_names <- resultsNames(wt_dds)
wt_res <- results(wt_dds, 
               name = res_names[2])
wt_resdf <- wt_res %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
  merge(g2s) %>%
  arrange(padj)
write_csv(wt_resdf, "results/wt_ESC_vs_NPC_res.csv")
write_csv(ko_resdf, "results/ko_ESC_vs_NPC_res.csv")
```


```{r}


# Plot a few example genes
goi_tpm <- tpm %>% 
  dplyr::select(gene_id, gene_name, everything()) %>%
  filter(gene_id == ko_resdf$gene_id[2]) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples) %>%
   unite(condition, cell_type, firre_ko, firre_induced)
g <- ggplot(goi_tpm, aes(x = condition, y = tpm))
g + geom_point()

goi_tpm <- tpm %>% 
  dplyr::select(gene_id, gene_name, everything()) %>%
  filter(gene_id == wt_resdf$gene_id[2]) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samples) %>%
   unite(condition, cell_type, firre_ko, firre_induced)
g <- ggplot(goi_tpm, aes(x = condition, y = tpm))
g + geom_point()


```

```{r}
# Okay, how many of these firre binders are different across cell types?
fb_ko_deg <- ko_resdf %>% filter(tolower(gene_name) %in% tolower(incprint$gene_name))
fb_wt_deg <- wt_resdf %>% filter(tolower(gene_name) %in% tolower(incprint$gene_name))

fb_ko_deg$firre_ko <- "KO"
fb_wt_deg$firre_ko <- "WT"

fb_deg <- bind_rows(fb_ko_deg, fb_wt_deg)


fb_degw <- fb_deg %>% 
  dplyr::select(gene_id, gene_name, log2FoldChange, padj, firre_ko) %>%
  pivot_wider(id_cols = c("gene_id", "gene_name"),
              names_from = firre_ko, values_from = c(log2FoldChange, padj))

g <- ggplot(fb_degw, aes(x = log2FoldChange_WT, y = log2FoldChange_KO, 
                         color = padj_WT < 0.01, label = gene_name))
g + geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point() + 
  geom_text_repel(data = fb_degw %>% filter(log2FoldChange_WT > 1.5 | log2FoldChange_WT < -2)) + 
  ggtitle("ESC vs NPC: Firre binders") + 
  theme_paperwhite()

# They seem to be perfectly correlated across WT and KO,
# so let's make a cutoff an decide which ones are DEG
fb_deg <- fb_degw %>% filter(padj_KO < 0.05, padj_WT < 0.05,
                             log2FoldChange_KO > 1,
                             log2FoldChange_WT > 1)
fb_deg$gene_name

write_csv(fb_degw, "results/firre_binders_DEG_ESC_vs_NPC.csv")
```

```{r}
# Now let's see how many are different between wt and ko in each cell type
firre_wtvsko_esc <- read_csv("../03_firre_knockout/results/firre_ko_vs_wt_esc_res.csv")
firre_wtvsko_npc <- read_csv("../03_firre_knockout/results/firre_ko_vs_wt_npc_res.csv")


fb_wtvsko_esc <- firre_wtvsko_esc %>% filter(tolower(gene_name) %in% tolower(incprint$gene_name))
fb_wtvsko_npc <- firre_wtvsko_npc %>% filter(tolower(gene_name) %in% tolower(incprint$gene_name))

fb_wtvsko_esc$cell_type <- "ESC"
fb_wtvsko_npc$cell_type <- "NPC"
fb_wtvsko <- bind_rows(fb_wtvsko_esc, fb_wtvsko_npc)
fb_wtvsko_w <- fb_wtvsko %>%
  dplyr::select(gene_id, gene_name, log2FoldChange, padj, cell_type) %>%
  pivot_wider(id_cols = c("gene_id", "gene_name"),
              names_from = cell_type, values_from = c(log2FoldChange, padj))

g <- ggplot(fb_wtvsko_w, aes(x = log2FoldChange_ESC, y = log2FoldChange_NPC))
g +  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point()

g <- ggplot(fb_wtvsko_w, aes(x = log2FoldChange_ESC, y = log2FoldChange_NPC,
                             color = padj_ESC < 0.05, label = gene_name))
g +geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) + 
  geom_point() + 
  geom_text_repel(data = fb_wtvsko_w %>% filter(padj_ESC < 0.05)) +
  ggtitle("Zero TP Firre WT vs KO: ESC sig")

g <- ggplot(fb_wtvsko_w, aes(x = log2FoldChange_ESC, y = log2FoldChange_NPC,
                             color = padj_NPC < 0.05, label = gene_name))
g + geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point() +
  geom_text_repel(data = fb_wtvsko_w %>% filter(padj_NPC < 0.05)) +
  ggtitle("Zero TP Firre WT vs KO: NPC sig")


fb_deg_wtvsko_esc <- fb_wtvsko_w %>% filter(padj_ESC < 0.01)
fb_deg_wtvsko_npc <- fb_wtvsko_w %>% filter(padj_NPC < 0.01)

write_csv(fb_wtvsko_w, "results/firre_binders_DEG_WT_vs_KO.csv")
```

```{r}
tfea_res <- read_csv("../04_tfea/results/firre_induction_tfea_res.csv")
fb_tfea_res <- tfea_res %>% filter(tolower(tf_name) %in% tolower(incprint$gene_name))
cat(paste(incprint$gene_name, collapse = "\n"))
si <- read.table("data/string_associated_with_firre_binders.txt")

fb_tfea_res <- tfea_res %>% filter(tolower(tf_name) %in% tolower(si$V1))

tfs_in_tfea <- data.frame("tf_name"= unique(tfea_res$tf_name))

system("cd data; wget https://stringdb-static.org/download/protein.links.v11.0/10090.protein.links.v11.0.txt.gz; gunzip 10090.protein.links.v11.0.txt.gz")
system("cd data; gunzip 10090.protein.links.v11.0.txt.gz")
```










