---
title: "tf analysis"
author: "Michael Smallegan"
date: "9/3/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
options(stringsAsFactors = FALSE)
library(tidyverse); library(RcisTarget)
```


```{r}
# okay, Rcis target just needs the gene names. 
load("../gene2symbol.RData")
```



```{r}
# let's consider making some gene lists now. Which lists do we want to check on?
# Well for one it would be nice to do the significant genes in each timepoint 
# and make the waterfall chart.
# However, if I don't have time for that or if this takes a long time to run than 
# I will for sure want to get the rank aggregated list. 
# now let's make some gene lists. 
tc <- read.csv("../timecourse_shrunken_lfc_results.csv")
kodat <- read.csv("../combined_shrunken_lfc_results.csv")
kodat <- kodat[which(!grepl("dxz4", kodat$contrast)),]
# let's also get the rank aggregation lists.

ra_all <- read.csv("../rank_aggreg_all_data.csv")  %>% arrange(Score)
ra_90 <- read.csv("rank_aggreg_tc90min_only_data.csv") %>% arrange(Score)

ra_all_top100 <- ra_all[1:100,c("geneName", "Score")]
ra_90_top100 <- ra_90[1:100,c("geneName", "Score")]


gl <- list("ra_all_top100" = ra_all_top100[which(ra_all_top100$geneName != ""),"geneName"],
           "ra_90_top100" = ra_90_top100[which(ra_90_top100$geneName != ""),"geneName"])

tc$timepoint <- unlist(lapply(tc$contrast, function(x) {
  as.numeric(unlist(strsplit(x, "_", fixed = TRUE))[[4]])
}))


doxko <- tc[grep("ko_dox", tc$contrast), ] %>%
  filter(padj < 0.1, abs(log2FoldChange) > 0.2)
doxwt <- tc[grep("wt_dox", tc$contrast), ] %>%
  filter(padj < 0.1, abs(log2FoldChange) > 0.2)
firreko <-  tc[grep("ko_firre", tc$contrast), ] %>%
  filter(padj < 0.1, abs(log2FoldChange) > 0.2)
firrewt <-  tc[grep("wt_firre", tc$contrast), ] %>%
  filter(padj < 0.1, abs(log2FoldChange) > 0.2)

# now let's go through each timepoint and make a list. 
tps <- seq(30,330, by = 30)
for(i in 1:length(tps)) {
  tdf <- doxko[which(doxko$timepoint == tps[[i]]), c("gene_name", "gene_id")]
  tdf <- tdf %>% as_tibble()
  gl <- c(gl, tdf[which(tdf$gene_name != ""),"gene_name"])
  names(gl)[length(gl)] <- paste0("doxko_",tps[[i]])
}

for(i in 1:length(tps)) {
  tdf <- doxwt[which(doxwt$timepoint == tps[[i]]), c("gene_name", "gene_id")]
  tdf <- tdf %>% as_tibble()
  gl <- c(gl, tdf[which(tdf$gene_name != ""),"gene_name"])
  names(gl)[length(gl)] <- paste0("doxwt_",tps[[i]])
}


for(i in 1:length(tps)) {
  tdf <- firrewt[which(firrewt$timepoint == tps[[i]]), c("gene_name", "gene_id")]
  tdf <- tdf %>% as_tibble()
  gl <- c(gl, tdf[which(tdf$gene_name != ""),"gene_name"])
  names(gl)[length(gl)] <- paste0("firrewt_",tps[[i]])
}


for(i in 1:length(tps)) {
  tdf <- firreko[which(firreko$timepoint == tps[[i]]), c("gene_name", "gene_id")]
  tdf <- tdf %>% as_tibble()
  gl <- c(gl, tdf[which(tdf$gene_name != ""),"gene_name"])
  names(gl)[length(gl)] <- paste0("firreko_",tps[[i]])
}



# https://resources.aertslab.org/cistarget/

```


```{r}
# https://resources.aertslab.org/cistarget/databases/mus_musculus/mm10/refseq_r80/mc9nr/gene_based/mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather
data(motifAnnotations_hgnc)
motifRankings <- importRankings("data/mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather")
?cisTarget

txtFile <- paste(file.path(system.file('examples', package='RcisTarget')),
                 "hypoxiaGeneSet.txt", sep="/")
geneLists <- list(hypoxia=read.table(txtFile, stringsAsFactors=FALSE)[,1])

load("../../11_short_timecourse_combined/results/short_vszero_sig.RData", verbose = T)
load("../../09_short_timecourse_overexpression/results/wt_overexp_short_vszero_sig.RData", verbose = T)
load("../../06_short_timecourse_rescue/results/ko_rescue_short_vszero_sig.RData", verbose = TRUE)
load("../../05_long_timecourse_rescue/results/ko_rescue_long_vszero_sig.RData", verbose = TRUE)
load("../../08_long_timecourse_overexpression/results/wt_overexp_long_vszero_sig.RData", verbose = TRUE)

all_sig <- bind_rows(short_vszero_sig %>%
  dplyr::select(gene_id, gene_name) %>%
  mutate(experiment = ("short_combined")),
wt_overexp_short_vszero_sig %>%
  dplyr::select(gene_id, gene_name) %>%
  mutate(experiment = ("short_OE")),
ko_rescue_short_vszero_sig %>%
  dplyr::select(gene_id, gene_name) %>%
  mutate(experiment = ("short_Res")),
ko_rescue_long_vszero_sig %>%
  dplyr::select(gene_id, gene_name) %>%
  mutate(experiment = ("long_Res")),
wt_overexp_long_vszero_sig %>%
  dplyr::select(gene_id, gene_name) %>%
  mutate(experiment = ("long_OE"))) %>%
  distinct()
# Turn into gene list 
gl <- list(short_combined = all_sig %>% filter(experiment == "short_combined") %>% pull(gene_name),
           short_OE = all_sig %>% filter(experiment == "short_OE") %>% pull(gene_name),
           short_Res = all_sig %>% filter(experiment == "short_Res") %>% pull(gene_name),
           long_Res = all_sig %>% filter(experiment == "long_Res") %>% pull(gene_name),
           long_OE = all_sig %>% filter(experiment == "long_OE") %>% pull(gene_name))
# https://github.com/aertslab/RcisTarget/raw/master/data/motifAnnotations_mgi.RData
load("data/motifAnnotations_mgi.RData", verbose = TRUE)
# Motif enrichment analysis:
motifEnrichmentTable_wGenes <- cisTarget(gl, motifRankings,
                               motifAnnot=motifAnnotations_mgi)
write_csv(motifEnrichmentTable_wGenes, "results/firre_responder_promoter_motif_enrichment.csv")
# Let's look at the overlaps for the short timecourse
tfs_short_OE <- motifEnrichmentTable_wGenes %>%
  filter(geneSet == "short_OE")

all_tfs <- data.frame(tfs = c(motifEnrichmentTable_wGenes$TF_highConf, motifEnrichmentTable_wGenes$TF_lowConf)) %>%
  filter(tfs != "") %>%
  mutate(tfs = gsub("\\(directAnnotation).", "", tfs),
         tfs = gsub("\\(inferredBy_Orthology).", "", tfs),
         tfs = gsub("\\(inferredBy_MotifSimilarity).", "", tfs),
         tfs = gsub("\\(inferredBy_MotifSimilarity_n_Orthology).", "", tfs)) %>%
  separate_rows(tfs, sep = "\\s*;\\s*") %>%
  separate_rows(tfs, sep = "\\s") %>%
  distinct() %>%
  filter(tfs != "")

all_tfs$tfs[tolower(all_tfs$tfs) %in% tolower(all_sig$gene_name)]

```


Tetracyclines such as doxycycline are known to inhibit the translation of mtDNA genes, but not nuclear genes. This leads to a disruption of the protein balance between nuclear mitochondrial genes and mtDNA mitochondrial proteins. This then results in changes in the nuclear transcription profile.

```{r}
# http://mitominer.mrc-mbu.cam.ac.uk/release-4.0/results.do?trail=%257Cquery
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4565776/
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4631686/
```

```{r}
#hmm , okay. Let's start with the doxko
doxko30 <- motifEnrichmentTable_wGenes[which(motifEnrichmentTable_wGenes$geneSet == 	
"doxko_30"),]
```



