---
title: "Long timecourse overexpression"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(seriation)
source("../util/_plot_theme.R")
source("../util/_util.R")
source("../01_setup/assumptions.R")

# Keep track of thresholds for figure naming
thresh <- paste0("pval", pval_thresh, "_l2fc", round(l2fc_thresh, 2))
```

```{r load, include=FALSE}
load("../01_setup/results/rnaseq_data.RData")
```

For this experiment, we induced the expression of Firre with an rTTA element by adding doxycycline to the cells. We see that Firre is indeed expressed in the KO background after the addition of doxycycline. The drug does instigate some gene expression changes on its own, so we will control for the effects by using a linear model which accounts for the effect of dox.

```{r long_timecourse}
if(!file.exists("results/wt_overexp_long.RData")) {

  # Filter to ESC KO long timecourse
  wt_overexp_long_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "long",
           firre_ko == "WT")
  wt_overexp_long_counts <- salmon_gene_counts[,wt_overexp_long_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(wt_overexp_long_samples) == colnames(wt_overexp_long_counts)))
  stopifnot(all(rownames(wt_overexp_long_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  wt_overexp_long_dds <- DESeqDataSetFromMatrix(countData = wt_overexp_long_counts, 
                                          colData = wt_overexp_long_samples, 
                                          design = ~ firre_induced + timepoint + timepoint*firre_induced)
  wt_overexp_long_dds <- DESeq(wt_overexp_long_dds, test = "LRT", reduced = ~ firre_induced + timepoint)
  
  # Compile results
  res_names <- resultsNames(wt_overexp_long_dds)
  dynamic_res <- res_names[grepl("firre_inducedfirre_induced.timepoint", res_names)]
  
  wt_overexp_long_lfc <- lapply(dynamic_res, function(x) {
    results(wt_overexp_long_dds, 
            name = x) %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Shrunken LFC results
  wt_overexp_long_shrnklfc <- lapply(dynamic_res, function(x) {
    lfcShrink(wt_overexp_long_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Calculate the maximum fold-change in any one timepoint
  wt_overexp_long_maxfc <- wt_overexp_long_shrnklfc %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  wt_overexp_long_shrnklfc <- wt_overexp_long_shrnklfc %>%
    left_join(wt_overexp_long_maxfc)
  
  save(wt_overexp_long_lfc, wt_overexp_long_shrnklfc, file = "results/wt_overexp_long.RData")
}

load("results/wt_overexp_long.RData", verbose = TRUE)
```

### Fold changes vs zero timepoint

This is without considering the control cell line.

```{r}
if(!file.exists("results/wt_overexp_long_vs_zero.RData")) {

  wt_overexp_long_vszero_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "long",
           firre_ko == "WT",
           firre_induced == "firre_induced")
  wt_overexp_long_vszero_counts <- salmon_gene_counts[,wt_overexp_long_vszero_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(wt_overexp_long_vszero_samples) == colnames(wt_overexp_long_vszero_counts)))
  stopifnot(all(rownames(wt_overexp_long_vszero_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  wt_overexp_long_vszero_dds <- DESeqDataSetFromMatrix(countData = wt_overexp_long_vszero_counts,
                                                 colData = wt_overexp_long_vszero_samples,
                                                 design = ~ timepoint)
  wt_overexp_long_vszero_dds <- DESeq(wt_overexp_long_vszero_dds)
  res_names <- resultsNames(wt_overexp_long_vszero_dds)
  
  vs_zero_res <- res_names[grepl("_vs_0", res_names)]
  wt_overexp_long_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(wt_overexp_long_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  save(wt_overexp_long_vszero_shrnklfc, file = "results/wt_overexp_long_vs_zero.RData")
}

load("results/wt_overexp_long_vs_zero.RData", verbose = T)
```

### Long timecourse call significant genes

We'll make the p-value cutoff based on the dox controlled model and the l2fc cutoff
based on the fold change vs zero.

```{r}
wt_overexp_long_dox_sig <- wt_overexp_long_shrnklfc %>% 
  filter(padj <= pval_thresh)

wt_overexp_long_vszero_sig <- wt_overexp_long_vszero_shrnklfc %>%
  filter(gene_id %in% wt_overexp_long_dox_sig$gene_id)

wt_overexp_long_vszero_maxfc <- wt_overexp_long_vszero_sig %>%
  group_by(gene_id) %>%
  summarize(max_fc = max(abs(log2FoldChange))) 

wt_overexp_long_vszero_sig <- wt_overexp_long_vszero_sig %>%
  left_join(wt_overexp_long_vszero_maxfc)

wt_overexp_long_vszero_sig <- wt_overexp_long_vszero_sig %>%
  filter(max_fc > l2fc_thresh)

save(wt_overexp_long_vszero_sig, file = "results/wt_overexp_long_vszero_sig.RData")
```

```{r}
load("results/wt_overexp_long_vszero_sig.RData", verbose = T)
wt_overexp_long_vszero_sig <- wt_overexp_long_vszero_sig %>%
  mutate(l2fc_shrunken = log2FoldChange)
make_timecourse_lfc_plot(deseq_res = wt_overexp_long_vszero_sig,
                         genes_to_include = unique(wt_overexp_long_vszero_sig$gene_id),
                         y_lim = c(-6,6),
                         draw_plot = FALSE)

ggsave(paste0("figures/wt_overexp_long_lfc_", thresh, ".pdf"),
       height = 3, width = 3.5, useDingbats = FALSE)
```

```{r}
# Overlap with long timecourse rescue
load("../05_long_timecourse_rescue/results/ko_rescue_long_vszero_sig.RData")

wt_overexp_sig_genes <- unique(wt_overexp_long_vszero_sig$gene_id)
ko_rescue_sig_genes <- unique(ko_rescue_long_vszero_sig$gene_id)

length(intersect(wt_overexp_sig_genes, ko_rescue_sig_genes))

length(wt_overexp_sig_genes)
length(ko_rescue_sig_genes)
length(combined_genes)

combined_genes <- unique(c(wt_overexp_sig_genes, ko_rescue_sig_genes))

mat <- cbind(
  overexp_long = combined_genes %in% wt_overexp_sig_genes,
  rescue_long = combined_genes %in% ko_rescue_sig_genes
)
fit2 <- euler(mat)

ovp <- calculate_overlap_pval(wt_overexp_sig_genes, ko_rescue_sig_genes, num_measured_genes)
pdf("figures/long_timecourse_rescue_overexp_overlap.pdf", height = 2, width = 2)
plot(fit2, quantities = TRUE, fontsize = 8)
grid.text(paste0("P=", signif(ovp, 3)), x = 0.1, y = 0.9)
dev.off()
```



### Long timecourse Firre responders heatmap

```{r fig.width=3, fig.height=5}
# Heatmap of fold-changes for DEGs in the rescue
# Check that there are no duplicate row names.
stopifnot(all(length(unique(wt_overexp_long_vszero_sig$gene_id)) == length(unique(wt_overexp_long_vszero_sig$gene_name))))

wt_overexp_long_lfc <- wt_overexp_long_vszero_sig %>%
  dplyr::select(gene_name, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()

# Add a zero column.
wt_overexp_long_lfc <- cbind(matrix(0, nrow = nrow(wt_overexp_long_lfc), ncol = 1), wt_overexp_long_lfc)
colnames(wt_overexp_long_lfc)[[1]] <- "0"

pdf(paste0("figures/long_overexpression_responders_heatmap_", thresh, ".pdf"),
    width = 3, height = 5)
ht1 <- Heatmap(wt_overexp_long_lfc, 
        name = "l2fc",
        cluster_columns = FALSE, show_row_names = FALSE, 
        col = colorRamp2(seq(-3,3,length.out = 100), col_pal10))
draw(ht1)
dev.off()
draw(ht1)
```

```{r}
load("results/wt_overexp_long_vszero_sig.RData")
# How many genes are changing in each timepoint?
deg_per_timepoint <- wt_overexp_long_vszero_sig %>%
  filter(abs(log2FoldChange) > l2fc_thresh) %>%
  mutate(direction = ifelse(log2FoldChange > 0, "up", "down")) %>%
  group_by(timepoint, direction) %>%
  summarize(ndeg = n()) %>%
  mutate(y = ifelse(direction == "up", ndeg, ndeg*-1))
deg_per_timepoint$timepoint <- factor(deg_per_timepoint$timepoint, labels = c("12", "24", "48", "96"))

ggplot(deg_per_timepoint, aes(x = timepoint, y = y, fill = direction)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 0) +
  geom_text(data = deg_per_timepoint %>% filter(direction == "up"), aes(label=ndeg), position=position_dodge(width=0.9), vjust=-0.25) +
  geom_text(data = deg_per_timepoint %>% filter(direction == "down"), aes(label=ndeg), position=position_dodge(width=0.9), vjust=1.1) +
  guides(fill = FALSE) +
  ylim(-65,75)
ggsave("figures/long_overexp_ndeg_per_timepoint.pdf",
       height = 2, width = 1.75)  
```
