---
title: "Long timecourse rescue"
author: "Michael Smallegan\n"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(seriation)
source("../util/_plot_theme.R")
source("../util/_util.R")
source("../01_setup/assumptions.R")
source("../util/figures.R")

if(!file.exists("data")) dir.create("data")
if(!file.exists("results")) dir.create("results")
if(!file.exists("figures")) dir.create("figures")

# Keep track of thresholds for figure naming
thresh <- paste0("pval", pval_thresh, "_l2fc", round(l2fc_thresh, 2))
```

```{r load, include=FALSE}
load("../01_setup/results/rnaseq_data.RData")
```

For this experiment, we induced the expression of Firre with an rTTA element by adding doxycycline to mouse embryonic stem cells. We see that Firre is indeed expressed in the KO background after the addition of doxycycline. The drug does instigate some gene expression changes on its own, so we will control for the effects by using a linear model which accounts for the effect of dox.

```{r run-deseq-lrt-firre-rescue-long}
if(!file.exists("results/ko_rescue_long.RData")) {
  
  # Filter to ESC KO long timecourse
  ko_rescue_long_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "long",
           firre_ko == "KO")
  ko_rescue_long_counts <- counts[,ko_rescue_long_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(ko_rescue_long_samples) == colnames(ko_rescue_long_counts)))
  stopifnot(all(rownames(ko_rescue_long_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  ko_rescue_long_dds <- DESeqDataSetFromMatrix(countData = ko_rescue_long_counts, 
                                          colData = ko_rescue_long_samples, 
                                          design = ~ firre_induced + timepoint + timepoint*firre_induced)
  ko_rescue_long_dds <- DESeq(ko_rescue_long_dds, test = "LRT", reduced = ~ firre_induced + timepoint)
  
  # Compile results
  res_names <- resultsNames(ko_rescue_long_dds)
  dynamic_res <- res_names[grepl("firre_inducedfirre_induced.timepoint", res_names)]
  
  ko_rescue_long_lfc <- lapply(dynamic_res, function(x) {
    results(ko_rescue_long_dds, 
            name = x) %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Shrunken LFC results
  ko_rescue_long_shrnklfc <- lapply(dynamic_res, function(x) {
    lfcShrink(ko_rescue_long_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Calculate the maximum fold-change in any one timepoint
  ko_rescue_long_maxfc <- ko_rescue_long_shrnklfc %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  ko_rescue_long_shrnklfc <- ko_rescue_long_shrnklfc %>%
    left_join(ko_rescue_long_maxfc)
  
  save(ko_rescue_long_lfc, ko_rescue_long_shrnklfc, file = "results/ko_rescue_long.RData")
}

load("results/ko_rescue_long.RData")
```

### Fold changes vs zero timepoint

This is without considering the control cell line.

```{r run-deseq-vszero-firre-rescue-long}
if(!file.exists("results/ko_rescue_long_vs_zero.RData")) {
  
  ko_rescue_long_vszero_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "long",
           firre_ko == "KO",
           firre_induced == "firre_induced")
  ko_rescue_long_vszero_counts <- counts[,ko_rescue_long_vszero_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(ko_rescue_long_vszero_samples) == colnames(ko_rescue_long_vszero_counts)))
  stopifnot(all(rownames(ko_rescue_long_vszero_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  ko_rescue_long_vszero_dds <- DESeqDataSetFromMatrix(countData = ko_rescue_long_vszero_counts,
                                                 colData = ko_rescue_long_vszero_samples,
                                                 design = ~ timepoint)
  ko_rescue_long_vszero_dds <- DESeq(ko_rescue_long_vszero_dds)
  res_names <- resultsNames(ko_rescue_long_vszero_dds)
  
  vs_zero_res <- res_names[grepl("_vs_0", res_names)]
  ko_rescue_long_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(ko_rescue_long_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  save(ko_rescue_long_vszero_shrnklfc, file = "results/ko_rescue_long_vs_zero.RData")
}

load("results/ko_rescue_long_vs_zero.RData", verbose = T)
```

### Long timecourse call significant genes

We'll make the p-value cutoff based on the dox controlled model and the l2fc cutoff based on the fold change vs zero.

```{r sig-firre-rescue-long}
ko_rescue_long_dox_sig <- ko_rescue_long_shrnklfc %>% 
  filter(padj <= pval_thresh)

ko_rescue_long_vszero_sig <- ko_rescue_long_vszero_shrnklfc %>%
  filter(gene_id %in% ko_rescue_long_dox_sig$gene_id)

ko_rescue_long_vszero_maxfc <- ko_rescue_long_vszero_sig %>%
  group_by(gene_id) %>%
  summarize(max_fc = max(abs(log2FoldChange))) 

ko_rescue_long_vszero_sig <- ko_rescue_long_vszero_sig %>%
  left_join(ko_rescue_long_vszero_maxfc)

ko_rescue_long_vszero_sig <- ko_rescue_long_vszero_sig %>%
  filter(max_fc > l2fc_thresh)

ko_rescue_long_vszero_sig <- ko_rescue_long_vszero_sig %>%
  mutate(l2fc_shrunken = log2FoldChange)

save(ko_rescue_long_vszero_sig, file = "results/ko_rescue_long_vszero_sig.RData")
```

```{r plot-firre-rescue-long}
## Fig. S1D
make_timecourse_lfc_plot(deseq_res = ko_rescue_long_vszero_sig,
                         genes_to_include = unique(ko_rescue_long_vszero_sig$gene_id),
                         y_lim = c(-6,6),
                         draw_plot = FALSE) +
  ggtitle("FirreRescue Firre response long",
          subtitle = "Fig. S1D")

ggsave(paste0("figures/ko_rescue_long_lfc_", thresh, ".pdf"),
       height = 3, width = 3.5, useDingbats = FALSE)
```

### Long timecourse Firre responders heatmap

```{r heatmap-firre-rescue-long, fig.width=3, fig.height=5}
# Heatmap of fold-changes for DEGs in the rescue
# Check that there are no duplicate row names.
stopifnot(all(length(unique(ko_rescue_long_vszero_sig$gene_id)) == length(unique(ko_rescue_long_vszero_sig$gene_name))))

ko_rescue_long_lfc <- ko_rescue_long_vszero_sig %>%
  dplyr::select(gene_name, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()

# Add a zero column.
ko_rescue_long_lfc <- cbind(matrix(0, nrow = nrow(ko_rescue_long_lfc), ncol = 1), ko_rescue_long_lfc)
colnames(ko_rescue_long_lfc)[[1]] <- "0"

row_ha = rowAnnotation(sig_in_ko = as.numeric(rownames(ko_rescue_long_lfc) %in% wtko_sig$gene_name),
                       rescued = as.numeric(rownames(ko_rescue_long_lfc) %in% rescued_genes_20),
                       col = list(sig_in_ko = c("1" = "black", "0" = "white"),
                                  rescued = c("1" = "black", "0" = "white")))

table(as.numeric(rownames(ko_rescue_long_lfc) %in% wtko_sig$gene_name))
table(as.numeric(rownames(ko_rescue_long_lfc) %in% rescued_genes_20))

pdf(paste0("figures/long_responders_heatmap_", thresh, ".pdf"), width = 3, height = 5)
h1 <- Heatmap(ko_rescue_long_lfc, 
        name = "l2fc",
        cluster_columns = FALSE, show_row_names = FALSE, 
        col = colorRamp2(seq(-3,3,length.out = 100), col_pal10),
        left_annotation = row_ha)
draw(h1)
dev.off()
draw(h1)
```

```{r}
load("results/ko_rescue_long_vszero_sig.RData")
# How many genes are changing in each timepoint?
deg_per_timepoint <- ko_rescue_long_vszero_sig %>%
  filter(abs(log2FoldChange) > l2fc_thresh) %>%
  mutate(direction = ifelse(log2FoldChange > 0, "up", "down")) %>%
  group_by(timepoint, direction) %>%
  summarize(ndeg = n()) %>%
  mutate(y = ifelse(direction == "up", ndeg, ndeg*-1))
deg_per_timepoint$timepoint <- factor(deg_per_timepoint$timepoint, labels = c("12", "24", "48", "96"))

ggplot(deg_per_timepoint, aes(x = timepoint, y = y, fill = direction)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 0) +
  geom_text(data = deg_per_timepoint %>% filter(direction == "up"), aes(label=ndeg), position=position_dodge(width=0.9), vjust=-0.25) +
  geom_text(data = deg_per_timepoint %>% filter(direction == "down"), aes(label=ndeg), position=position_dodge(width=0.9), vjust=1.1) +
  guides(fill = FALSE) +
  ylim(-150,200)
ggsave("figures/long_rescue_ndeg_per_timepoint.pdf",
       height = 2, width = 1.75)  
```

### Firre rescued genes heatmap

We can visualize the expression levels of the genes that were significantly differentially expressed in the Firre knockout and dynamically changing in response to Firre's induction in a knockout background.

```{r rescue_heatmap, fig.width=3.5, fig.height=6}
# Retrieve the set of genes that is DE in the KO and is significantly changing
# in the KO rescue experiment. 
# Retrieve the TPM -- take the mean TPM in each condition.
rescued_genes_20 <- rescued_df_closest %>% 
  filter(rescued == TRUE) %>% 
  pull(gene_id)

wtko_samples <- samples %>% 
    filter(cell_type == "ESC", 
           timepoint == 0, 
           timecourse_length == "long")

wtko_matrix <- tpm %>% 
  filter(gene_id %in% rescued_genes_20) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(wtko_samples) %>%
  drop_na() %>%
  group_by(gene_name, firre_ko) %>%
  summarize(tpm = mean(tpm)) %>%
  pivot_wider(names_from = "firre_ko", 
              values_from = "tpm")

ko_rescue_long_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "long",
           firre_ko == "KO")

ko_rescue_matrix <- tpm %>% 
  filter(gene_id %in% rescued_genes_20) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(ko_rescue_long_samples) %>%
  drop_na() %>%
  filter(timepoint != 0) %>%
  group_by(gene_name, timepoint) %>%
  summarize(tpm = mean(tpm)) %>%
  pivot_wider(names_from = "timepoint", 
              values_from = "tpm")

comb_matrix <- wtko_matrix %>% 
  left_join(ko_rescue_matrix) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()

comb_matrix <- comb_matrix[,c("KO", "720", "1440", "2880", "5760", "WT")]

comb_matrix_scaled <- t(scale(t(comb_matrix)))

p1 <- pheatmap::pheatmap(comb_matrix_scaled, cluster_cols = FALSE, cluster_rows = TRUE, show_rownames = TRUE,
                   color = col_pal10,
                   clustering_callback = olo_seriate,
                   clustering_distance_rows = "euclidean",
                   border_color = NA,
                   fontsize_row = 8,
                   breaks = seq(-1.5, 1.5, length.out = length(col_pal10)),
                   treeheight_row = 25)

ggsave(paste0("figures/long_rescue_heatmap_", thresh, ".pdf"), 
       plot = p1, width = 3.5, height = 5)
```

### Rescue PCA

```{r rescue_pca}
pca_dat <- prcomp(t(comb_matrix))
proportion_of_variance <- summary(pca_dat)$importance[2,1:2]

pca_df <- pca_dat$x[,1:2] %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  mutate(PC1 = PC1 * -1)

ggplot(pca_df, aes(x = PC1, y = PC2, label = sample)) +
  geom_point() +
  geom_text_repel() +
  xlab(paste0("PC1: ", round(proportion_of_variance["PC1"]*100), "%")) +
  ylab(paste0("PC2: ", round(proportion_of_variance["PC2"]*100), "%"))
ggsave(paste0("figures/long_rescue_pca_", thresh, ".pdf"), 
       width = 3.5, height = 0.9, 
       useDingbats = FALSE)
```
