---
title: "Long timecourse combined"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(seriation)
library(testit)
library(VennDiagram)
source("../util/_plot_theme.R")
source("../util/_util.R")
source("../01_setup/assumptions.R")
source("../util/figures.R")

# Keep track of thresholds for figure naming
thresh <- paste0("pval", pval_thresh, "_l2fc", round(l2fc_thresh, 2))
```

```{r load, include=FALSE}
load("../01_setup/results/rnaseq_data.RData", verbose = T)
```

# Overlap between OE and RESCUE long

```{r}
load("../05_long_timecourse_rescue/results/ko_rescue_long_vszero_sig.RData", 
     verbose = T)
load("../08_long_timecourse_overexpression/results/wt_overexp_long_vszero_sig.RData", 
     verbose = T)
```


# Here we are combining the WT and KO Firre transgene lines in a model
# That will account for both genotype, dox and timepoint
### results will be saved as "short_combined_dox_controlled.RData"

```{r long_timecourse}
if(!file.exists("results/combined_long_dox_controlled.RData")) {
  
  # Filter to short timecourse both WT and KO
  long_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "long")
  
  ls_check <- long_samples %>%
    group_by(experiment, firre_ko, firre_induced, timepoint) %>%
    summarize(count = n())
  assert(all(ls_check$count == 3))
  
  # getting salmoon counts for Deseeq
  long_counts <- salmon_gene_counts[,long_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(long_samples) == colnames(long_counts)))
  stopifnot(all(rownames(long_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  # LRT for dox control performed here
  combined_long_dds <- DESeqDataSetFromMatrix(countData = long_counts, 
                                              colData = long_samples, 
                                              design = ~ firre_ko + firre_induced + timepoint + timepoint*firre_induced)
  
  # running LRT for control lines
  combined_long_dds <- DESeq(combined_long_dds, test = "LRT", reduced = ~ firre_ko + firre_induced + timepoint)
  
  
  # Compile results
  res_names <- resultsNames(combined_long_dds)
  
  # grabbing just the Deseq results we want
  dynamic_res <- res_names[grepl("firre_inducedfirre_induced.timepoint", res_names)]
  
  # cleaning up each result file
  combined_long_lfc <- lapply(dynamic_res, function(x) {
    results(combined_long_dds, 
            name = x) %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  library(tidyverse)
  library(DESeq2)
  # Shrunken LFC results
  combined_long_shrnklfc <- lapply(dynamic_res, function(x) {
    lfcShrink(combined_long_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Calculate the maximum fold-change in any one timepoint
  combined_long_maxfc <- combined_long_shrnklfc %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  # merge max shrnklfc into dataframe
  combined_long_shrnklfc <- combined_long_shrnklfc %>%
    left_join(combined_long_maxfc)
  
  # saving
  save(combined_long_lfc, combined_long_shrnklfc, file = "results/combined_long_dox_controlled.RData")
}

load("results/combined_long_dox_controlled.RData", verbose = T)

# look at sig
sig_combined_long_dox_controlled <- combined_long_lfc %>%
  filter(padj < 0.05)

length(unique(sig_combined_long_dox_controlled$gene_id))

```

# Combined WT & KO  long time course Fold changes vs zero timepoint
# Deseq model does not have dox control line in it 

```{r Deseq combined WT and KO long time coruse versus 0 with out control line in model}

if(!file.exists("results/combined_long_vs_zero.RData")) {
  
  # calling out samples for long  WT and KO RNAseq
  combined_long_vszero_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "long",
           firre_induced == "firre_induced")
  
  # getting salmon counts for Deseq
  combined_long_vszero_counts <- salmon_gene_counts[,combined_long_vszero_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(combined_long_vszero_samples) == colnames(combined_long_vszero_counts)))
  stopifnot(all(rownames(combined_long_vszero_counts) == genes$gene_id))
  
  # DESeq2 -- model for both genotypes and time but not conrol line LRT test
  # NO LRT test 
  combined_long_vszero_dds <- DESeqDataSetFromMatrix(countData = combined_long_vszero_counts,
                                                     colData = combined_long_vszero_samples,
                                                     design = ~ firre_ko + timepoint)
  # running Deseq
  combined_long_vszero_dds <- DESeq(combined_long_vszero_dds)
  res_names <- resultsNames(combined_long_vszero_dds)
  
  
  # cleaning up result files & converting to shrnkLFC 
  
  vs_zero_res <- res_names[grepl("_vs_0", res_names)]
  combined_long_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(combined_long_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  # saving
  save(combined_long_vszero_shrnklfc, file = "results/combined_long_vs_zero.RData")
}

load("results/combined_long_vs_zero.RData", verbose = T)

```

# Here we are going to perform Deseq on the combined models versus 0 time point for all timepoints long and short

```{r combined WT and KO all timepoints no control line }

if(!file.exists("results/combined_long_vs_zero.RData")) {
  
  combined_long_all_timepoints_vszero_samples <- samples %>%
    filter(cell_type == "ESC",
           firre_induced == "firre_induced")
  # subsetting to combined_long_all_timepoints
  combined_long_all_timepoints_vszero_counts <- salmon_gene_counts[,combined_long_all_timepoints_vszero_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(combined_long_all_timepoints_vszero_samples) == colnames(combined_long_all_timepoints_vszero_counts)))
  
  stopifnot(all(rownames(combined_long_all_timepoints_vszero_counts) == genes$gene_id))
  
  # DESeq2 -- model across all time points versus 0 timepoint
  # NO LRT for dox controls 
  combined_long_all_timepoints_vszero_dds <- DESeqDataSetFromMatrix(countData = combined_long_all_timepoints_vszero_counts,
                                                                    colData = combined_long_all_timepoints_vszero_samples,
                                                                    design = ~ firre_ko + timepoint)
  # running Deseq
  combined_long_all_timepoints_vszero_dds <- DESeq(combined_long_all_timepoints_vszero_dds)
  
  # setting result names object to be filtered below
  res_names <- resultsNames(combined_long_all_timepoints_vszero_dds)
  
  # getting result files
  vs_zero_res <- res_names[grepl("_vs_0", res_names)]
  
  # creating shrunken lfc 
  combined_long_all_timepoints_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(combined_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  # saving 
  save(combined_long_all_timepoints_vszero_shrnklfc, file = "results/combined_long_vs_zero_all_time_points.RData")
}

load("results/combined_long_vs_zero_all_time_points.RData", verbose = T)
```

# Creating list of combined WT and KO long timecourse sig genes
# Will merge versus zero with dox controled and add dox to object name
# We will use Pval and shrnklfc thresholds 

```{r creating sig genes in combined WT and KO long time course}

# filtering to sig genes in dox control LRT test "short_shrnklfc"
# This means induced in WT and KO firre induced, but not in dox control lines
combined_long_dox_sig <- combined_long_shrnklfc %>% 
  filter(padj <= pval_thresh)

# Those genes that were in the combined short versus 0 sig genes
# Those genes that are also in the dox control line
combined_long_vszero_sig <- combined_long_vszero_shrnklfc %>%
  filter(gene_id %in% combined_long_dox_sig$gene_id)

# determining max fold change 
short_vszero_maxfc <- combined_long_vszero_sig %>%
  group_by(gene_id) %>%
  summarize(max_fc = max(abs(log2FoldChange))) 

# merging in new col for max fold change with short_vszero_sig
combined_long_vszero_sig <- combined_long_vszero_sig %>%
  left_join(short_vszero_maxfc)

# filtering on threshold
combined_long_vszero_sig <- combined_long_vszero_sig %>%
  filter(max_fc > l2fc_thresh)

# saving results
save(combined_long_vszero_sig, file = "results/combined_long_vszero_sig.RData")
```

# Plotting combined sig (dox controlled) genes with "make_timecourse_lfc_plot function
```{r plotting WT KO combined short sig dox controlled}

load("results/combined_long_vszero_sig.RData", verbose = T)

# changing lfc col name
combined_long_vszero_sig <- combined_long_vszero_sig %>%
  mutate(l2fc_shrunken = log2FoldChange)

# Plotting
make_timecourse_lfc_plot(deseq_res = combined_long_vszero_sig,
                         genes_to_include = unique(combined_long_vszero_sig$gene_id),
                         y_lim = c(-6,6),
                         draw_plot = FALSE)

ggsave(paste0("figures/combined_long_lfc_", thresh, ".pdf"),
       height = 3, width = 3.5, useDingbats = FALSE)

#TODO missing up adn down genes from long KO time course
```
### Result: there are more genes in combined long than short model
### Result: 108 are induced and 68 downregulated
### Result: WT long time course: 818 genes induced 757 down regulated
### Result: KO long time course: Ko long had about 500 genes total 



# Looking for rescue genes (firre should be positive control)
```{r finding overlap with rescue genes}

#TODO start here
#TODO check sig gene numbers
#TODO why can just load in 
# rescued_gene_ids_20 <- read.csv("../04_Firre_ko_long_time_course/results/rescued_sig_20.txt", col.names = "gene_id") 
# Then left joing with sig genes or %in%
#TODO Firre is not in combined_long_vszero_sig!! How is that possible
#TODO also not in combined_long_dox list? Variance differences?

# loading in long combined sig genes
combined_long_sig_genes <- unique(combined_long_vszero_sig$gene_id) 

# We need the KO vs WT data
load("../03_firre_ko_vs_wt/results/wtko.RData")

# Getting sig wt vs ko genes
wtko_res_shrnklfc <- wtko_res_shrnklfc %>%
  mutate(sig = padj <= pval_thresh & abs(log2FoldChange) > l2fc_thresh)

# filtering to sig genes
wtko_sig <- wtko_res_shrnklfc %>% 
  filter(sig == TRUE)

# IMPORTANT: we did not do a lfc threshold.
# So this means any gene that was sig in long time course that overlaps with wt vs ko
# 1,770 genes in long time course (with out fold change cut off )
rescued_genes <- wtko_sig$gene_id[wtko_sig$gene_id %in% 
                                    (combined_long_shrnklfc %>%
                                       filter(padj <= pval_thresh) %>% 
                                       pull(gene_id))]
# 64 overlaps of possible rescue genes

# From here on we are making objects that can be used generically with different 
# inputs above to create "rescued_genes"

# filtering to PVal and lfc thresholds 
rescue_wtko <- wtko_sig %>%
  filter(gene_id %in% rescued_genes) %>%
  dplyr::select(gene_id, gene_name, log2FoldChange) %>%
  dplyr::rename(l2fc_wtko = log2FoldChange)
# still at 64 genes in this object

# sig combined long time course genes and timepoint 
rescue_ltc <- combined_long_shrnklfc %>%
  filter(gene_id %in% rescued_genes) %>%
  dplyr::select(gene_id, gene_name, log2FoldChange, timepoint) %>%
  dplyr::rename(l2fc_ltc = log2FoldChange)

# merging wt vs ko sig and ko long time course sig genes
# There are 640 genes differentially expressed in long time course
rescued_df <- rescue_wtko %>%
  left_join(rescue_ltc)

# making a dist col to find distance of values in wt -vs ko and long Ko time course sig genes
rescued_df <- rescued_df %>%
  mutate(dist = abs((l2fc_wtko * -1) - l2fc_ltc))


# genes that are recriprocally regulated
# first finding overlap between wt vs ko and long time course 
rescued_df_closest <- rescued_df %>%
  group_by(gene_id, gene_name) %>%
  mutate(closest = min(dist) == dist,
         max_fc = max(abs(l2fc_ltc)) == abs(l2fc_ltc)) %>%
  filter(closest == TRUE)

# Genes that overlap in long time course wt vs ko and 20% of wild type value is restored
rescued_df_closest <-  rescued_df_closest %>%
  mutate(twenty_percent = l2fc_wtko * 0.2,
         rescued = (-1*l2fc_wtko - twenty_percent)>= l2fc_ltc & l2fc_ltc>= (-1*l2fc_wtko + twenty_percent) |
           (-1*l2fc_wtko + twenty_percent)>= l2fc_ltc & l2fc_ltc>= (-1*l2fc_wtko - twenty_percent))

# Genes that overlap in long time course wt vs ko and 20% of wild type value is restored
combined_long_rescued_genes_20 <- rescued_df_closest %>% 
  filter(rescued == TRUE) %>% 
  pull(gene_name)


```
### Result: Firre is not in this list -- suspicious and only gene is Tfap2a



# Plotting each of the combined long dox controlled sig genes across all timepoints 
```{r 0 combined sig genes across all timepoints }

#TODO start here
# Loading in results 
load("results/combined_long_vszero_sig.RData", verbose = T)
load("results/combined_long_vs_zero_all_time_points.RData", verbose = T)



sig_long_combined_vs_zero <- combined_long_vszero_shrnklfc %>% filter(gene_id %in% unique(combined_long_vszero_sig$gene_id))

#TODO this isn't loaded anywhere per the strategy of create and load etc...
# deseq_res <- sig_combined_vs_zero

genes_to_include <- unique(combined_long_vszero_sig$gene_id)

# Let's plot the downregulated genes separately
deg <- combined_long_vszero_sig %>%
  filter(gene_id %in% genes_to_include) %>%
  dplyr::select(gene_id, gene_name, baseMean, log2FoldChange, timepoint)

# Add a zero timepoint
deg_zero <- deg %>%
  dplyr::select(gene_id, gene_name, baseMean) %>%
  distinct() %>%
  mutate(log2FoldChange = 0,
         timepoint = 0) 

deg <- deg %>%
  bind_rows(deg_zero)

# proving information of up and down and max fold change 
num_deg <- deg  %>%
  group_by(gene_id) %>%
  summarize(max_fc = log2FoldChange[which.max(abs(log2FoldChange))]) %>%
  mutate(direction = ifelse(max_fc > 0, "up", "down")) %>%
  group_by(direction) %>%
  summarize(num_deg = length(unique(gene_id)))

label_x <- max(deg$timepoint / 2)

# Let's sort on the order of fold-change
deg_fc <- deg %>%
  group_by(gene_name) %>%
  summarize(max_fc = log2FoldChange[which.max(abs(log2FoldChange))]) %>%
  arrange(-max_fc)

deg$gene_name <- factor(deg$gene_name, levels = deg_fc$gene_name)

# Plotting each gene seperately across all time points
g1 <- ggplot(deg, aes(x = timepoint, y = log2FoldChange, group = gene_id)) +
  geom_hline(yintercept = 0) +
  geom_line(alpha = 0.2, color = "gray70") +
  geom_point(alpha = 0.4) +
  facet_wrap(~gene_name, scales = "free_y") +
  scale_x_continuous(breaks = c(0,330, 720, 1440, 2880, 5670), labels = c("0", "5.5", "12","24", "48","96") )
show(g1)

# Saving
ggsave("figures/short_and_long_individual_fc_profiles.pdf",
       height = 49, width = 49)
```

# Determining overlap of genes in long time course of WT and KO sepeately 
```{r overlap in short time course for WT and KO seperate} 

# Expressed genes
nrow(long_counts[rowMeans(long_counts) > 1,])

# Loading in the WT and KO short time courses vszero datasets 
# load("results/combined_long_vszero_sig.RData", verbose = T)
load("../05_long_timecourse_rescue/results/ko_rescue_long_vszero_sig.RData", 
     verbose = T)
load("../08_long_timecourse_overexpression/results/wt_overexp_long_vszero_sig.RData", 
     verbose = T)

# making a unique list of each backgrounds sig genes
ko_rescue_1 <- unique(ko_rescue_long_vszero_sig$gene_name) 
wt_overexp_2 <- unique(wt_overexp_long_vszero_sig$gene_name) 

table(ko_rescue_1 == "Firre")
table(wt_overexp_2 == "Firre")

# Remove Firre
ko_rescue_1 <- ko_rescue_1[ko_rescue_1 != "Firre"]
wt_overexp_2 <- wt_overexp_2[wt_overexp_2 != "Firre"]
# combined_3 <- unique(combined_long_vszero_sig$gene_name)

# finding overlap
individual_overlaps_12 <- intersect(ko_rescue_1, wt_overexp_2)

# how many overlap
length(individual_overlaps_12)

# determining if overlaps in each background
length(ko_rescue_1) - length(individual_overlaps_12)
length(wt_overexp_2) - length(individual_overlaps_12)

# all_overlaps_123 <- intersect(individual_overlaps_12, combined_3)
# length(all_overlaps_123)

# This might be useful so writting out:
# all_overlaps_123_df <- all_overlaps_123 %>% as.data.frame()
# write_csv2(all_overlaps_123_df, "results/sig_overlaps-KO-long_wt_long_combined-long.csv")

# finding other overlaps for venn
# ov_23 <- intersect(wt_overexp_2, combined_3)
# ov_13 <- intersect(ko_rescue_1, combined_3)
library(VennDiagram)
draw.pairwise.venn(area1 = length(ko_rescue_1),
                   area2 = length(wt_overexp_2),
                   cross.area = length(individual_overlaps_12))

all_genes <- unique(c(ko_rescue_1, wt_overexp_2))
ov_mat <- cbind(
  rescue = all_genes %in% ko_rescue_1,
  overexpr = all_genes %in% wt_overexp_2
)
library(eulerr)
fit2 <- euler(ov_mat)

pdf("figures/long_timecourse_overlap_venn.pdf")
plot(fit2, quantities = TRUE)
dev.off()
# # Plotting ven diagram
# draw.triple.venn(area1 = length(ko_rescue_1), 
#                  area2 = length(wt_overexp_2), 
#                  area3 = length(combined_3),
#                  n12 = length(individual_overlaps_12),
#                  n23 = length(ov_23), 
#                  n13 = length(ov_13),
#                  n123 = length(all_overlaps_123))

library(eulerr)
# TODO what the heck !
set.seed(1)
mat <- cbind(
  A = sample(c(TRUE, TRUE, FALSE), 50, TRUE),
  B = sample(c(TRUE, FALSE), 50, TRUE),
  C = sample(c(TRUE, FALSE, FALSE, FALSE), 50, TRUE)
)
all_genes <- unique(c(ko_rescue_1, wt_overexp_2, combined_3))
ov_mat <- cbind(
  rescue = all_genes %in% ko_rescue_1,
  overexpr = all_genes %in% wt_overexp_2,
  combined = all_genes %in% combined_3
)
fit2 <- euler(ov_mat)

# Plotting
pdf("figures/short_timecourse_overlap_venn.pdf")
plot(fit2, quantities = TRUE)
dev.off()
```
### Result: 17 genes overlap in all timecourses 



# Short timecourse Firre responders heatmap for genes sig in all three long time courses 

```{r fig.width=4, fig.height=4 Firre responders heatmap for combined WT KO and dox controlled}

# Let's look at the set of genes that overlap in the two genetic backgrounds
load("../05_Firre_ko_short_time_course/results/ko_rescue_short_vszero_sig.RData", verbose = T)
load("../07_wt_short_time_course/results/wt_overexp_short_vszero_sig.RData", verbose = T)

#TODO how to make this from the intersection of these three variables ...
ko_rescue_1 <- unique(ko_rescue_long_vszero_sig$gene_name) 
wt_overexp_2 <- unique(wt_overexp_long_vszero_sig$gene_id) 
combined_3 <- unique(combined_long_vszero_sig$gene_id)

# finding overlaps
overlapping_genes <- unique(combined_long_vszero_sig$gene_id)[unique(wt_overexp_long_vszero_sig$gene_id) %in% unique(ko_rescue_long_vszero_sig$gene_id)]

# Heatmap of fold-changes for DEGs in the rescue
# Check that there are no duplicate row names.
stopifnot(all(length(unique(combined_long_vszero_sig$gene_id)) == length(unique(combined_long_vszero_sig$gene_name))))

# converting to matrix for heatmap
combined_lfc <- combined_long_vszero_sig %>%
  dplyr::select(gene_name, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()

# getting order right
ordering <- combined_long_vszero_sig %>%
  filter(abs(log2FoldChange) > 0.2) %>%
  group_by(gene_name) %>%
  summarize(first_tp_de = min(timepoint),
            max_fc = max(log2FoldChange)) %>%
  arrange(first_tp_de,
          -max_fc)

# Add a zero column.
combined_lfc <- cbind(matrix(0, nrow = nrow(combined_lfc), ncol = 1), combined_lfc)
colnames(combined_lfc)[[1]] <- "0"

combined_lfc <- combined_lfc[ordering$gene_name,]

pdf(paste0("figures/combined_responders_heatmap_", thresh, ".pdf"), 
    width = 4, height = 30)
ht1 <- Heatmap(short_lfc, 
               name = "l2fc",
               cluster_columns = FALSE, show_row_names = TRUE, 
               cluster_rows = FALSE,
               col = colorRamp2(seq(-4,4,length.out = 100), col_pal10))
draw(ht1)
dev.off()
draw(ht1)

#TODO does this look right?
```
### Result: interesting Peg10 VLP is induced a lot by Firre in longer time course