---
title: "17_eclip"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
source("../util/_plot_theme.R")
```

```{r}
rnaseq_kd_files <- read.table("data/ENCSR413YAF_RNAseq_knockdowns.txt", skip = 1,
                              col.names = "url") %>%
  filter(grepl(".tsv", url))

dat <- read_tsv("data/ENCFF061TRZ.tsv")

# eCLIP processed files
# https://www.encodeproject.org/publication-data/ENCSR456FVU/
# eCLIP knockdowns
# https://www.encodeproject.org/publication-data/ENCSR413YAF/

```

```{r}
# Let's see which eCLIP datasets are binding Firre.
encode_eclip_datasets <- read_tsv("data/encode_cell_line_eclip.tsv", skip = 1) %>%
  clean_names()
```

```{r}
library(httr)
library(janitor)
## This will generate a request URL in the format that ENCODE requires to retrieve each of the columns listed in the field default parameter (accession, read_count, md5sum, etc.)
contstruct_query <- function(experiment_accession,
                             base_url = "https://www.encodeproject.org/report.tsv?",
                             file_format = "fastq",
                             type = "File",
                             status = "released",
                             fields = c("accession", "read_count", "md5sum",
                                        "controlled_by", "paired_end",
                                        "paired_with", "replicate", "target")) {
  query <- paste(list(paste0("type=", type),
                      paste0("status=", status),
                      paste0("file_format=", file_format),
                      paste0("dataset=%2Fexperiments%2F", experiment_accession, "%2F"),
                      map_chr(fields, ~paste0("field=", .))) %>%
                   flatten(),
                 collapse = "&")
  url <- paste0(base_url, query)
  return(url)
}
# This function actually makes the request and returns the data only (without the response headers) in a data.frame format.
encode_file_info <- function(experiment_accession,
                             base_url = "https://www.encodeproject.org/report.tsv?",
                             file_format = "bigWig",
                             type = "File",
                             status = "released",
                             fields = c("accession", "file_format", "file_format_type", "step_run", "filter_type", "filter_value", 
                                        "genome_annotation", "assembly", "run_type",
                                        "output_type", "biological_replicates", 
                                        "technical_replicates", "assay_title", "assay_term_name", "annotation_type", "md5sum",
                                        "controlled_by", "paired_end",
                                        "paired_with", "replicate", "target")) {
  path <- "report.tsv?"
  base_url <- modify_url("https://www.encodeproject.org/", path = path)
  url <- contstruct_query(experiment_accession,
                          base_url = base_url,
                          file_format,
                          type,
                          status,
                          fields)
  resp <- GET(url)
  if (http_error(resp)) {
    error_message <- content(resp, type = "text/html", encoding = "UTF-8") %>%
      xml_find_all("//p") %>%
      xml_text() %>%
      first()
    stop(
      sprintf(
        "ENCODE API request failed [%s]\n%s",
        status_code(resp),
        error_message
      ),
      call. = FALSE
    )
  }
  
  if (http_type(resp) != "text/tsv") {
    stop("API did not return text/tsv", call. = FALSE)
  }
  body <- read_tsv(content(resp, "text"), skip = 1) %>%
    clean_names()
  return(body)
}

eclip_bam_info <- encode_eclip_datasets %>% 
  dplyr::rename(experiment_accession = accession) %>%
  dplyr::select(experiment_accession, target_of_assay, biosample_term_name) %>%
  mutate(df = map(experiment_accession, ~ encode_file_info(.x, file_format = "bam"))) %>%
  unnest(df)

# Use the fastq info to get the paired endedness of the data.
eclip_fastq_info <- encode_eclip_datasets %>% 
  dplyr::rename(experiment_accession = accession) %>%
  dplyr::select(experiment_accession, target_of_assay, biosample_term_name) %>%
  mutate(df = map(experiment_accession, ~ encode_file_info(.x, file_format = "fastq"))) %>%
  unnest(df)
table(eclip_fastq_info$run_type)

# Okay, it looks like they are all paired-ended so we don't really need to worry about that.
eclip_bam_info <- eclip_bam_info %>% 
  filter(genome_assembly == "GRCh38")

eclip_bam_info <- eclip_bam_info %>%
  mutate(download_cmd = paste0("wget https://www.encodeproject.org/files/", 
                               accession, "/@@download/", accession, ".bam"))

write_csv(eclip_bam_info, "results/eclip_bam_info.csv")
eclip_bam_info$download_cmd[[1]]
writeLines(c("#!/bin/bash", eclip_bam_info %>% pull(download_cmd) %>% unique()), "data/eclip_bams/dl.sh")

# Write out an md5sum check file
eclip_bam_info <- eclip_bam_info %>%
  mutate(md5_cmd = paste0(md5sum, "  ", accession, ".bam"))
writeLines(eclip_bam_info %>% pull(md5_cmd) %>% unique(), "data/eclip_bams/md5.txt")
```

```{r}
encode_rnaseq <- read_tsv("data/encode_hepg2_k562_polyA_rnaseq.tsv", skip = 1) %>%
  clean_names()

rnaseq_info <- encode_rnaseq %>% 
  dplyr::rename(experiment_accession = accession) %>%
  dplyr::select(experiment_accession, target_of_assay, biosample_term_name) %>%
  mutate(df = map(experiment_accession, ~ encode_file_info(.x, file_format = "tsv"))) %>%
  unnest(df)

gene_quant <- rnaseq_info %>%
  filter(genome_assembly == "GRCh38", genome_annotation == "V29", output_type == "gene quantifications") %>%
  mutate(download_cmd = paste0("wget https://www.encodeproject.org/files/", 
                               accession, "/@@download/", accession, ".tsv"))
writeLines(c("#!/bin/bash", gene_quant %>% pull(download_cmd) %>% unique()), "data/rnaseq/gene_quant/dl.sh")



# v29 annotation file
# https://www.encodeproject.org/files/ENCFF159KBI/@@download/ENCFF159KBI.gtf.gz
gtf <- rtracklayer::import("data/rnaseq/ENCFF159KBI.gtf")
g2s <- gtf[gtf$type == "gene"] %>%
  as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

firre_gene_id <- g2s %>%
  filter(gene_name == "FIRRE") %>%
  pull(gene_id)

firre_quant <- data.frame("filepath" = list.files("data/rnaseq/gene_quant", pattern = ".tsv", full.names = T)) %>%
  mutate(accession = gsub("data/rnaseq/gene_quant/|.tsv", "", filepath)) %>%
  left_join(gene_quant %>% dplyr::select(experiment_accession, accession, biosample_term_name, biological_replicates)) %>%
  mutate(df = map(filepath, ~ read_tsv(.x) %>%
                    clean_names() %>% filter(gene_id == firre_gene_id))) %>%
  unnest(df) %>%
  left_join(encode_rnaseq %>% dplyr::rename(experiment_accession = accession) %>% 
              dplyr::select(experiment_accession, lab, biosample_treatment))

firre_quant$biosample_treatment[is.na(firre_quant$biosample_treatment)] <- "untrt"

table(firre_quant$biosample_treatment)
firre_quant$biosample_treatment <- factor(firre_quant$biosample_treatment, levels = c("untrt", "interferon alpha", "Interferon gamma"))
ggplot(firre_quant, aes(x = biosample_term_name, y = tpm, color = biosample_treatment, shape = biosample_treatment)) + 
  geom_jitter(width = 0.05) +
  ggtitle("FIRRE expression in HEPG2 and K562")
ggsave("firre_encode_hepg2_k562_expression_interferon_trt.pdf", useDingbats = FALSE)

write_csv(firre_quant, "results/firre_rnaseq_hepg2_k562.csv")



firre_quant <- firre_quant %>% filter(biosample_treatment == "untrt")


firre_median_expression <- firre_quant %>% group_by(biosample_term_name) %>%
  summarize(firre_median_expression = median(tpm))


ggplot(firre_quant, aes(x = biosample_term_name, y = tpm)) + 
  stat_boxplot(geom = "errorbar", width = 0.2) +
  geom_jitter(width = 0.05) +
  stat_summary(
    geom = "point",
    fun.y = "mean",
    col = "red",
    size = 2,
    fill = "red"
  ) +
  ggtitle("FIRRE expression in HEPG2 and K562")
ggsave("figures/firre_encode_hepg2_k562_expression.pdf", useDingbats = FALSE, height = 6, width = 4)
library(ggpubr)
ggerrorplot(firre_quant, x = "biosample_term_name", y = "tpm", desc_stat = "mean_sd", error.plot = "errorbar") +
  geom_jitter(width = 0.05)
firre_quant <- quant1 %>% filter(gene_id == firre_gene_id)

transcript_quant <- rnaseq_info %>%
  filter(genome_assembly == "GRCh38", genome_annotation == "V29", output_type == "transcript quantifications")
```


```{r}
# Let's grab just the Firre RNA from the Gencode GTF.
gtf <- rtracklayer::import("/scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v38/firre.gtf")
hmm <- gtf %>% as.data.frame()

library(Rsubread)
counts <- featureCounts("data/eclip_bams/ENCFF086HIT.bam", 
                        annot.ext = "/scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v38/firre.gtf",
                        isGTFAnnotationFile = TRUE, isPairedEnd = TRUE)


```

```{r}
# Read in the featurecounts that were run on fiji-viz
library(tidyverse)
firre_fc <- read_rds("results/eclip_firre_counts.rds")
firre_eclip_counts <- firre_fc$counts %>%
  as.data.frame() %>%
  pivot_longer(1:ncol(.), names_to = "file_accession", values_to = "count") %>%
  mutate(file_accession = gsub(".bam", "", file_accession))

## Need to figure out the number of reads in each sample in order to normalize...
```

```{bash}
ls bam_list.txt | while read f; do samtools view -c -F 260 ${f}; done;
```

```{r}
# Read in the bam counts -- mapped reads only
bam_order <- read.table("data/eclip_bams/bam_list.txt")
bam_counts <- read.table("data/eclip_bams/read_counts.txt")
colnames(bam_counts) <- "mapped_read_count"
colnames(bam_order) <- "bam_file"
read_counts <- bind_cols(bam_order, bam_counts)

read_counts <- read_counts %>%
  mutate(file_accession = gsub(".bam", "", bam_file)) %>%
  dplyr::select(-bam_file)

colnames(firre_eclip_counts)[2] <- "firre_count"

firre_eclip_counts <- firre_eclip_counts %>% left_join(read_counts)


info <- eclip_bam_info %>% dplyr::select(accession, target_of_assay, biosample_term_name) %>%
  dplyr::rename(file_accession = accession,
                rbp = target_of_assay,
                cell_line = biosample_term_name)
firre_eclip_counts <- firre_eclip_counts %>% left_join(info)



firre_eclip_counts <- firre_eclip_counts %>%
  mutate(cpm = firre_count / (mapped_read_count/1e6))


firre_median_expression <- firre_median_expression %>%
  dplyr::rename(cell_line = biosample_term_name)

firre_eclip_counts <- firre_eclip_counts %>%
  left_join(firre_median_expression)

write_csv(firre_eclip_counts, "results/firre_eclip_counts.csv")


ggplot(firre_eclip_counts, aes(x = cpm, color = cell_line)) + geom_density()

firre_eclip_counts <- firre_eclip_counts %>%
  group_by(cell_line) %>%
  mutate(zscore = scale(cpm))

# Get the TFs
# wget https://ars.els-cdn.com/content/image/1-s2.0-S0092867418301065-mmc2.xlsx
human_tfs <- read_csv("data/human_tfs.csv") %>%
  dplyr::rename(rbp = gene_name)


firre_eclip_counts <- firre_eclip_counts %>% left_join(human_tfs)
firre_eclip_counts$is_tf[is.na(firre_eclip_counts$is_tf)] <- "No"

firre_eclip_counts_w <- firre_eclip_counts %>%
  dplyr::select(cell_line, rbp, zscore, is_tf, dbd) %>%
  group_by(rbp, cell_line) %>%
  summarize(zscore = mean(zscore),
            is_tf = unique(is_tf),
            dbd = unique(dbd)) %>%
  pivot_wider(id_cols = c(is_tf, dbd, rbp), names_from = cell_line, values_from = zscore)

firre_eclip_counts_w <- firre_eclip_counts_w %>%
  mutate(mean_z = sum(K562, HepG2, na.rm = T) / 2)

firre_eclip_counts_matrix <- firre_eclip_counts_w %>%
  dplyr::select(rbp, K562, HepG2) %>%
  arrange(-K562) %>%
  column_to_rownames("rbp") %>%
  as.matrix()
firre_eclip_counts_matrix <- firre_eclip_counts_matrix[1:20,]
pdf("figures/firre_eclip_binders_ranked_by_K562.pdf", height = 6, width = 2.5)
pheatmap::pheatmap(firre_eclip_counts_matrix, cluster_rows = FALSE, cluster_column = FALSE)
dev.off()

firre_eclip_counts_matrix <- firre_eclip_counts_w %>%
  dplyr::select(rbp, K562, HepG2) %>%
  arrange(-HepG2) %>%
  column_to_rownames("rbp") %>%
  as.matrix()
firre_eclip_counts_matrix <- firre_eclip_counts_matrix[1:20,]
pdf("figures/firre_eclip_binders_ranked_by_HepG2.pdf", height = 6, width = 2.5)
pheatmap::pheatmap(firre_eclip_counts_matrix, cluster_rows = FALSE, cluster_column = FALSE)
dev.off()

write_csv(firre_eclip_counts, "results/firre_eclip_counts.csv")
```

```{r}
eclip_counts <- read_rds("results/eclip_counts.rds")
counts <- eclip_counts$counts %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(2:ncol(.), names_to = "file_accession", values_to = "count") %>%
  mutate(file_accession = gsub(".bam", "", file_accession))

length <- eclip_counts$annotation %>% as.data.frame() %>%
  dplyr::select(GeneID, Length) %>%
  dplyr::rename(gene_id = GeneID,
                gene_length = Length)

bam_order <- read.table("data/eclip_bams/bam_list.txt")
bam_counts <- read.table("data/eclip_bams/read_counts.txt")
colnames(bam_counts) <- "mapped_read_count"
colnames(bam_order) <- "bam_file"
read_counts <- bind_cols(bam_order, bam_counts)

read_counts <- read_counts %>%
  mutate(file_accession = gsub(".bam", "", bam_file)) %>%
  dplyr::select(-bam_file)


counts <- counts %>% left_join(length)
counts <- counts %>% left_join(read_counts)

info <- eclip_bam_info %>% dplyr::select(accession, target_of_assay, biosample_term_name) %>%
  dplyr::rename(file_accession = accession,
                rbp = target_of_assay,
                cell_line = biosample_term_name)

counts <- counts %>% left_join(info)




# Calculate the TPM
# Divide the read counts by the length of each gene in kilobases. This gives you reads per kilobase (RPK).
# Count up all the RPK values in a sample and divide this number by 1,000,000. This is your “per million” scaling factor.
# Divide the RPK values by the “per million” scaling factor. This gives you TPM.
# Might need to think about how to normalize this for eclip -- going to use total mapped reads instead.
# Also the option of using the total reads -- need to think about this more.

gtf <- rtracklayer::import("/scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v38/gencode.v38.annotation.gtf")

g2s <- gtf[gtf$type == "gene"] %>%
  as.data.frame() %>%
  dplyr::select(gene_id, gene_name, gene_type)

counts <- counts %>% left_join(g2s)

counts <- counts %>%
  mutate(rpk = count / (gene_length / 1e3),
         tpm = rpk / (mapped_read_count / 1e6))

ggplot(counts, aes(x = log10(tpm))) +
  geom_density()

# Let's remove all the zero counts. That will get rid of a bunch of genes
counts <- counts %>% filter(count > 0)
write_csv(counts, "results/encode_eclip_counts.csv")


counts <- read_csv("results/encode_eclip_counts.csv")
```

```{r}
# Let's use a bound / unbound cutoff even though it's a pretty continuous distribution... I 
# guess the usual cutoff is a tpm of 1
bound <- counts %>% filter(tpm > 10)

# let's just see how many RNAs are bound by each RBP
rbp_summary <- bound %>% 
  group_by(rbp) %>%
  summarize(n_genes_bound = length(unique(gene_id)))

# let's make a little plot just to show the number and types of RNA's bound.
gt <- table(bound$gene_type)
cat(gt[order(gt)])
cat(names(gt[order(gt)]))
# Let's read in a sanitized gene_type naming scheme
gene_type_abbr <- read_csv("data/gene_type_map.csv")

bound <- bound %>% left_join(gene_type_abbr)

rbp_summary <- rbp_summary %>%
  arrange(-n_genes_bound)
bound$rbp <- factor(bound$rbp, levels = rbp_summary$rbp)

gene_type_summary <- bound %>% 
  group_by(rbp, gene_type_abbrev) %>%
  summarize(count = length(unique(gene_id)))
ggplot(gene_type_summary, aes(x = rbp, y = count, fill = gene_type_abbrev)) +
  geom_bar(position = "stack", stat = "identity") + 
  coord_flip() + 
  ggtitle("eCLIP bound gene types",
          subtitle = "TPM > 10; HepG2 and K562 combined") +
  scale_fill_manual(values = rcartocolor::carto_pal(n = 11, name = "Bold"))
ggsave("figures/eclip_bound_rna_gene_types.pdf", height = 20, width = 7)


# KDM1A, HDAC1, RBFOX2, CTCF, SAFB2, HNRNPU, HNRNPL, HNRNPK, SAFB, EWSR1
ggplot(gene_type_summary %>% filter(rbp %in% c("KDM1A", "HDAC1", "RBFOX2", "CTCF", "SAFB2", "HNRNPU", "HNRNPL", "HNRNPK", "SAFB", "EWSR1")), aes(x = rbp, y = count, fill = gene_type_abbrev)) +
  geom_bar(position = "stack", stat = "identity") + 
  coord_flip() + 
  ggtitle("eCLIP bound gene types",
          subtitle = "TPM > 10; HepG2 and K562 combined") +
  scale_fill_manual(values = rcartocolor::carto_pal(n = 11, name = "Bold"))
ggsave("figures/eclip_bound_rna_gene_types_subset.pdf", height = 7, width = 8)



```




