---
title: "Firre timecourse long differential expression"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
params:
  alpha: 0.01
  lfc_threshold: 0.5
  data_dir: !r file.path("data")
  results_dir: !r file.path("results")
editor_options: 
  chunk_output_type: console
always_allow_html: yes
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, message=FALSE, warning=FALSE}
options(stringsAsFactors = FALSE)
source("../util/_setup.R")
source("../util/_util.R")
library(tidyverse); library(tximport); library(bcbioRNASeq); library(pals)
library(lemon); library(scales); library(ggrepel); library(colorspace)

library("BiocParallel")
register(MulticoreParam(12))
```


```{r load_data, message=FALSE}
# # Gene annotation data from script 00.
# load(paste0(params$data_dir,"/gene_annotations.RData"))
cc1 <- read.table("../../results/featureCounts/merged_gene_counts_long1.txt", header = T)
cc2 <- read.table("../../results/featureCounts/merged_gene_counts_long2.txt", header = T)

g2s <- cc1[,c(1,2)]
names(g2s) <- c("gene_id", "gene_name")

# I'm going to remove the gene name so we're just merging on one column
cc1 <- cc1 %>% select(-gene_name)
cc2 <- cc2 %>% select(-gene_name)
counts_combined <- merge(cc1, cc2)
# # Reduce Gencode annotation file to just gene information.
# genes <- rtracklayer::import("../../genomes/references/Mus_musculus/Gencode/M23/annotation/gencode.vM23.annotation.gtf")
# genes <- genes[which(genes$type == "gene")]
# rtracklayer::export(genes, "../../genomes/references/Mus_musculus/Gencode/M23/annotation/gencode.vM23.annotation.genes.gtf")
genes <- rtracklayer::import("/scratch/Shares/rinn/Michael/genomes/references/Mus_musculus/Gencode/M23/annotation/gencode.vM23.annotation.genes.gtf")

# Read in sample info.
samples <- read.csv("../../samplesheet.csv") 

# Here we're just going to analyze the long timecourse
# by itself.
samples <- samples %>% filter(experiment == "long")

# Factorize design variables
samples$firre_ko <- factor(samples$firre_ko, levels = c("WT", "KO"))

samples$timepoint <- factor(samples$timepoint, 
                            levels = c("0h", "12h", "24h", "48h", "96h"))

# Change from a boolean to a string
samples[which(samples$firre_induced), "firre_induced"] <- "firre_induced"
samples[which(samples$firre_induced == FALSE), "firre_induced"] <- "firre_uninduced"
samples$firre_induced <- factor(samples$firre_induced, levels = c("firre_uninduced", "firre_induced"))


rownames(samples) <- samples$sample_id


#  Make counts matrix.
names(counts_combined) <- sapply(names(counts_combined), function(x) {unlist(strsplit(x, "_"))[[1]]})
rownames(counts_combined) <- counts_combined$Geneid
counts <- as.matrix(counts_combined[,2:ncol(counts_combined)])

# Ensure that the ordering of the columns in the counts matrix 
# is the same as in the sample sheet.
# let's subset to just the samples which we would like to use
counts <- counts[,samples$sample_id]
stopifnot(all(rownames(samples) == colnames(counts)))

# Filter out genes with less than 15 counts.
keep <- rowSums(counts) >= 15
counts <- counts[keep, ]


# Ensure that the ordering of the rows (gene info) 
# is the same between the counts and the gencode
names(genes) <- genes$gene_id
genes <- genes[rownames(counts)]
stopifnot(all(names(genes) == rownames(counts)))
```

```{r}
# convert to tpm 
counts <- fCountReader("../../results/featureCounts/gene_counts/",
                       samples$sample_id, 
                       "_read1Aligned.sortedByCoord.out_gene.featureCounts.txt")
tpm <- countConverter(counts, return = "TPM") %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id")
write.csv(tpm, "./results/ftc_tpm.csv", row.names = FALSE)

fpkm <- countConverter(counts, return = "FPKM")  %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id")
write.csv(tpm, "./results/ftc_fpkm.csv", row.names = FALSE)
```


### 1. Genes changing in response to Firre 

```{r time_genes}
# DEseq results presents the estimated fold change for the 
# last variable in the design formula. 
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = samples, 
                              design = formula("~ timepoint + firre_induced + firre_induced*timepoint"),
                              rowData = genes)

dds <- DESeq(dds, 
             test="LRT", 
             reduced = ~ timepoint + firre_induced,
             parallel=TRUE,
             BPPARAM=MulticoreParam(12))

resultsNames(dds)
res <- results(dds, name = "timepoint96h.firre_inducedfirre_induced") %>% 
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s)

write.csv(res, "results/firre_responders_long.csv")
res <- res %>% filter(padj < 0.05)
firre_responder_long_genes <- res$gene_id
gl <- list("firre_responder_long_genes" = firre_responder_96_genes)
```


# Plots

Let's first look at Firre.

```{r}

ex_counts <- tpm %>%
  filter(gene_id %in% res$gene_id) %>%
  gather(key = "sample_id", value = "tpm", 2:ncol(.)) %>%
  merge(g2s) %>%
  merge(samples)

ex_counts <- merge(ex_counts, res %>% select(-gene_name))
ex_counts <- ex_counts %>% arrange(padj)
ex_counts$gene_name <- factor(ex_counts$gene_name, levels = unique(ex_counts$gene_name))

fr <- unique(ex_counts$gene_id)

goi <- ex_counts[which(ex_counts$gene_id %in% fr[1:10]),]

unique(goi$gene_id)
g <- ggplot(goi, aes(x = timepoint, y = tpm, color = firre_induced, group = firre_induced))
g + geom_point(alpha = 0.6) + 
  stat_summary(fun.y=mean, geom="line") +
  # scale_y_log10() + 
  facet_wrap(~ gene_name, scales = "free_y") + 
  # scale_color_manual(values = c("#414142", "#A83F4B"))
scale_color_manual(values = c("#414142", "#A81E2B"))
```

#### Fold change profile plots

```{r}

# let's first retrieve all the results. 
resdf <- results(dds, name = "timepoint12h.firre_inducedfirre_induced") %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s) %>%
  mutate(contrast = "timepoint12h.firre_inducedfirre_induced")

tps <- c("24h", "48h", "96h")
for(i in 1:length(tps)) {
  tresdf <- results(dds, name = paste0("timepoint",tps[[i]],".firre_inducedfirre_induced")) %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s) %>%
  mutate(contrast = paste0("timepoint",tps[[i]],".firre_inducedfirre_induced"))
  resdf <- bind_rows(resdf, tresdf)
}

resdf <- resdf %>% arrange(padj)
resdf$gene_name <- factor(resdf$gene_name, levels = unique(resdf$gene_name))

fr <- unique(ex_counts$gene_id)



unique(goi$gene_id)
unique(resdf$contrast)
resdf$timepoint <- sapply(resdf$contrast, function(x) {
  unlist(strsplit(unlist(strsplit(x, "int"))[[2]],".firre"))[[1]]
})
resdf$timepoint <- factor(resdf$timepoint, levels = c("12h", "24h", "48h", "96h"))

goi <- resdf[which(resdf$gene_id %in% fr[1:10]),]
g <- ggplot(goi, aes(x = timepoint, y = log2FoldChange))
g + geom_point(alpha = 0.6) + 
  # stat_summary(fun.y=mean, geom="line") +
  # scale_y_log10() + 
  # facet_wrap(~ gene_name, scales = "free_y") + 
  facet_wrap(~ gene_name)
  # scale_color_manual(values = c("#414142", "#A83F4B"))
scale_color_manual(values = c("#414142", "#A81E2B"))
```


############ THIS IS ALL PREVIOUS COPYPASTA
So this is the superset of all genes that are changing in time. 

There are `r nrow(dds)` genes expressed in the data and `nrow(res)` changing in time. 

### 2. Firre transgene responders

First let's start with the genes that respond the same in WT and KO in response to Firre. 

```{r}
# Create DESeq2 experiment objects.
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = samples, 
                              design = formula("~ timepoint + firre_induced + firre_induced*timepoint"),
                              rowData = genes)

dds <- DESeq(dds, 
             test="LRT", 
             reduced = ~ timepoint + firre_induced,
             parallel=TRUE,
             BPPARAM=MulticoreParam(12))


resultsNames(dds)
res <- results(dds, name = "timepoint330.firre_inducedfirre_induced") %>% 
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s)

write.csv(res, "results/differential_expression/firre_responders_same_in_both.csv")

res <- res %>%
  filter(padj < 0.05)

firre_responder_both_same <- res$gene_id
gl <- c(gl, "firre_responder_both_same" = list(firre_responder_both_same))
```

```{r}
# Let's look at some expamples: 
ex_counts <- counts[res$gene_id,] %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  gather(key = "id", value = "count", 2:ncol(.)) %>%
  merge(g2s) %>%
  merge(samples)

ex_counts <- merge(ex_counts, res %>% select(-gene_name))
ex_counts <- ex_counts %>% arrange(padj)

g <- ggplot(ex_counts, aes(x = timepoint, y = count, color = firre_induced, group = firre_induced))
g + geom_point() + stat_summary(fun.y=mean, geom="line") +
  scale_y_log10() + facet_wrap(~ gene_name, scales = "free_y")
```


Now let's look at all genes that respond to the Firre transgene in time.


```{r firretg_genes, message=FALSE}
# Create DESeq2 experiment objects.
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = samples, 
                              design = formula("~ timepoint + firre_ko + firre_induced + firre_ko*firre_induced + firre_induced*timepoint"),
                              rowData = genes)

dds <- DESeq(dds, 
             test="LRT", 
             reduced = ~ timepoint + firre_ko + firre_induced + firre_ko*firre_induced,
             parallel=TRUE,
             BPPARAM=MulticoreParam(12))


resultsNames(dds)
res <- results(dds, name = "timepoint330.firre_inducedfirre_induced") %>% 
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s)

write.csv(res, "results/differential_expression/firre_responders.csv")

res <- res %>%
  filter(padj < 0.05)
firre_responder_genes <- res$gene_id
gl <- c(gl, "firre_responder_genes" = list(firre_responder_genes))
```

```{r}
# Let's look at some expamples: 
ex_counts <- counts[res$gene_id,] %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  gather(key = "id", value = "count", 2:ncol(.)) %>%
  merge(g2s) %>%
  merge(samples)

ex_counts <- merge(ex_counts, res %>% select(-gene_name))
ex_counts <- ex_counts %>% arrange(padj)

g <- ggplot(ex_counts, aes(x = timepoint, y = count, color = firre_induced, group = firre_induced))
g + geom_point() + stat_summary(fun.y=mean, geom="line") +
  scale_y_log10() + facet_wrap(~ gene_name, scales = "free_y")
```





```{r}
# Some additional plotting functions to separate
# WT and KO samples in teh plots

# ordered_genes <- unique(ex_counts$gene_id)
# 
# ex_gene <- ex_counts %>% filter(gene_id %in% ordered_genes[1:10])
# ordered_gene_names <- unique(ex_counts$gene_name)
# ex_gene$gene_name <- factor(ex_gene$gene_name, levels = ordered_gene_names)
# g <- ggplot(ex_gene, aes(x = timepoint, y = count, color = firre_induced, group = firre_induced))
# g + geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10() + facet_wrap(gene_name~firre_ko, scales = "free_y", ncol = 4)

```


### 4. Firre KO responsive genes

```{r dox_responders}
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = samples, 
                              design = formula("~ timepoint + firre_induced + firre_ko + firre_induced*timepoint + firre_ko*firre_induced"),
                              rowData = genes)

dds <- DESeq(dds, test = "Wald",
             parallel=TRUE,
             BPPARAM=MulticoreParam(12))

resultsNames(dds)
res <- results(dds, name = "firre_ko_KO_vs_WT") %>% 
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s)


write.csv(res, "results/differential_expression/genes_changing_in_ko.csv")
res <- res %>%
  filter(padj < 0.05)


delta_KO_genes <- res$gene_id
gl <- c(gl, "delta_KO_genes" = list(delta_KO_genes))
```


There are `nrow(res)` changing between Firre WT and KO. I think there are so many because we have so many samples to test this. 

### 5. Genes changing as a results of the insertion of the transgene

```{r tg_responders}
# Let's subset to just the WT since that is where the line that has no transgene insertion is.
swt <- samples %>% filter(firre_ko == "WT")
cwt <- counts[,swt$id]
dds <- DESeqDataSetFromMatrix(countData = cwt, 
                              colData = swt, 
                              design = formula("~ timepoint + contains_itg + timepoint*contains_itg"),
                              rowData = genes)

dds <- DESeq(dds, test = "Wald",
             parallel=TRUE,
             BPPARAM=MulticoreParam(12))

resultsNames(dds)
res <- results(dds, name = "contains_itgTRUE") %>% 
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s)


write.csv(res, "results/differential_expression/transgene_insertion_responders.csv")
res <- res %>%
  filter(padj < 0.05)


tg_insertion_genes <- res$gene_id
gl <- c(gl, "tg_insertion_genes" = list(tg_insertion_genes))
```



### 6. Genes responding differently due to KO priming

```{r ko_priming_responders}
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = samples, 
                              design = formula("~ timepoint + firre_induced + firre_ko + firre_ko*timepoint"),
                              rowData = genes)

dds <- DESeq(dds, 
             test="LRT", 
             reduced = ~ timepoint + firre_induced + firre_ko,
             parallel=TRUE,
             BPPARAM=MulticoreParam(12))

resultsNames(dds)
res <- results(dds, name = "timepoint300.firre_koKO") %>% 
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  merge(g2s)


write.csv(res, "results/differential_expression/ko_priming_responders.csv")

res <- res %>%
  filter(padj < 0.05)


ko_priming_genes <- res$gene_id
gl <- c(gl, "ko_priming_genes" = list(ko_priming_genes))
```



```{r}
# Let's see how many genes that change in time are not in the ko genes
length(which(delta_time_genes %in% delta_KO_genes))
length(which(firre_responder_genes %in% delta_KO_genes))

same_same <- firre_responder_genes[which(!(firre_responder_genes %in% delta_KO_genes))]

samedf <- resdf[which(resdf$gene_id %in% same_same),]
```

```{r}
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_id == "Ppp1r18"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko, group = firre_ko)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Basp1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced", "contains_itg"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = contains_itg, group = contains_itg)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
#   # Umps
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Cyp1a1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko, group = firre_ko)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# 	
# # Dpf2	
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Gm15564"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko, group = firre_ko)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# # Mir7655
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Gfap"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Firre"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Apod"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Kdr"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# 
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Adgrg1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Ppp1r2"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Clock"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Nceh1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
```

```{r}
# betas <- coef(dds)
# colnames(betas)
# 
# 
# library(pheatmap)
# topGenes <- head(order(resTC$padj),20)
# mat <- betas[topGenes, -c(1,2)]
# rownames(mat) <- g2s[match(rownames(mat), g2s$gene_id), "gene_name"]
# thr <- 1
# mat[mat < -thr] <- -thr
# mat[mat > thr] <- thr
# pheatmap(mat, breaks=seq(from=-thr, to=thr, length=101),
#          cluster_col=FALSE)
```




Okay, now let's say we want to get the genes that are like Firre which has a large difference between the KO and WT condtion,
which means we want to model the interaction between firre_ko status and firre_induction and time?

For now, let's just see what the gene profiles look like if we swap Firre induction with firre_ko

Changing in time differentially in the KO and WT samples. did

```{r}
# dds <- DESeqDataSetFromMatrix(countData = counts$count, 
#                               colData = samples, 
#                               design = formula("~ timepoint + firre_ko + firre_induced + firre_ko*timepoint"),
#                               rowData = genes)
# 
# keep <- rowSums(counts(dds)) >= 15
# dds <- dds[keep, ]
# 
# dds <- DESeq(dds, test="LRT", reduced = ~ timepoint + firre_ko + firre_induced)
# 
# resTC <- results(dds)
# resultsNames(dds)
# 
# resdf <- resTC %>% as.data.frame() %>%
#   rownames_to_column(var = "gene_id") %>%
#   filter(padj < 0.05)
# 
# resdf <- merge(resdf, g2s)
```


```{r}
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Banp"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko, group = firre_ko)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# # Dido1
# # Cebpg
# # Nemp1
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Slc9a8"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko, group = firre_ko)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 	
# # Dppa5c	
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Zfp296"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko, group = firre_ko)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()

```



```{r}
# betas <- coef(dds)
# colnames(betas)
# 
# 
# library(pheatmap)
# topGenes <- head(order(resTC$padj),20)
# mat <- betas[topGenes, -c(1,2)]
# rownames(mat) <- g2s[match(rownames(mat), g2s$gene_id), "gene_name"]
# thr <- 1 
# mat[mat < -thr] <- -thr
# mat[mat > thr] <- thr
# pheatmap(mat, breaks=seq(from=-thr, to=thr, length=101),
#          cluster_col=FALSE)
```


Hmm, so these are genes that are changing differentially in response to Dox in the WT and KO background. 
Somehow Firre KO has changed the primed state of the cell, so that the responders are different? 

Okay, I still haven't captured Firre in this list though. 

```{r}
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Firre"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko, group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
```

For Firre, there's a large difference between WT and KO condition. Can I have a three way interaction?

```{r}
# dds <- DESeqDataSetFromMatrix(countData = counts$count, 
#                               colData = samples, 
#                               design = formula("~ timepoint + firre_ko + firre_induced + firre_ko*firre_induced*timepoint"),
#                               rowData = genes)
# 
# keep <- rowSums(counts(dds)) >= 15
# dds <- dds[keep, ]
# 
# dds <- DESeq(dds, test="LRT", reduced = ~ timepoint + firre_ko + firre_induced)
# 
# resTC <- results(dds)
# resultsNames(dds)
# 
# resdf <- resTC %>% as.data.frame() %>%
#   rownames_to_column(var = "gene_id") %>%
#   filter(padj < 0.05)
# 
# resdf <- merge(resdf, g2s)
```



```{r}
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Wdfy1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced,group = firre_ko)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
```

Hmm, this doesn't really look like it's changing in time... But is is changing at the baseline between the Firre WT with and without the
transgene incorporated.

So that's another category. Genes that changed in response to have the transgene inserted and remain changed. It's higher in the WT with no transgene. Hmm. 

```{r}
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Gm9732"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced,group = firre_ko)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
```

So there it seems it's just not in one sample.

```{r}
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Shf"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
```


```{r}
# betas <- coef(dds)
# colnames(betas)
# 
# 
# topGenes <- head(order(resTC$padj),20)
# mat <- betas[topGenes, -c(1,2)]
# rownames(mat) <- g2s[match(rownames(mat), g2s$gene_id), "gene_name"]
# thr <- 2
# mat[mat < -thr] <- -thr
# mat[mat > thr] <- thr
# pheatmap(mat, breaks=seq(from=-thr, to=thr, length=101),
#          cluster_col=FALSE)
```

```{r}
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Nceh1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Adgrg1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
```


Hmm, let's look at all the genes that are changing in time. 



```{r}
# dds <- DESeqDataSetFromMatrix(countData = counts$count, 
#                               colData = samples, 
#                               design = formula("~ timepoint + firre_ko + firre_induced + firre_ko*firre_induced"),
#                               rowData = genes)
# 
# keep <- rowSums(counts(dds)) >= 15
# dds <- dds[keep, ]
# 
# dds <- DESeq(dds, test="LRT", reduced = ~ firre_ko + firre_induced + firre_ko*firre_induced)
# 
# resTC <- results(dds)
# resultsNames(dds)
# 
# resdf <- resTC %>% as.data.frame() %>%
#   rownames_to_column(var = "gene_id") %>%
#   filter(padj < 0.05)
# 
# resdf <- merge(resdf, g2s)
```





```{r}
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Ppp1r10"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_induced,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Dido1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Prdm14"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# 
# # Lncenc1
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Lncenc1"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Myc"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
# 
# # Dpf2
# 
# fiss <- plotCounts(dds, which(dds@rowRanges$gene_name == "Dpf2"), 
#                    intgroup = c("timepoint","firre_ko", "firre_induced"), returnData = TRUE)
# fiss$timepoint <- as.numeric(as.character(fiss$timepoint))
# ggplot(fiss,
#   aes(x = timepoint, y = count, color = firre_ko,group = firre_induced)) + 
#   geom_point() + stat_summary(fun.y=mean, geom="line") +
#   scale_y_log10()
```

Hmm, firre is in this list, but it's p-val is not toward the bottom.


```{r}
# betas <- coef(dds)
# colnames(betas)
# 
# 
# topGenes <- head(order(resTC$padj),20)
# mat <- betas[topGenes, -c(1,2)]
# rownames(mat) <- g2s[match(rownames(mat), g2s$gene_id), "gene_name"]
# thr <- 1
# mat[mat < -thr] <- -thr
# mat[mat > thr] <- thr
# pheatmap(mat, breaks=seq(from=-thr, to=thr, length=101),
#          cluster_col=FALSE)
```





