---
title: "Identifying Firre responder genes"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Identifying Firre responder genes
### Firre induction vs control differential expression analysis

In this script we will analyze four experimental conditions to indentify
Firre responders and then look at the overlap in the responders across 
conditions.
We have two cell types and two genetic backgrounds for each: ESC - embryonic
stem cells, NPC - neural precursor cells and Firre KO and WT for each cell type.
In each genetic background (Firre KO & WT) we have a genomically incorporated
Firre transgene under the control of an rtTA (reverse tetracycline-controlled 
transactivator) element.
We will test whether turning on the expression of the Firre transgene (by adding
doxycycline) results in a gene being differentially expressed in comparison 
to the condition in which doxycycline is added, but Firre expression is 
not induced. Here we'll use the Likelihood ratio test from the DEseq2 package
to test whether Firre induces a change in gene expression any time after the 
zero timepoint. 

Briefly, we will use the likelihood ratio test to compare the fit for a model 
containing the time factor,
condition factor, and an interaction term to a reduced model without the 
interaction term: `time + firre_induced + time*firre_induced` vs 
`time + firre_induced`. The coefficients of the interaction term will 
give us the profile of the gene's expression relative to the control samples. 

Since it is likely that the genes influenced by Firre may change based on 
the genetic background and cell state, we will analyze each cell type and 
genotype separately. While the genes changing in each individual experiment
may be true Firre targets in that context, by comparing across contexts,
we will arrive at a set of genes that is robustly and reproducibly 
regulated by Firre. 

We will use the gene counts generated by Salmon	v0.14.1 (a la the
nf-core/rnaseq	v1.4.2 pipeline) using Gencode vM25 annotations and the mm10
reference genome.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(DESeq2)
library(gplots)
library(ggrepel)
source("../util/_plot_theme.R")
source("../util/_util.R")
```

```{r import}
### ANNOTATIONS
if(!file.exists("../util/gencode.vM25.annotation.genes.gtf")) {
  gencode <- rtracklayer::import(file.path("../../../../genomes/Mus_musculus/",
                                           "Gencode/M25/",
                                           "gencode.vM25.annotation.gtf"))
  genes <- gencode[gencode$type == "gene"]
  rtracklayer::export(genes, 
                      "../util/gencode.vM25.annotation.genes.gtf")
}
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id

g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

tx2gene <- read.csv("../../rnaseq/results/salmon/tx2gene.csv")

### COUNTS
tpm <- read.csv("../../rnaseq/results/salmon/salmon_merged_gene_tpm.csv") %>%
  merge(g2s) %>%
  dplyr::select(gene_id, gene_name, everything())

salmon_gene_counts <- read.csv(file.path("../../rnaseq/results/salmon/",
                                         "salmon_merged_gene_counts.csv")) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()
mode(salmon_gene_counts) <- "integer"

# Put in order
genes <- genes[rownames(salmon_gene_counts)]

### SAMPLE INFO
# Somehow the long timepoints didn't make it into this run.
# TODO: re-run with the long timepoints. Gaaah.
samples <- read.csv("../../rnaseq/samplesheet.csv") %>%
  filter(cell_type != "pMEF",
         date_sequenced != "Sep-18",
         timepoint_minutes < 400,
         timecourse_length == "short") %>%
  mutate(firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         timepoint_minutes = factor(timepoint_minutes, 
                                    levels = c(seq(0,360,30))),
         firre_induced = factor(firre_induced, 
                                levels = c("control", "firre_induced")))
rownames(samples) <- samples$sample_id
```

### DEseq2 

```{r deseq2}
# Since this takes a while to run, we'll cache this part
if(!file.exists("results/deseq_res.csv")) {
  
  ## FULL MODEL
  design_formula <- formula(paste0("~ timepoint_minutes + firre_induced + ",
                                   "firre_induced*timepoint_minutes"))
  ## REDUCED MODEL
  reduced_formula <- formula("~ timepoint_minutes + firre_induced")
  
  ## EXPERIMENTS
  # We're going to make 4 comparisons
  experiments <- c("ESC_KO","ESC_WT", "NPC_KO", "NPC_WT")
  
  ## RUN DESEQ2
  deseq_res <- lapply(experiments, run_timecourse_deseq,
                      counts = salmon_gene_counts, 
                      samples = samples,
                      genes = genes,
                      design_formula = design_formula,
                      reduced_formula = reduced_formula,
                      independent_filtering = FALSE)
  
  resdf <- bind_rows(deseq_res)
  write_csv(resdf, "results/deseq_res.csv")
}
resdf <- read.csv("results/deseq_res.csv")
```

#### Independent filtering

Something is kind of wonky with the independent filtering cutoff for NPC WT
I think it could be because not much is changing in that data or 
because Dox is causing a lot of changes, but DEseq2 has set really 
high filterThreshold.

For that reason, I will run the experiments without filtering in
DESeq2 and then filter at a filterThreshold of 100 baseMean counts.

| ORIGINAL DESEQ2 filterThresholds |
|----------------------------------|
| NPC WT = 969.99 |
| NPC KO = 0 |
| ESC WT = 3.35 |
| ESC KO = 79.98 |

For more on independent filtering, see: 
[DEseq2 vignette: independent filtering](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#independent-filtering-of-results)

```{r independent_filtering}
# Select a new independent filtering baseMean cutoff
resdf$pvalue[resdf$baseMean < 100] <- NA
resdf$padj <- NA

# Also, since these are independent experiments
# and since DEseq2 would typically do multiple
# hypothesis testing for each experiment indivudally
# we'll apply the testing correction per experiment
# here as well.
resdf <- resdf %>% group_by(cell_type, firre_ko) %>%
  mutate(padj = p.adjust(pvalue, method="BH"))
```

### P-value histogram

The pvalue histogram appears to be bimodal with a large number of p-values 
near 1. This would indicate that the test that we have is not the ideal 
statistical test. However, theoretically it is a good test for the question
we're trying to answer and Generalized linear models (GLM) as well as a 
likelihood ratio test with a full and reduced formula remains the gold standard
for this sort of analysis [ref](https://academic.oup.com/bib/article/20/1/288/4364840).
One alternative to explore is [impulseDE2](https://bioconductor.org/packages/release/bioc/vignettes/ImpulseDE2/inst/doc/ImpulseDE2_Tutorial.html).

```{r pval_hist, message=FALSE}
# First let's check out the distribution of p-values
hist(resdf$pvalue, main = "p-value histogram")

# One possibility is that we've got 
# a bunch of pvalues with really low counts
# At the upper end of the p-value scale
g <- ggplot(resdf, aes(x = log10(baseMean), color = pvalue > 0.9))
g + geom_density() + 
  ggtitle("p-value > 0.9 baseMean distribution")
# Looks like that is not the case. 
# One other thing that we could do is plot some of the 
# profiles of these genes that have pval close to 1 
# to look at what might be going on.
```

### Threshold DEGs

We'll look at the number of DEGs for a few different cutoffs.

```{r pval_thresholds}
sig_summary <- resdf %>% 
  mutate(pthresh_01 = padj < 0.01,
         pthresh_05 = padj < 0.05,
         pthresh_10 = padj < 0.10) %>%
  dplyr::select(gene_id, gene_name, cell_type, firre_ko,
                pthresh_01, pthresh_05, pthresh_10) %>%
  unite("experiment", cell_type, firre_ko) %>%
  group_by(experiment) %>%
  distinct() %>%
  summarize(pthresh_01 = length(which(pthresh_01 == TRUE)),
            pthresh_05 = length(which(pthresh_05 == TRUE)),
            pthresh_10 = length(which(pthresh_10 == TRUE)))
knitr::kable(sig_summary)
```


Here we'll make a cutoff at padj < 0.05 and look at the overlap between
experiments. It looks like despite very few gene changing in the NPC KO
condition there are still some that overlap with the other conditions.
It is a good sanity check that Firre is overlapping in all four experiments.
Notably, ADGRG1 is also overlapping in all. Shf is overlapping 
in three conditions and this was one of the early responders as noted
in previous analyses.

```{r sig_deg}
pthresh <- 0.05

sig_matrix <- resdf %>% 
  filter(padj < pthresh) %>%
  dplyr::select(gene_id, gene_name, cell_type, firre_ko) %>%
  unite("experiment", cell_type, firre_ko) %>%
  mutate(sig = 1) %>%
  distinct() %>%
  pivot_wider(names_from = experiment, values_from = sig, values_fill = 0) %>%
  dplyr::select(-gene_id) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()

sigdf <- sig_matrix %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_name")
write_csv(sigdf, "results/deg_firre_induced_vs_control.csv")

gplots::venn(sig_matrix, simplify = TRUE)

pdf("figures/venn.pdf")
gplots::venn(sig_matrix, simplify = TRUE)
dev.off()
```

## Time LFC Plot

Here we'll plot the fold-changes (GLM coefficients) for the interaction term
which corresponds to Firre induction vs control through time. This will 
give us some idea of the trends that the genes are following that are Firre 
responders.

It appears that most of the genes are moving in a consistent direction
and not transiently responding. It may be that the GLM / LRT method used
is better at indentifying these types of trends rather than transiently 
changing genes. For completeness, I will analyze individual timepoints
in a pairwise manner to get a list of genes that are changing transiently.

It seems that our overlapping gene

```{r time_lfc_plot}
# Here we'll retrieve the lfc values for just
# the dynamic firre induced vs control comparison (the interaction term).
sig_dyn <- resdf %>%
  get_dynamic_comparison_df() %>%
  filter(padj < 0.05)

# Let's label the two genes with the highest fold-changes in the
# last time point.
ltp <- sig_dyn %>% filter(timepoint == 330 | timepoint == 360) %>%
  arrange(-abs(log2FoldChange)) %>%
  group_by(cell_type, firre_ko) %>%
  dplyr::slice(1:3)

g <- ggplot(sig_dyn, aes(x = timepoint, y = log2FoldChange, group = gene_name))
g + geom_hline(yintercept = 0) +
  geom_point(alpha = 0.7) + 
  geom_line(color = "#424242", alpha = 0.2) + 
  facet_wrap(cell_type~firre_ko, scales = "free_y") +
  theme_paperwhite() +
  geom_text_repel(data = ltp,
                  aes(label = gene_name))
```

```{r}
# Let's try out some shrinkage
dds1 <- readRDS("results/ESC_KO_dds.rds")
resultsNames(dds1)
res1 <- lfcShrink(dds1, coef = "timepoint_minutes30.firre_inducedfirre_induced",
                  type = "apeglm", parallel = TRUE, BPPARAM=MulticoreParam(8))
```







```{r}



resdf_wide <- resdf %>% dplyr::select(gene_id,
                                      gene_name, cell_type,
                                      result_name, log2FoldChange,
                                      padj) %>%
  pivot_wider(id_cols = c("gene_id", "gene_name", "result_name"),
              names_from = "cell_type", values_from = c("log2FoldChange", "padj"))

sig_both <- resdf_wide %>% filter(padj_mESC < 0.01,
                                  padj_NPC < 0.01)
sig_npc <- resdf_wide %>% filter(padj_NPC < 0.05)
sig_esc <- resdf_wide %>% filter(padj_mESC < 0.05)
length(unique(sig_npc$gene_name))
length(unique(sig_esc$gene_name))
# Let's make some sweet plots.
goi <- "Firre"
goi <- "Supv3l1"
goi <- "Shf"
goi <- "Adgrg1"
goi <- "Lefty2"
goi <- "Gm47652"
goi <- "Il9r"
goi <- "Shf"
goi <- "Duox1"
goi <- "Gfap"
goi <- "Apod"
goi <- "Gm13040"
goi <- "Gm10039"

res_goi <- resdf %>% filter(gene_name == goi)
tpm_goi <- tpm %>% filter(gene_name == goi) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id",
               values_to = "tpm") %>%
  merge(samples) %>%
  unite("condition", cell_type, firre_ko, firre_induced, sep = ";",
        remove = FALSE)

g <- ggplot(tpm_goi, aes(x = timepoint_minutes, y = tpm, 
                         color = firre_induced,
                         group = condition))
g + geom_point() + 
  geom_smooth(method = "loess", formula = "y ~ x", se = FALSE) +
  facet_grid(cell_type~firre_ko, scales = "free_y") +
  theme_paperwhite() +
  scale_color_manual(values = c("#424242","#a8404c")) + 
  ggtitle(goi)


# Label the comparisons
res_goi$comparison <- "intercept"
res_goi[grep("timepoint_minutes_",res_goi$result_name),"comparison"] <- "vs_zero"
res_goi[res_goi$result_name == "firre_induced_firre_induced_vs_control",
        "comparison"] <- "static_firre_induction_vs_control"
res_goi[res_goi$result_name == "firre_ko_KO_vs_WT",
        "comparison"] <- "KO_vs_WT"
res_goi[grep(".firre_induced",res_goi$result_name,
             fixed = T),"comparison"] <- "dynamic_firre_induction_vs_control"

res_goi_dynamic <- res_goi %>% 
  filter(comparison == "dynamic_firre_induction_vs_control") 
res_goi_dynamic$timepoint <- sapply(res_goi_dynamic$result_name,
                                    function(x) {
                                      x <- gsub("timepoint_minutes", "", x)
                                      gsub(".firre_inducedfirre_induced", "", x,
                                           fixed = T)
                                    }) %>%
  as.numeric()
res_goi_dynamic$timepoint <- factor(res_goi_dynamic$timepoint, 
                                    levels = seq(30,360,30))
g <- ggplot(res_goi_dynamic, aes(y = log2FoldChange, x = timepoint))
g + geom_segment( aes(x= timepoint, xend= timepoint, y=0, yend=log2FoldChange)) +
  geom_point() + 
  geom_hline(yintercept = 0) +
  facet_grid(cell_type~.) + 
  theme_paperwhite() +
  ggtitle("Firre induction vs control")

res_goi_vs_zero <- res_goi %>% filter(comparison == "vs_zero")
res_goi_vs_zero$timepoint <- sapply(res_goi_vs_zero$result_name,
                                    function(x) {
                                      unlist(strsplit(x,"tes_"))[[2]]
                                    })
res_goi_vs_zero$timepoint <- factor(res_goi_vs_zero$timepoint, 
                                    levels = paste0(seq(30,360,30),"_vs_0"))

g <- ggplot(res_goi_vs_zero, aes(x = timepoint, y = log2FoldChange))
g + geom_hline(yintercept = 0) +
  geom_segment( aes(x= timepoint, xend= timepoint, y=0, yend=log2FoldChange)) +
  geom_point() +
  facet_grid(cell_type ~.) + 
  theme_paperwhite() + 
  ggtitle("Comparison to zero timepoint")


# Responds to Firre in time
# Is DE in response to Firre KO
```

```{r}
# # Aside for Adgrg1
# # https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz
# gtex <- read.table("data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct",
#                    sep = "\t",
#                    header = T,
#                    skip=2)
# grep("FIRRE", gtex$Description)
# grep("ADGRG1", gtex$Description)
# gtex_goi <- gtex[c(55114,42360),]
# gtex_goi <- gtex_goi %>%
#   pivot_longer(3:ncol(.),names_to = "sample",
#                values_to = "exp") %>%
#   dplyr::select(-Name) %>%
#   pivot_wider(names_from = Description, values_from = exp)
# library(ggpubr)
# gtex_goi$log10_firre <- log10(gtex_goi$FIRRE)
# gtex_goi$log10_adgrg1 <- log10(gtex_goi$ADGRG1)
# ggscatter(gtex_goi %>% filter(log10_firre > -3), x = "log10_firre", y = "log10_adgrg1") + 
#   stat_cor()
# g <- ggplot(gtex_goi %>% filter(), aes(x = log10(FIRRE), y = log10(ADGRG1)))
# g + stat_density_2d(aes(fill = ..level..), geom = "polygon") +
#   scale_fill_gradientn(colors = c("#FFEDA0", "#FEB24C", "#F03B20")) + 
#   theme_paperwhite() + 
#   stat_cor() + 
#   ggtitle("Co-expression across GTEX samples",
#           subtitle = "n = 17382")
# write_csv(gtex_goi, "results/gtex_firre_adgrg1.csv")
```





