---
title: "Does Firre induction rescue genes?"
author: "Michael Smallegan"
date: "9/24/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

```{r}
firre_wtvsko_esc <- read_csv("../03_firre_knockout/results/firre_ko_vs_wt_esc_res.csv")
firre_wtvsko_npc <- read_csv("../03_firre_knockout/results/firre_ko_vs_wt_npc_res.csv")

stc_res <- read_csv("../01_firre_induction_vs_control/results/deseq_res.csv")
ltc_res <- read_csv("../03_long_timecourse/results/deseq_res.csv")
```

Let's construct a matrix of sig genes to make some venn diagrams and see what
overlaps

```{r}
pthresh <- 0.01

firre_wtvsko_esc_sig_genes <- firre_wtvsko_esc %>%
  filter(padj < pthresh) %>%
  pull(gene_id)
firre_wtvsko_npc_sig_genes <- firre_wtvsko_npc %>%
  filter(padj < pthresh) %>%
  pull(gene_id)

stc_esc_sig_genes <- stc_res %>% filter(cell_type == "ESC", firre_ko == "KO",
                                        padj < 0.01) %>%
  pull(gene_id) %>%
  unique()
stc_npc_sig_genes <- stc_res %>% filter(cell_type == "NPC", firre_ko == "KO",
                                        padj < 0.01) %>%
  pull(gene_id) %>%
  unique()

ltc_esc_sig_genes <- ltc_res %>% filter(firre_ko == "KO", padj < 0.01) %>%
  pull(gene_id) %>%
  unique()
```

```{r}
# Let's just look at esc
esc_genes <- unique(c(firre_wtvsko_esc_sig_genes, stc_esc_sig_genes, ltc_esc_sig_genes))

esc_matrix <- matrix(nrow = length(esc_genes), ncol = 3)
rownames(esc_matrix) <- esc_genes
esc_matrix[,1] <- rownames(esc_matrix) %in% firre_wtvsko_esc_sig_genes
esc_matrix[,2] <- rownames(esc_matrix) %in% stc_esc_sig_genes
esc_matrix[,3] <- rownames(esc_matrix) %in% ltc_esc_sig_genes

colnames(esc_matrix) <- c("FKO", "STC", "LTC")
vennDiagram(esc_matrix)
plot(venneuler(esc_matrix))
```
```{r}
# Let's make some rld counts for all the ESC samples and look at whether there's
# trend over time of the firre induced samples back to the firre wt looking samples
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id

g2s <- genes %>% as.data.frame() %>%
  dplyr::select(gene_id, gene_name)

tx2gene <- read.csv("../../rnaseq/results/salmon/tx2gene.csv")


salmon_gene_counts <- read.csv(file.path("../../rnaseq/results/salmon/",
                                         "salmon_merged_gene_counts.csv")) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()
mode(salmon_gene_counts) <- "integer"
genes <- genes[rownames(salmon_gene_counts)]

samples <- read.csv("../../rnaseq/samplesheet.csv") %>%
  filter(cell_type == "ESC",
         date_sequenced != "2018-09") %>%
  mutate(firre_ko = factor(firre_ko, levels = c("WT", "KO")),
         timepoint_minutes = factor(timepoint_minutes, 
                                    levels = c(seq(0,360,30), 720, 1440, 2880, 5760)),
         firre_induced = factor(firre_induced, 
                                levels = c("control", "firre_induced"))) %>%
  unite(experiment, cell_type, firre_ko, remove = FALSE)
rownames(samples) <- samples$sample_id
counts <- salmon_gene_counts[,as.character(samples$sample_id)]
stopifnot(all(rownames(samples) == colnames(counts)))
dds <- DESeqDataSetFromMatrix(countData = counts, 
                                colData = samples, 
                                design = ~ firre_ko + firre_induced + timepoint_minutes,
                                rowData = genes)
dds <- dds[rowSums(counts(dds)) >= 20,]

if(!file.exists("results/esc_rld.RDS")) {
  rld <- rlog(dds, blind=TRUE)
  saveRDS(rld, "results/esc_rld.RDS")
}
rld <- readRDS("results/esc_rld.RDS")



```

```{r}
sampleDists <- dist(t(assay(rld)))
library("RColorBrewer")
library(pheatmap)
sampleDistMatrix <- as.matrix(sampleDists)

rownames(sampleDistMatrix) <- paste(rld$firre_ko, rld$firre_induced, rld$timecourse_length, rld$timepoint_minutes, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pdf("figures/esc_heatmap.pdf", height = 25, width = 30)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
dev.off()
plotPCA(rld, intgroup = c("firre_ko", "timecourse_length"))

rld_pca <- prcomp(t(assay(rld)))
rld_prcomps <- rld_pca$x %>% as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  select(sample_id, PC1, PC2, PC3)
rld_prcomps <- merge(samples, rld_prcomps)
stc_rld_prcomps <- rld_prcomps %>%
  filter(timecourse_length == "short",
         firre_induced == "firre_induced") %>%
  mutate(timepoint_minutes = as.numeric(as.character(timepoint_minutes)))

library(gganimate)
g <- ggplot(stc_rld_prcomps, aes(x = PC1, y = PC2, color = firre_ko))
g + geom_point() + transition_states(timepoint_minutes) + 
  labs(title = "{closest_state} min")

ltc_rld_prcomps <- rld_prcomps %>%
  filter(timecourse_length == "long") %>%
  mutate(timepoint_minutes = as.numeric(as.character(timepoint_minutes)))

g <- ggplot(ltc_rld_prcomps %>% filter(firre_induced == "firre_induced"), 
            aes(x = PC1, y = PC2, color = firre_ko)) + 
  geom_point() + transition_time(timepoint_minutes) + 
  geom_point(data = ltc_rld_prcomps %>% filter(timepoint_minutes == 0,
                                               firre_induced == "firre_induced") %>%
               dplyr::select(-timepoint_minutes), color = "red") +
  labs(title = "{round({frame_time},-2)} min") +
    geom_point(data = ltc_rld_prcomps %>% filter(timepoint_minutes == 5760,
                                             firre_induced == "control",
                                             timecourse_length == "long") %>%
                 dplyr::select(-timepoint_minutes), color = "black") +
    shadow_wake(wake_length = .2, size = TRUE, alpha = TRUE, colour = 'blue',
              fill = NULL, falloff = "cubic-in", wrap = TRUE,
              exclude_layer = NULL, exclude_phase = NULL)
animate(g, end_pause = 15, renderer = gifski_renderer(loop = TRUE))

g <- ggplot(ltc_rld_prcomps %>% filter(firre_induced == "control"), 
            aes(x = PC1, y = PC2, color = firre_ko)) + 
  geom_point() + transition_time(timepoint_minutes) + 
  geom_point(data = ltc_rld_prcomps %>% filter(timepoint_minutes == 0,
                                               firre_induced == "firre_induced") %>%
               dplyr::select(-timepoint_minutes), color = "red") +
  labs(title = "{round({frame_time},-2)} min") +
    geom_point(data = ltc_rld_prcomps %>% filter(timepoint_minutes == 0,
                                             firre_induced == "control",
                                             timecourse_length == "long") %>%
                 dplyr::select(-timepoint_minutes), color = "black") +
    shadow_wake(wake_length = .2, size = TRUE, alpha = TRUE, colour = 'blue',
              fill = NULL, falloff = "cubic-in", wrap = TRUE,
              exclude_layer = NULL, exclude_phase = NULL)
animate(g, end_pause = 15, renderer = gifski_renderer(loop = TRUE))

```






```{r}
sig_esc <- esc_resdf %>% filter(padj < 0.01)
sig_npc <- npc_resdf %>% filter(padj < 0.01)


sig_genes <- unique(c(sig_esc$gene_id, sig_npc$gene_id))
sig_matrix <- matrix(nrow = length(sig_genes), ncol = 2)
rownames(sig_matrix) <- sig_genes
sig_matrix[,1] <- rownames(sig_matrix) %in% sig_esc$gene_id
sig_matrix[,2] <- rownames(sig_matrix) %in% sig_npc$gene_id

colnames(sig_matrix) <- c("ESC", "NPC")

library(limma)
vennCounts(sig_matrix)
vennDiagram(sig_matrix)
```

```{r}
# Now the question is how best to reprsent the rescue?
# Start with the short timecourse resutls

esc_sig_short <- short_tc_res %>% filter(padj < 0.01, cell_type == "ESC")
esc_sig_short_genes <- unique(esc_sig_short$gene_id)
npc_sig_short <- short_tc_res %>% filter(padj < 0.01, cell_type == "NPC")
npc_sig_short_genes <- unique(npc_sig_short$gene_id)

short_ko_matrix_genes <- unique(c(rownames(sig_matrix), esc_sig_short_genes, 
                                  npc_sig_short_genes))
short_ko_matrix <- matrix(nrow = length(short_ko_matrix_genes), 
                          ncol = 4)
rownames(short_ko_matrix) <- short_ko_matrix_genes
short_ko_matrix[,1] <- rownames(short_ko_matrix) %in% sig_esc$gene_id
short_ko_matrix[,2] <- rownames(short_ko_matrix) %in% sig_npc$gene_id
short_ko_matrix[,3] <- rownames(short_ko_matrix) %in% esc_sig_short_genes
short_ko_matrix[,4] <- rownames(short_ko_matrix) %in% npc_sig_short_genes
colnames(short_ko_matrix) <- c("ESC_FKO", "NPC_FKO", "ESC_FI", "NPC_FI")
vennDiagram(short_ko_matrix)

esc_short_ko_matrix <- short_ko_matrix[,c("ESC_FKO", "ESC_FI")]
vennDiagram(esc_short_ko_matrix)
npc_short_ko_matrix <- short_ko_matrix[,c("NPC_FKO", "NPC_FI")]
vennDiagram(npc_short_ko_matrix)
```


```{r}
# What if we just took the significant genes?
stc_samples <- samples %>% filter(timecourse_length == "short")
rld_sig <- assay(rld) 
rld_sig <- rld_sig[rownames(esc_matrix), stc_samples$sample_id]
pca_sig <- prcomp(t(rld_sig))
pca_sig_df <- pca_sig$x %>% as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  select(sample_id, PC1, PC2, PC3, PC4, PC5) %>%
  merge(samples)

pca_sig_df$timepoint_minutes <- as.numeric(as.character(pca_sig_df$timepoint_minutes))
g <- ggplot(pca_sig_df %>% filter(timecourse_length == "short"), 
            aes(x = PC1, y = PC2, color = firre_induced)) + geom_point() +
  transition_time(timepoint_minutes) + 
  geom_point(data = pca_sig_df %>% filter(timepoint_minutes == 0) %>%
               dplyr::select(-timepoint_minutes), color = "red") +
  labs(title = "{round({frame_time},-2)} min") +
    shadow_wake(wake_length = .2, size = TRUE, alpha = TRUE, colour = 'blue',
              fill = NULL, falloff = "cubic-in", wrap = TRUE,
              exclude_layer = NULL, exclude_phase = NULL)
animate(g)

rownames(esc_matrix)
rownames(assay(rld))
rld_pca <- prcomp(t(assay(rld)))

g <- ggplot(pca_sig_df, aes(x = PC1, y = PC2, color = timecourse_length))
g + geom_point()  

  
```

```{r}
# Let's get a little weird for a sec.
# We want to transform the firre wt samples so that they start from the same place 
# as the ko samples. 
zero_wt <- stc_samples %>% filter(firre_ko == "WT", timepoint_minutes == 0)
mean_exp_wt_zero <- rowMeans(rld_sig[,zero_wt$sample_id])
zero_ko <- stc_samples %>% filter(firre_ko == "KO", timepoint_minutes == 0)
mean_exp_ko_zero <- rowMeans(rld_sig[,zero_ko$sample_id])
ko_to_wt_transform <- mean_exp_wt_zero - mean_exp_ko_zero

# So now let's take the ko samples and add this transformation
ko_samples <- stc_samples %>% filter(firre_ko == "KO")
rld_ko <- rld_sig[, ko_samples$sample_id]
rld_kot <- rld_ko + ko_to_wt_transform

rld_kot_pca <- predict(pca_sig, t(rld_kot)) %>%
  as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  select(sample_id, PC1, PC2, PC3) %>%
  merge(stc_samples)

rld_kot_pca$timepoint_minutes <- as.numeric(as.character(rld_kot_pca$timepoint_minutes))
rld_kot_pca$transformed <- TRUE
pca_sig_df$transformed <- FALSE
ckot_pca <- bind_rows(rld_kot_pca, pca_sig_df)
ckot_pca <- ckot_pca %>% filter(!(firre_ko == "KO" & transformed == FALSE))
g <- ggplot(ckot_pca , aes(x = PC1, y = PC2, color = firre_ko))
g + geom_point()


g <- ggplot(ckot_pca %>% filter(firre_induced == "firre_induced"), 
            aes(x = PC1, y = PC2, color = firre_ko)) + geom_point() +
  transition_time(timepoint_minutes) + 
  geom_point(data = ckot_pca %>% filter(timepoint_minutes == 0) %>%
               dplyr::select(-timepoint_minutes), color = "red") +
  labs(title = "{round({frame_time},-2)} min") +
    shadow_wake(wake_length = .2, size = TRUE, alpha = TRUE, colour = 'blue',
              fill = NULL, falloff = "cubic-in", wrap = TRUE,
              exclude_layer = NULL, exclude_phase = NULL)
animate(g)

g <- ggplot(ckot_pca %>% filter(firre_induced == "control"), 
            aes(x = PC1, y = PC2, color = firre_ko)) + geom_point() +
  transition_time(timepoint_minutes) + 
  geom_point(data = ckot_pca %>% filter(timepoint_minutes == 0) %>%
               dplyr::select(-timepoint_minutes), color = "red") +
  labs(title = "{round({frame_time},-2)} min") +
    shadow_wake(wake_length = .2, size = TRUE, alpha = TRUE, colour = 'blue',
              fill = NULL, falloff = "cubic-in", wrap = TRUE,
              exclude_layer = NULL, exclude_phase = NULL)
animate(g)



```

```{r}

rld_sig <- assay(rld) 
rld_sig <- rld_sig[rownames(esc_matrix), samples$sample_id]
pca_sig <- prcomp(t(rld_sig))
pca_sig_df <- pca_sig$x %>% as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  select(sample_id, PC1, PC2, PC3, PC4, PC5) %>%
  merge(samples)
pca_sig_df$timepoint_minutes <- as.numeric(as.character(pca_sig_df$timepoint_minutes))
g <- ggplot(pca_sig_df %>% filter(timecourse_length == "long", firre_induced == "firre_induced"), 
            aes(x = PC1, y = PC2, color = firre_induced)) + geom_point()  +
  transition_time(timepoint_minutes) +
   geom_point(data = pca_sig_df %>% filter(timepoint_minutes == 0,
                                           firre_ko == "KO",
                                           timecourse_length == "long") %>%
               dplyr::select(-timepoint_minutes), color = "blue") +
     geom_point(data = pca_sig_df %>% filter(timepoint_minutes == 0,
                                           firre_ko == "WT",
                                           timecourse_length == "long") %>%
               dplyr::select(-timepoint_minutes), color = "red") +
  labs(title = "{round({frame_time},-2)} min") +
    shadow_wake(wake_length = .2, size = TRUE, alpha = TRUE, colour = 'blue',
              fill = NULL, falloff = "cubic-in", wrap = TRUE,
              exclude_layer = NULL, exclude_phase = NULL)
animate(g)
```

```{r}

# Let's do a weird transform to center up the long timecourse and the short timecourse
rld_assay <- assay(rld)
rld_assay <- rld_assay[rownames(esc_matrix),]
zero_ko_short_samples <- samples %>% filter(timepoint_minutes == 0, timecourse_length == "short", firre_ko == "KO")
zero_ko_long_samples <- samples %>% filter(timepoint_minutes == 0, timecourse_length == "long", firre_ko == "KO")
zero_ko_short <- rowMeans(rld_assay[,zero_ko_short_samples$sample_id])
zero_ko_long <- rowMeans(rld_assay[,zero_ko_long_samples$sample_id])

long_to_short_transform <- zero_ko_short - zero_ko_long

long_samples <- samples %>% filter(timecourse_length == "long")
rld_long <- rld_assay[,long_samples$sample_id]

rld_longt <- rld_long + long_to_short_transform

short_samples <- samples %>% filter(timecourse_length == "short")
rld_short <- rld_assay[,short_samples$sample_id]

rld_ct <- cbind(rld_short, rld_longt)

pca_lt <- prcomp(t(rld_ct))
pca_lt_df <- pca_lt$x %>% as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  select(sample_id, PC1, PC2) %>%
  merge(samples) %>%
  mutate(timepoint_minutes = as.numeric(as.character(timepoint_minutes)))

g <- ggplot(pca_lt_df, aes(x = PC1, y = PC2, color = firre_induced))
g + geom_point()

# Let's make a pca just separating the firre ko vs wt zero timepoints

zero_samples <- samples %>% filter(timepoint_minutes == 0)
rld_zero <- rld_ct[rownames(esc_matrix), zero_samples$sample_id]
# rld_zero <- rld_zero[, samples$sample_id]
pca_zero <- prcomp(t(rld_zero))
pca_zero_df <- pca_zero$x %>% as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  select(sample_id, PC1, PC2) %>%
  merge(zero_samples) %>%
  mutate(timepoint_minutes = as.numeric(as.character(timepoint_minutes)))



nz_rld <- rld_ct[,samples %>% filter(timepoint_minutes != 0) %>% pull(sample_id)]
nz_pca <- predict(pca_zero, t(nz_rld)) %>%
  as.data.frame() %>%
  rownames_to_column("sample_id") %>%
  select(sample_id, PC1, PC2, PC3) %>%
  merge(samples) %>%
  mutate(timepoint_minutes = as.numeric(as.character(timepoint_minutes)))

c_pca <- bind_rows(pca_zero_df, nz_pca)

c_pca <- c_pca %>% unite(condition, firre_ko, firre_induced, remove = FALSE)
g <- ggplot(c_pca %>% filter(timecourse_length == "short"), aes(x = PC1, y = PC2, color = condition))
g + geom_point()  +
  transition_time(timepoint_minutes)
  
```








