---
title: "Untitled"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(ggrepel)
library(gganimate)
source("../util/_plot_theme.R")
```

```{r}
# cp /scratch/Shares/public/singularity/tfea-3.0.img /scratch/Shares/rinn/Michael/firre_timecourse/analysis/04_tfea/bin


```

Let's generate a ranked list for the 30 minute timepoint for Firre vs control

```{r}
# Define promoter regions
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id
proms <- promoters(genes, upstream = 1500, downstream = 1500)


res_shrunkendf <- read_csv(file.path("../01_firre_induction_vs_control",
                                     "results/deseq_res_shrunken.csv"))
dynamic_res <- res_shrunkendf %>%
  filter(grepl(".firre_inducedfirre_induced", result_name))
unique(dynamic_res$result_name)
dynamic_res$timepoint <- gsub("timepoint_minutes", "", dynamic_res$result_name)
dynamic_res$timepoint <- gsub(".firre_inducedfirre_induced", "", dynamic_res$timepoint)
dynamic_res$timepoint <- as.numeric(dynamic_res$timepoint)

dynamic_res <- dynamic_res %>%
  unite(experiment, cell_type, firre_ko, remove = F)
experiments <- unique(dynamic_res$experiment)

dir.create("results/firre_induction_vs_control_rankings")
dir.create("results/tfea_results")

for(j in 1:length(experiments)) {
  exp_res <- dynamic_res %>%
    filter(experiment == experiments[j])
  timepoints <- unique(exp_res$timepoint)
  for(i in 1:length(timepoints)) {
    
    tp_order <- exp_res %>%
      filter(timepoint == timepoints[i],
             !is.na(padj)) %>%
      arrange(-log2FoldChange)
    tp_promoters <- proms[tp_order$gene_id] %>%
      as.data.frame() %>%
      dplyr::select(seqnames, start, end)
    
    if(any(tp_promoters$start < 0)) {
      warning(paste("Promoter start < 0 in ", experiments[j], " tp: ", timepoints[i]))
      tp_promoters[which(tp_promoters$start < 0), "start"] <- 1
    }
    
    cat("#chrom	start	stop \n",file=paste0("results/firre_induction_vs_control_rankings/",
                                           experiments[j],"_tp",timepoints[i],".bed"))
    write.table(tp_promoters, paste0("results/firre_induction_vs_control_rankings/",
                                     experiments[j],"_tp",timepoints[i],".bed"),
                row.names = FALSE, col.names = FALSE, sep = "\t",
                append = TRUE, quote = FALSE)
    dir.create(paste0("results/tfea_results/", 
                      experiments[j],"_tp",timepoints[i]))
    
    create_sbatch_script(experiments[j], timepoints[i])
  }
}



```

```{r}
# Create tfea batch jobs
dir.create("bin/tfea_sbatch")
create_sbatch_script <- function(experiment, timepoint) {
  
line <- paste0("#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=tfea_", experiment, "_", timepoint, "
#SBATCH --mail-type=NONE
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=6gb
#SBATCH --time=12:00:00
#SBATCH --output=", experiment, "_", timepoint, "_tfea.out
#SBATCH --error=", experiment, "_", timepoint, "_tfea.err

module load singularity/3.1.1

BASE_DIR='/scratch/Shares/rinn/Michael/firre_timecourse/analysis/04_tfea'

singularity exec \\
--bind $BASE_DIR/results/firre_induction_vs_control_rankings:/mnt/rankings \\
--bind $BASE_DIR/results/tfea_results/", experiment, "_tp", timepoint,":/mnt/output \\
--bind $BASE_DIR/data/motif_databases/MOUSE/:/mnt/mouse_meme \\
--bind /scratch/Shares/rinn/genomes/Mus_musculus/Gencode/M25:/mnt/genome \\
	$BASE_DIR/bin/tfea.sif TFEA \\
--output /mnt/output \\
--ranked_file /mnt/rankings/", experiment, "_tp", timepoint, ".bed \\
--label1 firre_induced --label2 dox_control \\
--genomefasta /mnt/genome/GRCm38.p6.genome.fa \\
--fimo_motifs /mnt/mouse_meme/HOCOMOCOv11_full_MOUSE_mono_meme_format.meme")
writeLines(line, 
           con = paste0("bin/tfea_sbatch/", experiment,"_tp",timepoint, ".sbatch"))


}

```

```{r}
# Create sumit script
cat(paste0("#!/bin/bash \n",
              paste(paste0("sbatch ", list.files("bin/tfea_sbatch/", pattern = ".sbatch")),
              collapse = "\n")),
    file = "bin/tfea_sbatch/submit.sh")
```

```{r}
# Read in results
tfea_res_dirs <- list.dirs("results/tfea_results/", full.names = TRUE,
                           recursive = FALSE)
tfea_res_files <- paste0(tfea_res_dirs, "/results.txt")

tfea_res <- read.table(tfea_res_files[1],
                   col.names = c("TF", "e_score", "corrected_e_score", "events",
                                 "GC", "FPKM", "padj", "corrected_padj"))
tfea_res$condition <- gsub("/results.txt", "", 
                       gsub("results/tfea_results//", "", 
                            tfea_res_files[1]))

for(i in 2:length(tfea_res_files)) {
  tmp_res <- read.table(tfea_res_files[i],
                   col.names = c("TF", "e_score", "corrected_e_score", "events",
                                 "GC", "FPKM", "padj", "corrected_padj"))
  tmp_res$condition <- gsub("/results.txt", "", 
                       gsub("results/tfea_results//", "", 
                            tfea_res_files[i]))
  tfea_res <- bind_rows(tfea_res, tmp_res)
}

tfea_res <- tfea_res %>% separate(condition, into = c("cell_type",
                                                      "firre_ko",
                                                      "timepoint"), 
                                  remove = FALSE)
tfea_res$timepoint <- as.numeric(gsub("tp", "", tfea_res$timepoint))
tfea_res$tf_name <- sapply(tfea_res$TF, function(x) {
  unlist(strsplit(x, "_"))[[1]]
})

esc_ko_tfea_res <- tfea_res %>% filter(cell_type == "ESC", firre_ko == "KO")
sig_esc_ko_tfs <- esc_ko_tfea_res %>% filter(corrected_padj < 0.1) %>%
  dplyr::select(TF) %>%
  distinct()
esc_ko_tfea_res$sig <- esc_ko_tfea_res$TF %in% sig_esc_ko_tfs$TF
table(esc_ko_tfea_res$sig)
esc_ko_tfea_res$sig_timepoint <- esc_ko_tfea_res$corrected_padj < 0.1
g <- ggplot(esc_ko_tfea_res %>% filter(sig == TRUE), aes(x = timepoint, 
                                                         y = corrected_e_score, 
                                                         group = TF,
                                 color = sig_timepoint))
g + geom_point() + geom_line()

g <- ggplot(esc_ko_tfea_res %>% filter(timepoint == 30),
            aes(x = events, y = corrected_e_score, label = TF))
g + geom_point() + geom_text_repel(data = esc_ko_tfea_res %>% filter(timepoint == 30,
                                                                     sig_timepoint == TRUE))

g <- ggplot(esc_ko_tfea_res %>% filter(timepoint == 330),
            aes(x = log10(events), y = corrected_e_score, label = TF))
g + geom_point() + geom_text_repel(data = esc_ko_tfea_res %>% filter(timepoint == 330,
                                                                     sig_timepoint == TRUE))


library(gganimate)
library(ggrepel)
p <- ggplot(esc_ko_tfea_res,
            aes(x = log10(events), y = corrected_e_score, label = tf_name)) + 
  geom_point(data = esc_ko_tfea_res %>% filter(sig == FALSE),
             alpha = 0.1) + 
    geom_point(data = esc_ko_tfea_res %>% filter(sig == TRUE),
             color = "#a8404c") + 
  geom_text_repel(data = esc_ko_tfea_res %>% filter(sig_timepoint == TRUE),
            nudge_x = 0.1, nudge_y = 0.01) +
  theme_paperwhite() +
  labs(title = 'ESC KO: {closest_state} min') +
  transition_states(timepoint, transition_length = 10, state_length = 30) +
  ease_aes('linear')
animate(p, renderer = gifski_renderer(),
        nframes = 200)
# anim_save("esc_ko_tfea_ma.gif")

# ggplot(gapminder, aes(gdpPercap, lifeExp, size = pop, colour = country)) +
#   geom_point(alpha = 0.7, show.legend = FALSE) +
#   scale_colour_manual(values = country_colors) +
#   scale_size(range = c(2, 12)) +
#   scale_x_log10() +
#   facet_wrap(~continent) +
#   # Here comes the gganimate specific bits
#   labs(title = 'Year: {frame_time}', x = 'GDP per capita', y = 'life expectancy') +
#   transition_time(year) +
#   ease_aes('linear')
```




