---
title: "Untitled"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(ggrepel)
library(gganimate)
source("../util/_plot_theme.R")
```

```{r}
create_sbatch_script <- function(experiment, timepoint, rankings_dir, output_dir,
                                 script_dir) {
  
line <- paste0("#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=tfea_", experiment, "_", timepoint, "
#SBATCH --mail-type=NONE
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=6gb
#SBATCH --time=4:00:00
#SBATCH --output=", experiment, "_", timepoint, "_tfea.out
#SBATCH --error=", experiment, "_", timepoint, "_tfea.err

module load singularity/3.1.1

BASE_DIR='/scratch/Shares/rinn/Michael/firre_timecourse/analysis/04_tfea'

singularity exec \\
--bind $BASE_DIR/", rankings_dir,":/mnt/rankings \\
--bind $BASE_DIR/", output_dir, experiment, "_tp", timepoint,":/mnt/output \\
--bind $BASE_DIR/data/motif_databases/MOUSE/:/mnt/mouse_meme \\
--bind /scratch/Shares/rinn/genomes/Mus_musculus/Gencode/M25:/mnt/genome \\
	$BASE_DIR/bin/tfea.sif TFEA \\
--output /mnt/output \\
--ranked_file /mnt/rankings/", experiment, "_tp", timepoint, ".bed \\
--label1 lb1 --label2 lb2 \\
--genomefasta /mnt/genome/GRCm38.p6.genome.fa \\
--fimo_motifs /mnt/mouse_meme/HOCOMOCOv11_full_MOUSE_mono_meme_format.meme")
writeLines(line, 
           con = paste0(script_dir, experiment,"_tp",timepoint, ".sbatch"))
}

```

Let's generate a ranked list for the 30 minute timepoint for Firre vs control

```{r}
# Define promoter regions
genes <- rtracklayer::import(file.path("../util/",
                                       "gencode.vM25.annotation.genes.gtf"))
names(genes) <- genes$gene_id
proms <- promoters(genes, upstream = 1500, downstream = 1500)


res_shrunkendf <- read_csv(file.path("../01_firre_induction_vs_control",
                                     "results/deseq_res_shrunken.csv"))
dynamic_res <- res_shrunkendf %>%
  filter(grepl(".firre_inducedfirre_induced", result_name))
unique(dynamic_res$result_name)
dynamic_res$timepoint <- gsub("timepoint_minutes", "", dynamic_res$result_name)
dynamic_res$timepoint <- gsub(".firre_inducedfirre_induced", "", dynamic_res$timepoint)
dynamic_res$timepoint <- as.numeric(dynamic_res$timepoint)

dynamic_res <- dynamic_res %>%
  unite(experiment, cell_type, firre_ko, remove = F)
experiments <- unique(dynamic_res$experiment)

dir.create("results/firre_induction_vs_control_rankings")
dir.create("results/tfea_results")

for(j in 1:length(experiments)) {
  exp_res <- dynamic_res %>%
    filter(experiment == experiments[j])
  timepoints <- unique(exp_res$timepoint)
  for(i in 1:length(timepoints)) {
    
    tp_order <- exp_res %>%
      filter(timepoint == timepoints[i],
             !is.na(padj)) %>%
      arrange(-log2FoldChange)
    tp_promoters <- proms[tp_order$gene_id] %>%
      as.data.frame() %>%
      dplyr::select(seqnames, start, end)
    
    if(any(tp_promoters$start < 0)) {
      warning(paste("Promoter start < 0 in ", experiments[j], " tp: ", timepoints[i]))
      tp_promoters[which(tp_promoters$start < 0), "start"] <- 1
    }
    
    cat("#chrom	start	stop \n",file=paste0("results/firre_induction_vs_control_rankings/",
                                           experiments[j],"_tp",timepoints[i],".bed"))
    write.table(tp_promoters, paste0("results/firre_induction_vs_control_rankings/",
                                     experiments[j],"_tp",timepoints[i],".bed"),
                row.names = FALSE, col.names = FALSE, sep = "\t",
                append = TRUE, quote = FALSE)
    dir.create(paste0("results/tfea_results/", 
                      experiments[j],"_tp",timepoints[i]))
    
    create_sbatch_script(experiments[j], timepoints[i], 
                         "results/firre_induction_vs_control_rankings",
                         "results/tfea_results/",
                         "bin/tfea_sbatch/")
  }
}

# Create sumit script
cat(paste0("#!/bin/bash \n",
              paste(paste0("sbatch ", list.files("bin/tfea_sbatch/", pattern = ".sbatch")),
              collapse = "\n")),
    file = "bin/tfea_sbatch/submit.sh")
```

Dox control prep for TFEA

```{r}
res_shrunkendf <- read.csv(file.path("../02_doxycycline_responders/",
                                     "results/deseq_dox_control_res_shrunken.csv"))
unique(res_shrunkendf$result_name)

res_shrunkendf$timepoint <- gsub("timepoint_minutes_", "", res_shrunkendf$result_name)
res_shrunkendf$timepoint <- gsub("_vs_0", "", res_shrunkendf$timepoint)
res_shrunkendf$timepoint <- as.numeric(res_shrunkendf$timepoint)

res_shrunkendf <- res_shrunkendf %>%
  unite(experiment, cell_type, firre_ko, remove = F)
experiments <- unique(res_shrunkendf$experiment)

dir.create("results/dox_vs_zero_rankings")
dir.create("results/dox_tfea_results")
dir.create("bin/dox_tfea_sbatch")

for(j in 1:length(experiments)) {
  exp_res <- res_shrunkendf %>%
    filter(experiment == experiments[j])
  timepoints <- unique(exp_res$timepoint)
  for(i in 1:length(timepoints)) {
    
    tp_order <- exp_res %>%
      filter(timepoint == timepoints[i],
             !is.na(padj)) %>%
      arrange(-log2FoldChange)
    tp_promoters <- proms[tp_order$gene_id] %>%
      as.data.frame() %>%
      dplyr::select(seqnames, start, end)
    
    if(any(tp_promoters$start < 0)) {
      warning(paste("Promoter start < 0 in ", experiments[j], " tp: ", timepoints[i]))
      tp_promoters[which(tp_promoters$start < 0), "start"] <- 1
    }
    
    cat("#chrom	start	stop \n",file=paste0("results/dox_vs_zero_rankings/",
                                           experiments[j],"_tp",timepoints[i],".bed"))
    write.table(tp_promoters, paste0("results/dox_vs_zero_rankings/",
                                     experiments[j],"_tp",timepoints[i],".bed"),
                row.names = FALSE, col.names = FALSE, sep = "\t",
                append = TRUE, quote = FALSE)
    dir.create(paste0("results/dox_tfea_results/", 
                      experiments[j],"_tp",timepoints[i]))
    
    create_sbatch_script(experiments[j], timepoints[i],
                           "results/dox_vs_zero_rankings/",
                         "results/dox_tfea_results/",
                         "bin/dox_tfea_sbatch/")
  }
}

# Create submit script
cat(paste0("#!/bin/bash \n",
              paste(paste0("sbatch ", list.files("bin/dox_tfea_sbatch/", pattern = ".sbatch")),
              collapse = "\n")),
    file = "bin/dox_tfea_sbatch/submit.sh")
```


## Firre response TFEA results

```{r}
# Read in results
tfea_res_dirs <- list.dirs("results/tfea_results/", full.names = TRUE,
                           recursive = FALSE)
tfea_res_files <- paste0(tfea_res_dirs, "/results.txt")

tfea_res <- read.table(tfea_res_files[1],
                   col.names = c("TF", "e_score", "corrected_e_score", "events",
                                 "GC", "FPKM", "padj", "corrected_padj"))
tfea_res$condition <- gsub("/results.txt", "", 
                       gsub("results/tfea_results//", "", 
                            tfea_res_files[1]))

for(i in 2:length(tfea_res_files)) {
  tmp_res <- read.table(tfea_res_files[i],
                   col.names = c("TF", "e_score", "corrected_e_score", "events",
                                 "GC", "FPKM", "padj", "corrected_padj"))
  tmp_res$condition <- gsub("/results.txt", "", 
                       gsub("results/tfea_results//", "", 
                            tfea_res_files[i]))
  tfea_res <- bind_rows(tfea_res, tmp_res)
}

tfea_res <- tfea_res %>% separate(condition, into = c("cell_type",
                                                      "firre_ko",
                                                      "timepoint"), 
                                  remove = FALSE)
tfea_res$timepoint <- as.numeric(gsub("tp", "", tfea_res$timepoint))
tfea_res$tf_name <- sapply(tfea_res$TF, function(x) {
  unlist(strsplit(x, "_"))[[1]]
})


tfea_res <- tfea_res %>%
  group_by(cell_type, firre_ko, TF) %>%
  mutate(sig = any(corrected_padj <= 0.1)) %>%
  ungroup() %>%
  mutate(sig_timepoint = corrected_padj <= 0.1)

esc_tfea <- tfea_res %>% filter(cell_type == "ESC")

p <- ggplot(esc_tfea,
            aes(x = log10(events), y = corrected_e_score, label = tf_name, 
                group = TF)) + 
  geom_hline(yintercept = 0) +
  geom_text_repel(data = esc_tfea %>% filter(sig_timepoint == TRUE),
                  nudge_x = 0.1, nudge_y = 0.01, segment.alpha = 0) +
  geom_point(data = esc_tfea %>% filter(sig == FALSE),
             alpha = 0.1) + 
  geom_point(data = esc_tfea %>% filter(sig == TRUE),
             aes(color = sig_timepoint)) + 
  scale_color_manual(values = c("#424242","#a8404c")) +
  theme_paperwhite() +
  facet_grid(~firre_ko) +
  theme(legend.position = "none") +
  labs(title = 'ESC: {closest_state} min') +
  transition_states(timepoint, transition_length = 10, state_length = 30) +
  ease_aes('cubic-in-out') +
  enter_grow() +
  exit_shrink()
animate(p, renderer = gifski_renderer(),
        nframes = 200)
anim_save("figures/firre_response_ESC_TFEA.gif")

npc_tfea <- tfea_res %>% filter(cell_type == "NPC")

p <- ggplot(npc_tfea,
            aes(x = log10(events), y = corrected_e_score, label = tf_name, 
                group = TF)) + 
  geom_hline(yintercept = 0) +
  geom_text_repel(data = npc_tfea %>% filter(sig_timepoint == TRUE),
                  nudge_x = 0.1, nudge_y = 0.01, segment.alpha = 0) +
  geom_point(data = npc_tfea %>% filter(sig == FALSE),
             alpha = 0.1) + 
  geom_point(data = npc_tfea %>% filter(sig == TRUE),
             aes(color = sig_timepoint)) + 
  scale_color_manual(values = c("#424242","#a8404c")) +
  theme_paperwhite() +
  facet_grid(~firre_ko) +
  theme(legend.position = "none") +
  labs(title = 'NPC: {closest_state} min') +
  transition_states(timepoint, transition_length = 10, state_length = 30) +
  ease_aes('cubic-in-out') +
  enter_grow() +
  exit_shrink()
animate(p, renderer = gifski_renderer(),
        nframes = 200)
anim_save("figures/firre_response_NPC_TFEA.gif")
```




